{"cells":[{"metadata":{},"cell_type":"markdown","source":"## About this Notebook"},{"metadata":{},"cell_type":"markdown","source":"Researchers and engineers all over the globe are scrambling to develop new vaccines and diagnostic tools for COVID-19. One aspect of research is the analysis and quantification of virus proteins. Thus, this notebook serves the purpose of engineering features from a protein sequence that can be used as inputs to a classifier or regressor. In this illustration, we will use the protein sequence of the **Human SARS coronavirus (SARS-CoV) spike glycoprotein**, which was obtained from UniProt."},{"metadata":{},"cell_type":"markdown","source":"Disclaimer: I am the author of the `protlearn` package that is used throughout this notebook. The documentation of it can be found [here](https://github.com/tadorfer/ProtLearn)."},{"metadata":{"trusted":true},"cell_type":"code","source":"# install protlearn\n!pip install protlearn","execution_count":null,"outputs":[]},{"metadata":{"_uuid":"8f2839f25d086af736a60e9eeb907d3b93b6e0e5","_cell_guid":"b1076dfc-b9ad-4769-8c92-a6c4dae69d19","trusted":true},"cell_type":"code","source":"import pandas as pd \nimport protlearn as pl\nimport matplotlib.pyplot as plt","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"## Get protein sequence"},{"metadata":{},"cell_type":"markdown","source":"Get the amino acid sequence of the Human SARS coronavirus (SARS-CoV) spike glycoprotein from [UniProt](https://www.uniprot.org/uniprot/P59594.fasta). In this case, the sequence was copied from the UniProt website. Alternatively, the sequence can also be downloaded and imported into Python using the `biopython` package."},{"metadata":{"trusted":true},"cell_type":"code","source":"spike = \"MFIFLLFLTLTSGSDLDRCTTFDDVQAPNYTQHTSSMRGVYYPDEIFRSDTLYLTQDLFL\\\n         PFYSNVTGFHTINHTFGNPVIPFKDGIYFAATEKSNVVRGWVFGSTMNNKSQSVIIINNS\\\n         TNVVIRACNFELCDNPFFAVSKPMGTQTHTMIFDNAFNCTFEYISDAFSLDVSEKSGNFK\\\n         HLREFVFKNKDGFLYVYKGYQPIDVVRDLPSGFNTLKPIFKLPLGINITNFRAILTAFSP\\\n         AQDIWGTSAAAYFVGYLKPTTFMLKYDENGTITDAVDCSQNPLAELKCSVKSFEIDKGIY\\\n         QTSNFRVVPSGDVVRFPNITNLCPFGEVFNATKFPSVYAWERKKISNCVADYSVLYNSTF\\\n         FSTFKCYGVSATKLNDLCFSNVYADSFVVKGDDVRQIAPGQTGVIADYNYKLPDDFMGCV\\\n         LAWNTRNIDATSTGNYNYKYRYLRHGKLRPFERDISNVPFSPDGKPCTPPALNCYWPLND\\\n         YGFYTTTGIGYQPYRVVVLSFELLNAPATVCGPKLSTDLIKNQCVNFNFNGLTGTGVLTP\\\n         SSKRFQPFQQFGRDVSDFTDSVRDPKTSEILDISPCSFGGVSVITPGTNASSEVAVLYQD\\\n         VNCTDVSTAIHADQLTPAWRIYSTGNNVFQTQAGCLIGAEHVDTSYECDIPIGAGICASY\\\n         HTVSLLRSTSQKSIVAYTMSLGADSSIAYSNNTIAIPTNFSISITTEVMPVSMAKTSVDC\\\n         NMYICGDSTECANLLLQYGSFCTQLNRALSGIAAEQDRNTREVFAQVKQMYKTPTLKYFG\\\n         GFNFSQILPDPLKPTKRSFIEDLLFNKVTLADAGFMKQYGECLGDINARDLICAQKFNGL\\\n         TVLPPLLTDDMIAAYTAALVSGTATAGWTFGAGAALQIPFAMQMAYRFNGIGVTQNVLYE\\\n         NQKQIANQFNKAISQIQESLTTTSTALGKLQDVVNQNAQALNTLVKQLSSNFGAISSVLN\\\n         DILSRLDKVEAEVQIDRLITGRLQSLQTYVTQQLIRAAEIRASANLAATKMSECVLGQSK\\\n         RVDFCGKGYHLMSFPQAAPHGVVFLHVTYVPSQERNFTTAPAICHEGKAYFPREGVFVFN\\\n         GTSWFITQRNFFSPQIITTDNTFVSGNCDVVIGIINNTVYDPLQPELDSFKEELDKYFKN\\\n         HTSPDVDLGDISGINASVVNIQKEIDRLNEVAKNLNESLIDLQELGKYEQYIKWPWYVWL\\\n         GFIAGLIAIVMVTILLCCMTSCCSCLKGACSCGSCCKFDEDDSEPVLKGVKLHYT\"","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"# delete empty spaces\nspike = spike.replace(\" \", \"\")","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"len(spike)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"As `protlearn` was designed for datasets containing multiple sequences, it requires the input to be a Pandas DataFrame with column name 'Sequence'."},{"metadata":{"trusted":true},"cell_type":"code","source":"# build dataframe\nspike = pd.DataFrame(data=[spike], columns=['Sequence'])","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"## Feature Engineering"},{"metadata":{},"cell_type":"markdown","source":"Information on the AAIndex used below can be accessed [here](https://www.genome.jp/aaindex/)."},{"metadata":{},"cell_type":"markdown","source":"#### Compute Features"},{"metadata":{"trusted":true},"cell_type":"code","source":"# compute the frequency of each amino acid in the sequence\ncomp = pl.composition(spike)\ncomp","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"# compute the ngram compositions of the sequence (in this case, only the dipeptide composition)\nng = pl.ngram_composition(spike)\nng","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"# compute aaindex1\naa1 = pl.aaindex1(spike)\naa1","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"# compute aaindex1\naa2 = pl.aaindex2(spike)\naa2","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"# compute aaindex3\naa3 = pl.aaindex3(spike)\naa3","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"#### Concatenate features"},{"metadata":{"trusted":true},"cell_type":"code","source":"features = pd.concat([comp, ng, aa1, aa2, aa3], axis=1)\nfeatures","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"# the protlearn package automatically removes columns with NaNs \nfeatures.isna().sum().sum()","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"## Visualizations"},{"metadata":{},"cell_type":"markdown","source":"#### Amino acid composition"},{"metadata":{"trusted":true},"cell_type":"code","source":"pl.viz_composition(spike)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"#### N-gram composition"},{"metadata":{"trusted":true},"cell_type":"code","source":"plt.figure(figsize=(20,12))\nplt.subplot(2,1,1)\npl.viz_ngram(spike, method='relative', ngram=2, top=20, xtick_rotation=60)\nplt.subplot(2,1,2)\npl.viz_ngram(spike, method='relative', ngram=3, top=5, xtick_rotation=60)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"The above plot shows the top 20% of the dipeptide composition and the top 5% of the tripeptide composition of the sequence."},{"metadata":{},"cell_type":"markdown","source":"## Conclusion"},{"metadata":{},"cell_type":"markdown","source":"This notebook provides some basic feature engineering techniques for protein sequences using the `protlearn` package. If you have any **criticism / feedback / or suggestions** for additional functions, I would highly welcome that :)"}],"metadata":{"kernelspec":{"language":"python","display_name":"Python 3","name":"python3"},"language_info":{"pygments_lexer":"ipython3","nbconvert_exporter":"python","version":"3.6.4","file_extension":".py","codemirror_mode":{"name":"ipython","version":3},"name":"python","mimetype":"text/x-python"}},"nbformat":4,"nbformat_minor":4}