{"cells":[{"cell_type":"markdown","metadata":{"_cell_guid":"1317c6bf-3520-6baf-439c-ac1bb5eddaa2"},"source":"# Clearly, uploading a jupyter notebook does't work well. Just check this:    \nhttps://github.com/ShengyaoChen/bioinformatics/blob/master/proteome_analysis/breast_cancer_proteosome_machine_learning.ipynb\n\n**Sorry about that.**\n\n# Prediction of Breast Cancaer Sub-types according to Proteome data\n\n### Rationale\nThere are 4 different breast cancer sub-types: 'basal-lie', 'HER2', 'LumA', 'LumB'. **The goal here is to predict breast cancer sub-types according to proteome data and advance the prediction precision** \n\nOne key goal is to **extracting the correct feature sets (protein sets in this case)** for the prediction. It is a combination of application of domain knowledge and statistical analysis. In the breast_cancer_proteosome_analysis, I directly chose the PAM50 genes as the feature sets. Here, I will explore the 'statistical way' to extract features using scikit learn.\n\nnotes: **sklearn.feature_selection**. Notice here I didn't try feature union here, which is pretty handy most of the time. (I tried, the performance is not getting better.)\n\nAnother key goal is to finetune a ML pipeline to get better performance. I've already explored this datasets with 2 machine learning models: K-neighbors and svm.SVC. However, I didn't explore all the way to finetune the models to get better performance. \n\nThe third key point is utilizing **pipeline**. Pipelines work by allowing for a linear sequence of data transforms to be chained together culminating in a modeling process that can be evaluated. The goal is to ensure that all of the steps in the pipeline are constrained to the data available for the evaluation. **This is very important.**\n\n### Outline\n1. Prep data\n2. Feature Extraction\n3. Pipeline & Gridsearch\n\n### Result\nThe performance that we get is much better than the previous model. The prediction accuracy is 85%.   \nThe F1 score summary is:    \n{'Basal-like': 0.90909090909090906, 'HER2': 1.0, 'LumB': 0.83333333333333326, 'LumA': 0.72727272727272718}"}],"metadata":{"_change_revision":0,"_is_fork":false,"kernelspec":{"display_name":"Python 3","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.5.2"}},"nbformat":4,"nbformat_minor":0}