{"cells":[{"metadata":{},"cell_type":"markdown","source":"![](https://image.slidesharecdn.com/karenfent-191031210734/95/enabling-biobankscale-genomic-processing-with-spark-sql-53-638.jpg?cb=1572556092)slideshare.net"},{"metadata":{},"cell_type":"markdown","source":"#Variant Call Format (VCF)"},{"metadata":{"_uuid":"8f2839f25d086af736a60e9eeb907d3b93b6e0e5","_cell_guid":"b1076dfc-b9ad-4769-8c92-a6c4dae69d19","trusted":true,"collapsed":true,"_kg_hide-output":true,"_kg_hide-input":true},"cell_type":"code","source":"# This Python 3 environment comes with many helpful analytics libraries installed\n# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python\n# For example, here's several helpful packages to load\n\nimport numpy as np # linear algebra\nimport pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)\nimport matplotlib.pyplot as plt\nimport seaborn as sns\n\n# other imports\nimport sys\n\n# Input data files are available in the read-only \"../input/\" directory\n# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory\n\nimport os\nfor dirname, _, filenames in os.walk('/kaggle/input'):\n    for filename in filenames:\n        print(os.path.join(dirname, filename))\n\n# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using \"Save & Run All\" \n# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"#Code by Alistair Miles  http://alimanfoo.github.io/2018/04/09/selecting-variants.html"},{"metadata":{"trusted":true,"_kg_hide-output":true,"collapsed":true},"cell_type":"code","source":"!pip install scikit-allel","execution_count":null,"outputs":[]},{"metadata":{"_kg_hide-output":true,"trusted":true},"cell_type":"code","source":"import allel\nallel.__version__","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"#Start with the scikit-allel function read_vcf()"},{"metadata":{"_kg_hide-output":true,"trusted":true},"cell_type":"code","source":"#callset = allel.read_vcf('../input/end-als/end-als/genomics-data/AnswerALS_subset_annovar.hg38_anno_and_geno.no_intergenic.vcf')","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"#That snippet above took so long (more than 15 minutes to read a VCF) and is consuming My RAM.\nThen I commented it."},{"metadata":{},"cell_type":"markdown","source":"#The callset object returned by read_vcf() is a Python dictionary (dict). It contains several NumPy arrays, each of which can be accessed via a key. Here are the available keys:"},{"metadata":{"_kg_hide-output":true,"trusted":true},"cell_type":"code","source":"#sorted(callset.keys())","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"#The ‘samples’ array contains sample identifiers extracted from the header line in the VCF file."},{"metadata":{"trusted":true},"cell_type":"code","source":"#callset['samples']","execution_count":null,"outputs":[]},{"metadata":{"_kg_hide-output":true,"trusted":true},"cell_type":"code","source":"import numpy as np\nnp.__version__","execution_count":null,"outputs":[]},{"metadata":{"_kg_hide-output":true,"trusted":true,"collapsed":true},"cell_type":"code","source":"!pip install zarr","execution_count":null,"outputs":[]},{"metadata":{"_kg_hide-output":true,"trusted":true},"cell_type":"code","source":"import zarr\nzarr.__version__","execution_count":null,"outputs":[]},{"metadata":{"_kg_hide-output":true,"trusted":true},"cell_type":"code","source":"!pip install numcodecs","execution_count":null,"outputs":[]},{"metadata":{"_kg_hide-output":true,"trusted":true},"cell_type":"code","source":"import numcodecs\nnumcodecs.__version__","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"#Extract data from a VCF"},{"metadata":{"trusted":true},"cell_type":"code","source":"vcf_path = '../input/end-als/end-als/genomics-data/AnswerALS_subset_annovar.hg38_anno_and_geno.no_intergenic.vcf'\n!ls -lh {vcf_path}","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"#Extract data and convert to Zarr format, use the `vcf_to_zarr`() function from scikit-allel.\nThat conversion will make life easier downstream."},{"metadata":{"trusted":true},"cell_type":"code","source":"zarr_path = '../input/end-als/end-als/genomics-data/AnswerALS_subset_annovar.hg38_anno_and_geno.no_intergenic.vcf'","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"#The original snippet used data from chromosome 22. End ALS didn't mention any chromossome in the file\n#Trying to make it work, I removed the group and the fields. Unfortunately, it didn't work.\n\n#allel.vcf_to_zarr(vcf_path, zarr_path, group='22', fields='*', log=sys.stdout, overwrite=True)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"allel.vcf_to_zarr(vcf_path, zarr_path, group='*', fields='*', log=sys.stdout, overwrite=True)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"#After FSPathExistNotDir: path exists but is not a directory: %r  Everything went wrong.\n\n#Maybe anyone can fix it."},{"metadata":{},"cell_type":"markdown","source":"#Open the Zarr data "},{"metadata":{"_kg_hide-output":true,"trusted":true,"collapsed":true},"cell_type":"code","source":"callset = zarr.open_group(zarr_path, mode='r')\ncallset.tree(expand=True)","execution_count":null,"outputs":[]},{"metadata":{"_kg_hide-output":true,"trusted":true,"collapsed":true},"cell_type":"code","source":"gt_zarr = callset['22/calldata/GT']\ngt_zarr.info","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"#Loading data for a gene"},{"metadata":{"_kg_hide-output":true,"trusted":true,"collapsed":true},"cell_type":"code","source":"pos = allel.SortedIndex(callset['22/variants/POS'])\npos","execution_count":null,"outputs":[]},{"metadata":{"_kg_hide-output":true,"trusted":true,"collapsed":true},"cell_type":"code","source":"loc_region = pos.locate_range(20000000, 20100000)\nloc_region","execution_count":null,"outputs":[]},{"metadata":{"_kg_hide-output":true,"trusted":true},"cell_type":"code","source":"gt_region = allel.GenotypeArray(gt_zarr[loc_region])\ngt_region","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"#Filtering variants"},{"metadata":{"_kg_hide-output":true,"trusted":true},"cell_type":"code","source":"multi_allelic = callset['22/variants/MULTI_ALLELIC'][:]\nmulti_allelic","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"#Extract genotype data for these variants"},{"metadata":{"_kg_hide-output":true,"trusted":true},"cell_type":"code","source":"gt = allel.GenotypeArray(gt_zarr)\ngt","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"#Extract genotypes for the selection"},{"metadata":{"_kg_hide-output":true,"trusted":true},"cell_type":"code","source":"gt_variant_selection = gt.compress(loc_variant_selection, axis=0)\ngt_variant_selection","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"#Use Dask if data is larger and/or your computer doesn’t have much RAM"},{"metadata":{"_kg_hide-output":true,"trusted":true},"cell_type":"code","source":"gt_dask = allel.GenotypeDaskArray(gt_zarr)\ngt_dask","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"#Apply the selection, using almost the same syntax, except that when working via Dask we need to call the compute() method to get the final result:"},{"metadata":{"_kg_hide-output":true,"trusted":true},"cell_type":"code","source":"gt_variant_selection = gt_dask.compress(loc_variant_selection, axis=0).compute()\ngt_variant_selection","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"#I hope to learn a little bit how to work with vcf files, scikit-allel and BioInformatics."}],"metadata":{"kernelspec":{"language":"python","display_name":"Python 3","name":"python3"},"language_info":{"pygments_lexer":"ipython3","nbconvert_exporter":"python","version":"3.6.4","file_extension":".py","codemirror_mode":{"name":"ipython","version":3},"name":"python","mimetype":"text/x-python"}},"nbformat":4,"nbformat_minor":4}