{"cells":[{"metadata":{},"cell_type":"markdown","source":"#Inborn errors of type I IFN immunity in patients with life-threatening COVID-19\n\nQian Zhang, Paul Bastard, Zhiyong Liu, Jérémie Le Pen, Marcela Moncada-Velez, Jie Chen.\n\nCitation authors and affiliations (at the end of this Notebook)\n\nScience  23 Oct 2020: Vol. 370, Issue 6515, eabd4570 -DOI: 10.1126/science.abd4570\n\n\nTHE GENETICS UNDERLYING SEVERE COVID-19\n\nThe immune system is complex and involves many genes, including those that encode cytokines known as interferons (IFNs). Individuals that lack specific IFNs can be more susceptible to infectious diseases.\n\nFurthermore, the autoantibody system dampens IFN response to prevent damage from pathogen-induced inflammation. Two studies now examine the likelihood that genetics affects the risk of severe coronavirus disease 2019 (COVID-19) through components of this system.\n\nQ. Zhang et al. used a candidate gene approach and identified patients with severe COVID-19 who have mutations in genes involved in the regulation of type I and III IFN immunity. They found enrichment of these genes in patients and conclude that genetics may determine the clinical course of the infection.\n\n\n<p style=\"color:purple; background-color: #ffccff;\">Autoantibodies were not found either in infected people who were asymptomatic or had milder phenotype or in healthy individuals</p>\n\nBastard et al. identified individuals with high titers of neutralizing autoantibodies against type I IFN-α2 and IFN-ω in about 10% of patients with severe COVID-19 pneumonia. These autoantibodies were not found either in infected people who were asymptomatic or had milder phenotype or in healthy individuals. Together, these studies identify a means by which individuals at highest risk of life-threatening COVID-19 can be identified.\nhttps://science.sciencemag.org/content/370/6515/eabd4570"},{"metadata":{},"cell_type":"markdown","source":"![](https://www.hhmi.org/sites/default/files/news/2020/infection_schematic_715x375.jpg)\nhttps://www.hhmi.org/news/some-severe-covid-19-cases-linked-to-genetic-mutations-or-antibodies-that-attack-the-body"},{"metadata":{"_uuid":"8f2839f25d086af736a60e9eeb907d3b93b6e0e5","_cell_guid":"b1076dfc-b9ad-4769-8c92-a6c4dae69d19","trusted":true,"_kg_hide-output":true,"collapsed":true},"cell_type":"code","source":"# This Python 3 environment comes with many helpful analytics libraries installed\n# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python\n# For example, here's several helpful packages to load\n\nimport numpy as np # linear algebra\nimport pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)\nimport matplotlib.pyplot as plt\nimport seaborn as sns\nimport plotly.offline as py\nimport plotly.graph_objs as go\nimport plotly.offline as py\nimport plotly.express as px\n\n# Input data files are available in the read-only \"../input/\" directory\n# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory\n\nimport os\nfor dirname, _, filenames in os.walk('/kaggle/input'):\n    for filename in filenames:\n        print(os.path.join(dirname, filename))\n\n# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using \"Save & Run All\" \n# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"<p style=\"color:purple; background-color: #ffccff;\">Three risk factors for severe disease: being male, being elderly, and having other medical conditions.</p>\n\nClinical outcomes of human severe acute respiratory syndrome coronavirus infection range from silent infection to lethal disease. Epidemiological studies have identified three risk factors for severe disease: being male, being elderly, and having other medical conditions. \n\nHowever, interindividual clinical variability remains huge in each demographic category. Discovering the root cause and detailed molecular, cellular, and tissue- and body-level mechanisms underlying life-threatening COVID-19 is of the utmost biological and medical importance. https://science.sciencemag.org/content/370/6515/eabd4570"},{"metadata":{},"cell_type":"markdown","source":"<p style=\"color:purple; background-color: #ffccff;\">Demographic and genetic data for the COVID-19 cohort.</p>\n\n(A) Age and sex distribution of patients with life-threatening COVID-19. (B) PCA of patient (with or without LOF variants in the 13 candidate genes) and control cohorts (patients with mild or asymptomatic disease and individuals from the 1000 Genomes Project).\n\n![](https://science.sciencemag.org/content/sci/370/6515/eabd4570/F2.large.jpg)\nhttps://science.sciencemag.org/content/370/6515/eabd4570/tab-figures-data"},{"metadata":{"_uuid":"d629ff2d2480ee46fbb7e2d37f6b5fab8052498a","_cell_guid":"79c7e3d0-c299-4dcb-8224-4455121ee9b0","trusted":true,"_kg_hide-input":true},"cell_type":"code","source":"nRowsRead = 1000 # specify 'None' if want to read whole file\ndf = pd.read_csv('../input/cusersmarildownloadsinterferoncsv/interferon.csv', delimiter=';', encoding = \"ISO-8859-1\", nrows = nRowsRead)\ndf.dataframeName = 'cusersmarildownloadsinterferoncsv/interferon.csv'\nnRow, nCol = df.shape\nprint(f'There are {nRow} rows and {nCol} columns')\ndf.head()","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"<p style=\"color:purple; background-color: #ffccff;\">Testing the hypothesis that inborn errors of immunity may cause life-threatening Covid-19</p>\n\nThe authors established the COVID Human Genetic Effort (www.covidhge.com) to test the general hypothesis that life-threatening COVID-19 in some or most patients may be caused by monogenic inborn errors of immunity to SARS-CoV-2 with incomplete or complete penetrance.\n\nThey sequenced the exome or genome of 659 patients of various ancestries with life-threatening COVID-19 pneumonia and 534 subjects with asymptomatic or benign infection. \n\nThey tested the specific hypothesis that inborn errors of Toll-like receptor 3 (TLR3)– and interferon regulatory factor 7 (IRF7)–dependent type I interferon (IFN) immunity that underlie life-threatening influenza pneumonia also underlie life-threatening COVID-19 pneumonia. \n\nThey considered three loci identified as mutated in patients with life-threatening influenza: TLR3, IRF7, and IRF9. \nhttps://science.sciencemag.org/content/370/6515/eabd4570"},{"metadata":{"_kg_hide-output":true,"trusted":true,"collapsed":true},"cell_type":"code","source":"df.isnull().sum()","execution_count":null,"outputs":[]},{"metadata":{"_kg_hide-output":true,"trusted":true},"cell_type":"code","source":"# categorical features with missing values\ncategorical_nan = [feature for feature in df.columns if df[feature].isna().sum()>0 and df[feature].dtypes=='O']\nprint(categorical_nan)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"<p style=\"color:purple; background-color: #ffccff;\">Type I IFN administration May be of therapeutic benefit in selected patients</p>\n\nAt least 3.5% of patients with life-threatening COVID-19 pneumonia had known (AR IRF7 and IFNAR1 deficiencies or AD TLR3, TICAM1, TBK1, and IRF3 deficiencies) or new (AD UNC93B1, IRF7, IFNAR1, and IFNAR2 deficiencies) genetic defects at eight of the 13 candidate loci involved in the TLR3- and IRF7-dependent induction and amplification of type I IFNs.\n\nThis discovery reveals essential roles for both the double-stranded RNA sensor TLR3 and type I IFN cell-intrinsic immunity in the control of SARS-CoV-2 infection.\n\nType I IFN administration may be of therapeutic benefit in selected patients, at least early in the course of SARS-CoV-2 infection. https://science.sciencemag.org/content/370/6515/eabd4570"},{"metadata":{"_kg_hide-output":true,"trusted":true},"cell_type":"code","source":"# replacing missing values in categorical features\nfor feature in categorical_nan:\n    df[feature] = df[feature].fillna('None')","execution_count":null,"outputs":[]},{"metadata":{"_kg_hide-output":true,"trusted":true,"collapsed":true},"cell_type":"code","source":"df[categorical_nan].isna().sum()","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"<p style=\"color:purple; background-color: #ffccff;\">Demographic and ethnic information about the patients and controls.</p>\n\n(A) Gender distribution in patients with life-threatening COVID-19 and auto-Abs to type I IFNs, patients with life-threatening COVID-19 and without auto-Abs to type I IFNs, and individuals with asymptomatic or mild SARS-CoV-2. (B) Age distribution in patients with life-threatening COVID-19 and auto-Abs to type I IFNs, patients with life-threatening COVID-19 and without auto-Abs to type I IFNs, and individuals with asymptomatic or mild SARS-CoV-2. yo, years old. (C) PCA on 49 patients with life-threatening COVID-19 and auto-Abs against type I IFNs. EUR, Europeans; AFR, Africans; EAS, East-Asians.\n\n![](https://science.sciencemag.org/content/sci/370/6515/eabd4585/F5.large.jpg)\nhttps://science.sciencemag.org/content/370/6515/eabd4585","attachments":{}},{"metadata":{"_kg_hide-output":true,"trusted":true},"cell_type":"code","source":"# Lets first handle numerical features with nan value\nnumerical_nan = [feature for feature in df.columns if df[feature].isna().sum()>1 and df[feature].dtypes!='O']\nnumerical_nan","execution_count":null,"outputs":[]},{"metadata":{"_kg_hide-output":true,"trusted":true},"cell_type":"code","source":"df[numerical_nan].isna().sum()","execution_count":null,"outputs":[]},{"metadata":{"_kg_hide-output":true,"trusted":true},"cell_type":"code","source":"#Replacing the numerical Missing Values\n\nfor feature in numerical_nan:\n    ## We will replace by using median since there are outliers\n    median_value=df[feature].median()\n    \n    df[feature].fillna(median_value,inplace=True)\n    \ndf[numerical_nan].isnull().sum()","execution_count":null,"outputs":[]},{"metadata":{"_kg_hide-output":true,"trusted":true},"cell_type":"code","source":"from sklearn.preprocessing import LabelEncoder\n\n#fill in mean for floats\nfor c in df.columns:\n    if df[c].dtype=='float16' or  df[c].dtype=='float32' or  df[c].dtype=='float64':\n        df[c].fillna(df[c].mean())\n\n#fill in -999 for categoricals\ndf = df.fillna(-999)\n# Label Encoding\nfor f in df.columns:\n    if df[f].dtype=='object': \n        lbl = LabelEncoder()\n        lbl.fit(list(df[f].values))\n        df[f] = lbl.transform(list(df[f].values))\n        \nprint('Labelling done.')","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"df = pd.get_dummies(df)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"#Codes by Puru Behl https://www.kaggle.com/accountstatus/traffic-collision-data-analysis/notebook"},{"metadata":{"trusted":true},"cell_type":"code","source":"sns.distplot(df['Age'])","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"m=df['Sex'].mean()","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"#Codes by Puru Behl https://www.kaggle.com/accountstatus/traffic-collision-data-analysis/notebook\n\nfig=px.bar(df,x='IFNG+874',y='Sex')\nfig.add_shape(type='line',x0=0,x1=32,y1=m,y0=m,line=dict(        color=\"Red\",\n        width=4,\n        dash=\"dashdot\",))","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"df_grp = df.groupby([\"IFNG+874\",\"Sex\"])[[\"unstimulated IFN-y pg/ml\",\"PHA IFN-y pg/ml\",\"Ag IFN-y pg/ml\"]].sum().reset_index()\ndf_grp.head()","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"x=np.asarray(list(df_grp.index)).astype('float').reshape(-1,1)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"y=df_grp['Sex'].values","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"from sklearn.linear_model import LinearRegression\n\nlr=LinearRegression()\nlr.fit(x,df_grp['Sex'].values)\npred=lr.predict(x)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"fig=go.Figure()\nfig.add_trace(go.Bar(name='IFNG+874 vs Sex', x=df_grp.index, y=df_grp['Sex'].values))\nfig.add_trace(go.Scatter(name='line of best fit', x=df_grp.index, y=pred, mode='lines'))","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"px.line(data_frame=df_grp,x=df_grp.index,y='Sex',template='plotly_dark')","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"px.bar(data_frame=df_grp,x=df_grp.index,y='Sex',template='plotly_white', color_discrete_sequence=px.colors.diverging.balance)","execution_count":null,"outputs":[]},{"metadata":{"_kg_hide-input":true,"trusted":true},"cell_type":"code","source":"px.bar(data_frame=df_grp,x=df_grp.index,y='IFNG+874',template='plotly_dark', color_discrete_sequence=px.colors.diverging.Armyrose, title='Interferon Gamma and +874 gene')","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"<p style=\"color:purple; background-color: #ffccff;\">Unstimulated Interferon Gamma picograms per millilitre</p>"},{"metadata":{"_kg_hide-input":true,"trusted":true},"cell_type":"code","source":"px.bar(data_frame=df_grp,x=df_grp.index,y='unstimulated IFN-y pg/ml',template='plotly_dark', color_discrete_sequence=px.colors.diverging.Tealrose, title='Unstimulated IFN-y pg/ml')","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"<p style=\"color:purple; background-color: #ffccff;\">Phytohaemagutinin (PHA) Interferon Gamma picograms per millilitre</p>"},{"metadata":{"_kg_hide-input":true,"trusted":true},"cell_type":"code","source":"px.bar(data_frame=df_grp,x=df_grp.index,y='PHA IFN-y pg/ml',template='plotly_dark', color_discrete_sequence=px.colors.diverging.Portland, title='Phytohaemagutinin (PHA) Interferon Gamma picograms per millilitre')","execution_count":null,"outputs":[]},{"metadata":{"_kg_hide-input":true,"trusted":true},"cell_type":"code","source":"px.bar(data_frame=df_grp,x=df_grp.index,y='Ag IFN-y pg/ml',template='plotly_dark', color_discrete_sequence=px.colors.diverging.Spectral, title= 'Antigen Interferon Gamma pg/ml')","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"#Inborn errors of type I IFN immunity in patients with life-threatening COVID-19\n\nQian Zhang1, View ORCID ProfilePaul Bastard2,3,*, Zhiyong Liu1,*, View ORCID ProfileJérémie Le Pen4,*, View ORCID ProfileMarcela Moncada-Velez1,*, Jie Chen1,*, View ORCID ProfileMasato Ogishi1,*, View ORCID ProfileIra K. D. Sabli5,*, View ORCID ProfileStephanie Hodeib5,*, View ORCID ProfileCecilia Korol2,*, Jérémie Rosain2,3,*, View ORCID ProfileKaya Bilguvar6,*, Junqiang Ye7,*, View ORCID ProfileAlexandre Bolze8,*, View ORCID ProfileBenedetta Bigio1,*, View ORCID ProfileRui Yang1,*, View ORCID ProfileAndrés Augusto Arias1,9,10,*, View ORCID ProfileQinhua Zhou1,*, Yu Zhang11,12,*, Fanny Onodi13, Sarantis Korniotis13, Léa Karpf13, Quentin Philippot2,3, View ORCID ProfileMarwa Chbihi2,3, View ORCID ProfileLucie Bonnet-Madin14, View ORCID ProfileKarim Dorgham15, View ORCID ProfileNikaïa Smith16, View ORCID ProfileWilliam M. Schneider4, View ORCID ProfileBrandon S. Razooky4, View ORCID ProfileHans-Heinrich Hoffmann4, View ORCID ProfileEleftherios Michailidis4, Leen Moens17, View ORCID ProfileJi Eun Han1, View ORCID ProfileLazaro Lorenzo2,3, View ORCID ProfileLucy Bizien2,3, View ORCID ProfilePhilip Meade18, View ORCID ProfileAnna-Lena Neehus2,3, View ORCID ProfileAileen Camille Ugurbil1, Aurélien Corneau19, View ORCID ProfileGaspard Kerner2,3, View ORCID ProfilePeng Zhang1, View ORCID ProfileFranck Rapaport1, Yoann Seeleuthner2,3, View ORCID ProfileJeremy Manry2,3, View ORCID ProfileCecile Masson20, Yohann Schmitt20, View ORCID ProfileAgatha Schlüter21, View ORCID ProfileTom Le Voyer2,3, View ORCID ProfileTaushif Khan22, Juan Li1, View ORCID ProfileJacques Fellay23,24,25, View ORCID ProfileLucie Roussel26, Mohammad Shahrooei27,28, View ORCID ProfileMohammed F. Alosaimi29, Davood Mansouri30,31,32, Haya Al-Saud33, View ORCID ProfileFahd Al-Mulla34, View ORCID ProfileFeras Almourfi33, Saleh Zaid Al-Muhsen35, View ORCID ProfileFahad Alsohime29, View ORCID ProfileSaeed Al Turki36,37, View ORCID ProfileRana Hasanato29, View ORCID ProfileDiederik van de Beek38, Andrea Biondi39, View ORCID ProfileLaura Rachele Bettini39, Mariella D’Angio’39, View ORCID ProfilePaolo Bonfanti40, Luisa Imberti41, Alessandra Sottini41, Simone Paghera41, Eugenia Quiros-Roldan42, Camillo Rossi43, View ORCID ProfileAndrew J. Oler44, View ORCID ProfileMiranda F. Tompkins45, View ORCID ProfileCamille Alba45, Isabelle Vandernoot46, Jean-Christophe Goffard47, View ORCID ProfileGuillaume Smits46, Isabelle Migeotte48, Filomeen Haerynck49, View ORCID ProfilePere Soler-Palacin50, View ORCID ProfileAndrea Martin-Nalda50, View ORCID ProfileRoger Colobran51, Pierre-Emmanuel Morange52, Sevgi Keles53, View ORCID ProfileFatma Çölkesen54, Tayfun Ozcelik55, View ORCID ProfileKadriye Kart Yasar56, View ORCID ProfileSevtap Senoglu56, View ORCID ProfileŞemsi Nur Karabela56, View ORCID ProfileCarlos Rodríguez-Gallego57,58, View ORCID ProfileGiuseppe Novelli59, Sami Hraiech60, View ORCID ProfileYacine Tandjaoui-Lambiotte61,62, Xavier Duval63,64, View ORCID ProfileCédric Laouénan63,64,65, COVID-STORM Clinicians†, COVID Clinicians†, Imagine COVID Group†, French COVID Cohort Study Group†, CoV-Contact Cohort†, Amsterdam UMC Covid-19 Biobank†, COVID Human Genetic Effort†, NIAID-USUHS/TAGC COVID Immunity Group†, Andrew L. Snow66, Clifton L. Dalgard45,67, Joshua D. Milner68, Donald C. Vinh26, Trine H. Mogensen69,70, Nico Marr22,71, András N. Spaan1,72, Bertrand Boisson1,2,3, Stéphanie Boisson-Dupuis1,2,3, Jacinta Bustamante1,2,3,73, Anne Puel1,2,3, Michael J. Ciancanelli1,74, Isabelle Meyts17,75, Tom Maniatis7,76, Vassili Soumelis13,77, Ali Amara14, Michel Nussenzweig78,79, Adolfo García-Sastre18,80,81,82, Florian Krammer18, Aurora Pujol21, Darragh Duffy16, Richard P. Lifton83,84,85,‡, Shen-Ying Zhang1,2,3,‡, Guy Gorochov15,‡, Vivien Béziat1,2,3,‡, Emmanuelle Jouanguy1,2,3,‡, Vanessa Sancho-Shimizu5,‡, Charles M. Rice4,‡, Laurent Abel1,2,3,‡, Luigi D. Notarangelo11,12,§, Aurélie Cobat1,2,3,§, Helen C. Su11,12,§, Jean-Laurent Casanova1,2,3,79,86,§,\n \n\nScience  23 Oct 2020:\n\nVol. 370, Issue 6515, eabd4570\n\nDOI: 10.1126/science.abd4570\nhttps://science.sciencemag.org/content/370/6515/eabd4570"},{"metadata":{"_kg_hide-input":true,"trusted":true},"cell_type":"code","source":"#Code by Olga Belitskaya https://www.kaggle.com/olgabelitskaya/sequential-data/comments\nfrom IPython.display import display,HTML\nc1,c2,f1,f2,fs1,fs2=\\\n'#2B3A67','#42a7f5','Akronim','Smokum',30,15\ndef dhtml(string,fontcolor=c1,font=f1,fontsize=fs1):\n    display(HTML(\"\"\"<style>\n    @import 'https://fonts.googleapis.com/css?family=\"\"\"\\\n    +font+\"\"\"&effect=3d-float';</style>\n    <h1 class='font-effect-3d-float' style='font-family:\"\"\"+\\\n    font+\"\"\"; color:\"\"\"+fontcolor+\"\"\"; font-size:\"\"\"+\\\n    str(fontsize)+\"\"\"px;'>%s</h1>\"\"\"%string))\n    \n    \ndhtml('Marília Prata, @mpwolke was Here.' )","execution_count":null,"outputs":[]}],"metadata":{"kernelspec":{"language":"python","display_name":"Python 3","name":"python3"},"language_info":{"pygments_lexer":"ipython3","nbconvert_exporter":"python","version":"3.6.4","file_extension":".py","codemirror_mode":{"name":"ipython","version":3},"name":"python","mimetype":"text/x-python"}},"nbformat":4,"nbformat_minor":4}