{"cells":[{"metadata":{},"cell_type":"markdown","source":"### What is genotype?\n\nIn a broad sense, the term \"genotype\" refers to the genetic makeup of an organism; in other words, it describes an organism's complete set of genes. In a more narrow sense, the term can be used to refer to the alleles, or variant forms of a gene, that are carried by an organism. Humans are diploid organisms, which means that they have two alleles at each genetic position, or locus, with one allele inherited from each parent. Each pair of alleles represents the genotype of a specific gene."},{"metadata":{},"cell_type":"markdown","source":"### Genotype Matrix"},{"metadata":{"trusted":true},"cell_type":"code","source":"from IPython.display import YouTubeVideo\n\nYouTubeVideo('zNoi5SV1KLM', width=800, height=600)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"import numpy as np\nimport pandas as pd\nimport matplotlib.pyplot as plt \nimport seaborn as sns\nimport plotly.express as px\nimport plotly.graph_objects as go","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"data_path = \"/kaggle/input/family-genome-dataset/\"\ndf1 = pd.read_csv(data_path + \"Child 2 Genome.csv\")\ndf2 = pd.read_csv(data_path + \"Father Genome.csv\")\ndf3 = pd.read_csv(data_path + \"Child 3 Genome.csv\")\ndf4 = pd.read_csv(data_path + \"Child 1 Genome.csv\")\ndf5 = pd.read_csv(data_path + \"Mother Genome.csv\")","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"df1.head()","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"df1.info()","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"df1['genotype'].unique()","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"df1['genotype'].value_counts()","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"genotype_name = df1['genotype'].unique()\ngenotype_total = df1['genotype'].value_counts()\n\nfig = go.Figure([go.Bar(x=genotype_name, y=genotype_total)])\nfig.show()","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"chromosome_name = df1['chromosome'].unique()\nchromosome_total = df1['chromosome'].value_counts()\n\nfig = go.Figure([go.Bar(x=chromosome_name, y=chromosome_total)])\nfig.show()","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"fig = px.imshow(df1)\nfig.show()","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"df2.head()","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"","execution_count":null,"outputs":[]}],"metadata":{"kernelspec":{"language":"python","display_name":"Python 3","name":"python3"},"language_info":{"pygments_lexer":"ipython3","nbconvert_exporter":"python","version":"3.6.4","file_extension":".py","codemirror_mode":{"name":"ipython","version":3},"name":"python","mimetype":"text/x-python"}},"nbformat":4,"nbformat_minor":4}