{"cells":[{"metadata":{"trusted":true},"cell_type":"code","source":"import pandas as pd\nimport numpy as np\nimport matplotlib.pyplot as plt\nimport seaborn as sns\nfrom sklearn.preprocessing import StandardScaler\nfrom sklearn.preprocessing import MinMaxScaler \nfrom sklearn.preprocessing import RobustScaler\nfrom sklearn.model_selection import train_test_split\nfrom sklearn.model_selection import GridSearchCV\nfrom sklearn.model_selection import RandomizedSearchCV\nfrom sklearn.tree import DecisionTreeClassifier\nfrom sklearn.ensemble import RandomForestClassifier\nfrom xgboost import XGBRegressor\nfrom sklearn.feature_selection import RFECV\nfrom sklearn.pipeline import Pipeline\nfrom sklearn.metrics import classification_report\nfrom sklearn.metrics import confusion_matrix\nfrom sklearn import metrics\nimport os\nfor dirname, _, filenames in os.walk('/kaggle/input'):\n    for filename in filenames:\n        print(os.path.join(dirname, filename))","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"gen_var = pd.read_csv('/kaggle/input/clinvar-conflicting/clinvar_conflicting.csv')","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"gen_var.head()","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"tot_cols = gen_var.columns","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"# INFO of the data set\ngen_var.info()","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"# Find the nulls\ngen_var.isnull().sum().sort_values(ascending = False)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"# Drop columns with nulls more than 65000\ncols_with_max_nulls = ['MOTIF_SCORE_CHANGE', 'HIGH_INF_POS', 'MOTIF_POS', 'MOTIF_NAME', 'DISTANCE', 'SSR', 'CLNSIGINCL', 'CLNDNINCL', 'CLNDISDBINCL']\ngen_var.drop(cols_with_max_nulls, axis = 1, inplace = True)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"cols = gen_var.isnull().sum().sort_values(ascending = False)\ncols.index","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"null_cols = []\nfor col in cols.index:\n    if gen_var[col].isnull().sum() > 0:\n        null_cols.append(col)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"for col in null_cols:\n    print(col+' '+str(len(gen_var[col].unique())))\n    print(gen_var[col].unique())\n    print()","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"null_cols","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"plt.figure(figsize = (12, 10))\nsns.heatmap(gen_var.corr(), annot = True, linewidths=.5, cmap = plt.cm.Accent_r)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"def correlation(dataset, threshold):\n    col_corr = set()  # Set of all the names of correlated columns\n    corr_matrix = dataset.corr()\n    for i in range(len(corr_matrix.columns)):\n        for j in range(i):\n            if abs(corr_matrix.iloc[i, j]) > threshold: # we are interested in absolute coeff value\n                colname = corr_matrix.columns[i]  # getting the name of column\n                col_corr.add(colname)\n    return col_corr","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"corr_features = correlation(gen_var.corr(), 0.8)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"# Drop the correlated features\ngen_var.drop(corr_features, axis = 1, inplace = True)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"gen_var.shape\nnull_cols","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"# INTRON\ngen_var.drop('INTRON', axis = 1, inplace = True)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"# Polyphen \nprint(gen_var['PolyPhen'].mode())\ngen_var['PolyPhen'].fillna(value = 'benign', inplace = True)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"# SIFT\nprint(gen_var['SIFT'].isnull().sum())\nprint(gen_var['SIFT'].unique())\nprint(gen_var['SIFT'].mode())\ngen_var['SIFT'].fillna('deleterious', inplace = True)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"# BLOSUM62\nprint(gen_var['BLOSUM62'].isnull().sum())\nprint(gen_var['BLOSUM62'].unique())\nprint(gen_var['BLOSUM62'].mode())\ngen_var['BLOSUM62'] = gen_var['BLOSUM62'].fillna(1.0)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"# CLNVI\nprint(gen_var['CLNVI'].isnull().sum())\nprint(gen_var['CLNVI'].unique())\ngen_var.drop('CLNVI', axis = 1, inplace = True)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"# BAM_EDIT\nprint(gen_var['BAM_EDIT'].isnull().sum())\nprint(gen_var['BAM_EDIT'].unique())\nprint(gen_var['BAM_EDIT'].mode())\ngen_var['BAM_EDIT'] = gen_var['BAM_EDIT'].fillna('OK')","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"# Codons\nprint(gen_var['Codons'].isnull().sum())\nprint(gen_var['Codons'].unique())\nprint(gen_var['Codons'].mode())\ngen_var['Codons'] = gen_var['Codons'].fillna('cGg/cAg')","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"# Amino_acids\nprint(gen_var['Amino_acids'].isnull().sum())\nprint(gen_var['Amino_acids'].unique())\nmode_Amino_acids = gen_var['Amino_acids'].mode()\nprint(mode_Amino_acids)\ngen_var['Amino_acids'] = gen_var['Amino_acids'].fillna('A')\ngen_var[gen_var['Amino_acids'] == 'A']","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"# Protein_position\nprint(gen_var['Protein_position'].isnull().sum())\nprint(gen_var['Protein_position'].unique())\nmode_Protein_position = gen_var['Protein_position'].mode()\nprint(mode_Protein_position)\ngen_var['Protein_position'].fillna('1', inplace = True)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"# Codons\nprint(gen_var['Codons'].isnull().sum())\nprint(gen_var['Codons'].unique())\nmode_Codons = gen_var['Codons'].mode()\nmode_Codons\ngen_var['Codons'].fillna('cGg/cAg', inplace = True)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"# CDS_position\nprint(gen_var['CDS_position'].isnull().sum())\nprint(gen_var['CDS_position'].unique())\nmode_Codons = gen_var['CDS_position'].mode()\ngen_var['CDS_position'].fillna('1', inplace = True)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"# EXON\nprint(gen_var['EXON'].isnull().sum())\nprint(gen_var['EXON'].unique())\nmode_Codons = gen_var['EXON'].mode()\ngen_var['EXON'].fillna('16/16', inplace = True)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":" 'CADD_RAW',\n 'CADD_PHRED',\n 'MC',\n 'SYMBOL',\n 'BIOTYPE',\n 'Feature_type',\n 'STRAND',\n 'Feature'","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"# cDNA_position\nprint(gen_var['cDNA_position'].isnull().sum())\nprint(gen_var['cDNA_position'].unique())\nmode_Codons = gen_var['cDNA_position'].mode()\nmode_Codons\ngen_var['cDNA_position'].fillna('852', inplace = True)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"# LoFtool\nprint(gen_var['LoFtool'].isnull().sum())\nprint(gen_var['LoFtool'].unique())\nmode_Codons = gen_var['LoFtool'].mode()\ngen_var['LoFtool'].fillna(0.971, inplace = True)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"# CADD_PHRED\nprint(gen_var['CADD_PHRED'].isnull().sum())\nprint(gen_var['CADD_PHRED'].unique())\nmode_Codons = gen_var['CADD_PHRED'].mode()\ngen_var['CADD_PHRED'].fillna(34.0, inplace = True)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"# MC\nprint(gen_var['MC'].isnull().sum())\n# print(gen_var['MC'].unique())\nmode_Codons = gen_var['MC'].mode()\nmode_Codons\ngen_var['MC'].fillna('SO:0001583|missense_variant', inplace = True)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"# SYMBOL\nprint(gen_var['SYMBOL'].isnull().sum())\nprint(gen_var['SYMBOL'].unique())\nmode_Codons = gen_var['SYMBOL'].mode()\nmode_Codons\ngen_var['SYMBOL'].fillna('TTN', inplace = True)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"# BIOTYPE\nprint(gen_var['BIOTYPE'].isnull().sum())\nprint(gen_var['BIOTYPE'].unique())\nmode_Codons = gen_var['BIOTYPE'].mode()\nmode_Codons\ngen_var['BIOTYPE'].fillna('protein_coding', inplace = True)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"# Feature_type\nprint(gen_var['Feature_type'].isnull().sum())\nprint(gen_var['Feature_type'].unique())\nmode_Codons = gen_var['Feature_type'].mode()\nmode_Codons\ngen_var['Feature_type'].fillna('Transcript', inplace = True)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"# Feature\nprint(gen_var['Feature'].isnull().sum())\nprint(gen_var['Feature'].unique())\nmode_Codons = gen_var['Feature'].mode()\nmode_Codons\ngen_var['Feature'].fillna('NM_001267550.1', inplace = True)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"# Feature\nprint(gen_var['STRAND'].isnull().sum())\nprint(gen_var['STRAND'].unique())\nmode_Codons = gen_var['STRAND'].mode()\nmode_Codons\ngen_var['STRAND'].fillna(-1.0, inplace = True)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"gen_var.isnull().sum().sort_values(ascending = False)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"obj_cols = []\nfor col in gen_var.columns:\n    if gen_var[col].dtypes == 'object':\n        obj_cols.append(col)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"for col in obj_cols:\n    if len(gen_var[col].unique()) < 5:\n        print(col+' '+str(len(gen_var[col].unique())))\n        print(gen_var[col].unique())","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"encode_IMPACT = {\n    'LOW' : 0, 'MODERATE' : 1, 'MODIFIER' : 2, 'HIGH' : 3\n}\n\nencode_Feature_type = {\n    'Transcript' : 0, 'MotifFeature' : 1\n}\n\nencode_BIOTYPE = {\n    'protein_coding' : 0, 'misc_RNA' : 1\n}\n\nencode_BAM_EDIT = {\n    'OK' : 1, 'FAILED' : 0\n}\n\nencode_PolyPhen = {\n    'benign' : 0, 'probably_damaging' : 1, 'possibly_damaging' : 2, 'unknown' : 3\n}\n\nencode_CHROM = {\n    '1' : 1, '2' : 2, '3' : 3, '4' : 4, '5' : 5, '6' : 6, '7' : 7, '8' : 8, '9' : 9, '10' : 10, '11' : 11, \n    '12' : 12, '13' : 13, '14' : 14, '15' : 15, '16' : 16, '17' : 17, '18' : 18, '19' : 19, '20' : 20, '21' : 21, '22' : 22,\n    'X' : 23, 'MT' : 24\n}","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"cat_cols = []\nfor col in obj_cols:\n    if len(gen_var[col].unique()) > 5 :\n        if gen_var[col].dtypes == 'object' :\n            cat_cols.append(col)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"from sklearn.preprocessing import LabelEncoder","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"cat_cols.remove('CHROM')\ncat_cols.append('SIFT')\nle = LabelEncoder()\nfor col in cat_cols:\n    print(col)\n    gen_var[col] = le.fit_transform(gen_var[col].values)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"gen_var.info()","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"gen_var['IMPACT'].replace(encode_IMPACT, inplace = True)\ngen_var['Feature_type'].replace(encode_Feature_type, inplace = True)\ngen_var['BIOTYPE'].replace(encode_BIOTYPE, inplace = True)\ngen_var['BAM_EDIT'].replace(encode_BAM_EDIT, inplace = True)\ngen_var['PolyPhen'].replace(encode_PolyPhen, inplace = True)\ngen_var['CHROM'].replace(encode_CHROM, inplace = True)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"gen_var.info()","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"plt.figure(figsize = (15, 10))\nsns.heatmap(gen_var.corr(), linewidths = .5, annot = True, cmap = plt.cm.Accent_r)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"gen_var.drop(['CHROM', 'ALT'], axis = 1, inplace = True)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"gen_var.shape","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"X = gen_var.drop('CLASS', axis = 1)\nY = pd.DataFrame(gen_var['CLASS'])","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.3, random_state = 42)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"# # Create a pipeline\n# pipe = Pipeline([(\"classifier\", RandomForestClassifier())])\n# # Create dictionary with candidate learning algorithms and their hyperparameters\n# grid_param = [\n#                 {\"classifier\": [RandomForestClassifier()],\n#                  \"classifier__n_estimators\": [10, 100, 1000],\n#                  \"classifier__max_depth\":[5,8,15,25,30,None],\n#                  \"classifier__min_samples_leaf\":[1,2,5,10,15,100],\n#                  \"classifier__max_leaf_nodes\": [2, 5,10]}]\n# # create a gridsearch of the pipeline, the fit the best model\n# gridsearch = GridSearchCV(pipe, grid_param, cv=5, verbose=0,n_jobs=-1) # Fit grid search\n# best_model = gridsearch.fit(X_train, Y_train)\n\n# print(best_model.best_estimator_)\n# print(\"The mean accuracy of the model is:\",best_model.score(X_test, Y_test))","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"# Creating an empty Dataframe with Scores\ndf_accur_roc_score_importance = pd.DataFrame(columns=['Roc_Auc_Score'])\ndf_accur_score = pd.DataFrame(columns=['Accuracy_Score'])","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"dt = DecisionTreeClassifier()\nrf = RandomForestClassifier(n_estimators=100, n_jobs=-1, random_state=42)\nX_std_scaled = StandardScaler().fit_transform(X)\nX_norm_scled = MinMaxScaler().fit_transform(X)\nX_robust_scaled = RobustScaler().fit_transform(X)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"X_sc_train, X_sc_test, Y_sc_train, Y_sc_test = train_test_split(X_std_scaled, Y, test_size = 0.3, random_state = 40)\nX_rb_train, X_rb_test, Y_rb_train, Y_rb_test = train_test_split(X_robust_scaled, Y, test_size = 0.3, random_state = 40)\nX_nm_train, X_nm_test, Y_nm_train, Y_nm_test = train_test_split(X_norm_scled, Y, test_size = 0.3, random_state = 40)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"dt.fit(X_sc_train, Y_sc_train)\nY_dt_pred = dt.predict(X_sc_test)\nprint(classification_report(Y_sc_test, Y_dt_pred))\nconfusion_matrix(Y_sc_test, Y_dt_pred)\n\nfpr, tpr, _ = metrics.roc_curve(Y_sc_test, Y_dt_pred)\n\nauc_score = metrics.auc(fpr, tpr)\ndf_accur_roc_score_importance.loc['Standardized Decision Tree'] = [auc_score]\ndf_accur_score.loc['Standardized Decision Tree'] = [metrics.accuracy_score(Y_sc_test, Y_dt_pred)]","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"dt.fit(X_rb_train, Y_rb_train)\nY_dt_pred = dt.predict(X_rb_test)\nprint(classification_report(Y_rb_test, Y_dt_pred))\nconfusion_matrix(Y_rb_test, Y_dt_pred)\nfpr, tpr, _ = metrics.roc_curve(Y_rb_test, Y_dt_pred)\n\nauc_score = metrics.auc(fpr, tpr)\ndf_accur_roc_score_importance.loc['Robust Decision Tree'] = [auc_score]\ndf_accur_score.loc['Robust Decision Tree'] = [metrics.accuracy_score(Y_rb_test, Y_dt_pred)]","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"dt.fit(X_nm_train, Y_nm_train)\nY_dt_pred = dt.predict(X_nm_test)\nprint(classification_report(Y_nm_test, Y_dt_pred))\nconfusion_matrix(Y_nm_test, Y_dt_pred)\nfpr, tpr, _ = metrics.roc_curve(Y_nm_test, Y_dt_pred)\n\nauc_score = metrics.auc(fpr, tpr)\ndf_accur_roc_score_importance.loc['Normalized Decision Tree'] = [auc_score]\ndf_accur_score.loc['Normalized Decision Tree'] = [metrics.accuracy_score(Y_nm_test, Y_dt_pred)]","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"rf.fit(X_sc_train, Y_sc_train)\nY_rf_pred = rf.predict(X_sc_test)\nprint(classification_report(Y_sc_test, Y_rf_pred))\nconfusion_matrix(Y_sc_test, Y_rf_pred)\nfpr, tpr, _ = metrics.roc_curve(Y_sc_test, Y_rf_pred)\n\nauc_score = metrics.auc(fpr, tpr)\ndf_accur_roc_score_importance.loc['Standardized Random Forest'] = [auc_score]\ndf_accur_score.loc['Standardized Random Forest'] = [metrics.accuracy_score(Y_sc_test, Y_dt_pred)]","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"rf.fit(X_rb_train, Y_rb_train)\nY_dt_pred = rf.predict(X_rb_test)\nprint(classification_report(Y_rb_test, Y_dt_pred))\nconfusion_matrix(Y_rb_test, Y_dt_pred)\nfpr, tpr, _ = metrics.roc_curve(Y_rb_test, Y_dt_pred)\n\nauc_score = metrics.auc(fpr, tpr)\ndf_accur_roc_score_importance.loc['Robust Random Forest'] = [auc_score]\ndf_accur_score.loc['Robust Random Forest'] = [metrics.accuracy_score(Y_rb_test, Y_dt_pred)]","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"rf.fit(X_nm_train, Y_nm_train)\nY_dt_pred = rf.predict(X_nm_test)\nprint(classification_report(Y_nm_test, Y_dt_pred))\nconfusion_matrix(Y_nm_test, Y_dt_pred)\nfpr, tpr, _ = metrics.roc_curve(Y_nm_test, Y_dt_pred)\n\nauc_score = metrics.auc(fpr, tpr)\ndf_accur_roc_score_importance.loc['Normalized Random Forest'] = [auc_score]\ndf_accur_score.loc['Normalized Random Forest'] = [metrics.accuracy_score(Y_nm_test, Y_dt_pred)]","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"from sklearn.neighbors import KNeighborsClassifier\nknn = KNeighborsClassifier()\nknn.fit(X_sc_train, Y_sc_train)\nY_dt_pred = knn.predict(X_sc_test)\nprint(classification_report(Y_sc_test, Y_dt_pred))\nconfusion_matrix(Y_sc_test, Y_dt_pred)\nfpr, tpr, _ = metrics.roc_curve(Y_sc_test, Y_dt_pred)\n\nauc_score = metrics.auc(fpr, tpr)\ndf_accur_roc_score_importance.loc['Standardized KNN'] = [auc_score]\ndf_accur_score.loc['Standardized KNN'] = [metrics.accuracy_score(Y_sc_test, Y_dt_pred)]","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"knn.fit(X_rb_train, Y_rb_train)\nY_dt_pred = knn.predict(X_rb_test)\nprint(classification_report(Y_rb_test, Y_dt_pred))\nconfusion_matrix(Y_rb_test, Y_dt_pred)\nfpr, tpr, _ = metrics.roc_curve(Y_rb_test, Y_dt_pred)\n\nauc_score = metrics.auc(fpr, tpr)\ndf_accur_roc_score_importance.loc['Robust KNN'] = [auc_score]\ndf_accur_score.loc['Robust KNN'] = [metrics.accuracy_score(Y_rb_test, Y_dt_pred)]","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"knn.fit(X_nm_train, Y_nm_train)\nY_dt_pred = knn.predict(X_nm_test)\nprint(classification_report(Y_nm_test, Y_dt_pred))\nconfusion_matrix(Y_nm_test, Y_dt_pred)\nfpr, tpr, _ = metrics.roc_curve(Y_nm_test, Y_dt_pred)\n\nauc_score = metrics.auc(fpr, tpr)\ndf_accur_roc_score_importance.loc['Normalized KNN'] = [auc_score]\ndf_accur_score.loc['Normalized KNN'] = [metrics.accuracy_score(Y_nm_test, Y_dt_pred)]","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"import xgboost as xgb\n\nxgb_model = xgb.XGBClassifier(n_estimators=150,random_state=0,learning_rate=0.1,eta=0.4,booster=\"gbtree\",\n                              base_score=0.8,colsample_bylevel=0.9009229642844634,gamma=0.49967765132613584,\n                        max_depth=6,min_child_weight=7,reg_lambda=0.27611902459972926,subsample=0.9300916052594785)\nxgb_model.fit(X_sc_train, Y_sc_train)\nY_dt_pred = xgb_model.predict(X_sc_test)\nprint(classification_report(Y_sc_test, Y_dt_pred))\nconfusion_matrix(Y_sc_test, Y_dt_pred)\nfpr, tpr, _ = metrics.roc_curve(Y_sc_test, Y_dt_pred)\n\nauc_score = metrics.auc(fpr, tpr)\ndf_accur_roc_score_importance.loc['Standardized XGB'] = [auc_score]\ndf_accur_score.loc['Standardized XGB'] = [metrics.accuracy_score(Y_sc_test, Y_dt_pred)]","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"xgb_model.fit(X_rb_train, Y_rb_train)\nY_dt_pred = xgb_model.predict(X_rb_test)\nprint(classification_report(Y_rb_test, Y_dt_pred))\nconfusion_matrix(Y_rb_test, Y_dt_pred)\nfpr, tpr, _ = metrics.roc_curve(Y_rb_test, Y_dt_pred)\n\nauc_score = metrics.auc(fpr, tpr)\ndf_accur_roc_score_importance.loc['Robust XGB'] = [auc_score]\ndf_accur_score.loc['Robust XGB'] = [metrics.accuracy_score(Y_rb_test, Y_dt_pred)]","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"xgb_model.fit(X_nm_train, Y_nm_train)\nY_dt_pred = xgb_model.predict(X_nm_test)\nprint(classification_report(Y_nm_test, Y_dt_pred))\nconfusion_matrix(Y_nm_test, Y_dt_pred)\nfpr, tpr, _ = metrics.roc_curve(Y_nm_test, Y_dt_pred)\n\nauc_score = metrics.auc(fpr, tpr)\ndf_accur_roc_score_importance.loc['Normalized XGB'] = [auc_score]\ndf_accur_score.loc['Normalized XGB'] = [metrics.accuracy_score(Y_nm_test, Y_dt_pred)]","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"df_accur_roc_score_importance.sort_values(by=['Roc_Auc_Score'],ascending=False).plot(kind='bar', y='Roc_Auc_Score',figsize=(20,8),color='#79ccb3', rot=0,title=\"Model outputs by roc score\")\nplt.xticks(rotation='vertical')","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"df_accur_score.sort_values(by=['Accuracy_Score'],ascending=False).plot(kind='bar', y='Accuracy_Score',figsize=(20,8),color='#79ccb3', rot=0,title=\"Model outputs by accuracy score\")\nplt.xticks(rotation='vertical')","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"df_accur_score.sort_values(by = 'Accuracy_Score', ascending = False)","execution_count":null,"outputs":[]}],"metadata":{"kernelspec":{"language":"python","display_name":"Python 3","name":"python3"},"language_info":{"pygments_lexer":"ipython3","nbconvert_exporter":"python","version":"3.6.4","file_extension":".py","codemirror_mode":{"name":"ipython","version":3},"name":"python","mimetype":"text/x-python"}},"nbformat":4,"nbformat_minor":4}