{"cells":[{"metadata":{},"cell_type":"markdown","source":"# What is about ?\n\nAnalysis of genes correlations for Stoeckius et al 2017 single cell RNA sequences data - 1,774 cells in total by 11757 genes\n\nThere is large correlated component consisting of mainly RPL/RPS genes:\nhttps://en.wikipedia.org/wiki/Ribosomal_protein \n\nAnother corrleated group: \n['ALOX12', 'ARHGAP6', 'C19orf33', 'CALD1', 'CD36', 'CD9', 'CLU', 'GNG11',\n       'GP1BA', 'ITGA2B', 'LTBP1', 'MGLL', 'PF4', 'PPBP', 'PRKAR2B', 'PROS1',\n       'ROGDI', 'SDPR', 'TREML1', 'WFDC1']\n       \n\n\nPay attention data here are integer counts - non-normalized - that might affect the results, if normalization is imposed.\n\n\n\nBased on:\n\nhttps://www.kaggle.com/alexandervc/correlation-analysis-scrnaseq-nestorova2016\n\nhttps://www.kaggle.com/alexandervc/moa-correlation-analysis-use-igraph","execution_count":null},{"metadata":{"_uuid":"8f2839f25d086af736a60e9eeb907d3b93b6e0e5","_cell_guid":"b1076dfc-b9ad-4769-8c92-a6c4dae69d19","trusted":true},"cell_type":"code","source":"# This Python 3 environment comes with many helpful analytics libraries installed\n# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python\n# For example, here's several helpful packages to load\n\nimport numpy as np # linear algebra\nimport pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)\n\n# Input data files are available in the read-only \"../input/\" directory\n# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory\n\nimport os\nfor dirname, _, filenames in os.walk('/kaggle/input'):\n    for filename in filenames:\n        print(os.path.join(dirname, filename))\n\n# You can write up to 5GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using \"Save & Run All\" \n# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"import time\n\nimport matplotlib.pyplot as plt\nimport seaborn as sns\n","execution_count":null,"outputs":[]},{"metadata":{"_uuid":"d629ff2d2480ee46fbb7e2d37f6b5fab8052498a","_cell_guid":"79c7e3d0-c299-4dcb-8224-4455121ee9b0","trusted":true},"cell_type":"code","source":"df = pd.read_csv('/kaggle/input/single-cell-rna-seq-from-stoeckius-et-al-2017/GSE100866_CD8_merged-RNA_umi.csv', index_col = 0)\n# The dataset is single cell rna seq from the Cite-seq paper. \n# Cells were classified by flow cytometry as low, medium, or high cd8 expression cells (columns), with rows corresponding to RNA counts. 1,774 cells in total by 11757 genes. \n# Stoeckius et al 2017.\n\ndf = df.T\ndf","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"X = df.values\nX.shape","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"# First look on correlations","execution_count":null},{"metadata":{"trusted":true},"cell_type":"code","source":"t0 = time.time()\ncorr_matr = np.corrcoef(X.T) # Hint - use numpy , pandas is MUCH SLOWER   (df.corr() )\nprint(time.time() - t0, 'seconds passed')\nprint(np.min(corr_matr ), 'minimal correlation' )\ncorr_matr_abs = np.abs( corr_matr )\nprint(np.mean(corr_matr_abs ), 'average absolute correlation' )\nprint(np.median(corr_matr_abs), 'median absolute correlation' )\nprint(np.min(corr_matr_abs ), 'min absolute correlation' )\nprint(np.std(corr_matr_abs ), 'std absolute correlation' )","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"v = corr_matr.flatten()\nplt.figure(figsize=(14,8))\nt0 = time.time()\nplt.hist(v, bins = 50)\nplt.title('correlation coefficients distribution')\nplt.show()\nprint(time.time() - t0, 'seconds passed')\n\nprint(np.min(corr_matr ), 'minimal correlation' )\nprint(np.mean(corr_matr_abs ), 'average absolute correlation' )\nprint(np.median(corr_matr_abs), 'median absolute correlation' )\nprint(np.min(corr_matr_abs ), 'min absolute correlation' )\nprint(np.std(corr_matr_abs ), 'std absolute correlation' )\nfor t in [0.5,0.6, 0.7,0.8,0.9,0.95,0.97,0.98,.99]:\n    print( ((np.abs(v) < 0.99999999) & (np.abs(v) > t)).sum()/2 , 'number of pairs correlated more than', t  )\nv.shape","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"\n\nplt.figure(figsize=(14,8))\nt0 = time.time()\nsns.heatmap(corr_matr_abs).set_title('Correlation (abs) heatmap')\nprint(time.time() - t0, 'seconds passed')\n\n","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"t0=time.time()\nw,v = np.linalg.eig(corr_matr)\nprint(time.time()-t0)\nw = np.real(w.ravel())\nw = np.sort(w)[::-1]\nplt.plot(w,'*-')\nplt.title('Eigenvalues of correlation matrix ')\nplt.show()\n","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"\n# Analysis with graphs\n\nTake a correlation matrix and change it with threshold - if above - put 1, if below put 0. Construct a graph from it - nodes are connected if 1 in matrix. Look at that graph for several thresholds and see the pattern - there seems to be only one group of correlated features - all other stands separate.\n","execution_count":null},{"metadata":{"trusted":true},"cell_type":"code","source":"\n\nimport igraph\n\n","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"corr_matr_abs_bool = corr_matr_abs > 0.6\ncorr_matr_abs_bool = corr_matr_abs_bool# [:772 ,:772 ]\ncorr_matr_abs_bool = np.triu(corr_matr_abs_bool,1) # Take upper triangular part \ng = igraph.Graph().Adjacency(corr_matr_abs_bool.tolist())\ng.to_undirected(mode = 'collapse')\nprint( corr_matr_abs_bool.astype(int) )\n\nprint('Number of nodes ', g.vcount())\nprint('Number of edges ', g.ecount() )\nprint('Number of weakly connected compoenents', len( g.clusters(mode='WEAK')))\n\n\nprint('Sizes of connected components large than 5 nodes')\nc = 0\nfor t in list(g.clusters(mode='WEAK') ):\n    if len(t) <= 5: continue\n    c+=1 \n    print(len(t) )\nprint('count components large than 5 nodes:', c )\n    \nvisual_style = {}\nvisual_style[\"vertex_color\"] = ['green' for v in g.vs]\n#visual_style[\"vertex_label\"] = range(g.vcount()) \nvisual_style[\"vertex_size\"] = 2\nigraph.plot(g,bbox = (800,500), **visual_style )","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"\n\n# Plot the largest component separately \nfor t in list(g.clusters(mode='WEAK') ):\n    if len(t) <= 30: continue\n    print(t)\n    g2 = g.subgraph(t)\n    index_save = t.copy()\n    \nprint('Number of nodes ', g2.vcount())\nprint('Number of edges ', g2.ecount() )    \nprint(df.columns[index_save])    \nigraph.plot(g2,bbox = (800,200), **visual_style )\n\n","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"verbose = 0\ndf_stat = pd.DataFrame() # dict_save_largest_component_size = {} \ni = 0\nfor correlation_threshold in [0.9, 0.8, 0.7, 0.6, 0.5, 0.4] :\n    t0 = time.time()\n    print()\n    print(correlation_threshold , 'correlation_threshold ')\n    corr_matr_abs_bool = corr_matr_abs > correlation_threshold\n    corr_matr_abs_bool = corr_matr_abs_bool# [:772 ,:772 ]\n    corr_matr_abs_bool = np.triu(corr_matr_abs_bool,1) # Take upper triangular part \n    g = igraph.Graph().Adjacency(corr_matr_abs_bool.tolist())\n    g.to_undirected(mode = 'collapse')\n    if verbose >= 10:\n        print( corr_matr_abs_bool.astype(int) )\n        print('Number of nodes ', g.vcount())\n        print('Number of edges ', g.ecount() )\n        print('Number of weakly connected compoenents', len( g.clusters(mode='WEAK')))\n\n\n    list_clusters_nodes_lists = list( g.clusters(mode='WEAK') )\n    list_clusers_size = [len(t) for t in list_clusters_nodes_lists ]\n    list_clusers_size = np.sort(list_clusers_size)[::-1]\n    print('Top 5 cluster sizes:', list_clusers_size[:5] , 'seconds passed:', np.round(time.time()-t0 , 2))\n    #dict_save_largest_component_size[correlation_threshold ] = list_clusers_size[0]\n    for t  in list_clusters_nodes_lists:\n        if len(t) == list_clusers_size[0]:\n            print('50 Genes in largest correlated group:')\n            print(df.columns[t[:50]])\n    i += 1\n    df_stat.loc[i,'correlation threshold'] = correlation_threshold\n    df_stat.loc[i,'Largest Component Size'] = list_clusers_size[0]\n    df_stat.loc[i,'Second Component Size'] = list_clusers_size[1]\n    \ndf_stat\n","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"","execution_count":null,"outputs":[]}],"metadata":{"kernelspec":{"language":"python","display_name":"Python 3","name":"python3"},"language_info":{"pygments_lexer":"ipython3","nbconvert_exporter":"python","version":"3.6.4","file_extension":".py","codemirror_mode":{"name":"ipython","version":3},"name":"python","mimetype":"text/x-python"}},"nbformat":4,"nbformat_minor":4}