{"cells":[{"cell_type":"markdown","metadata":{"_cell_guid":"11ecdbc1-79a2-4418-af92-fe137896b08b"},"source":"#Getting Started with Biopython\n*Myles O'Neill - based on http://biopython.org/DIST/docs/tutorial/Tutorial.html*\n\nBiopython is a library we can use to analyze bioinformatic data. Lets have a look at some of the things we can do with it!\n\nThis script will go through a selection of tutorial excercises from the Biopython cookbook. I've tweaked them a bit to work with the Drosophila dataset we are using. This is an adapation of the tutorial with a bunch of tweaks, overall it should make it easier to get started with this data on Kaggle.\n\nFirst lets import numpy, pandas, and Bio (biopython):"},{"cell_type":"code","execution_count":null,"metadata":{"_cell_guid":"2932d230-b761-4e6c-a398-51544e5f3814"},"outputs":[],"source":"import numpy as np # linear algebra\nimport pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)\nimport Bio\nprint(\"Biopython v\" + Bio.__version__)"},{"cell_type":"markdown","metadata":{"_cell_guid":"a38748bb-8565-4001-ba04-566150920e08"},"source":"## 2.2  Working with sequences\n\nDisputably (of course!), the central object in bioinformatics is the sequence. Thus, we’ll start with a quick introduction to the Biopython mechanisms for dealing with sequences, the Seq object.\n\nMost of the time when we think about sequences we have in my mind a string of letters like ‘AGTACACTGGT’. You can create such Seq object with this sequence as follows:\n\n"},{"cell_type":"code","execution_count":null,"metadata":{"_cell_guid":"d603893c-a141-42aa-9ef4-024293313509"},"outputs":[],"source":"from Bio.Seq import Seq\nmy_seq = Seq(\"AGTACACTGGT\")\nprint(my_seq)\nmy_seq.alphabet"},{"cell_type":"markdown","metadata":{"_cell_guid":"2d024f71-fa7b-4e26-94a4-ff73348bfc04"},"source":"What we have here is a sequence object with a generic alphabet - reflecting the fact we have not specified if this is a DNA or protein sequence (okay, a protein with a lot of Alanines, Glycines, Cysteines and Threonines!).\n\nIn addition to having an alphabet, the Seq object differs from the Python string in the methods it supports. You can’t do this with a plain string:"},{"cell_type":"code","execution_count":null,"metadata":{"_cell_guid":"2919ba52-baed-4582-b93d-10d7e63d44a3"},"outputs":[],"source":"print(my_seq + \" - Sequence\")\nprint(my_seq.complement() + \" - Complement\")\nprint(my_seq.reverse_complement() + \" - Reverse Complement\")"},{"cell_type":"markdown","metadata":{"_cell_guid":"0fdac57e-bd5c-4739-9e71-dba0536df21c"},"source":"The next most important class is the SeqRecord or Sequence Record. This holds a sequence (as a Seq object) with additional annotation including an identifier, name and description. The Bio.SeqIO module for reading and writing sequence file formats works with SeqRecord objects.\n\nThis covers the basic features and uses of the Biopython sequence class. Now that you’ve got some idea of what it is like to interact with the Biopython libraries, it’s time to delve into the fun, fun world of dealing with biological file formats!"},{"cell_type":"markdown","metadata":{"_cell_guid":"ecc1a0da-0634-4659-9ebd-69a55df12b8b"},"source":"## 2.4  Parsing sequence file formats\n\nA large part of much bioinformatics work involves dealing with the many types of file formats designed to hold biological data. These files are loaded with interesting biological data, and a special challenge is parsing these files into a format so that you can manipulate them with some kind of programming language. However the task of parsing these files can be frustrated by the fact that the formats can change quite regularly, and that formats may contain small subtleties which can break even the most well designed parsers.\n\nFor this example, lets grab the first 6 sequences in the genomic data set"},{"cell_type":"code","execution_count":null,"metadata":{"_cell_guid":"b1603b5a-0ec9-4b18-bae8-1bf56836f253"},"outputs":[],"source":"from Bio import SeqIO\ncount = 0\nsequences = [] # Here we are setting up an array to save our sequences for the next step\n\nfor seq_record in SeqIO.parse(\"../input/genome.fa\", \"fasta\"):\n    if (count < 6):\n        sequences.append(seq_record)\n        print(\"Id: \" + seq_record.id + \" \\t \" + \"Length: \" + str(\"{:,d}\".format(len(seq_record))) )\n        print(repr(seq_record.seq) + \"\\n\")\n        count = count + 1"},{"cell_type":"code","execution_count":null,"metadata":{"_cell_guid":"53e4d376-a619-4a21-85cf-14f7058aebee"},"outputs":[],"source":"# Lets set these sequences up for easy access later\n\nchr2L = sequences[0].seq\nchr2R = sequences[1].seq\nchr3L = sequences[2].seq\nchr3R = sequences[3].seq\nchr4 = sequences[4].seq\nchrM = sequences[5].seq"},{"cell_type":"markdown","metadata":{"_cell_guid":"5129973a-987e-40db-ac37-c848cbd3cb0a"},"source":"## 3.2  Sequences act like strings\n\nIn many ways, we can deal with Seq objects as if they were normal Python strings, for example getting the length, or iterating over the elements:\n\nLets start by printing the length of the first sequence we grabbed before:"},{"cell_type":"code","execution_count":null,"metadata":{"_cell_guid":"8f525ccf-a864-41b8-8890-2ca3d572545a"},"outputs":[],"source":"print(len(chr2L))"},{"cell_type":"markdown","metadata":{"_cell_guid":"259a2c84-86e7-4f77-be1a-fda5231670ae"},"source":"You can access elements of the sequence in the same way as for strings:"},{"cell_type":"code","execution_count":null,"metadata":{"_cell_guid":"70108e95-fa5d-44fa-8d6a-bc9b607aa5dc"},"outputs":[],"source":"print(\"First Letter: \" + chr2L[0])\nprint(\"Third Letter: \" + chr2L[2])\nprint(\"Last Letter: \" + chr2L[-1])"},{"cell_type":"markdown","metadata":{"_cell_guid":"ce6ae462-de9e-4c16-aff5-11a1db9356f7"},"source":"The Seq object has a .count() method, just like a string. Note that this means that like a Python string, this gives a non-overlapping count:"},{"cell_type":"code","execution_count":null,"metadata":{"_cell_guid":"4f178fc6-cba7-46f0-b558-2a69a2f10335"},"outputs":[],"source":"print(\"AAAA\".count(\"AA\"))\nprint(Seq(\"AAAA\").count(\"AA\"))"},{"cell_type":"markdown","metadata":{"_cell_guid":"363d8f86-a921-4c77-a9a4-c9e33f44fe4c"},"source":"For some biological uses, you may actually want an overlapping count (i.e. 3 in this trivial example). When searching for single letters, this makes no difference."},{"cell_type":"markdown","metadata":{"_cell_guid":"888aaf4f-f661-46dd-9fdb-de96e53651c4"},"source":"Lets count the number of G shown in the sequence"},{"cell_type":"code","execution_count":null,"metadata":{"_cell_guid":"0ce1c0b7-1b1e-4aa4-8d51-bef3946e2e94"},"outputs":[],"source":"print(\"Length:\\t\" + str(len(chr2L)))\nprint(\"G Count:\\t\" + str(chr2L.count(\"G\")))"},{"cell_type":"markdown","metadata":{"_cell_guid":"b89d62d4-d80b-489b-84f6-dee5aad6deb9"},"source":"The GC Content of a DNA sequence is important and relates to how stable the molecule will be. We can calculate it manually like this:"},{"cell_type":"code","execution_count":null,"metadata":{"_cell_guid":"09d50139-31a9-4f6d-bb4e-81e5809c250e"},"outputs":[],"source":"print(\"GC%:\\t\\t\" + str(100 * float((chr2L.count(\"G\") + chr2L.count(\"C\")) / len(chr2L) ) ))"},{"cell_type":"markdown","metadata":{"_cell_guid":"95d891ca-6390-4999-9fb9-1f1059c604cf"},"source":"While you could use the above snippet of code to calculate a GC%, note that the Bio.SeqUtils module has several GC functions already built. For example:"},{"cell_type":"code","execution_count":null,"metadata":{"_cell_guid":"b139a051-a39e-4545-a063-c4ecf10f34d4"},"outputs":[],"source":"from Bio.SeqUtils import GC\nprint(\"GC% Package:\\t\" + str(GC(chr2L)))"},{"cell_type":"markdown","metadata":{"_cell_guid":"ef0fda8a-cd2f-48d5-a26a-6c6c42b3b9e4"},"source":"But wait a minute! Why aren't those two numbers the name?\n\nThe answer lies in the sequence. We only used capital G/C characters, but in the actual sequence there are lowercase g/c characters. In addition, there are also S and s characters which represent an ambiguous G OR C character - but which are being counted for GC content by the package. Lets add those and check again:\n\n"},{"cell_type":"code","execution_count":null,"metadata":{"_cell_guid":"e08fbc39-a3e9-4c2c-928f-48a8c8362f80"},"outputs":[],"source":"print(\"GgCcSs%:\\t\" + str(100 * float((chr2L.count(\"G\") + chr2L.count(\"g\") + chr2L.count(\"C\") + chr2L.count(\"c\") + chr2L.count(\"S\") + chr2L.count(\"s\") ) / len(chr2L) ) ))\nprint(\"GC% Package:\\t\" + str(GC(chr2L)))"},{"cell_type":"markdown","metadata":{"_cell_guid":"e4503e44-9ec4-4c8f-885c-b5ef2191304f"},"source":"Much better!"},{"cell_type":"markdown","metadata":{"_cell_guid":"8b34cb8c-7a06-40d4-a90e-4324a48e6d7f"},"source":"## 3.3  Slicing a sequence\n\nA more complicated example, let’s get a slice of the sequence:"},{"cell_type":"code","execution_count":null,"metadata":{"_cell_guid":"2b506fd7-5a89-4b6f-bbfe-7be0669c2ad6"},"outputs":[],"source":"print(chr2L[4:12])"},{"cell_type":"markdown","metadata":{"_cell_guid":"c034a869-76bc-4aa3-93ea-530b1e29a97c"},"source":"When you do a slice the first item is included (i.e. 4 in this case) and the last is excluded (12 in this case)\n\nThe second thing to notice is that the slice is performed on the sequence data string, but the new object produced is another Seq object which retains the alphabet information from the original Seq object.\n\nAlso like a Python string, you can do slices with a start, stop and stride (the step size, which defaults to one). For example, we can get the first, second and third codon positions of this DNA sequence:\n\nWe are using a short subset of chr2L since we don't want to go printing millions of characters"},{"cell_type":"code","execution_count":null,"metadata":{"_cell_guid":"06a1bc1e-ddd4-4f8c-8841-dd41cc07fac0"},"outputs":[],"source":"chr2LSHORT = chr2L[0:20]\nprint(\"Short chr2L: \" + chr2LSHORT)\n\nprint(\"Codon Pos 1: \" + chr2LSHORT[0::3])\nprint(\"Codon Pos 2: \" + chr2LSHORT[1::3])\nprint(\"Codon Pos 3: \" + chr2LSHORT[2::3])"},{"cell_type":"markdown","metadata":{"_cell_guid":"eb936c97-3976-4894-95d3-2793b2b210aa"},"source":"Another stride trick you might have seen with a Python string is the use of a -1 stride to reverse the string. You can do this with a Seq object too:"},{"cell_type":"code","execution_count":null,"metadata":{"_cell_guid":"498df722-ff7c-410d-b673-6bde0ee085d6"},"outputs":[],"source":"print(\"Reversed: \" + chr2LSHORT[::-1])"},{"cell_type":"markdown","metadata":{"_cell_guid":"a34335e0-6ecf-4dfa-8792-637c61e49bce"},"source":"## 3.5  Concatenating or adding sequences\n\nNaturally, you can in principle add any two Seq objects together - just like you can with Python strings to concatenate them. However, you can’t add sequences with incompatible alphabets, such as a protein sequence and a DNA sequence:"},{"cell_type":"code","execution_count":null,"metadata":{"_cell_guid":"69b0c0a5-0940-4388-b390-1256edc3866c"},"outputs":[],"source":"chr2LSHORT = chr2L[0:20]\nprint(\"Short chr2L: \" + chr2LSHORT)\n\nchr2RSHORT = chr2R[0:20]\nprint(\"Short chr2R: \" + chr2RSHORT)\n\nconcat = chr2LSHORT + chr2RSHORT\nprint(\"Concat: \" + concat)"},{"cell_type":"markdown","metadata":{"_cell_guid":"18155aaf-9507-48c1-9d6b-d3b776817527"},"source":"If you really want to concat sequences from different alphabets, you’d have to first give both sequences generic alphabets:"},{"cell_type":"code","execution_count":null,"metadata":{"_cell_guid":"511a6e90-d712-46cc-8858-4cdb4db2aef3"},"outputs":[],"source":"from Bio.Alphabet import IUPAC\nprotein_seq = Seq(\"EVRNAK\", IUPAC.protein)\ndna_seq = Seq(\"ACGT\", IUPAC.unambiguous_dna)\n\n# This will fail since they have different alphabets\n# print(protein_seq + dna_seq)\n\n# Error: Incompatible alphabets IUPACProtein() and IUPACUnambiguousDNA()\n\n# But if we give them the same generic alphabet it works\n\nfrom Bio.Alphabet import generic_alphabet\nprotein_seq.alphabet = generic_alphabet\ndna_seq.alphabet = generic_alphabet\nprint(protein_seq + dna_seq)"},{"cell_type":"markdown","metadata":{"_cell_guid":"b016faa6-3e78-48d8-b4f2-4c7d001163f2"},"source":"You may often have many sequences to add together, which can be done with a for loop like this:"},{"cell_type":"code","execution_count":null,"metadata":{"_cell_guid":"4d5e1a33-b7d5-450b-aad4-6be7d130971e"},"outputs":[],"source":"from Bio.Alphabet import generic_dna\n\nlist_of_seqs = [Seq(\"ACGT\", generic_dna), Seq(\"AACC\", generic_dna), Seq(\"GGTT\", generic_dna)]\nconcatenated = Seq(\"\", generic_dna)\nfor s in list_of_seqs:\n    concatenated += s\nprint(concatenated)"},{"cell_type":"markdown","metadata":{"_cell_guid":"94d10247-4bee-40df-a7a7-bab06cc79e24"},"source":"Or, a more elegant approach is to the use built in sum function with its optional start value argument (which otherwise defaults to zero):"},{"cell_type":"code","execution_count":null,"metadata":{"_cell_guid":"f0e019b7-b7c7-4228-b22a-05f9dedc5ccb"},"outputs":[],"source":"list_of_seqs = [Seq(\"ACGT\", generic_dna), Seq(\"AACC\", generic_dna), Seq(\"GGTT\", generic_dna)]\nprint(sum(list_of_seqs, Seq(\"\", generic_dna)))"},{"cell_type":"markdown","metadata":{"_cell_guid":"de1011ae-ca16-4972-abc3-82dbbd32a641"},"source":"Unlike the Python string, the Biopython Seq does not (currently) have a .join method."},{"cell_type":"markdown","metadata":{"_cell_guid":"0f8452e8-9ace-4e6a-92ea-02b35d8feab8"},"source":"## 3.6  Changing case"},{"cell_type":"code","execution_count":null,"metadata":{"_cell_guid":"4cdd747c-85dc-4223-807b-b1fae295423a"},"outputs":[],"source":"dna_seq = Seq(\"acgtACGT\", generic_dna)\nprint(\"Original: \" + dna_seq)\nprint(\"Upper: \" + dna_seq.upper())\nprint(\"Lower: \" + dna_seq.lower())"},{"cell_type":"markdown","metadata":{"_cell_guid":"8f50132c-8d9f-47ff-98cc-24e921587a03"},"source":"These are useful for doing case insensitive matching:"},{"cell_type":"code","execution_count":null,"metadata":{"_cell_guid":"dea72af3-a650-4668-adc2-4645cdc75dc6"},"outputs":[],"source":"print(\"GTAC\" in dna_seq)\nprint(\"GTAC\" in dna_seq.upper())"},{"cell_type":"markdown","metadata":{"_cell_guid":"421fa369-e00e-49f5-8dd8-497f1a3fe416"},"source":"## 3.7  Nucleotide sequences and (reverse) complements\n\nFor nucleotide sequences, you can easily obtain the complement or reverse complement of a Seq object using its built-in methods:\n"},{"cell_type":"code","execution_count":null,"metadata":{"_cell_guid":"17caf43b-edb9-4d02-a705-dff03ede276a"},"outputs":[],"source":"print(\"Original: \" + chr2LSHORT)\nprint(\"Complement: \" + chr2LSHORT.complement())\nprint(\"Reverse Complement: \" + chr2LSHORT.reverse_complement())"},{"cell_type":"markdown","metadata":{"_cell_guid":"c3c3f6f0-1db9-40c2-bcbf-a1a7a0b802c0"},"source":"In all of these operations, the alphabet property is maintained. This is very useful in case you accidentally end up trying to do something weird like take the (reverse)complement of a protein sequence (which will give an error)"},{"cell_type":"markdown","metadata":{"_cell_guid":"0c689fa0-e383-4ca3-b565-46930fc53c9e"},"source":"## 3.8  Transcription\n\nThe actual biological transcription process works from the template strand, doing a reverse complement (TCAG → CUGA) to give the mRNA. However, in Biopython and bioinformatics in general, we typically work directly with the coding strand because this means we can get the mRNA sequence just by switching T → U."},{"cell_type":"code","execution_count":null,"metadata":{"_cell_guid":"0bfdf64a-2534-4285-b7ea-1e1b995ff6a8"},"outputs":[],"source":"print(\"Coding DNA: \" + chr2LSHORT)\ntemplate_dna = chr2LSHORT.reverse_complement()\nprint(\"Template DNA: \" + template_dna)"},{"cell_type":"markdown","metadata":{"_cell_guid":"5cc0e69e-91b7-456a-9dbe-c7388f430246"},"source":"Biology Note: (remember by convention nucleotide sequences are normally read from the 5’ to 3’ direction)\n\nNow let’s transcribe the coding strand into the corresponding mRNA, using the Seq object’s built in transcribe method:\n"},{"cell_type":"code","execution_count":null,"metadata":{"_cell_guid":"2b0df5c1-48af-453a-9306-b5de5240c561"},"outputs":[],"source":"messenger_rna = chr2LSHORT.transcribe()\nprint(\"Messenger RNA: \" + messenger_rna)"},{"cell_type":"markdown","metadata":{"_cell_guid":"104767c4-72fc-416d-b301-c5d3a12fd4f0"},"source":"As you can see, all this does is switch T → U, and adjust the alphabet."},{"cell_type":"markdown","metadata":{"_cell_guid":"9ed10e61-4627-41ed-903e-41af79464432"},"source":"## 3.9  Translation\n\nUsing a new example, let’s translate this mRNA into the corresponding protein sequence - again taking advantage of one of the Seq object’s biological methods:\n\n"},{"cell_type":"code","execution_count":null,"metadata":{"_cell_guid":"749df605-a812-4d65-a6c0-b871f154bfae"},"outputs":[],"source":"from Bio.Alphabet import IUPAC\nmessenger_rna = Seq(\"AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG\", IUPAC.unambiguous_rna)\nprint(\"Messenger RNA: \" + messenger_rna)\nprint(\"Protein Sequence: \" + messenger_rna.translate())"},{"cell_type":"markdown","metadata":{"_cell_guid":"090060e6-3b1d-4048-bda1-a5f3ca6a4e87"},"source":"You can also translate directly from the coding strand DNA sequence:\n"},{"cell_type":"code","execution_count":null,"metadata":{"_cell_guid":"408b3ced-989e-452a-ba9f-bca3d1535bff"},"outputs":[],"source":"coding_dna = Seq(\"ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG\", IUPAC.unambiguous_dna)\nprint(\"Coding DNA: \" + coding_dna)\nprint(\"Protein Sequence: \" + coding_dna.translate())"},{"cell_type":"markdown","metadata":{"_cell_guid":"2a13807a-5c87-4e2f-9c8b-b92aa4d92300"},"source":"You should notice in the above protein sequences that in addition to the end stop character, there is an internal stop as well. This was a deliberate choice of example, as it gives an excuse to talk about some optional arguments, including different translation tables (Genetic Codes).\n\nThe translation tables available in Biopython are based on those from the NCBI (see the next section of this tutorial). By default, translation will use the standard genetic code (NCBI table id 1). Suppose we are dealing with a mitochondrial sequence. We need to tell the translation function to use the relevant genetic code instead:\n\nNCBI Tables: http://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi"},{"cell_type":"code","execution_count":null,"metadata":{"_cell_guid":"efdb0095-eab5-42d0-867b-d6e79baa45e4"},"outputs":[],"source":"print(\"Vertebrate Mitochondrial Table Result: \" + coding_dna.translate(table=\"Vertebrate Mitochondrial\"))"},{"cell_type":"markdown","metadata":{"_cell_guid":"dcb20cf4-b446-4364-afb3-4b3e023ecee7"},"source":"You can also specify the table using the NCBI table number which is shorter, and often included in the feature annotation of GenBank files:"},{"cell_type":"code","execution_count":null,"metadata":{"_cell_guid":"d398034a-d4b4-4a91-a1c0-a09996f3a013"},"outputs":[],"source":"print (\"Table 2 Result: \" + coding_dna.translate(table=2))"},{"cell_type":"markdown","metadata":{"_cell_guid":"471c26b2-4545-46c5-b15b-26533e46e3cf"},"source":"Now, you may want to translate the nucleotides up to the first in frame stop codon, and then stop (as happens in nature):"},{"cell_type":"code","execution_count":null,"metadata":{"_cell_guid":"95f5090f-734b-4d2d-add9-37a2ce20b263"},"outputs":[],"source":"print(\"Standard Translation: \" + coding_dna.translate())\nprint(\"Stop as in Biology: \" + coding_dna.translate(to_stop=True))\nprint(\"Table 2 Translation: \" + coding_dna.translate(table=2))\nprint(\"Table 2 Translation with Stop: \" + coding_dna.translate(table=2, to_stop=True))"},{"cell_type":"markdown","metadata":{"_cell_guid":"ba06f769-3065-40d4-bbb8-2ebd7399cb91"},"source":"Notice that when you use the to_stop argument, the stop codon itself is not translated - and the stop symbol is not included at the end of your protein sequence."},{"cell_type":"markdown","metadata":{"_cell_guid":"014bc1ad-ad3c-4268-bc16-f6fab33f6b8c"},"source":"Now, suppose you have a complete coding sequence CDS, which is to say a nucleotide sequence (e.g. mRNA – after any splicing) which is a whole number of codons (i.e. the length is a multiple of three), commences with a start codon, ends with a stop codon, and has no internal in-frame stop codons. In general, given a complete CDS, the default translate method will do what you want (perhaps with the to_stop option). However, what if your sequence uses a non-standard start codon? This happens a lot in bacteria – for example the gene yaaX in E. coli K12:"},{"cell_type":"code","execution_count":null,"metadata":{"_cell_guid":"65282521-7c37-418a-b88c-89dfa2c827f2"},"outputs":[],"source":"from Bio.Seq import Seq\nfrom Bio.Alphabet import generic_dna\ngene = Seq(\"GTGAAAAAGATGCAATCTATCGTACTCGCACTTTCCCTGGTTCTGGTCGCTCCCATGGCAGCACAGGCTGCGGAAATTACGTTAGTCCCGTCAGTAAAATTACAGATAGGCGATCGTGATAATCGTGGCTATTACTGGGATGGAGGTCACTGGCGCGACCACGGCTGGTGGAAACAACATTATGAATGGCGAGGCAATCGCTGGCACCTACACGGACCGCCGCCACCGCCGCGCCACCATAAGAAAGCTCCTCATGATCATCACGGCGGTCATGGTCCAGGCAAACATCACCGCTAA\", generic_dna)\nprint(gene)\nprint(\"Bacterial Translation With Stop: \" + gene.translate(table=\"Bacterial\", to_stop=True))"},{"cell_type":"markdown","metadata":{"_cell_guid":"09cb0734-6b6c-465a-8695-1c4a1566db5d"},"source":"In the bacterial genetic code GTG is a valid start codon, and while it does normally encode Valine, if used as a start codon it should be translated as methionine. This happens if you tell Biopython your sequence is a complete CDS:"},{"cell_type":"code","execution_count":null,"metadata":{"_cell_guid":"299d7910-7893-41ea-a640-6fba48f7e8ac"},"outputs":[],"source":"print (\"Bacterial Translation of CDS: \" + gene.translate(table=\"Bacterial\", cds=True))"},{"cell_type":"markdown","metadata":{"_cell_guid":"c4593188-1855-4dac-b3b2-d0f410afe6ce"},"source":"In addition to telling Biopython to translate an alternative start codon as methionine, using this option also makes sure your sequence really is a valid CDS (you’ll get an exception if not)."},{"cell_type":"markdown","metadata":{"_cell_guid":"9adbbc0e-69aa-45df-8c1e-3837e4f01339"},"source":"## 3.11  Comparing Seq objects\n\nSequence comparison is actually a very complicated topic, and there is no easy way to decide if two sequences are equal. The basic problem is the meaning of the letters in a sequence are context dependent - the letter “A” could be part of a DNA, RNA or protein sequence. Biopython uses alphabet objects as part of each Seq object to try to capture this information - so comparing two Seq objects could mean considering both the sequence strings and the alphabets.\n\nFor example, you might argue that the two DNA Seq objects Seq(\"ACGT\", IUPAC.unambiguous_dna) and Seq(\"ACGT\", IUPAC.ambiguous_dna) should be equal, even though they do have different alphabets. Depending on the context this could be important.\n\nThis gets worse – suppose you think Seq(\"ACGT\", IUPAC.unambiguous_dna) and Seq(\"ACGT\") (i.e. the default generic alphabet) should be equal. Then, logically, Seq(\"ACGT\", IUPAC.protein) and Seq(\"ACGT\") should also be equal. Now, in logic if A=B and B=C, by transitivity we expect A=C. So for logical consistency we’d require Seq(\"ACGT\", IUPAC.unambiguous_dna) and Seq(\"ACGT\", IUPAC.protein) to be equal – which most people would agree is just not right. This transitivity also has implications for using Seq objects as Python dictionary keys.\n\nNow, in everyday use, your sequences will probably all have the same alphabet, or at least all be the same type of sequence (all DNA, all RNA, or all protein). What you probably want is to just compare the sequences as strings – which you can do explicitly:"},{"cell_type":"code","execution_count":null,"metadata":{"_cell_guid":"c9fcf1c7-ce82-4cf5-8c6a-f2b36338bd58"},"outputs":[],"source":"from Bio.Seq import Seq\nfrom Bio.Alphabet import IUPAC\nseq1 = Seq(\"ACGT\", IUPAC.unambiguous_dna)\nseq2 = Seq(\"ACGT\", IUPAC.ambiguous_dna)\nprint(str(seq1) == str(seq2))\nprint(str(seq1) == str(seq1))"},{"cell_type":"markdown","metadata":{"_cell_guid":"6f48e418-178c-4fbf-acdc-cd63b280e7a8"},"source":"So, what does Biopython do? Well, as of Biopython 1.65, sequence comparison only looks at the sequence, essentially ignoring the alphabet:"},{"cell_type":"code","execution_count":null,"metadata":{"_cell_guid":"1a35fa31-567c-4f02-8e38-75c6427aae93"},"outputs":[],"source":"print(seq1 == seq2)\nprint(seq1 == \"ACGT\")"},{"cell_type":"markdown","metadata":{"_cell_guid":"f8513161-aff7-4a1f-aebb-a12d5194aa00"},"source":"Thats all for now, hope this tutorial run through was helpful! You can find lots more info about biopython here: http://biopython.org/DIST/docs/tutorial/Tutorial.html"}],"metadata":{"_change_revision":0,"_is_fork":false,"kernelspec":{"display_name":"Python 3","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.6.0"}},"nbformat":4,"nbformat_minor":0}