{"cells":[{"metadata":{},"cell_type":"markdown","source":"# Hacking COVID-19 "},{"metadata":{},"cell_type":"markdown","source":"First I imported the raw COVID-19 genome and stored it in a string. Next, I was interested in how much raw information the corona virus contained. To figure that out, I took the genome and encoded it from a string to bytes (encoded via UTF-8 formatting), and it turns out the corona virus contains exactly 29KB of data, uncompressed, which is about the same amount of data as a non formatted written essay. Below you can see the full genome, and the associated file size."},{"metadata":{"_uuid":"8f2839f25d086af736a60e9eeb907d3b93b6e0e5","_cell_guid":"b1076dfc-b9ad-4769-8c92-a6c4dae69d19","trusted":true},"cell_type":"code","source":"filename = '../input/sars-coronavirus-accession/SARS_CORONAVIRUS_NC_045512_sequence.fasta'\n\ncovid_genome = \"\" \nwith open(filename) as f: \n    genome = f.read() \n\ncovid_genome = genome.replace(\">NC_045512.2 Wuhan seafood market pneumonia virus isolate Wuhan-Hu-1, complete genome\", \"\").replace(\"\\n\", \"\")\nprint(genome)\n","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"b = covid_genome.encode()\n\nprint(\"Corona Virus Uncompressed File Size: \" + str(len(b) / 1000) + \" KB\")","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"# Transcription and Translation"},{"metadata":{},"cell_type":"markdown","source":"Next, I performmed both transcription and translation on the genome. To start, I transcribed the genome by replacing the Thymine nucleotides with Uracil, since RNA cannot contain Thymine. Next, I imported a dataset contianing a list of codons and their associated nucleotide sequences. I matched the codon, and then inserted the amino acid single letter identifier into an array. Below you can see the translated genome, and the amino acid sequence."},{"metadata":{"trusted":true},"cell_type":"code","source":"transcribed_genome = covid_genome.replace('T', 'U')\ncovid_codons = [transcribed_genome[i:i+3] for i in range(0, len(transcribed_genome), 3)]\nprint(covid_codons)    ","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"import pandas as pd\ndf = pd.read_csv('../input/codons/Codons - Sheet1 (1).csv')\n\namino_acids = []\n\nfor codon in covid_codons: \n    try: \n        index = df.index[df['codon'] == codon][0]\n        amino_acids.append(df.at[index, 'letter'])\n    except: \n        continue \n\nprotien_sequence = ''.join(amino_acids)\nprint(protien_sequence)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"# Amino Acids Analysis"},{"metadata":{},"cell_type":"markdown","source":"Next, I analyzed the amino acid sequence. To do so, I once again translated the string of letters into bytes, and evaluated how much information it contained. About 9.9 kilobytes. Finally, I counted the amount of amino acids and plotted them on a nice bar graph for visualization. "},{"metadata":{"trusted":true},"cell_type":"code","source":"x = protien_sequence.encode() \nprint(str(len(x) / 1000) + ' KB')","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"unique_acids = [] \nfor acid in protien_sequence: \n    if acid not in unique_acids: \n        unique_acids.append(acid)\ndata = {}\nfor acid in unique_acids: \n    data[acid] = protien_sequence.count(acid)\ndf = pd.read_csv('../input/codons/Codons - names.csv')\ndf.tail()\nfor letter in data.keys():\n    try: \n        index = df.index[df['One Letter Code'] == letter][0]\n        val = data[letter]\n        new_title = df.at[index, 'name']\n        del data[letter]\n        data[new_title] = val \n    except: \n        continue \n\n#remove stop/start codons\ndel data['K']\ndel data['O']\n    ","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"import matplotlib.pyplot as plt\nfig = plt.figure() \nax = fig.add_axes([0 ,1, 3.5, 2])\nax.set_title('COVID-19 Amino Acids', fontsize=20)\nax.bar(data.keys(), data.values())\nplt.show()","execution_count":null,"outputs":[]}],"metadata":{"kernelspec":{"language":"python","display_name":"Python 3","name":"python3"},"language_info":{"pygments_lexer":"ipython3","nbconvert_exporter":"python","version":"3.6.4","file_extension":".py","codemirror_mode":{"name":"ipython","version":3},"name":"python","mimetype":"text/x-python"}},"nbformat":4,"nbformat_minor":4}