{"cells":[{"metadata":{},"cell_type":"markdown","source":"## Collecting Corona Virus Proteins from Uniprot Database\n\n**What is Uniprot Database?**\n\nUniProt is a freely accessible database of protein sequence and functional information, many entries being derived from genome sequencing projects. It contains a large amount of information about the biological function of proteins derived from the research literature.[source: Wikipedia](https://en.wikipedia.org/wiki/UniProt)\n\n**Why is this protein list important to current COVID-19 Dataset?**\n\nInorder to gather the information about the biomolecular mechanism from the scientific literature (COVID-19 Dataset), one need to have the list of associated Proteins, Genes, Pathways, Drugs etc. This notebook presents the steps to gather Corona Virus associated proteins, Gene names and associated Pathways from Uniprot database. These lits could be useful to look at the textual documents for further NLP processing and to present the entity relationship."},{"metadata":{},"cell_type":"markdown","source":"### 1. Getting Data"},{"metadata":{"_uuid":"d629ff2d2480ee46fbb7e2d37f6b5fab8052498a","collapsed":true,"_cell_guid":"79c7e3d0-c299-4dcb-8224-4455121ee9b0","trusted":false},"cell_type":"markdown","source":"#### Step -I\nGp to Uniprot Database (https://www.uniprot.org/) and select UniprotKB in search bar. Then inter corona virus into the search bar.\n"},{"metadata":{},"cell_type":"markdown","source":"![](https://github.com/Vasuji/COVID19/blob/master/img/uniprot-search.png?raw=true)"},{"metadata":{},"cell_type":"markdown","source":"#### Step -II:\nAfter you hit search operation, you will get a table like disply of the result. It is multi page table. \n\n![](https://github.com/Vasuji/COVID19/blob/master/img/uniprot-table.png?raw=true)"},{"metadata":{},"cell_type":"markdown","source":"#### Step-III:\nLook at the right most task bar of this table. You can see pen like icon through which you get next window. You can make a selection of the information you want to gather (e.g., Name, Gene, Pathways).\n\n![](https://github.com/Vasuji/COVID19/blob/master/img/pen.png?raw=true)"},{"metadata":{},"cell_type":"markdown","source":"#### Step - IV\n\nOnce you are done with selection of information, you can go back to previous table and hit download button. You can select the format of the data. Excel file download is one option.\n\n![](https://github.com/Vasuji/COVID19/blob/master/img/uniprot-download.png?raw=true)\n"},{"metadata":{},"cell_type":"markdown","source":"### 2. Data Wrangling"},{"metadata":{},"cell_type":"markdown","source":"#### What After getting Protein Data?\n\nLets play around with this data"},{"metadata":{"_cell_guid":"","_uuid":"","trusted":true},"cell_type":"code","source":"import pandas as pd\nimport matplotlib.pyplot as plt\n%matplotlib inline\nfile_path = \"../input/corona-virus-proteins-from-uniprot-database/corona.csv\"\ndf = pd.read_csv(file_path)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"df.head(5)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"**There are total 21,876 proteins from different sources**"},{"metadata":{"trusted":true},"cell_type":"code","source":"df.shape","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"#### Q: What are the different Organisms? Can you find the top 50 Organisms?"},{"metadata":{"trusted":true},"cell_type":"code","source":"df_organism = pd.DataFrame(df.groupby(\"Organism\").count()['Entry'])\ndf_organism = df_organism.sort_values(by = \"Entry\", ascending = False)\ndf_organism[0:20].plot.barh(figsize = [15,10], fontsize =20)\nplt.gca().invert_yaxis()","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"df_organism[0:20]","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"#### Q: What are the different Viral hosts? Can you find top Virus hosts?"},{"metadata":{"trusted":true},"cell_type":"code","source":"df['Virus hosts'] = df['Virus hosts'].apply(lambda x: str(x)[0:50] )\ndf_host = pd.DataFrame(df.groupby(\"Virus hosts\").count()['Entry'])\ndf_host = df_host.sort_values(by = \"Entry\", ascending = False)\ndf_host[1:20].plot.barh(figsize = [15,10], fontsize =20)\nplt.gca().invert_yaxis()","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"df_host[1:20]","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"### 3. Cleaning Protein Names, Synonyms and abbreviations"},{"metadata":{"trusted":true},"cell_type":"code","source":"def filter(line):\n    proteins = set()\n    line = str(line)\n    line = line.lower()\n    \n    '''for lines without () or [] terms'''\n    if \"(\" not in line or \"[\" not in line:\n        proteins.add(line.strip().replace(' ', '_'))\n        \n        \n    '''for line including () terms'''    \n    if '(' in line:\n        start = 0\n        open_in = line.find('(')\n        tmp = line[start:open_in].strip().replace(' ', '_')\n        proteins.add(tmp)\n        while open_in >=0:\n            start = open_in+1\n            end = line.find(')', start)\n            proteins.add(line[start:end].strip().replace(' ', '_'))\n            open_in = line.find('(', end)\n     \n    '''for lines including [] trems'''\n    if '[' in line:\n        raw = line[line.find('['):line.find(']')]\n        #print(\"THIS IS RAW:\", raw[15:-1])\n        raw = raw[15:-1]\n        lraw = raw.split(\"; \")\n        for item in lraw:\n            #print(item)\n            if '(' in item:\n                start = 0\n                open_in = item.find('(')\n                tmp = item[start:open_in].strip().replace(' ', '_')\n                proteins.add(tmp)\n            else:\n                proteins.add(item.strip().replace(' ', '_'))\n    return proteins","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"allProteins = []\ni = 0\nfor u,p in zip(df['Entry'],df['Protein names']):\n    print(u,\"|\",p)\n    print(\"------------\")\n    print(u,\"|\",filter(p))\n    print(\"===================================================\")\n    i += 1\n    if i>4:\n        break","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"allProteins = []\nfor u,p in zip(df['Entry'],df['Protein names']):\n    allProteins.append({\"id\":u, \"names\":list(filter(p))})","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"allProteins[0:5]","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"import json\nwith open(\"virus-proteins.json\", 'w') as fn:\n    json.dump(allProteins,fn)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"***Note: Bringing more detail about how to create data indexing and search of Proteins, Genes etc from COVID19 dataset in the next notebook***"}],"metadata":{"kernelspec":{"language":"python","display_name":"Python 3","name":"python3"},"language_info":{"pygments_lexer":"ipython3","nbconvert_exporter":"python","version":"3.6.4","file_extension":".py","codemirror_mode":{"name":"ipython","version":3},"name":"python","mimetype":"text/x-python"}},"nbformat":4,"nbformat_minor":4}