{"cells":[{"metadata":{},"cell_type":"markdown","source":"V8 OVA_Prostate dataset from openml \n\nV6,7 - reformat code - add compraison with Gauss and uniform datasets with the same size \n\nV5 - single-cell-rna-seq-from-stoeckius-et-al-2017/GSE100866_CD8_merged-RNA_umi\n\nV4 - GSE67123_v6_scrna_10x_143_23548_Mouse_Embryo_HSCs_invivo_fromCytotrace.csv\n\nV3 - human_liver from ARCHS4\n\nV2 - GSE90047_Mouse_Liver_Hepatoblast_invivo\n\nV1 - Dexter300 dataset \n"},{"metadata":{"_uuid":"8f2839f25d086af736a60e9eeb907d3b93b6e0e5","_cell_guid":"b1076dfc-b9ad-4769-8c92-a6c4dae69d19","trusted":true},"cell_type":"code","source":"# This Python 3 environment comes with many helpful analytics libraries installed\n# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python\n# For example, here's several helpful packages to load\n\nimport numpy as np # linear algebra\nimport pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)\n\n# Input data files are available in the read-only \"../input/\" directory\n# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory\n\nimport os\nfor dirname, _, filenames in os.walk('/kaggle/input'):\n    for filename in filenames:\n        print(os.path.join(dirname, filename))\n\n# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using \"Save & Run All\" \n# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"from sklearn.neighbors import kneighbors_graph\nimport time\nimport matplotlib.pyplot as plt\n\ntry:\n    import cudf\n    from cuml.neighbors import NearestNeighbors\n    #from cuml.datasets import make_blobs    \nexcept:\n    print('GPU library cannot be imported. Turn ON GPU')\n    ","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"# Load Data"},{"metadata":{"trusted":true},"cell_type":"code","source":"pip install openml\n","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"import openml\n\nt0 = time.time()\nopenml_df = openml.datasets.list_datasets(output_format=\"dataframe\")\nprint(time.time()-t0,'seconds passed ')\n\nif 0: ## The same can be done with BIGGER lines of code\n    openml_list = openml.datasets.list_datasets()  # returns a dict\n    # Show a nice table with some key data properties\n    openml_df = pd.DataFrame.from_dict(openml_list, orient=\"index\")\n    openml_df = datalist[[\"did\", \"name\", \"NumberOfInstances\", \"NumberOfFeatures\", \"NumberOfClasses\"]]\n\n    print(f\"First 10 of {len(datalist)} datasets...\")\n    openml_df.head(n=10)\n\n\nopenml_df\n\n# Among all datasets there is subset of 45 datasets with  mircoarray Gene Expressions (mainly cancer) \n# can be selected by number of features == 10937\ndf_gemler = openml_df[ ( openml_df.NumberOfFeatures == 10937)    ].sort_values([\"NumberOfInstances\"], ascending=False)#.head(n=20)\nprint(df_gemler.shape)\ndf_gemler","execution_count":null,"outputs":[]},{"metadata":{"_uuid":"d629ff2d2480ee46fbb7e2d37f6b5fab8052498a","_cell_guid":"79c7e3d0-c299-4dcb-8224-4455121ee9b0","trusted":true},"cell_type":"code","source":"dataset = '' \n\nif dataset == 'Dexter300from_scikit_hubness':\n    f = '/kaggle/input/hubness-for-high-dimensional-datasets/y_Dexter300from_scikit_hubness.csv'\n    y = pd.read_csv(f ,  index_col= 0  )\n    y\n    f = '/kaggle/input/hubness-for-high-dimensional-datasets/X_Dexter300from_scikit_hubness.csv'\n    X = pd.read_csv(f ,  index_col= 0  )\n    X\n\nif dataset == 'GSE90047_Mouse_Liver_Hepatoblast_invivo':\n    f = '/kaggle/input/trajectory-inference-single-cell-rna-seq/GSE90047_Mouse_Liver_Hepatoblast_invivo.csv'\n    X = pd.read_csv(f ,  index_col= 0  )\n    X\n\nif dataset == 'human_liver':\n    f = '/kaggle/input/human-liver-rnaseq-gene-expression-903-samples/human_liver.tsv'\n    X = pd.read_csv(f ,  index_col= 0 , sep = '\\t' )\n    X = X.transpose()\n    X\n\nif dataset == 'GSE67123_v6_scrna_10x_143_23548_Mouse_Embryo_HSCs_invivo_fromCytotrace':\n    f = '/kaggle/input/genes-expressions-datasets-collection/GSE67123_v6_scrna_10x_143_23548_Mouse_Embryo_HSCs_invivo_fromCytotrace.csv'\n    X = pd.read_csv(f ,  index_col= 0 )#  , sep = '\\t' )\n    X\n    \nif dataset == 'single-cell-rna-seq-from-stoeckius-et-al-2017/GSE100866_CD8_merged-RNA_umi':\n    f = '/kaggle/input/single-cell-rna-seq-from-stoeckius-et-al-2017/GSE100866_CD8_merged-RNA_umi.csv'\n    X = pd.read_csv(f ,  index_col= 0 )#  , sep = '\\t' )\n    X = X.transpose()\n    X\n\ndataset = openml.datasets.get_dataset(1146)\nX, y, categorical_indicator, attribute_names = dataset.get_data(\n    dataset_format=\"dataframe\", target=dataset.default_target_attribute )\nX","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"from sklearn.decomposition import PCA\n\nt0=time.time()\nfig = plt.figure(figsize = (20,4) )\n\nX2 = PCA().fit_transform(X)\nplt.scatter(X2[:,0],X2[:,1]    )\nplt.title('PCA')\nplt.grid()\nplt.show()\n\nprint(time.time()-t0,'seconds passed')\n","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"# Function - degrees of KNN graph"},{"metadata":{"trusted":true},"cell_type":"code","source":"#distribution_type = 'Gauss' # 'uniform'\n\ndef get_vec_degs(X, n_neighbors = 5, method = 'sklearn', verbose = 1 ): \n    #verbose = 0\n    #method = 'sklearn'\n    #n_neighbors = 5\n\n    t00=time.time()\n\n    if method in ['GPU']:\n        X_cudf = cudf.DataFrame(X)\n        model = NearestNeighbors(n_neighbors=n_neighbors+1)\n        model.fit(X)\n        # get 3 nearest neighbors\n        m = model.kneighbors_graph(X_cudf)#, include_self=False) #  no include_self\n        vec_degs = np.sum(m,axis = 0 )\n        vec_degs = vec_degs.get()\n\n        vec_degs = np.squeeze(np.array(vec_degs.ravel()))\n        vec_degs -= np.ones_like(vec_degs ) # no include_self - so that is a way round\n\n    else: #if method in ['sklearn']:\n        m = kneighbors_graph(X, n_neighbors= n_neighbors, mode='connectivity' , include_self=False) #'distance' mode=  'connectivity'\n        vec_degs = np.sum(m,axis = 0 )\n        vec_degs = np.squeeze(np.array(vec_degs.ravel()))\n\n    if verbose >= 1:\n        print( np.round(time.time()-t00,1),'seconds passed total')    \n    \n    return vec_degs","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"# Historgrams for data only "},{"metadata":{"trusted":true},"cell_type":"code","source":"t00 = time.time()\ndf_stat = pd.DataFrame()\n\nfor n_neighbors in [5,10]:\n    list_vec_degs = []\n    list_info = []\n    vec_degs = get_vec_degs(X, n_neighbors = n_neighbors, method = 'sklearn', verbose = 1 )\n    list_vec_degs.append(vec_degs)\n    list_info.append('Data')\n\n    fig = plt.figure(figsize=(20,10))\n\n    fig.add_subplot(2,1,1)\n    c = 0\n    for vec_degs in list_vec_degs[:1]:\n        c += 1\n        t0 = time.time()\n\n        bins = np.arange(50)# np.max(vec_degs))\n        h = np.histogram(vec_degs,bins = bins, density= True)\n\n        label = ' n_neighbors'+str(n_neighbors) + ' ' + list_info[c-1]\n        plt.plot(h[0],'*-',label = label)\n\n        st = pd.Series(vec_degs).describe()\n        st.name = label\n        df_stat = df_stat.join( st, how = 'outer'  )\n    plt.title(' n_neighbors'+str(n_neighbors)  )\n    plt.legend()\n    plt.grid()\n\n\n    fig.add_subplot(2,1,2)\n    c = 0\n    for vec_degs in list_vec_degs[:1]:\n        c += 1\n        t0 = time.time()\n\n        bins = np.arange(50)# np.max(vec_degs))\n        h = np.histogram(vec_degs,bins = bins, density= True)\n\n        label = ' n_neighbors'+str(n_neighbors) + ' ' + list_info[c-1]\n        plt.loglog(h[0],'*-',label = label)\n    plt.title(' n_neighbors'+str(n_neighbors)  )\n    plt.legend()\n    plt.grid()\n\ndf_stat","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"# Plot histograms for data and benchamrks"},{"metadata":{"trusted":true},"cell_type":"code","source":"t00 = time.time()\ndf_stat = pd.DataFrame()\n\nfor n_neighbors in [5,10]:\n    \n    list_vec_degs = []\n    list_info = []\n    vec_degs = get_vec_degs(X, n_neighbors = n_neighbors, method = 'sklearn', verbose = 1 )\n    list_vec_degs.append(vec_degs)\n    list_info.append('Data')\n\n    X2 = np.random.randn(*X.shape)\n    vec_degs = get_vec_degs(X2, n_neighbors = n_neighbors, method = 'sklearn', verbose = 1 )\n    list_vec_degs.append(vec_degs)\n    list_info.append('Gauss')\n\n    X2 = np.random.rand(*X.shape)\n    vec_degs = get_vec_degs(X2, n_neighbors = n_neighbors, method = 'sklearn', verbose = 1 )\n    list_vec_degs.append(vec_degs)\n    list_info.append('Uniform')\n    \n\n    fig = plt.figure(figsize=(20,10))\n\n    fig.add_subplot(2,1,1)\n    c = 0\n    for vec_degs in list_vec_degs:\n        c += 1\n        t0 = time.time()\n\n        bins = np.arange(50)# np.max(vec_degs))\n        h = np.histogram(vec_degs,bins = bins, density= True)\n\n        label = ' n_neighbors'+str(n_neighbors) + ' ' + list_info[c-1]\n        plt.plot(h[0],'*-',label = label)\n\n        st = pd.Series(vec_degs).describe()\n        st.name = label\n        df_stat = df_stat.join( st, how = 'outer'  )\n    plt.title(' n_neighbors'+str(n_neighbors)  )\n    plt.legend()\n    plt.grid()\n\n\n    fig.add_subplot(2,1,2)\n    c = 0\n    for vec_degs in list_vec_degs:\n        c += 1\n        t0 = time.time()\n\n        bins = np.arange(50)# np.max(vec_degs))\n        h = np.histogram(vec_degs,bins = bins, density= True)\n\n        label = ' n_neighbors'+str(n_neighbors) + ' ' + list_info[c-1]\n        plt.loglog(h[0],'*-',label = label)\n    plt.title(' n_neighbors'+str(n_neighbors)  )\n    plt.legend()\n    plt.grid()\n\ndf_stat","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"# Further comparaison with Gauss and uniform of other dimension"},{"metadata":{"trusted":true},"cell_type":"code","source":" list_vec_degs = []\nlist_info = []\nvec_degs = get_vec_degs(X, n_neighbors = 5, method = 'sklearn', verbose = 1 )\nlist_vec_degs.append(vec_degs)\nlist_info.append('Data')\n\nX2 = np.random.randn(*X.shape)\nvec_degs = get_vec_degs(X2, n_neighbors = 5, method = 'sklearn', verbose = 1 )\nlist_vec_degs.append(vec_degs)\nlist_info.append('Gauss')\n\nX2 = np.random.rand(*X.shape)\nvec_degs = get_vec_degs(X2, n_neighbors = 5, method = 'sklearn', verbose = 1 )\nlist_vec_degs.append(vec_degs)\nlist_info.append('Uniform')\n\nX2 = np.random.randn(X.shape[0], int( X.shape[0]/4 ))\nvec_degs = get_vec_degs(X2, n_neighbors = 5, method = 'sklearn', verbose = 1 )\nlist_vec_degs.append(vec_degs)\nlist_info.append('Gauss dim->dim/4')\n\nX2 = np.random.rand(X.shape[0], int( X.shape[0]/4 ))\nvec_degs = get_vec_degs(X2, n_neighbors = 5, method = 'sklearn', verbose = 1 )\nlist_vec_degs.append(vec_degs)\nlist_info.append('Uniform dim->dim/4')\n\nX2.shape","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"t00 = time.time()\ndf_stat = pd.DataFrame()\n\nfor n_neighbors in [5]:\n\n    fig = plt.figure(figsize=(20,10))\n\n    #fig.add_subplot(2,1,1)\n    c = 0\n    for vec_degs in list_vec_degs:\n        c += 1\n        t0 = time.time()\n\n        bins = np.arange(50)# np.max(vec_degs))\n        h = np.histogram(vec_degs,bins = bins, density= True)\n\n        label = ' n_neighbors'+str(n_neighbors) + ' ' + list_info[c-1]\n        plt.plot(h[0],'*-',label = label)\n\n        st = pd.Series(vec_degs).describe()\n        st.name = label\n        df_stat = df_stat.join( st, how = 'outer'  )\n    plt.title(' n_neighbors'+str(n_neighbors)  )\n    plt.legend()\n    plt.grid()\n\n\n    fig = plt.figure(figsize=(20,10)) # fig.add_subplot(2,1,2)\n    c = 0\n    for vec_degs in list_vec_degs:\n        c += 1\n        t0 = time.time()\n\n        bins = np.arange(50)# np.max(vec_degs))\n        h = np.histogram(vec_degs,bins = bins, density= True)\n\n        label = ' n_neighbors'+str(n_neighbors) + ' ' + list_info[c-1]\n        plt.loglog(h[0],'*-',label = label)\n    plt.title(' n_neighbors'+str(n_neighbors)  )\n    plt.legend()\n    plt.grid()\n\ndf_stat","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"","execution_count":null,"outputs":[]}],"metadata":{"kernelspec":{"language":"python","display_name":"Python 3","name":"python3"},"language_info":{"pygments_lexer":"ipython3","nbconvert_exporter":"python","version":"3.6.4","file_extension":".py","codemirror_mode":{"name":"ipython","version":3},"name":"python","mimetype":"text/x-python"}},"nbformat":4,"nbformat_minor":4}