{"metadata":{"kernelspec":{"language":"python","display_name":"Python 3","name":"python3"},"language_info":{"pygments_lexer":"ipython3","nbconvert_exporter":"python","version":"3.6.4","file_extension":".py","codemirror_mode":{"name":"ipython","version":3},"name":"python","mimetype":"text/x-python"}},"nbformat_minor":4,"nbformat":4,"cells":[{"cell_type":"code","source":"# This Python 3 environment comes with many helpful analytics libraries installed\n# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python\n# For example, here's several helpful packages to load\n\nimport numpy as np # linear algebra\nimport pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)\n\n# Input data files are available in the read-only \"../input/\" directory\n# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory\n\nimport os\nfor dirname, _, filenames in os.walk('/kaggle/input'):\n    for filename in filenames:\n        print(os.path.join(dirname, filename))\n\n# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using \"Save & Run All\" \n# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session","metadata":{"_uuid":"8f2839f25d086af736a60e9eeb907d3b93b6e0e5","_cell_guid":"b1076dfc-b9ad-4769-8c92-a6c4dae69d19","execution":{"iopub.status.busy":"2021-07-31T20:15:23.944148Z","iopub.execute_input":"2021-07-31T20:15:23.944962Z","iopub.status.idle":"2021-07-31T20:15:23.977006Z","shell.execute_reply.started":"2021-07-31T20:15:23.944847Z","shell.execute_reply":"2021-07-31T20:15:23.976008Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"import time\nt0start = time.time()\nt0 = time.time()\n\n\n!pip install --user magic-impute\n\nimport magic\nimport scprep # scprep provides an all-in-one framework for loading, preprocessing, and plotting matrices in Python, with a focus on single-cell genomics.\n\n\nimport numpy as np\nimport pandas as pd\nimport matplotlib\nimport matplotlib.pyplot as plt\n\n# Matplotlib command for Jupyter notebooks only\n%matplotlib inline\n\n!pip install scanpy\nimport scanpy as sc\nimport anndata\n\nimport scipy \n\n\n\nimport pandas as pd\nimport numpy as np\nimport os\nimport sys\n\nprint(np.round(time.time()-t0,1) , 'seconds passed')\n","metadata":{"execution":{"iopub.status.busy":"2021-07-31T20:23:29.126983Z","iopub.execute_input":"2021-07-31T20:23:29.127404Z","iopub.status.idle":"2021-07-31T20:23:54.787572Z","shell.execute_reply.started":"2021-07-31T20:23:29.127369Z","shell.execute_reply":"2021-07-31T20:23:54.786555Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"import os\n# for dirname, _, filenames in os.walk('/kaggle/input'):\n#     for filename in filenames:\n#         print(os.path.join(dirname, filename))\npath_loc = '/kaggle/input/scrnaseq-bone-marrow-collection-of-datasets/'\nll = os.listdir(path_loc)\n\ndf_stat = pd.DataFrame()\nidx = 0\nfor fn in ll:\n    if fn[-4:] != 'h5ad': continue\n    str_data_inf = fn[:-5]\n    \n    l = str_data_inf.split('_')\n    df_stat.loc[idx,'GSE'] = l[0]\n    df_stat.loc[idx,'Organism'] = l[1]\n    df_stat.loc[idx,'Cells Info'] = l[2]\n    df_stat.loc[idx,'Cells SubInfo'] = l[3]\n    df_stat.loc[idx,'Year'] = int( l[4][1:] ); df_stat['Year'] = df_stat['Year'].astype(int)\n    df_stat.loc[idx,'n_cells'] = int( l[5][:-5] ); df_stat['n_cells'] = df_stat['n_cells'].astype(int)\n    df_stat.loc[idx,'Source'] = l[6]\n    df_stat.loc[idx,'Counts or Preprocessed Etc'] = l[7]\n    df_stat.loc[idx,'First Author'] = l[8]\n    df_stat.loc[idx,'Last Author'] = l[9]\n \n    t0 = time.time()\n    adata = sc.read(path_loc + fn )\n    df_stat.loc[idx,'Load Seconds'] = np.round(time.time()-t0,1)\n    \n    idx += 1\ndf_stat.sort_values('GSE')\n    ","metadata":{"execution":{"iopub.status.busy":"2021-07-31T20:45:30.930523Z","iopub.execute_input":"2021-07-31T20:45:30.930942Z","iopub.status.idle":"2021-07-31T20:45:45.838478Z","shell.execute_reply.started":"2021-07-31T20:45:30.930907Z","shell.execute_reply":"2021-07-31T20:45:45.837369Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"","metadata":{},"execution_count":null,"outputs":[]}]}