{"metadata":{"kernelspec":{"language":"python","display_name":"Python 3","name":"python3"},"language_info":{"pygments_lexer":"ipython3","nbconvert_exporter":"python","version":"3.6.4","file_extension":".py","codemirror_mode":{"name":"ipython","version":3},"name":"python","mimetype":"text/x-python"}},"nbformat_minor":4,"nbformat":4,"cells":[{"cell_type":"markdown","source":"# What is about ?\n\nHere we load data and make simple EDA. \n\nData from Nature 2020 paper: \nhttps://www.nature.com/articles/s41586-021-03232-9\n\"Spatiotemporal dissection of the cell cycle with single-cell proteogenomics\"\nEmma Lundberg et.al.\n\nRemarks on data:\nLoom file: 1152 cells Ã— 58884 genes expression matrix.  Stored in sparse matrix format. \n\n**PROBLEM WITH LOOM file** - we canNOT corectly merge it with csv file with information on cell cycle phases - \nthe file SraRunTable.txt  (the order of rows in files is different and no key to merge - conlusions made from visulization and compararaison with CSV file with count matrix).\n\nBetter to use CSV count matrix file - it does not have such problem. We will use it in next versions of the notebook.\n\n\n\n\n\n\n","metadata":{}},{"cell_type":"markdown","source":"# Import and install modules","metadata":{}},{"cell_type":"code","source":"# This Python 3 environment comes with many helpful analytics libraries installed\n# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python\n# For example, here's several helpful packages to load\n\nimport numpy as np # linear algebra\nimport pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)\n\n# Input data files are available in the read-only \"../input/\" directory\n# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory\n\nimport os\nfor dirname, _, filenames in os.walk('/kaggle/input'):\n    for filename in filenames:\n        print(os.path.join(dirname, filename))\n\n# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using \"Save & Run All\" \n# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session","metadata":{"_uuid":"8f2839f25d086af736a60e9eeb907d3b93b6e0e5","_cell_guid":"b1076dfc-b9ad-4769-8c92-a6c4dae69d19","execution":{"iopub.status.busy":"2021-06-19T18:06:03.170436Z","iopub.execute_input":"2021-06-19T18:06:03.170835Z","iopub.status.idle":"2021-06-19T18:06:03.182718Z","shell.execute_reply.started":"2021-06-19T18:06:03.170804Z","shell.execute_reply":"2021-06-19T18:06:03.181321Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"import time\nt0start = time.time()\n\nimport pandas as pd\nimport numpy as np\nimport os\nimport sys\n\nimport matplotlib.pyplot as plt\nimport matplotlib as mpl\nmpl.rcParams['figure.dpi'] = 70\nplt.style.use('dark_background')\n\nimport seaborn as sns\n\nfrom sklearn.decomposition import PCA","metadata":{"execution":{"iopub.status.busy":"2021-06-19T18:06:03.303412Z","iopub.execute_input":"2021-06-19T18:06:03.304144Z","iopub.status.idle":"2021-06-19T18:06:03.312614Z","shell.execute_reply.started":"2021-06-19T18:06:03.304087Z","shell.execute_reply":"2021-06-19T18:06:03.311275Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"!pip install scanpy\nimport scanpy as sc\nimport anndata\n\n!pip install loompy # scanpy needs it to load loom files \n","metadata":{"execution":{"iopub.status.busy":"2021-06-19T18:06:03.42454Z","iopub.execute_input":"2021-06-19T18:06:03.424991Z","iopub.status.idle":"2021-06-19T18:06:17.486172Z","shell.execute_reply.started":"2021-06-19T18:06:03.424954Z","shell.execute_reply":"2021-06-19T18:06:17.484764Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"","metadata":{"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# Load data","metadata":{}},{"cell_type":"code","source":"str_data_inf = ' Lundberg2020 u2os LoomFile' # Cell type u2os","metadata":{"execution":{"iopub.status.busy":"2021-06-19T18:06:17.488186Z","iopub.execute_input":"2021-06-19T18:06:17.488507Z","iopub.status.idle":"2021-06-19T18:06:17.496951Z","shell.execute_reply.started":"2021-06-19T18:06:17.488474Z","shell.execute_reply":"2021-06-19T18:06:17.495843Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"fn = '/kaggle/input/single-cell-rnaseq-data-related-to-cell-cycle/u2os.loom'\n\nimport time\nt0 = time.time()\nadata = sc.read_loom(fn)\nadata.obs['n.umi'] = np.asarray(adata.X.sum(axis = 1)).ravel()\nadata_orig = adata.copy()\nprint(np.round(time.time() - t0,1), 'seconds')\nadata","metadata":{"execution":{"iopub.status.busy":"2021-06-19T18:06:17.500219Z","iopub.execute_input":"2021-06-19T18:06:17.500592Z","iopub.status.idle":"2021-06-19T18:06:30.366921Z","shell.execute_reply.started":"2021-06-19T18:06:17.500558Z","shell.execute_reply":"2021-06-19T18:06:30.36589Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# Look on data","metadata":{}},{"cell_type":"code","source":"adata.var","metadata":{"execution":{"iopub.status.busy":"2021-06-19T18:06:30.368702Z","iopub.execute_input":"2021-06-19T18:06:30.369007Z","iopub.status.idle":"2021-06-19T18:06:30.390107Z","shell.execute_reply.started":"2021-06-19T18:06:30.368976Z","shell.execute_reply":"2021-06-19T18:06:30.388907Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"adata.obs","metadata":{"execution":{"iopub.status.busy":"2021-06-19T18:06:30.391604Z","iopub.execute_input":"2021-06-19T18:06:30.391972Z","iopub.status.idle":"2021-06-19T18:06:30.407116Z","shell.execute_reply.started":"2021-06-19T18:06:30.391937Z","shell.execute_reply":"2021-06-19T18:06:30.405754Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"adata.X","metadata":{"execution":{"iopub.status.busy":"2021-06-19T18:06:30.408496Z","iopub.execute_input":"2021-06-19T18:06:30.408834Z","iopub.status.idle":"2021-06-19T18:06:30.422311Z","shell.execute_reply.started":"2021-06-19T18:06:30.408804Z","shell.execute_reply":"2021-06-19T18:06:30.421096Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"adata.X[:15,:15].toarray()","metadata":{"execution":{"iopub.status.busy":"2021-06-19T18:06:30.424115Z","iopub.execute_input":"2021-06-19T18:06:30.424549Z","iopub.status.idle":"2021-06-19T18:06:30.440591Z","shell.execute_reply.started":"2021-06-19T18:06:30.424496Z","shell.execute_reply":"2021-06-19T18:06:30.439455Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"adata.layers","metadata":{"execution":{"iopub.status.busy":"2021-06-19T18:06:30.443332Z","iopub.execute_input":"2021-06-19T18:06:30.443744Z","iopub.status.idle":"2021-06-19T18:06:30.449538Z","shell.execute_reply.started":"2021-06-19T18:06:30.443703Z","shell.execute_reply":"2021-06-19T18:06:30.448694Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"adata.layers['matrix']","metadata":{"execution":{"iopub.status.busy":"2021-06-19T18:06:30.451526Z","iopub.execute_input":"2021-06-19T18:06:30.451964Z","iopub.status.idle":"2021-06-19T18:06:30.466828Z","shell.execute_reply.started":"2021-06-19T18:06:30.451928Z","shell.execute_reply":"2021-06-19T18:06:30.466017Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"adata.layers['matrix'][:15,:15].toarray()","metadata":{"execution":{"iopub.status.busy":"2021-06-19T18:06:30.468259Z","iopub.execute_input":"2021-06-19T18:06:30.468876Z","iopub.status.idle":"2021-06-19T18:06:30.485439Z","shell.execute_reply.started":"2021-06-19T18:06:30.46884Z","shell.execute_reply":"2021-06-19T18:06:30.484526Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"adata.layers['ambiguous']","metadata":{"execution":{"iopub.status.busy":"2021-06-19T18:06:30.487651Z","iopub.execute_input":"2021-06-19T18:06:30.487989Z","iopub.status.idle":"2021-06-19T18:06:30.501541Z","shell.execute_reply.started":"2021-06-19T18:06:30.487959Z","shell.execute_reply":"2021-06-19T18:06:30.500336Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"adata.layers['ambiguous'][:15,:15].toarray()","metadata":{"execution":{"iopub.status.busy":"2021-06-19T18:06:30.502826Z","iopub.execute_input":"2021-06-19T18:06:30.503208Z","iopub.status.idle":"2021-06-19T18:06:30.516889Z","shell.execute_reply.started":"2021-06-19T18:06:30.503172Z","shell.execute_reply":"2021-06-19T18:06:30.515773Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"adata.layers['spanning']","metadata":{"execution":{"iopub.status.busy":"2021-06-19T18:06:30.518293Z","iopub.execute_input":"2021-06-19T18:06:30.51883Z","iopub.status.idle":"2021-06-19T18:06:30.528611Z","shell.execute_reply.started":"2021-06-19T18:06:30.518697Z","shell.execute_reply":"2021-06-19T18:06:30.527117Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"adata.layers['spanning'][:15,:15].toarray()","metadata":{"execution":{"iopub.status.busy":"2021-06-19T18:06:30.529937Z","iopub.execute_input":"2021-06-19T18:06:30.530269Z","iopub.status.idle":"2021-06-19T18:06:30.547028Z","shell.execute_reply.started":"2021-06-19T18:06:30.53024Z","shell.execute_reply":"2021-06-19T18:06:30.5458Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"adata.layers['spliced']","metadata":{"execution":{"iopub.status.busy":"2021-06-19T18:06:30.548856Z","iopub.execute_input":"2021-06-19T18:06:30.549206Z","iopub.status.idle":"2021-06-19T18:06:30.56187Z","shell.execute_reply.started":"2021-06-19T18:06:30.549177Z","shell.execute_reply":"2021-06-19T18:06:30.560743Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"adata.layers['spliced'][:15,:15].toarray()","metadata":{"execution":{"iopub.status.busy":"2021-06-19T18:06:30.563731Z","iopub.execute_input":"2021-06-19T18:06:30.564397Z","iopub.status.idle":"2021-06-19T18:06:30.581145Z","shell.execute_reply.started":"2021-06-19T18:06:30.564351Z","shell.execute_reply":"2021-06-19T18:06:30.579752Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"adata.layers['unspliced']","metadata":{"execution":{"iopub.status.busy":"2021-06-19T18:06:30.582407Z","iopub.execute_input":"2021-06-19T18:06:30.582679Z","iopub.status.idle":"2021-06-19T18:06:30.591522Z","shell.execute_reply.started":"2021-06-19T18:06:30.582652Z","shell.execute_reply":"2021-06-19T18:06:30.590308Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"adata.layers['unspliced'][:15,:15].toarray()","metadata":{"execution":{"iopub.status.busy":"2021-06-19T18:06:30.593597Z","iopub.execute_input":"2021-06-19T18:06:30.594722Z","iopub.status.idle":"2021-06-19T18:06:30.606297Z","shell.execute_reply.started":"2021-06-19T18:06:30.594661Z","shell.execute_reply":"2021-06-19T18:06:30.605105Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# EDA","metadata":{}},{"cell_type":"markdown","source":"## Expressions  per cells","metadata":{}},{"cell_type":"code","source":"adata.obs['n.umi'] = np.asarray(adata.X.sum(axis = 1)).ravel()\n\nfig = plt.figure(figsize = (20,6)); c = 0\nc+=1; fig.add_subplot(1,2,c);\nplt.plot(np.sort(adata.obs['n.umi'])) \nplt.title('Expression per cell')\nplt.xlabel('cells sorted')\nplt.ylabel('Counts')\n\nc+=1; fig.add_subplot(1,2,c);\nplt.hist(np.sort(adata.obs['n.umi']), bins = 100) \nplt.title('Expression per cell')\n\nplt.show()\n\nadata.obs['n.umi'].describe()","metadata":{"execution":{"iopub.status.busy":"2021-06-19T18:06:30.607982Z","iopub.execute_input":"2021-06-19T18:06:30.608678Z","iopub.status.idle":"2021-06-19T18:06:31.106707Z","shell.execute_reply.started":"2021-06-19T18:06:30.60863Z","shell.execute_reply":"2021-06-19T18:06:31.105559Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"## Expression per gene","metadata":{}},{"cell_type":"code","source":"v = np.asarray(adata.X.sum(axis = 0)).ravel()\ndisplay(pd.Series(v).describe())\nv = np.log10(1+v)\nfig = plt.figure(figsize = (20,6)); c = 0\nc+=1; fig.add_subplot(1,2,c);\nplt.plot(np.sort(v)) \nplt.title('LOG10 Expression per gene')\nplt.xlabel('genes sorted')\nplt.ylabel('Log10 (1+Counts) ')\n\nc+=1; fig.add_subplot(1,2,c);\nplt.hist(np.sort(v), bins = 30) \nplt.title('LOG10 Expression per gene')\n\nplt.show()\n\npd.Series(v).describe()","metadata":{"execution":{"iopub.status.busy":"2021-06-19T18:06:31.108163Z","iopub.execute_input":"2021-06-19T18:06:31.108468Z","iopub.status.idle":"2021-06-19T18:06:31.514502Z","shell.execute_reply.started":"2021-06-19T18:06:31.108439Z","shell.execute_reply":"2021-06-19T18:06:31.513408Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"v = np.asarray(adata.X.sum(axis = 0)).ravel()\nadata.var['counts'] = v\nadata.var.sort_values('counts',ascending = False ).head(50) # ['counts']","metadata":{"execution":{"iopub.status.busy":"2021-06-19T18:06:31.51561Z","iopub.execute_input":"2021-06-19T18:06:31.515909Z","iopub.status.idle":"2021-06-19T18:06:31.582656Z","shell.execute_reply.started":"2021-06-19T18:06:31.515881Z","shell.execute_reply":"2021-06-19T18:06:31.581536Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# Visualizations ","metadata":{}},{"cell_type":"markdown","source":"## PCA (without preprocessing)","metadata":{}},{"cell_type":"code","source":"import scipy\nfrom sklearn.decomposition import TruncatedSVD\nfrom sklearn.decomposition import PCA\nimport time\nt0 = time.time()\nif scipy.sparse.issparse(adata.X):\n    reducer = TruncatedSVD(n_components=2, n_iter=7, random_state=42)\nelse:\n    reducer = PCA(n_components=2)\n    \nr = reducer.fit_transform(adata.X)\n\nfig = plt.figure(figsize = (25,12))\nplt.title(str_data_inf + ' PCA  n_cells: ' + str(adata.X.shape[0]) +\\\n            ' n_genes: ' + str(adata.X.shape[1])  + ' '  , fontsize = 20   )#' \nax = sns.scatterplot(x=r[:,0],y=r[:,1])# ,   hue= color_by)#, alpha = 0.7, marker = '.', legend = None)\nplt.xlabel('PCA1' , fontsize = 20 )\nplt.ylabel('PCA2' , fontsize = 20 )\nplt.show()\nprint( np.round( time.time() - t0,1) , ' seconds passed ' ) \n","metadata":{"execution":{"iopub.status.busy":"2021-06-19T18:06:31.584025Z","iopub.execute_input":"2021-06-19T18:06:31.584336Z","iopub.status.idle":"2021-06-19T18:06:34.559368Z","shell.execute_reply.started":"2021-06-19T18:06:31.584305Z","shell.execute_reply":"2021-06-19T18:06:34.558103Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"## UMAP (without preprocessing)","metadata":{}},{"cell_type":"code","source":"import scipy\nimport umap \nt0 = time.time()\nreducer = umap.UMAP()\nr = reducer.fit_transform(adata.X)\n\nfig = plt.figure(figsize = (25,12))\nplt.title(str_data_inf + ' UMAP n_cells: ' + str(adata.X.shape[0]) +\\\n            ' n_genes: ' + str(adata.X.shape[1])  + ' '  , fontsize = 20   )#' \nax = sns.scatterplot(x=r[:,0],y=r[:,1])# ,   hue= color_by)#, alpha = 0.7, marker = '.', legend = None)\nplt.xlabel('UMAP1' , fontsize = 20 )\nplt.ylabel('UMAP2' , fontsize = 20 )\nplt.show()\nprint( np.round( time.time() - t0,1) , ' seconds passed ' ) \n","metadata":{"execution":{"iopub.status.busy":"2021-06-19T18:06:34.564249Z","iopub.execute_input":"2021-06-19T18:06:34.5647Z","iopub.status.idle":"2021-06-19T18:06:56.663328Z","shell.execute_reply.started":"2021-06-19T18:06:34.564653Z","shell.execute_reply":"2021-06-19T18:06:56.662237Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"import scipy\nimport umap \nt0 = time.time()\n#reducer = umap.UMAP()\nreducer = umap.UMAP(n_neighbors = 250,min_dist = 0.9)\n\nr = reducer.fit_transform(adata.X)\n\nfig = plt.figure(figsize = (25,12))\nplt.title(str_data_inf + ' UMAP(250,0.9) n_cells: ' + str(adata.X.shape[0]) +\\\n            ' n_genes: ' + str(adata.X.shape[1])  + ' '  , fontsize = 20   )#' \nax = sns.scatterplot(x=r[:,0],y=r[:,1])# ,   hue= color_by)#, alpha = 0.7, marker = '.', legend = None)\nplt.xlabel('UMAP1' , fontsize = 20 )\nplt.ylabel('UMAP2' , fontsize = 20 )\nplt.show()\nprint( np.round( time.time() - t0,1) , ' seconds passed ' ) \n","metadata":{"execution":{"iopub.status.busy":"2021-06-19T18:06:56.665125Z","iopub.execute_input":"2021-06-19T18:06:56.665434Z","iopub.status.idle":"2021-06-19T18:07:20.022903Z","shell.execute_reply.started":"2021-06-19T18:06:56.665403Z","shell.execute_reply":"2021-06-19T18:07:20.021674Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"## PCA colored by genes","metadata":{}},{"cell_type":"code","source":"plot_mode = 'PCA'# 'Phase_plot':\nn_x_subplots = 3\ngenes_processing_mode = 'discretize_3bins_'# 'median_binarize_'\n\nmask = np.ones( adata.X.shape[0]).astype(bool) # \n\nimport scipy\nfrom sklearn.decomposition import TruncatedSVD\nfrom sklearn.decomposition import PCA\nimport time\nt0 = time.time()\nif scipy.sparse.issparse(adata.X):\n    reducer = TruncatedSVD(n_components=2, n_iter=7, random_state=42)\nelse:\n    reducer = PCA(n_components=2)\n    \nr = reducer.fit_transform(adata.X)\n\nc = 0\nfor color_by_mode in  ['n.umi',#'cell_cycle_phase', #  'PCA1ALL', 'n.umi', 'threshold_binarize_5_PCA1ALL',\n    'E2F1','FOXM1']:#,'PCNA','TOP2A', 'CCNE1', 'CCNE2', 'CDK2', 'CCNB1','CCNB2','CCNB3','CCNA2',  ]: # , 'pct_counts_mt']:\n    if c % n_x_subplots == 0:\n        fig = plt.figure(figsize = (20,5) ); c = 0\n        plt.suptitle(str_data_inf + ' PCA  n_cells: ' + str(adata.X.shape[0]) +\\\n            ' n_genes: ' + str(adata.X.shape[1])  + ' '  , fontsize = 20   )#' \n        c = 0\n    c += 1; fig.add_subplot(1,n_x_subplots ,c)\n    plt.title(str(color_by_mode), fontsize = 20)\n    \n    color_by_field_name = color_by_mode\n    if color_by_mode in adata.obs:\n        color_by = (adata[mask].obs[color_by_field_name])\n    elif color_by_mode in adata.var.index:\n        I_gene = np.where(adata.var.index == color_by_mode)[0]\n        v = adata[mask].X[:,I_gene]\n        if scipy.sparse.issparse(adata.X):\n            v = np.asarray(v.toarray()).ravel()\n        if genes_processing_mode == 'median_binarize_':\n            median_loc = np.median( v  )\n            color_by = np.asarray (v  > median_loc ).ravel()\n        elif genes_processing_mode == 'discretize_3bins_':\n            t1 = np.percentile(v,33)    \n            t2 = np.percentile(v,66)\n            color_by = np.asarray (v  > t1 ).ravel().astype(int) + np.asarray (v  > t2 ).ravel().astype(int)\n        else:\n            color_by = v\n    elif 'median_binarize_' in color_by_mode:\n        color_by_field_name = color_by_mode[16:]\n        if color_by_field_name in adata.obs:\n            color_by = (adata[mask].obs[color_by_field_name]) > np.median( (adata.obs[color_by_field_name]) )\n        else: color_by_mode = None\n    elif 'threshold_binarize_' in color_by_mode:\n        color_by_field_name = color_by_mode.split('_')[3]\n        threshold_binarize = float( color_by_mode.split('_')[2] )\n        if color_by_field_name in adata.obs:\n            color_by = (adata[mask].obs[color_by_field_name]) > threshold_binarize\n        else: color_by_mode = None\n    else:\n        color_by_mode = None\n\n    v1 = r[:,0]\n    v2 = r[:,1]\n    if color_by_mode is None:\n        ax = sns.scatterplot(x=v1, y = v2)# ,  hue= color_by,   alpha = 0.8, marker = '.')#, legend=None)\n    else:\n        if color_by_field_name == 'cell_cycle_phase':\n            if len(np.unique(color_by)) == 3:\n                color_palette = ['red', 'green','blue' ]\n            else:\n                color_palette = \"viridis\"# sns.color_palette(\"tab10\")\n        else:\n            color_palette = \"viridis\"# sns.color_palette(\"tab10\")        #color_by = (adata.obs[color_by_field_name]) > np.median( adata.obs[color_by_field_name].values ) \n        ax = sns.scatterplot(x=v1, y = v2,  hue= color_by, palette = color_palette )#, palette = \"viridis\")# sns.color_palette(\"viridis\", as_cmap=True),\n                            #)# ,   alpha = 0.8, marker = '.')#, )#, legend=None)\n        plt.setp(ax.get_legend().get_texts(), fontsize='20') # for legend text\n        plt.setp(ax.get_legend().get_title(), fontsize='20') # for legend title\n\n        \nplt.show()\nprint( np.round( time.time() - t0,1) , ' seconds passed ' ) \n","metadata":{"execution":{"iopub.status.busy":"2021-06-19T18:07:20.024704Z","iopub.execute_input":"2021-06-19T18:07:20.025134Z","iopub.status.idle":"2021-06-19T18:07:23.83744Z","shell.execute_reply.started":"2021-06-19T18:07:20.025078Z","shell.execute_reply":"2021-06-19T18:07:23.836138Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"## Different axes PCA colored by genes ","metadata":{}},{"cell_type":"code","source":"plot_mode = 'PCA'# 'Phase_plot':\nn_x_subplots = 3\ngenes_processing_mode = 'discretize_3bins_'# 'median_binarize_'\n\nmask = np.ones( adata.X.shape[0]).astype(bool) # \n\nimport scipy\nfrom sklearn.decomposition import TruncatedSVD\nfrom sklearn.decomposition import PCA\nimport time\nt0 = time.time()\nif scipy.sparse.issparse(adata.X):\n    reducer = TruncatedSVD(n_components=10, n_iter=7, random_state=42)\nelse:\n    reducer = PCA(n_components=10)\n    \nr = reducer.fit_transform(adata.X)\n\nc = 0\nfor i,j in [(0,1),(2,3),(3,4),(5,6)]:#,(1,3),(2,3),(3,4)]:\n    for color_by_mode in  ['n.umi',#'cell_cycle_phase', #  'PCA1ALL', 'n.umi', 'threshold_binarize_5_PCA1ALL',\n        'E2F1','FOXM1']:#,'PCNA','TOP2A', 'CCNE1', 'CCNE2', 'CDK2', 'CCNB1','CCNB2','CCNB3','CCNA2',  ]: # , 'pct_counts_mt']:\n        if c % n_x_subplots == 0:\n            fig = plt.figure(figsize = (20,5) ); c = 0\n            plt.suptitle(str_data_inf + ' PCA  n_cells: ' + str(adata.X.shape[0]) +\\\n                ' n_genes: ' + str(adata.X.shape[1])  + ' '  , fontsize = 20   )#' \n            c = 0\n        c += 1; fig.add_subplot(1,n_x_subplots ,c)\n        plt.title(str(color_by_mode) + ' PCA '+str(i)+','+str(j), fontsize = 20)\n\n        color_by_field_name = color_by_mode\n        if color_by_mode in adata.obs:\n            color_by = (adata[mask].obs[color_by_field_name])\n        elif color_by_mode in adata.var.index:\n            I_gene = np.where(adata.var.index == color_by_mode)[0]\n            v = adata[mask].X[:,I_gene]\n            if scipy.sparse.issparse(adata.X):\n                v = np.asarray(v.toarray()).ravel()\n            if genes_processing_mode == 'median_binarize_':\n                median_loc = np.median( v  )\n                color_by = np.asarray (v  > median_loc ).ravel()\n            elif genes_processing_mode == 'discretize_3bins_':\n                t1 = np.percentile(v,33)    \n                t2 = np.percentile(v,66)\n                color_by = np.asarray (v  > t1 ).ravel().astype(int) + np.asarray (v  > t2 ).ravel().astype(int)\n            else:\n                color_by = v\n        elif 'median_binarize_' in color_by_mode:\n            color_by_field_name = color_by_mode[16:]\n            if color_by_field_name in adata.obs:\n                color_by = (adata[mask].obs[color_by_field_name]) > np.median( (adata.obs[color_by_field_name]) )\n            else: color_by_mode = None\n        elif 'threshold_binarize_' in color_by_mode:\n            color_by_field_name = color_by_mode.split('_')[3]\n            threshold_binarize = float( color_by_mode.split('_')[2] )\n            if color_by_field_name in adata.obs:\n                color_by = (adata[mask].obs[color_by_field_name]) > threshold_binarize\n            else: color_by_mode = None\n        else:\n            color_by_mode = None\n\n        v1 = r[:,i]\n        v2 = r[:,j]\n        if color_by_mode is None:\n            ax = sns.scatterplot(x=v1, y = v2)# ,  hue= color_by,   alpha = 0.8, marker = '.')#, legend=None)\n        else:\n            if color_by_field_name == 'cell_cycle_phase':\n                if len(np.unique(color_by)) == 3:\n                    color_palette = ['red', 'green','blue' ]\n                else:\n                    color_palette = \"viridis\"# sns.color_palette(\"tab10\")\n            else:\n                color_palette = \"viridis\"# sns.color_palette(\"tab10\")        #color_by = (adata.obs[color_by_field_name]) > np.median( adata.obs[color_by_field_name].values ) \n            ax = sns.scatterplot(x=v1, y = v2,  hue= color_by, palette = color_palette )#, palette = \"viridis\")# sns.color_palette(\"viridis\", as_cmap=True),\n                                #)# ,   alpha = 0.8, marker = '.')#, )#, legend=None)\n            plt.setp(ax.get_legend().get_texts(), fontsize='20') # for legend text\n            plt.setp(ax.get_legend().get_title(), fontsize='20') # for legend title\n\n        \nplt.show()\nprint( np.round( time.time() - t0,1) , ' seconds passed ' ) \n","metadata":{"execution":{"iopub.status.busy":"2021-06-19T18:07:23.839189Z","iopub.execute_input":"2021-06-19T18:07:23.839618Z","iopub.status.idle":"2021-06-19T18:07:32.012347Z","shell.execute_reply.started":"2021-06-19T18:07:23.839572Z","shell.execute_reply":"2021-06-19T18:07:32.011269Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"## Plots in genes axes","metadata":{}},{"cell_type":"code","source":"\nmask = np.ones( adata.X.shape[0]).astype(bool) # \n\ngene1 = 'E2F1' \ngene2 = 'FOXM1'\n\ncolor_by_mode = 'n.umi' # ,#'cell_cycle_phase', #  'PCA1ALL', 'n.umi', 'threshold_binarize_5_PCA1ALL',\nif color_by_mode in adata.obs:\n    color_by_field_name = color_by_mode\n    color_by = (adata[mask].obs[color_by_field_name])\nelse:\n    color_by_mode = None\n\n\n\nI_gene = np.where(adata.var.index == gene1)[0]\nv = adata[mask].X[:,I_gene]\nif scipy.sparse.issparse(adata.X):\n    v = np.asarray(v.toarray()).ravel()\nv1 = v\nI_gene = np.where(adata.var.index == gene2)[0]\nv = adata[mask].X[:,I_gene]\nif scipy.sparse.issparse(adata.X):\n    v = np.asarray(v.toarray()).ravel()\nv2 = v\n\nfig = plt.figure(figsize = (20,10) ); c = 0\nplt.title(str_data_inf + ' Genes axes   n_cells: ' + str(adata.X.shape[0]) +\\\n    ' n_genes: ' + str(adata.X.shape[1])  + ' '  , fontsize = 20   )#' \n\nif color_by_mode is None:\n    ax = sns.scatterplot(x=v1,y=v2)# ,   hue= color_by)#, alpha = 0.7, marker = '.', legend = None)\nelse:\n    ax = sns.scatterplot(x=v1,y=v2,   hue= color_by)#, alpha = 0.7, marker = '.', legend = None)\n    plt.setp(ax.get_legend().get_texts(), fontsize='15') # for legend text\n    plt.setp(ax.get_legend().get_title(), fontsize='15') # for legend title    plt.xlabel('PCA1' , fontsize = 20 )\n    \nplt.xlabel(gene1,  fontsize = 20)\nplt.ylabel(gene2,  fontsize = 20)\nplt.show()   \n    \n","metadata":{"execution":{"iopub.status.busy":"2021-06-19T18:07:32.013727Z","iopub.execute_input":"2021-06-19T18:07:32.014021Z","iopub.status.idle":"2021-06-19T18:07:32.553361Z","shell.execute_reply.started":"2021-06-19T18:07:32.013992Z","shell.execute_reply":"2021-06-19T18:07:32.552511Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"mask = np.ones( adata.X.shape[0]).astype(bool) # \n\ngene1 = 'E2F1' \ngene2 = 'FOXM1'\nn_x_subplots = 4\n\nc = 0 \nfor gene1,gene2 in [ ('E2F1','FOXM1'), ('PCNA','TOP2A'), ('CCNB1', 'CCNB2'), ('CCNE1', 'CCNB1'),\n                   ('E2F1','PCNA'),('FOXM1','PCNA'), ('E2F1','TOP2A'),('FOXM1','TOP2A'), ]: \n    #,'PCNA','TOP2A', 'CCNE1', 'CCNE2', 'CDK2', 'CCNB1','CCNB2','CCNB3','CCNA2',  ]: # , 'pct_counts_mt']:\n    color_by_mode = 'n.umi' # ,#'cell_cycle_phase', #  'PCA1ALL', 'n.umi', 'threshold_binarize_5_PCA1ALL',\n    if color_by_mode in adata.obs:\n        color_by_field_name = color_by_mode\n        color_by = (adata[mask].obs[color_by_field_name])\n    else:\n        color_by_mode = None\n    \n    if c % n_x_subplots == 0:\n        fig = plt.figure(figsize = (20,6) ); c = 0\n        plt.suptitle(str_data_inf + ' Genes axes   n_cells: ' + str(adata.X.shape[0]) +\\\n            ' n_genes: ' + str(adata.X.shape[1])  + ' color by '+ str(color_by_mode) , fontsize = 20   )#' \n        c = 0\n        \n    c += 1; fig.add_subplot(1,n_x_subplots ,c)\n    plt.title(gene1 + ' ' + gene2) # str(color_by_mode) + ' PCA '+str(i)+','+str(j))\n\n    plt.legend()\n    I_gene = np.where(adata.var.index == gene1)[0]\n    v = adata[mask].X[:,I_gene]\n    if scipy.sparse.issparse(adata.X):\n        v = np.asarray(v.toarray()).ravel()\n    v1 = v\n    I_gene = np.where(adata.var.index == gene2)[0]\n    v = adata[mask].X[:,I_gene]\n    if scipy.sparse.issparse(adata.X):\n        v = np.asarray(v.toarray()).ravel()\n    v2 = v\n\n\n    if color_by_mode is None:\n        ax = sns.scatterplot(x=v1,y=v2)# ,   hue= color_by)#, alpha = 0.7, marker = '.', legend = None)\n    else:\n        ax = sns.scatterplot(x=v1,y=v2,   hue= color_by)#, alpha = 0.7, marker = '.', legend = None)\n        plt.setp(ax.get_legend().get_texts(), fontsize='15') # for legend text\n        plt.setp(ax.get_legend().get_title(), fontsize='15') # for legend title    plt.xlabel('PCA1' , fontsize = 20 )\n        ax.legend().set_visible(False)\n\n    plt.xlabel(gene1,  fontsize = 20)\n    plt.ylabel(gene2,  fontsize = 20)\nplt.show()  ","metadata":{"execution":{"iopub.status.busy":"2021-06-19T18:07:32.554578Z","iopub.execute_input":"2021-06-19T18:07:32.554878Z","iopub.status.idle":"2021-06-19T18:07:35.474849Z","shell.execute_reply.started":"2021-06-19T18:07:32.55485Z","shell.execute_reply":"2021-06-19T18:07:35.473554Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"","metadata":{"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# Preprocessing - normalizations / log / filtering ","metadata":{}},{"cell_type":"code","source":"adata_orig.var_names_make_unique()\nadata_orig.var_names_make_unique()","metadata":{"execution":{"iopub.status.busy":"2021-06-19T18:07:35.476473Z","iopub.execute_input":"2021-06-19T18:07:35.476796Z","iopub.status.idle":"2021-06-19T18:07:35.537786Z","shell.execute_reply.started":"2021-06-19T18:07:35.476765Z","shell.execute_reply":"2021-06-19T18:07:35.536564Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"# First standard preprocessing\n\nlist_genes2include_mandotory = [ 'E2F1','FOXM1'] # can specify genes which we will keep even if they are not top variable\n# Params: \nn_top_genes_to_keep = 10000\nthreshold_pct_counts_mt = 32\nmin_count = 150_000\nmax_count = 420_000# 12000\nif 'LoomFile' not in str_data_inf:\n    threshold_pct_counts_mt = 25\n    min_count = 300_000\n    max_count = 700_000# 12000\n    \n\nprint(adata_orig.X.sum())\n\n# ################################################################################################\n# Preprocessing first step: filter CELLs by  counts and level of MT-percent\n# thresholds are set visually looking on violin plots \n# Examples:\n# threshold_pct_counts_mt = 20 - 40  min_count = 500 - 1000; max_count = 10000 - 12000; \n# These thresholds depends on cell line dataset\n\n\nadata = adata_orig.copy()\nprint(adata)\n\n# Calculate and plot statistics of counts per cell, percent of mitochondrial genes (high percent is bad for cell)\nadata.var['mt'] = adata.var_names.str.startswith('MT-')  # annotate the group of mitochondrial genes as 'mt'\n#sv.pp.remove_duplicate_cells(adata)\nsc.pp.calculate_qc_metrics(adata, qc_vars=['mt'], percent_top=None, log1p=False, inplace=True)\nsc.pl.violin(adata, ['n_genes_by_counts', 'n.umi', 'total_counts', 'pct_counts_mt'],jitter=1.9, multi_panel=True)\nmedian_count = np.median(adata.obs['total_counts'])\nprint('Median total counts =',median_count)\n# min_count = np.max((median_count/2,5000))\nprint('min_count=',min_count,'max_count=',max_count)\n\nprint('Look at total_counts va MT-percent, expect some linear dependence - but does not happen: ')\nsc.pl.scatter(adata, x='total_counts', y='pct_counts_mt')\n\n# Filtering out cells which beyond thresholds\ninds1 = np.where((adata.obs['total_counts']>min_count) & (adata.obs['total_counts']<max_count))\ninds2 = np.where(adata.obs['pct_counts_mt']<threshold_pct_counts_mt)\nprint(len(inds1[0]),'samples pass the count filter')\nprint(len(inds2[0]),' samples pass the mt filter')\nind_samples = np.intersect1d(inds1[0],inds2[0])\nprint('Samples selected',len(ind_samples))\nadata.uns['ind_samples'] = ind_samples\n\n# Here we cut cells. Filtering out those with counts too low or too big\nadata = adata[ind_samples,:]\n\n# ################################################################################################\n# Preprocessing second step: \n# 1) normalization to some value, i.e. median of the total counts\n# 2) taking logs\n# 3) keeping only higly variable genes\n\nsc.pp.normalize_total(adata, target_sum=np.median(adata.obs[\"total_counts\"]))\nsc.pp.log1p(adata)\nsc.pp.highly_variable_genes(adata,n_top_genes=n_top_genes_to_keep,n_bins=20)\nind_genes = np.where(adata.var['highly_variable'])[0]\nind_genes2 = np.where(adata.var.index.isin(list_genes2include_mandotory ) )[0]\nind_genes = list( set( ind_genes )|set( ind_genes2 ) )\nadata = adata[:,ind_genes]\n\nprint('Violin plots after filtering cells and genes')\nsc.pl.violin(adata, ['n_genes_by_counts','n.umi', 'total_counts', 'pct_counts_mt'],jitter=1.9, multi_panel=True)\n","metadata":{"execution":{"iopub.status.busy":"2021-06-19T18:07:35.539461Z","iopub.execute_input":"2021-06-19T18:07:35.539775Z","iopub.status.idle":"2021-06-19T18:07:39.731567Z","shell.execute_reply.started":"2021-06-19T18:07:35.539745Z","shell.execute_reply":"2021-06-19T18:07:39.730144Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# Visualizations after preprocessing","metadata":{}},{"cell_type":"markdown","source":"## Various PCA","metadata":{}},{"cell_type":"code","source":"import scipy\nfrom sklearn.decomposition import TruncatedSVD\nfrom sklearn.decomposition import PCA\nimport time\nt0 = time.time()\nif scipy.sparse.issparse(adata.X):\n    reducer = TruncatedSVD(n_components=2, n_iter=7, random_state=42)\nelse:\n    reducer = PCA(n_components=2)\n    \nr = reducer.fit_transform(adata.X)\n\nfig = plt.figure(figsize = (25,12))\nplt.title(str_data_inf + ' PCA  n_cells: ' + str(adata.X.shape[0]) +\\\n            ' n_genes: ' + str(adata.X.shape[1])  + ' '  , fontsize = 20   )#' \nax = sns.scatterplot(x=r[:,0],y=r[:,1])# ,   hue= color_by)#, alpha = 0.7, marker = '.', legend = None)\nplt.xlabel('PCA1' , fontsize = 20 )\nplt.ylabel('PCA2' , fontsize = 20 )\nplt.show()\nprint( np.round( time.time() - t0,1) , ' seconds passed ' ) \n\n# ################################################################################################################\n# PCA colored by genes n.umi etc\n# ################################################################################################################\n\nplot_mode = 'PCA'# 'Phase_plot':\nn_x_subplots = 3\ngenes_processing_mode = 'discretize_3bins_'# 'median_binarize_'\n\nmask = np.ones( adata.X.shape[0]).astype(bool) # \n\nimport scipy\nfrom sklearn.decomposition import TruncatedSVD\nfrom sklearn.decomposition import PCA\nimport time\nt0 = time.time()\nif scipy.sparse.issparse(adata.X):\n    reducer = TruncatedSVD(n_components=2, n_iter=7, random_state=42)\nelse:\n    reducer = PCA(n_components=2)\n    \nr = reducer.fit_transform(adata.X)\n\nc = 0\nfor color_by_mode in  ['n.umi',#'cell_cycle_phase', #  'PCA1ALL', 'n.umi', 'threshold_binarize_5_PCA1ALL',\n    'E2F1','FOXM1']:#,'PCNA','TOP2A', 'CCNE1', 'CCNE2', 'CDK2', 'CCNB1','CCNB2','CCNB3','CCNA2',  ]: # , 'pct_counts_mt']:\n    if c % n_x_subplots == 0:\n        fig = plt.figure(figsize = (20,5) ); c = 0\n        plt.suptitle(str_data_inf + ' PCA  n_cells: ' + str(adata.X.shape[0]) +\\\n            ' n_genes: ' + str(adata.X.shape[1])  + ' '  , fontsize = 20   )#' \n        c = 0\n    c += 1; fig.add_subplot(1,n_x_subplots ,c)\n    plt.title(str(color_by_mode), fontsize = 20)\n    \n    color_by_field_name = color_by_mode\n    if color_by_mode in adata.obs:\n        color_by = (adata[mask].obs[color_by_field_name])\n    elif color_by_mode in adata.var.index:\n        I_gene = np.where(adata.var.index == color_by_mode)[0]\n        v = adata[mask].X[:,I_gene]\n        if scipy.sparse.issparse(adata.X):\n            v = np.asarray(v.toarray()).ravel()\n        if genes_processing_mode == 'median_binarize_':\n            median_loc = np.median( v  )\n            color_by = np.asarray (v  > median_loc ).ravel()\n        elif genes_processing_mode == 'discretize_3bins_':\n            t1 = np.percentile(v,33)    \n            t2 = np.percentile(v,66)\n            color_by = np.asarray (v  > t1 ).ravel().astype(int) + np.asarray (v  > t2 ).ravel().astype(int)\n        else:\n            color_by = v\n    elif 'median_binarize_' in color_by_mode:\n        color_by_field_name = color_by_mode[16:]\n        if color_by_field_name in adata.obs:\n            color_by = (adata[mask].obs[color_by_field_name]) > np.median( (adata.obs[color_by_field_name]) )\n        else: color_by_mode = None\n    elif 'threshold_binarize_' in color_by_mode:\n        color_by_field_name = color_by_mode.split('_')[3]\n        threshold_binarize = float( color_by_mode.split('_')[2] )\n        if color_by_field_name in adata.obs:\n            color_by = (adata[mask].obs[color_by_field_name]) > threshold_binarize\n        else: color_by_mode = None\n    else:\n        color_by_mode = None\n\n    v1 = r[:,0]\n    v2 = r[:,1]\n    if color_by_mode is None:\n        ax = sns.scatterplot(x=v1, y = v2)# ,  hue= color_by,   alpha = 0.8, marker = '.')#, legend=None)\n    else:\n        if color_by_field_name == 'cell_cycle_phase':\n            if len(np.unique(color_by)) == 3:\n                color_palette = ['red', 'green','blue' ]\n            else:\n                color_palette = \"viridis\"# sns.color_palette(\"tab10\")\n        else:\n            color_palette = \"viridis\"# sns.color_palette(\"tab10\")        #color_by = (adata.obs[color_by_field_name]) > np.median( adata.obs[color_by_field_name].values ) \n        ax = sns.scatterplot(x=v1, y = v2,  hue= color_by, palette = color_palette )#, palette = \"viridis\")# sns.color_palette(\"viridis\", as_cmap=True),\n                            #)# ,   alpha = 0.8, marker = '.')#, )#, legend=None)\n        plt.setp(ax.get_legend().get_texts(), fontsize='20') # for legend text\n        plt.setp(ax.get_legend().get_title(), fontsize='20') # for legend title\n\n        \nplt.show()\nprint( np.round( time.time() - t0,1) , ' seconds passed ' ) \n\n# ################################################################################################################\n# PCA different axes \n# ################################################################################################################\n\nplot_mode = 'PCA'# 'Phase_plot':\nn_x_subplots = 3\ngenes_processing_mode = 'discretize_3bins_'# 'median_binarize_'\n\nmask = np.ones( adata.X.shape[0]).astype(bool) # \n\nimport scipy\nfrom sklearn.decomposition import TruncatedSVD\nfrom sklearn.decomposition import PCA\nimport time\nt0 = time.time()\nif scipy.sparse.issparse(adata.X):\n    reducer = TruncatedSVD(n_components=10, n_iter=7, random_state=42)\nelse:\n    reducer = PCA(n_components=10)\n    \nr = reducer.fit_transform(adata.X)\n\nc = 0\nfor i,j in [(2,3),(3,4),(5,6)]:#(0,1),(1,3),(2,3),(3,4)]:\n    for color_by_mode in  ['n.umi',#'cell_cycle_phase', #  'PCA1ALL', 'n.umi', 'threshold_binarize_5_PCA1ALL',\n        'E2F1','FOXM1']:#,'PCNA','TOP2A', 'CCNE1', 'CCNE2', 'CDK2', 'CCNB1','CCNB2','CCNB3','CCNA2',  ]: # , 'pct_counts_mt']:\n        if c % n_x_subplots == 0:\n            fig = plt.figure(figsize = (20,5) ); c = 0\n            plt.suptitle(str_data_inf + ' PCA  n_cells: ' + str(adata.X.shape[0]) +\\\n                ' n_genes: ' + str(adata.X.shape[1])  + ' '  , fontsize = 20   )#' \n            c = 0\n        c += 1; fig.add_subplot(1,n_x_subplots ,c)\n        plt.title(str(color_by_mode) + ' PCA '+str(i)+','+str(j), fontsize = 20)\n\n        color_by_field_name = color_by_mode\n        if color_by_mode in adata.obs:\n            color_by = (adata[mask].obs[color_by_field_name])\n        elif color_by_mode in adata.var.index:\n            I_gene = np.where(adata.var.index == color_by_mode)[0]\n            v = adata[mask].X[:,I_gene]\n            if scipy.sparse.issparse(adata.X):\n                v = np.asarray(v.toarray()).ravel()\n            if genes_processing_mode == 'median_binarize_':\n                median_loc = np.median( v  )\n                color_by = np.asarray (v  > median_loc ).ravel()\n            elif genes_processing_mode == 'discretize_3bins_':\n                t1 = np.percentile(v,33)    \n                t2 = np.percentile(v,66)\n                color_by = np.asarray (v  > t1 ).ravel().astype(int) + np.asarray (v  > t2 ).ravel().astype(int)\n            else:\n                color_by = v\n        elif 'median_binarize_' in color_by_mode:\n            color_by_field_name = color_by_mode[16:]\n            if color_by_field_name in adata.obs:\n                color_by = (adata[mask].obs[color_by_field_name]) > np.median( (adata.obs[color_by_field_name]) )\n            else: color_by_mode = None\n        elif 'threshold_binarize_' in color_by_mode:\n            color_by_field_name = color_by_mode.split('_')[3]\n            threshold_binarize = float( color_by_mode.split('_')[2] )\n            if color_by_field_name in adata.obs:\n                color_by = (adata[mask].obs[color_by_field_name]) > threshold_binarize\n            else: color_by_mode = None\n        else:\n            color_by_mode = None\n\n        v1 = r[:,i]\n        v2 = r[:,j]\n        if color_by_mode is None:\n            ax = sns.scatterplot(x=v1, y = v2)# ,  hue= color_by,   alpha = 0.8, marker = '.')#, legend=None)\n        else:\n            if color_by_field_name == 'cell_cycle_phase':\n                if len(np.unique(color_by)) == 3:\n                    color_palette = ['red', 'green','blue' ]\n                else:\n                    color_palette = \"viridis\"# sns.color_palette(\"tab10\")\n            else:\n                color_palette = \"viridis\"# sns.color_palette(\"tab10\")        #color_by = (adata.obs[color_by_field_name]) > np.median( adata.obs[color_by_field_name].values ) \n            ax = sns.scatterplot(x=v1, y = v2,  hue= color_by, palette = color_palette )#, palette = \"viridis\")# sns.color_palette(\"viridis\", as_cmap=True),\n                                #)# ,   alpha = 0.8, marker = '.')#, )#, legend=None)\n            plt.setp(ax.get_legend().get_texts(), fontsize='20') # for legend text\n            plt.setp(ax.get_legend().get_title(), fontsize='20') # for legend title\n\n        \nplt.show()\nprint( np.round( time.time() - t0,1) , ' seconds passed ' ) ","metadata":{"execution":{"iopub.status.busy":"2021-06-19T18:07:39.733526Z","iopub.execute_input":"2021-06-19T18:07:39.734005Z","iopub.status.idle":"2021-06-19T18:07:50.145409Z","shell.execute_reply.started":"2021-06-19T18:07:39.733956Z","shell.execute_reply":"2021-06-19T18:07:50.14424Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"## UMAP (with several param choices)","metadata":{}},{"cell_type":"code","source":"import scipy\nimport umap \nt0 = time.time()\n#reducer = umap.UMAP(n_neighbors = 250,min_dist = 0.9)\nreducer = umap.UMAP()\n\nr = reducer.fit_transform(adata.X)\n\nfig = plt.figure(figsize = (25,12))\nplt.title(str_data_inf + ' UMAP n_cells: ' + str(adata.X.shape[0]) +\\\n            ' n_genes: ' + str(adata.X.shape[1])  + ' '  , fontsize = 20   )#' \nax = sns.scatterplot(x=r[:,0],y=r[:,1])# ,   hue= color_by)#, alpha = 0.7, marker = '.', legend = None)\nplt.xlabel('UMAP1' , fontsize = 20 )\nplt.ylabel('UMAP2' , fontsize = 20 )\nplt.show()\nprint( np.round( time.time() - t0,1) , ' seconds passed ' ) \n\n############################################################################\n# UMAP change params \n############################################################################\nprint('change umap params')\n\nimport scipy\nimport umap \nt0 = time.time()\n#reducer = umap.UMAP()\nreducer = umap.UMAP(n_neighbors = 250,min_dist = 0.9)\n\nr = reducer.fit_transform(adata.X)\n\nfig = plt.figure(figsize = (25,12))\nplt.title(str_data_inf + ' UMAP(250,0.9) n_cells: ' + str(adata.X.shape[0]) +\\\n            ' n_genes: ' + str(adata.X.shape[1])  + ' '  , fontsize = 20   )#' \nax = sns.scatterplot(x=r[:,0],y=r[:,1])# ,   hue= color_by)#, alpha = 0.7, marker = '.', legend = None)\nplt.xlabel('UMAP1' , fontsize = 20 )\nplt.ylabel('UMAP2' , fontsize = 20 )\nplt.show()\nprint( np.round( time.time() - t0,1) , ' seconds passed ' ) \n\n############################################################################\n# UMAP change params \n############################################################################\nprint('change umap params')\n\nimport scipy\nimport umap \nt0 = time.time()\n#reducer = umap.UMAP()\nreducer = umap.UMAP(n_neighbors = 20,min_dist = 0.2)\n\nr = reducer.fit_transform(adata.X)\n\nfig = plt.figure(figsize = (25,12))\nplt.title(str_data_inf + ' UMAP(250,0.9) n_cells: ' + str(adata.X.shape[0]) +\\\n            ' n_genes: ' + str(adata.X.shape[1])  + ' '  , fontsize = 20   )#' \nax = sns.scatterplot(x=r[:,0],y=r[:,1])# ,   hue= color_by)#, alpha = 0.7, marker = '.', legend = None)\nplt.xlabel('UMAP1' , fontsize = 20 )\nplt.ylabel('UMAP2' , fontsize = 20 )\nplt.show()\nprint( np.round( time.time() - t0,1) , ' seconds passed ' ) ","metadata":{"execution":{"iopub.status.busy":"2021-06-19T18:07:50.146631Z","iopub.execute_input":"2021-06-19T18:07:50.146932Z","iopub.status.idle":"2021-06-19T18:09:00.381308Z","shell.execute_reply.started":"2021-06-19T18:07:50.146903Z","shell.execute_reply":"2021-06-19T18:09:00.380437Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"import scipy\nfrom sklearn.decomposition import TruncatedSVD\nfrom sklearn.decomposition import PCA\nimport time\nt0 = time.time()\nreducer = umap.UMAP()    \nr = reducer.fit_transform(adata.X)\n\nc = 0\ni,j = 0,1\nfor color_by_mode in  ['n.umi',#'cell_cycle_phase', #  'PCA1ALL', 'n.umi', 'threshold_binarize_5_PCA1ALL',\n    'E2F1','FOXM1']:#,'PCNA','TOP2A', 'CCNE1', 'CCNE2', 'CDK2', 'CCNB1','CCNB2','CCNB3','CCNA2',  ]: # , 'pct_counts_mt']:\n    if c % n_x_subplots == 0:\n        fig = plt.figure(figsize = (20,5) ); c = 0\n        plt.suptitle(str_data_inf + ' PCA  n_cells: ' + str(adata.X.shape[0]) +\\\n            ' n_genes: ' + str(adata.X.shape[1])  + ' '  , fontsize = 20   )#' \n        c = 0\n    c += 1; fig.add_subplot(1,n_x_subplots ,c)\n    plt.title(str(color_by_mode) + ' UMAP ', fontsize = 20)\n\n    color_by_field_name = color_by_mode\n    if color_by_mode in adata.obs:\n        color_by = (adata[mask].obs[color_by_field_name])\n    elif color_by_mode in adata.var.index:\n        I_gene = np.where(adata.var.index == color_by_mode)[0]\n        v = adata[mask].X[:,I_gene]\n        if scipy.sparse.issparse(adata.X):\n            v = np.asarray(v.toarray()).ravel()\n        if genes_processing_mode == 'median_binarize_':\n            median_loc = np.median( v  )\n            color_by = np.asarray (v  > median_loc ).ravel()\n        elif genes_processing_mode == 'discretize_3bins_':\n            t1 = np.percentile(v,33)    \n            t2 = np.percentile(v,66)\n            color_by = np.asarray (v  > t1 ).ravel().astype(int) + np.asarray (v  > t2 ).ravel().astype(int)\n        else:\n            color_by = v\n    elif 'median_binarize_' in color_by_mode:\n        color_by_field_name = color_by_mode[16:]\n        if color_by_field_name in adata.obs:\n            color_by = (adata[mask].obs[color_by_field_name]) > np.median( (adata.obs[color_by_field_name]) )\n        else: color_by_mode = None\n    elif 'threshold_binarize_' in color_by_mode:\n        color_by_field_name = color_by_mode.split('_')[3]\n        threshold_binarize = float( color_by_mode.split('_')[2] )\n        if color_by_field_name in adata.obs:\n            color_by = (adata[mask].obs[color_by_field_name]) > threshold_binarize\n        else: color_by_mode = None\n    else:\n        color_by_mode = None\n\n    v1 = r[:,i]\n    v2 = r[:,j]\n    if color_by_mode is None:\n        ax = sns.scatterplot(x=v1, y = v2)# ,  hue= color_by,   alpha = 0.8, marker = '.')#, legend=None)\n    else:\n        if color_by_field_name == 'cell_cycle_phase':\n            if len(np.unique(color_by)) == 3:\n                color_palette = ['red', 'green','blue' ]\n            else:\n                color_palette = \"viridis\"# sns.color_palette(\"tab10\")\n        else:\n            color_palette = \"viridis\"# sns.color_palette(\"tab10\")        #color_by = (adata.obs[color_by_field_name]) > np.median( adata.obs[color_by_field_name].values ) \n        ax = sns.scatterplot(x=v1, y = v2,  hue= color_by, palette = color_palette )#, palette = \"viridis\")# sns.color_palette(\"viridis\", as_cmap=True),\n                            #)# ,   alpha = 0.8, marker = '.')#, )#, legend=None)\n        plt.setp(ax.get_legend().get_texts(), fontsize='20') # for legend text\n        plt.setp(ax.get_legend().get_title(), fontsize='20') # for legend title\n\n        \nplt.show()\nprint( np.round( time.time() - t0,1) , ' seconds passed ' ) ","metadata":{"execution":{"iopub.status.busy":"2021-06-19T18:09:00.382827Z","iopub.execute_input":"2021-06-19T18:09:00.383422Z","iopub.status.idle":"2021-06-19T18:09:24.274466Z","shell.execute_reply.started":"2021-06-19T18:09:00.383376Z","shell.execute_reply":"2021-06-19T18:09:24.27344Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"## Plots in genes axes","metadata":{}},{"cell_type":"code","source":"mask = np.ones( adata.X.shape[0]).astype(bool) # \n\ngene1 = 'E2F1' \ngene2 = 'FOXM1'\nn_x_subplots = 4\n\nc = 0 \nfor gene1,gene2 in [ ('E2F1','FOXM1'), ('PCNA','TOP2A'), ('CCNB1', 'CCNB2'), ('CCNE1', 'CCNB1'),\n                   ('E2F1','PCNA'),('FOXM1','PCNA'), ('E2F1','TOP2A'),('FOXM1','TOP2A'), ]: \n    #,'PCNA','TOP2A', 'CCNE1', 'CCNE2', 'CDK2', 'CCNB1','CCNB2','CCNB3','CCNA2',  ]: # , 'pct_counts_mt']:\n    color_by_mode = 'n.umi' # ,#'cell_cycle_phase', #  'PCA1ALL', 'n.umi', 'threshold_binarize_5_PCA1ALL',\n    if color_by_mode in adata.obs:\n        color_by_field_name = color_by_mode\n        color_by = (adata[mask].obs[color_by_field_name])\n    else:\n        color_by_mode = None\n    \n    if c % n_x_subplots == 0:\n        fig = plt.figure(figsize = (20,6) ); c = 0\n        plt.suptitle(str_data_inf + ' Genes axes   n_cells: ' + str(adata.X.shape[0]) +\\\n            ' n_genes: ' + str(adata.X.shape[1])  + ' color by '+ str(color_by_mode) , fontsize = 20   )#' \n        c = 0\n        \n    c += 1; fig.add_subplot(1,n_x_subplots ,c)\n    plt.title(gene1 + ' ' + gene2) # str(color_by_mode) + ' PCA '+str(i)+','+str(j))\n\n    plt.legend()\n    I_gene = np.where(adata.var.index == gene1)[0]\n    v = adata[mask].X[:,I_gene]\n    if scipy.sparse.issparse(adata.X):\n        v = np.asarray(v.toarray()).ravel()\n    v1 = v\n    I_gene = np.where(adata.var.index == gene2)[0]\n    v = adata[mask].X[:,I_gene]\n    if scipy.sparse.issparse(adata.X):\n        v = np.asarray(v.toarray()).ravel()\n    v2 = v\n\n\n    if color_by_mode is None:\n        ax = sns.scatterplot(x=v1,y=v2)# ,   hue= color_by)#, alpha = 0.7, marker = '.', legend = None)\n    else:\n        ax = sns.scatterplot(x=v1,y=v2,   hue= color_by)#, alpha = 0.7, marker = '.', legend = None)\n        plt.setp(ax.get_legend().get_texts(), fontsize='15') # for legend text\n        plt.setp(ax.get_legend().get_title(), fontsize='15') # for legend title    plt.xlabel('PCA1' , fontsize = 20 )\n        ax.legend().set_visible(False)\n\n    plt.xlabel(gene1,  fontsize = 20)\n    plt.ylabel(gene2,  fontsize = 20)\nplt.show()  ","metadata":{"execution":{"iopub.status.busy":"2021-06-19T18:09:24.275989Z","iopub.execute_input":"2021-06-19T18:09:24.276336Z","iopub.status.idle":"2021-06-19T18:09:29.262362Z","shell.execute_reply.started":"2021-06-19T18:09:24.276303Z","shell.execute_reply":"2021-06-19T18:09:29.261239Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"print(np.round(time.time()-t0start,1),  np.round( (time.time()-t0start)/60,1),  np.round( (time.time()-t0start)/3600,1), \n      'seconds, munutes, hours  passed')\n","metadata":{"execution":{"iopub.status.busy":"2021-06-19T18:09:29.263845Z","iopub.execute_input":"2021-06-19T18:09:29.264172Z","iopub.status.idle":"2021-06-19T18:09:29.270825Z","shell.execute_reply.started":"2021-06-19T18:09:29.264138Z","shell.execute_reply":"2021-06-19T18:09:29.270063Z"},"trusted":true},"execution_count":null,"outputs":[]}]}