{"metadata":{"kernelspec":{"language":"python","display_name":"Python 3","name":"python3"},"language_info":{"pygments_lexer":"ipython3","nbconvert_exporter":"python","version":"3.6.4","file_extension":".py","codemirror_mode":{"name":"ipython","version":3},"name":"python","mimetype":"text/x-python"}},"nbformat_minor":4,"nbformat":4,"cells":[{"cell_type":"markdown","source":"# What is about \n\nLoad and EDA for scRNA-seq data.\n\n","metadata":{}},{"cell_type":"code","source":"# This Python 3 environment comes with many helpful analytics libraries installed\n# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python\n# For example, here's several helpful packages to load\n\nimport numpy as np # linear algebra\nimport pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)\n\n# Input data files are available in the read-only \"../input/\" directory\n# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory\n\nimport os\nfor dirname, _, filenames in os.walk('/kaggle/input'):\n    for filename in filenames:\n        print(os.path.join(dirname, filename))\n\n# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using \"Save & Run All\" \n# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session","metadata":{"_uuid":"8f2839f25d086af736a60e9eeb907d3b93b6e0e5","_cell_guid":"b1076dfc-b9ad-4769-8c92-a6c4dae69d19","execution":{"iopub.status.busy":"2021-06-30T08:40:06.879977Z","iopub.execute_input":"2021-06-30T08:40:06.880389Z","iopub.status.idle":"2021-06-30T08:40:06.899848Z","shell.execute_reply.started":"2021-06-30T08:40:06.880302Z","shell.execute_reply":"2021-06-30T08:40:06.898874Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# Import and install modules ","metadata":{}},{"cell_type":"code","source":"import time\nt0start = time.time()\n\nimport pandas as pd\nimport numpy as np\nimport os\nimport sys\n\nimport matplotlib.pyplot as plt\nimport matplotlib as mpl\nmpl.rcParams['figure.dpi'] = 70\nplt.style.use('dark_background')\n\nimport seaborn as sns\n\nfrom sklearn.decomposition import PCA","metadata":{"execution":{"iopub.status.busy":"2021-06-30T08:40:06.901623Z","iopub.execute_input":"2021-06-30T08:40:06.901895Z","iopub.status.idle":"2021-06-30T08:40:07.982667Z","shell.execute_reply.started":"2021-06-30T08:40:06.901869Z","shell.execute_reply":"2021-06-30T08:40:07.981833Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"!pip install scanpy\nimport scanpy as sc\nimport anndata\n\n!pip install loompy # scanpy needs it to load loom files ","metadata":{"execution":{"iopub.status.busy":"2021-06-30T08:40:07.986785Z","iopub.execute_input":"2021-06-30T08:40:07.987521Z","iopub.status.idle":"2021-06-30T08:40:35.165995Z","shell.execute_reply.started":"2021-06-30T08:40:07.987449Z","shell.execute_reply":"2021-06-30T08:40:35.16492Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# Load data , show some info ","metadata":{}},{"cell_type":"code","source":"","metadata":{"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"#fn = '/kaggle/input/single-cell-rnaseq-data-related-to-cell-cycle/u2os.loom'\n#fn = '/kaggle/input/single-cell-rna-sequencing-data-scc9-cell-line/SCC9_UPPER_AERODIGESTIVE_TRACT_proc.h5ad'\n#fn = '/kaggle/input/single-cell-rna-sequencing-data-scc9-cell-line/SCC9_UPPER_AERODIGESTIVE_TRACT_raw.h5ad'\n#fn = '/kaggle/input/scrna-d085t8-mouse-retina-tumor/D085T8.loom'\n#fn = '/kaggle/input/scrna-d085t8-mouse-retina-tumor/D085T8.txt'\n#path = '/kaggle/input/human-glioblastoma-dataset/' # matrix.mtx\n#path = '/kaggle/input/scrnaseq-early-developing-mouse-retina-cells/'\n#fn = '/kaggle/input/scrna-d085t8-mouse-retina-tumor/D085T8.txt'\n#fn = '/kaggle/input/scrnaseq-early-developing-mouse-retina-cells/raw_count_matrix.mtx'\nfn = '/kaggle/input/rds-scrnaseq-data/scrnaseq_Klein_mouse_embryo.csv'\nstr_data_inf = ' Klein mouse ESC ' # set information string on data - will be used in titles of plots etc \n\nimport time\nt0 = time.time()\nif fn.endswith('.mtx') :\n    #adata = sc.read_10x_mtx(path)\n    adata = sc.read_mtx(fn)\n    \n    f = '/kaggle/input/scrnaseq-early-developing-mouse-retina-cells/genes.csv'\n    df = pd.read_csv(f)\n    print(df.shape)\n    list_genes_upper = [ t.upper() for t in df['gene'].values ]\n    adata.var = pd.DataFrame(index = list_genes_upper)# df['gene']) \n    #adata.var\n\nelif fn.endswith('.loom') :\n    adata = sc.read_loom(fn)\nelif fn.endswith('.h5ad') :\n    adata = sc.read_h5ad(fn)\nelif fn.endswith('.csv'):\n    df = pd.read_csv(fn,  index_col = 0) # \n    df = df.T\n    display(df)\n    var = pd.DataFrame(index = df.columns)\n    obs = pd.DataFrame(index = df.index )\n    adata =  anndata.AnnData(df.values,  obs=obs, var=var ) #\nelif fn.endswith('.txt'):\n    df = pd.read_csv(fn, sep = '\\t', index_col = 0)\n    df = df.T\n    display(df)\n    var = pd.DataFrame(index = df.columns)\n    obs = pd.DataFrame(index = df.index )\n    adata =  anndata.AnnData(df.values,  obs=obs, \n                         var=var ) #\n    #print('New non-sparse adata: ', adata)\n\n# For simplicity mouse genes - > human genes \nlist_genes_upper = [t.upper() for t in adata.var.index ]\nadata.var.index = list_genes_upper\n    \nprint('Data loaded. ', np.round(time.time() - t0,1), 'seconds passed')\nif ('obs' in dir(adata) ) and ('n.umi' not in adata.obs.columns):\n    if ('obs' in dir(adata) ) and ('TotalUMIs' in  adata.obs.columns):\n        adata.obs['n.umi'] = adata.obs['TotalUMIs'] \n    elif np.sum(adata.X) == int( np.sum(adata.X) ) :\n        adata.obs['n.umi'] = adata.X.sum(axis = 1) \n\n\nadata_orig = adata.copy() \nprint(np.sum(adata.X))\nadata","metadata":{"execution":{"iopub.status.busy":"2021-06-30T08:40:35.168035Z","iopub.execute_input":"2021-06-30T08:40:35.16832Z","iopub.status.idle":"2021-06-30T08:40:44.903484Z","shell.execute_reply.started":"2021-06-30T08:40:35.168293Z","shell.execute_reply":"2021-06-30T08:40:44.902643Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"# Show some info on loaded data \nprint('Data type: ', type(adata.X))\nprint('First elements of the count matrix:')\nimport scipy\nif scipy.sparse.issparse(adata.X):\n    print(adata.X[:5,:5].toarray())\nelse:\n    print(adata.X[:5,:5])\n\nif 'obs' in dir(adata):\n    print()\n    print('adata.obs.head(3):')\n    display( adata.obs.head(3))\nif 'var' in dir(adata):\n    print()\n    print('adata.var.head(3):')\n    display( adata.var.head(3))\n\nprint()\nprint('adata')\nprint(adata)","metadata":{"execution":{"iopub.status.busy":"2021-06-30T08:40:44.904605Z","iopub.execute_input":"2021-06-30T08:40:44.904869Z","iopub.status.idle":"2021-06-30T08:40:44.92749Z","shell.execute_reply.started":"2021-06-30T08:40:44.904844Z","shell.execute_reply":"2021-06-30T08:40:44.926324Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# EDA","metadata":{}},{"cell_type":"markdown","source":"## Expression per cell","metadata":{}},{"cell_type":"code","source":"","metadata":{"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"fig = plt.figure(figsize = (20,6)); c = 0\nc+=1; fig.add_subplot(1,2,c);\nv = np.asarray(adata.X.sum(axis = 1)).ravel() # adata.obs['n.umi']\nplt.plot(np.sort(v )) \nplt.title('Expression per cell')\nplt.xlabel('cells sorted')\nplt.ylabel('Expression Total')\n\nc+=1; fig.add_subplot(1,2,c);\nplt.hist(np.sort(v), bins = 100) \nplt.title('Expression per cell')\n\nplt.show()\n\npd.Series(v).describe()\n\n\nfig = plt.figure(figsize = (20,6)); c = 0\nc+=1; fig.add_subplot(1,2,c);\nv = np.asarray((adata.X != 0).sum(axis = 1)).ravel() # adata.obs['n.umi']\nplt.plot(np.sort(v )) \nplt.title('Count genes expressed per cell')\nplt.xlabel('cells sorted')\nplt.ylabel('Count expressed genes')\n\nc+=1; fig.add_subplot(1,2,c);\nplt.hist(np.sort(v), bins = 100) \nplt.title('Count genes expressed per cell')\n\nplt.show()\n\npd.Series(v).describe()\n\n","metadata":{"execution":{"iopub.status.busy":"2021-06-30T08:40:44.928861Z","iopub.execute_input":"2021-06-30T08:40:44.929252Z","iopub.status.idle":"2021-06-30T08:40:46.190396Z","shell.execute_reply.started":"2021-06-30T08:40:44.929211Z","shell.execute_reply":"2021-06-30T08:40:46.189665Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"## Expression per gene","metadata":{}},{"cell_type":"code","source":"v = np.asarray(adata.X.sum(axis = 0)).ravel()\ndisplay(pd.Series(v).describe())\nv = np.log10(1+v)\nfig = plt.figure(figsize = (20,6)); c = 0\nc+=1; fig.add_subplot(1,2,c);\nplt.plot(np.sort(v)) \nplt.title('LOG10 Expression per gene')\nplt.xlabel('genes sorted')\nplt.ylabel('Log10 (1+Counts) ')\n\nc+=1; fig.add_subplot(1,2,c);\nplt.hist(np.sort(v), bins = 30) \nplt.title('LOG10 Expression per gene')\n\nplt.show()\n\npd.Series(v).describe()","metadata":{"execution":{"iopub.status.busy":"2021-06-30T08:40:46.191543Z","iopub.execute_input":"2021-06-30T08:40:46.191887Z","iopub.status.idle":"2021-06-30T08:40:46.548659Z","shell.execute_reply.started":"2021-06-30T08:40:46.191858Z","shell.execute_reply":"2021-06-30T08:40:46.547703Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"## Show to expressed genes","metadata":{}},{"cell_type":"code","source":"v = np.asarray(adata.X.sum(axis = 0)).ravel()\nadata.var['Expression Sum'] = v\nv = np.asarray( (adata.X != 0) .sum(axis = 0)).ravel()\nadata.var['Expressed Cell'] = v\nadata.var.sort_values('Expression Sum',ascending = False ).head(30) # ['counts']\n","metadata":{"execution":{"iopub.status.busy":"2021-06-30T08:40:46.549819Z","iopub.execute_input":"2021-06-30T08:40:46.550076Z","iopub.status.idle":"2021-06-30T08:40:46.69119Z","shell.execute_reply.started":"2021-06-30T08:40:46.550051Z","shell.execute_reply":"2021-06-30T08:40:46.690369Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"## Correlation analysis","metadata":{}},{"cell_type":"code","source":"number_top_genes_to_consider = 10_000\nv = np.asarray(adata.X.sum(axis = 0 ) ).ravel()\nv.shape\nI = np.argsort(v)\nI = I[::-1]\nI = I[:number_top_genes_to_consider]\n\nmask = np.ones( adata.X.shape[0]).astype(bool) \nt0 = time.time()\nif scipy.sparse.issparse(adata.X):\n    corr_matr = np.corrcoef(adata[mask].X[:,I].toarray().T) # Hint - use numpy , pandas is MUCH SLOWER   (df.corr() )\nelse:\n    corr_matr = np.corrcoef(adata[mask].X[:,I].T) # Hint - use numpy , pandas is MUCH SLOWER   (df.corr() )\nprint('Calculated correlation matrix. Shape: ',corr_matr.shape,'  seconds passed:' , np.round( time.time() - t0,1) )\n\nprint()\nprint(np.min(corr_matr ), 'minimal correlation' )\ncorr_matr_abs = np.abs( corr_matr )\nprint(np.mean(corr_matr_abs ), 'average absolute correlation' )\nprint(np.median(corr_matr_abs), 'median absolute correlation' )\nprint(np.min(corr_matr_abs ), 'min absolute correlation' )\nprint(np.std(corr_matr_abs ), 'std absolute correlation' )\nprint()\n\nv = corr_matr.flatten()\nplt.figure(figsize=(14,8))\nt0 = time.time()\nplt.hist(v, bins = 50)\nplt.title('correlation coefficients distribution')\nplt.show()\nprint(time.time() - t0, 'seconds passed')\n\nprint(np.min(corr_matr ), 'minimal correlation' )\nprint(np.mean(corr_matr_abs ), 'average absolute correlation' )\nprint(np.median(corr_matr_abs), 'median absolute correlation' )\nprint(np.min(corr_matr_abs ), 'min absolute correlation' )\nprint(np.std(corr_matr_abs ), 'std absolute correlation' )\nfor t in [0.5,0.6, 0.7,0.8,0.9,0.95,0.97,0.98,.99]:\n    print( ((np.abs(v) < 0.99999999) & (np.abs(v) > t)).sum()/2 , 'number of pairs correlated more than', t  )\nv.shape\n\nprint()\nprint()\nfor threshold_corr in [0.99,0.98,00.97,0.95,0.93,0.90,0.85,0.8,0.75,0.7,0.65,0.6,0.55,.5,0.45,0.4,0.35,0.3]: \n    a,b = np.where( np.triu(np.abs(corr_matr),1) > threshold_corr )\n    if len(a) > 50: break\nprint('for threshold:', threshold_corr,'we have ',len(a),' correlated pairs' )\n\n\nd = pd.DataFrame()\nfor i in range(len(a)):\n    Iai, Ibi =  I[a[i]], I[b[i]]\n    d.loc[i,'Gene1'] = adata[mask].var.index[Iai] \n    d.loc[i,'Gene2'] = adata[mask].var.index[Ibi ] \n    d.loc[i,'Correlation'] = corr_matr[a[i],b[i]] \n    d.loc[i,'Correlation Abs'] = np.abs(d.loc[i,'Correlation'] )\n    d.loc[i,'Gene1 Expression Sum'] = np.asarray(adata[mask].X[:,Iai].sum(axis = 0)).ravel()[0]\n    d.loc[i,'Gene2 Expression Sum'] = np.asarray(adata[mask].X[:,Ibi].sum(axis = 0)).ravel()[0]\n    \nprint(time.time() - t0, 'seconds passed')\n    \nd.sort_values('Correlation Abs',ascending = False).head(30)","metadata":{"execution":{"iopub.status.busy":"2021-06-30T08:40:46.693649Z","iopub.execute_input":"2021-06-30T08:40:46.693915Z","iopub.status.idle":"2021-06-30T08:41:20.030166Z","shell.execute_reply.started":"2021-06-30T08:40:46.69389Z","shell.execute_reply":"2021-06-30T08:41:20.029463Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"#clustermap\nimport seaborn as sns\nt0 = time.time()\nsns.clustermap(np.abs(corr_matr[:3000,:3000]),cmap='vlag');\nprint( np.round(time.time()- t0,1), ' seconds passed.')","metadata":{"execution":{"iopub.status.busy":"2021-06-30T08:41:20.031972Z","iopub.execute_input":"2021-06-30T08:41:20.032451Z","iopub.status.idle":"2021-06-30T08:41:55.912513Z","shell.execute_reply.started":"2021-06-30T08:41:20.032421Z","shell.execute_reply":"2021-06-30T08:41:55.911434Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"t0=time.time()\nw,v = np.linalg.eig(corr_matr[:2000,:2000])\nprint(time.time()-t0, 'seconds passed')\nw = np.real(w.ravel())\nw = np.sort(w)[::-1]\nplt.plot(w,'*-')\nplt.title('Eigenvalues of correlation matrix ')\nplt.show()","metadata":{"execution":{"iopub.status.busy":"2021-06-30T08:41:55.913894Z","iopub.execute_input":"2021-06-30T08:41:55.914201Z","iopub.status.idle":"2021-06-30T08:42:04.809365Z","shell.execute_reply.started":"2021-06-30T08:41:55.914172Z","shell.execute_reply":"2021-06-30T08:42:04.808399Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# Visualizations","metadata":{}},{"cell_type":"markdown","source":"## PCA","metadata":{}},{"cell_type":"code","source":"import scipy\nfrom sklearn.decomposition import TruncatedSVD\nfrom sklearn.decomposition import PCA\nimport time\nt0 = time.time()\nif scipy.sparse.issparse(adata.X):\n    reducer = TruncatedSVD(n_components=2, n_iter=7, random_state=42)\nelse:\n    reducer = PCA(n_components=2)\n    \nr = reducer.fit_transform(adata.X)\n\nfig = plt.figure(figsize = (25,12))\nplt.title(str_data_inf + ' PCA  n_cells: ' + str(adata.X.shape[0]) +\\\n            ' n_genes: ' + str(adata.X.shape[1])  + ' '  , fontsize = 20   )#' \nif 'n.umi' not in adata.obs.columns: \n    ax = sns.scatterplot(x=v1, y = v2)#,  hue= color_by)#, palette = color_palette )#, palette = \"viridis\")# sns.color_palette(\"viridis\", as_cmap=True),\n                        #)# ,   alpha = 0.8, marker = '.')#, )#, legend=None)\nelse:\n    color_by = adata.obs['n.umi'].values\n    ax = sns.scatterplot(x=r[:,0], y = r[:,1],  hue= color_by)#, palette = color_palette )#, palette = \"viridis\")# sns.color_palette(\"viridis\", as_cmap=True),\n                        #)# ,   alpha = 0.8, marker = '.')#, )#, legend=None)\n    plt.setp(ax.get_legend().get_texts(), fontsize='20') # for legend text\n    plt.setp(ax.get_legend().get_title(), fontsize='20') # for legend title\nplt.xlabel('PCA1' , fontsize = 20 )\nplt.ylabel('PCA2' , fontsize = 20 )\nplt.show()\nprint( np.round( time.time() - t0,1) , ' seconds passed ' ) ","metadata":{"execution":{"iopub.status.busy":"2021-06-30T08:42:04.810852Z","iopub.execute_input":"2021-06-30T08:42:04.811272Z","iopub.status.idle":"2021-06-30T08:42:07.979312Z","shell.execute_reply.started":"2021-06-30T08:42:04.811231Z","shell.execute_reply":"2021-06-30T08:42:07.978128Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"## UMAP","metadata":{}},{"cell_type":"code","source":"import scipy\nimport umap \nt0 = time.time()\nreducer = umap.UMAP()\nr = reducer.fit_transform(adata.X)\n\nfig = plt.figure(figsize = (25,12))\nplt.title(str_data_inf + ' UMAP n_cells: ' + str(adata.X.shape[0]) +\\\n            ' n_genes: ' + str(adata.X.shape[1])  + ' '  , fontsize = 20   )#' \nif 'n.umi' not in adata.obs.columns: \n    ax = sns.scatterplot(x=v1, y = v2)#,  hue= color_by)#, palette = color_palette )#, palette = \"viridis\")# sns.color_palette(\"viridis\", as_cmap=True),\n                        #)# ,   alpha = 0.8, marker = '.')#, )#, legend=None)\nelse:\n    color_by = adata.obs['n.umi'].values\n    ax = sns.scatterplot(x=r[:,0], y = r[:,1],  hue= color_by)#, palette = color_palette )#, palette = \"viridis\")# sns.color_palette(\"viridis\", as_cmap=True),\n                        #)# ,   alpha = 0.8, marker = '.')#, )#, legend=None)\n    plt.setp(ax.get_legend().get_texts(), fontsize='20') # for legend text\n    plt.setp(ax.get_legend().get_title(), fontsize='20') # for legend title\n\nplt.xlabel('UMAP1' , fontsize = 20 )\nplt.ylabel('UMAP2' , fontsize = 20 )\nplt.show()\nprint( np.round( time.time() - t0,1) , ' seconds passed ' ) ","metadata":{"execution":{"iopub.status.busy":"2021-06-30T08:42:07.980678Z","iopub.execute_input":"2021-06-30T08:42:07.981002Z","iopub.status.idle":"2021-06-30T08:45:22.714726Z","shell.execute_reply.started":"2021-06-30T08:42:07.980973Z","shell.execute_reply":"2021-06-30T08:45:22.713746Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"## Different axes PCA colored by genes","metadata":{}},{"cell_type":"code","source":"plot_mode = 'PCA'# 'Phase_plot':\nn_x_subplots = 3\ngenes_processing_mode = 'discretize_3bins_'# 'median_binarize_'\n\nmask = np.ones( adata.X.shape[0]).astype(bool) # \n\nimport scipy\nfrom sklearn.decomposition import TruncatedSVD\nfrom sklearn.decomposition import PCA\nimport time\nt0 = time.time()\nif scipy.sparse.issparse(adata.X):\n    reducer = TruncatedSVD(n_components=10, n_iter=7, random_state=42)\nelse:\n    reducer = PCA(n_components=10)\n    \nr = reducer.fit_transform(adata.X)\n\nc = 0\nfor i,j in [(0,1),(2,3),(3,4),(5,6)]:#,(1,3),(2,3),(3,4)]:\n    for color_by_mode in  ['n.umi',#'cell_cycle_phase', #  'PCA1ALL', 'n.umi', 'threshold_binarize_5_PCA1ALL',\n        'E2F1','FOXM1']:#,'PCNA','TOP2A', 'CCNE1', 'CCNE2', 'CDK2', 'CCNB1','CCNB2','CCNB3','CCNA2',  ]: # , 'pct_counts_mt']:\n        if c % n_x_subplots == 0:\n            fig = plt.figure(figsize = (20,5) ); c = 0\n            plt.suptitle(str_data_inf + ' PCA  n_cells: ' + str(adata.X.shape[0]) +\\\n                ' n_genes: ' + str(adata.X.shape[1])  + ' '  , fontsize = 20   )#' \n            c = 0\n        c += 1; fig.add_subplot(1,n_x_subplots ,c)\n        plt.title(str(color_by_mode) + ' PCA '+str(i+1)+','+str(j+1), fontsize = 20)\n\n        color_by_field_name = color_by_mode\n        if color_by_mode in adata.obs:\n            color_by = (adata[mask].obs[color_by_field_name])\n        elif color_by_mode in adata.var.index:\n            I_gene = np.where(adata.var.index == color_by_mode)[0]\n            v = adata[mask].X[:,I_gene]\n            if scipy.sparse.issparse(adata.X):\n                v = np.asarray(v.toarray()).ravel()\n            if genes_processing_mode == 'median_binarize_':\n                median_loc = np.median( v  )\n                color_by = np.asarray (v  > median_loc ).ravel()\n            elif genes_processing_mode == 'discretize_3bins_':\n                t1 = np.percentile(v,33)    \n                t2 = np.percentile(v,66)\n                color_by = np.asarray (v  > t1 ).ravel().astype(int) + np.asarray (v  > t2 ).ravel().astype(int)\n            else:\n                color_by = v\n        elif 'median_binarize_' in color_by_mode:\n            color_by_field_name = color_by_mode[16:]\n            if color_by_field_name in adata.obs:\n                color_by = (adata[mask].obs[color_by_field_name]) > np.median( (adata.obs[color_by_field_name]) )\n            else: color_by_mode = None\n        elif 'threshold_binarize_' in color_by_mode:\n            color_by_field_name = color_by_mode.split('_')[3]\n            threshold_binarize = float( color_by_mode.split('_')[2] )\n            if color_by_field_name in adata.obs:\n                color_by = (adata[mask].obs[color_by_field_name]) > threshold_binarize\n            else: color_by_mode = None\n        else:\n            color_by_mode = None\n\n        v1 = r[:,i]\n        v2 = r[:,j]\n        if color_by_mode is None:\n            ax = sns.scatterplot(x=v1, y = v2)# ,  hue= color_by,   alpha = 0.8, marker = '.')#, legend=None)\n        else:\n            if color_by_field_name == 'cell_cycle_phase':\n                if len(np.unique(color_by)) == 3:\n                    color_palette = ['red', 'green','blue' ]\n                else:\n                    color_palette = \"viridis\"# sns.color_palette(\"tab10\")\n            else:\n                color_palette = \"viridis\"# sns.color_palette(\"tab10\")        #color_by = (adata.obs[color_by_field_name]) > np.median( adata.obs[color_by_field_name].values ) \n            ax = sns.scatterplot(x=v1, y = v2,  hue= color_by, palette = color_palette )#, palette = \"viridis\")# sns.color_palette(\"viridis\", as_cmap=True),\n                                #)# ,   alpha = 0.8, marker = '.')#, )#, legend=None)\n            plt.setp(ax.get_legend().get_texts(), fontsize='20') # for legend text\n            plt.setp(ax.get_legend().get_title(), fontsize='20') # for legend title\n\n        \nplt.show()\nprint( np.round( time.time() - t0,1) , ' seconds passed ' ) ","metadata":{"execution":{"iopub.status.busy":"2021-06-30T08:45:22.715994Z","iopub.execute_input":"2021-06-30T08:45:22.716259Z","iopub.status.idle":"2021-06-30T08:45:29.868177Z","shell.execute_reply.started":"2021-06-30T08:45:22.716233Z","shell.execute_reply":"2021-06-30T08:45:29.867099Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"## Plots in genes axes","metadata":{}},{"cell_type":"code","source":"mask = np.ones( adata.X.shape[0]).astype(bool) # \n\ngene1 = 'E2F1' \ngene2 = 'FOXM1'\nn_x_subplots = 4\n\nc = 0 \nfor gene1,gene2 in [ ('E2F1','FOXM1'), ('PCNA','TOP2A'), ('CCNB1', 'CCNB2'), ('CCNE1', 'CCNB1'),\n                   ('E2F1','PCNA'),('FOXM1','PCNA'), ('E2F1','TOP2A'),('FOXM1','TOP2A'), ]: \n    #,'PCNA','TOP2A', 'CCNE1', 'CCNE2', 'CDK2', 'CCNB1','CCNB2','CCNB3','CCNA2',  ]: # , 'pct_counts_mt']:\n    if gene1 not in adata.var.index : continue\n    if gene2 not in adata.var.index : continue\n        \n    color_by_mode = 'n.umi' # ,#'cell_cycle_phase', #  'PCA1ALL', 'n.umi', 'threshold_binarize_5_PCA1ALL',\n    if color_by_mode in adata.obs:\n        color_by_field_name = color_by_mode\n        color_by = (adata[mask].obs[color_by_field_name])\n    else:\n        color_by_mode = None\n    \n    if c % n_x_subplots == 0:\n        fig = plt.figure(figsize = (20,6) ); c = 0\n        plt.suptitle(str_data_inf + ' Genes axes   n_cells: ' + str(adata.X.shape[0]) +\\\n            ' n_genes: ' + str(adata.X.shape[1])  + ' color by '+ str(color_by_mode) , fontsize = 20   )#' \n        c = 0\n        \n    c += 1; fig.add_subplot(1,n_x_subplots ,c)\n    plt.title(gene1 + ' ' + gene2) # str(color_by_mode) + ' PCA '+str(i)+','+str(j))\n\n    plt.legend()\n    I_gene = np.where(adata.var.index == gene1)[0]\n    v = adata[mask].X[:,I_gene]\n    if scipy.sparse.issparse(adata.X):\n        v = np.asarray(v.toarray()).ravel()\n    else:\n        v = np.asarray(v).ravel()\n        \n    v1 = v\n    I_gene = np.where(adata.var.index == gene2)[0]\n    v = adata[mask].X[:,I_gene]\n    if scipy.sparse.issparse(adata.X):\n        v = np.asarray(v.toarray()).ravel()\n    else:\n        v = np.asarray(v).ravel()\n    v2 = v\n\n\n    if color_by_mode is None:\n        ax = sns.scatterplot(x=v1,y=v2)# ,   hue= color_by)#, alpha = 0.7, marker = '.', legend = None)\n    else:\n        ax = sns.scatterplot(x=v1,y=v2,   hue= color_by)#, alpha = 0.7, marker = '.', legend = None)\n        plt.setp(ax.get_legend().get_texts(), fontsize='15') # for legend text\n        plt.setp(ax.get_legend().get_title(), fontsize='15') # for legend title    plt.xlabel('PCA1' , fontsize = 20 )\n        ax.legend().set_visible(False)\n\n    plt.xlabel(gene1,  fontsize = 20)\n    plt.ylabel(gene2,  fontsize = 20)\nplt.show()  ","metadata":{"execution":{"iopub.status.busy":"2021-06-30T08:45:29.869324Z","iopub.execute_input":"2021-06-30T08:45:29.869616Z","iopub.status.idle":"2021-06-30T08:45:33.22127Z","shell.execute_reply.started":"2021-06-30T08:45:29.869589Z","shell.execute_reply":"2021-06-30T08:45:33.220379Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# Preprocessing","metadata":{}},{"cell_type":"markdown","source":"## Auxilliary functions","metadata":{}},{"cell_type":"code","source":"def _smooth_adata_by_pooling(adata, X_embed, n_neighbours=10, copy=False):\n    # adata_pooled = adata.copy if copy else adata\n    nbrs = NearestNeighbors(n_neighbors=n_neighbours).fit(X_embed)\n    print( 'done: nbrs = NearestNeighbors(n_neighbors=n_neighbours).fit(X_embed) ')\n    distances, indices = nbrs.kneighbors(X_embed)\n    print( 'done: distances, indices = nbrs.kneighbors(X_embed) ')\n    nd_array_X = _get_nd_array(adata.X)\n    print('done: nd_array_X = _get_nd_array(adata.X)' )\n    adata.X = _smooth_matrix_by_pooling(nd_array_X , indices)\n    print( 'done: adata.X = _smooth_matrix_by_pooling(_get_nd_array(adata.X), indices) ')\n    if \"matrix\" in adata.layers:\n        adata.layers[\"matrix\"] = _smooth_matrix_by_pooling(\n            _get_nd_array(adata.layers[\"matrix\"]), indices\n        )\n    if \"spliced\" in adata.layers:\n        adata.layers[\"spliced\"] = _smooth_matrix_by_pooling(\n            _get_nd_array(adata.layers[\"spliced\"]), indices\n        )\n    if \"unspliced\" in adata.layers:\n        adata.layers[\"unspliced\"] = _smooth_matrix_by_pooling(\n            _get_nd_array(adata.layers[\"unspliced\"]), indices\n        )\n\n    adata.uns[\"scycle\"] = True\n\n\ndef _smooth_matrix_by_pooling(matrix, indices):\n    matrix_pooled = matrix.copy()\n    t0 = time.time()\n    #print(indices.shape,  matrix.shape, type(matrix) )\n    for i in range(len(indices)):\n        #print(i,len(indices), time.time()-t0 )\n        matrix_pooled[i, :] = np.mean(matrix[indices[i], :], axis=0)\n    return matrix_pooled\n\n\ndef _get_nd_array(arr):\n    x = None\n    if str(type(arr)):\n        x = arr\n    else:\n        print('start:  x = arr.toarray()')\n        x = arr.toarray()\n        print('finished:  x = arr.toarray()')\n    return x\n\ndef calc_scores(anndata,signature_dict):\n    matrix = anndata.to_df().to_numpy()\n    scores_dic = {}\n    for key in signature_dict:\n        names = np.array(signature_dict[key])\n        inds = np.where(np.isin(anndata.var_names,names))[0]\n        matrix_sel = matrix[:,inds]\n        scores = np.mean(matrix_sel,axis=1)\n        scores_dic[key] = scores\n    return scores_dic\n\ndef find_nonproliferative_cells(adata,estimation_fraction_nonproliferating_cells=0.3, \n                                number_of_nodes=30,max_number_of_iterations = 20,\n                               number_of_sigmas=3.0,Mu=1.0):\n    all_markers = list(adata.uns['S-phase_genes'])+list(adata.uns['G2-M_genes'])\n    Xccm = adata[:,list(set(all_markers)&set(adata.var_names))].X\n    cc_score = np.array(list(np.mean(Xccm,axis=1)))\n\n    ind_sorted_prolif = np.argsort(cc_score)\n    ind_nonprolif = ind_sorted_prolif[0:int(len(adata)*estimation_fraction_nonproliferating_cells)]\n    adata.obs['proliferating'] = np.empty(len(adata)).astype(np.int)\n    adata.obs['proliferating'][:] = 1\n    adata.obs['proliferating'][ind_nonprolif] = 0\n    \n    sc.pl.scatter(adata,x='S-phase',y='G2-M',color='proliferating')\n\n    fraction_nonprolif_old = estimation_fraction_nonproliferating_cells\n\n    for i in range(max_number_of_iterations):\n        X_elpigraph_training = adata.obs[['S-phase','G2-M']].to_numpy().astype(np.float64)\n        u = X_elpigraph_training.copy()\n        ind_prolif = np.where(np.array(adata.obs['proliferating'])==1)[0]        \n        X_elpigraph_training = X_elpigraph_training[ind_prolif,:]\n\n        egr = elpigraph.computeElasticPrincipalCircle(X_elpigraph_training,number_of_nodes,\n                                                      Mu=Mu,drawPCAView=False,verbose=False)\n        partition, dists = elpigraph.src.core.PartitionData(X = u, NodePositions = egr[0]['NodePositions'], \n                                                                MaxBlockSize = 100000000, TrimmingRadius = np.inf,\n                                                                SquaredX = np.sum(u**2,axis=1,keepdims=1))\n\n        mndist = np.mean(dists[ind_prolif])\n        #plt.hist(dists,bins=100)\n        #plt.show()\n        #plt.hist(dists[ind_prolif],bins=100)\n        #plt.show()        \n        intervaldist = np.std(dists[ind_prolif])*number_of_sigmas\n        tt1 = np.array(1-adata.obs['proliferating'])\n        tt2 = np.zeros(len(tt1)).astype(np.int)\n        for k,d in enumerate(dists):\n            if d>mndist+intervaldist:\n                tt2[k] = 1\n        nonprolif_new = np.array(tt1) * np.array(tt2)\n        adata.obs['proliferating'] = 1-nonprolif_new\n        #adata.obs['dists'] = dists\n        #sc.pl.scatter(adata,x='S-phase',y='G2-M',color='dists')\n        fraction_nonprolif = 1-np.sum(adata.obs['proliferating'])/len(adata)\n        print('\\n\\n===========\\nIteration',i,'Fraction of non-proliferating cells:',fraction_nonprolif,'\\n==============\\n\\n\\n')    \n        if np.abs(fraction_nonprolif-fraction_nonprolif_old)<0.01:\n            break\n        fraction_nonprolif_old = fraction_nonprolif\n        \ndef _compute_principal_circle(adata,number_of_nodes=30,n_components = 30):\n    #driving_feature = np.array(adata_orig[adata.uns['ind_samples']].obs['total_counts'])\n    #driving_feature_weight = 1.0\n    #egr, partition, X, Xp = _compute_principal_circle(adata,number_of_nodes=40,\n    #                                                  driving_feature=driving_feature,driving_feature_weight=0.1)\n\n    X = adata.X\n    X_prolif = adata.X[np.where(adata.obs['proliferating']==1)[0],:]\n    #driving_feature = driving_feature[adata.obs[np.where(adata.obs['proliferating']==1)[0]]]\n\n    mn_prolif = np.mean(X_prolif,axis=0)\n    pca = PCA(n_components=n_components)\n    u = pca.fit_transform(X_prolif)\n    v = pca.components_.T\n    #X_pca = adata.obsm['X_pca'].astype(np.float64)\n    X_elpigraph_training = u.astype(np.float64)\n    #std = np.std(X_elpigraph_training)\n    #driving_feature = stats.zscore(driving_feature)*std*driving_feature_weight\n    #driving_feature = driving_feature.reshape(-1,1)\n    #X_elpigraph_training = np.concatenate([X_elpigraph_training,driving_feature],axis=1)\n    #print('std=',std)\n    #print('std df=',np.std(driving_feature))\n\n    #print(X_elpigraph_training.shape)\n    egr = elpigraph.computeElasticPrincipalCircle(X_elpigraph_training,number_of_nodes,Mu=0.2)\n    partition, dists = elpigraph.src.core.PartitionData(X = X_elpigraph_training, NodePositions = egr[0]['NodePositions'], \n                                                            MaxBlockSize = 100000000, TrimmingRadius = np.inf,\n                                                            SquaredX = np.sum(X_elpigraph_training**2,axis=1,keepdims=1))\n    Xp = (X-mn_prolif)@v\n    #number_of_driving_features = driving_feature.shape[1]\n    #node_positions_reduced = egr[0]['NodePositions'][:,:-number_of_driving_features]\n    nodep = egr[0]['NodePositions']\n    partition, dists = elpigraph.src.core.PartitionData(X = Xp, NodePositions = nodep, \n                                                           MaxBlockSize = 100000000, TrimmingRadius = np.inf,\n                                                           SquaredX = np.sum(Xp**2,axis=1,keepdims=1))\n\n    #nodep = egr[0]['NodePositions'][:,:-number_of_driving_features]\n    #edges = egr[0]['Edges'][0]\n    #X_elpigraph_training = X_elpigraph_training[:,:-number_of_driving_features]\n    #egr[0]['NodePositions'] = nodep\n    return egr, partition, X_elpigraph_training, Xp\n\n\ndef _compute_principal_curve(adata,number_of_nodes=30, Mu=0.3, Lambda=0.001,n_components=30):\n    X = adata.X\n    X_prolif = adata.X[np.where(adata.obs['proliferating']==1)[0],:]\n\n    mn_prolif = np.mean(X_prolif,axis=0)\n    pca = PCA(n_components=n_components)\n    u = pca.fit_transform(X_prolif)\n    v = pca.components_.T\n    #X_pca = adata.obsm['X_pca'].astype(np.float64)\n    \n    adata.varm['pc_components_elpigraph'] = v\n    adata.varm['mean_point_elpigraph'] = mn_prolif\n    \n    X_elpigraph_training = u.astype(np.float64)\n\n    egr, starting_node = compute_principal_curve_from_circle(X_elpigraph_training,\n                                                             n_nodes=number_of_nodes,\n                                                            produceTree=False,Mu=Mu,Lambda=Lambda)\n\n    partition_, dists = elpigraph.src.core.PartitionData(X = X_elpigraph_training, NodePositions = egr[0]['NodePositions'], \n                                                            MaxBlockSize = 100000000, TrimmingRadius = np.inf,\n                                                            SquaredX = np.sum(X_elpigraph_training**2,axis=1,keepdims=1))\n    Xp = (X-mn_prolif)@v\n    partition_, dists = elpigraph.src.core.PartitionData(X = Xp, NodePositions = egr[0]['NodePositions'], \n                                                            MaxBlockSize = 100000000, TrimmingRadius = np.inf,\n                                                            SquaredX = np.sum(Xp**2,axis=1,keepdims=1))\n\n    nodep = egr[0]['NodePositions']\n    edges = egr[0]['Edges'][0]\n    return egr, partition_, starting_node,X_elpigraph_training, Xp\n\ndef compute_principal_curve_from_circle(X,n_nodes=30,Mu=0.1,Lambda=0.01, produceTree=False):\n    egr = elpigraph.computeElasticPrincipalCircle(X,int(n_nodes/2),Mu=Mu)\n    nodep = egr[0]['NodePositions']\n    edges = egr[0]['Edges'][0]\n    partition_, dists = elpigraph.src.core.PartitionData(X = X, NodePositions = nodep, \n                                                            MaxBlockSize = 100000000, TrimmingRadius = np.inf,\n                                                            SquaredX = np.sum(X**2,axis=1,keepdims=1))\n    node_sizes = np.array([len(np.where(partition_==i)[0]) for i in range(nodep.shape[0])])\n    node_min = np.argmin(node_sizes)\n    edges_open = edges.copy()\n    k = 0\n    starting_node = -1\n    while k<edges_open.shape[0]:\n        e = edges_open[k,:]\n        if (e[0]==node_min)|(e[1]==node_min):\n            edges_open = np.delete(edges_open,k,axis=0)\n            if e[0]==node_min:\n                starting_node = e[1]\n            if e[1]==node_min:\n                starting_node = e[0]\n        else:\n            k = k+1\n    nodep_open = np.delete(nodep,node_min,axis=0)\n    if starting_node>node_min:\n        starting_node = starting_node-1\n    for e in edges_open:\n        if e[0]>node_min:\n            e[0]=e[0]-1\n        if e[1]>node_min:\n            e[1]=e[1]-1\n    if produceTree:\n        egrl = elpigraph.computeElasticPrincipalTree(X,n_nodes,InitNodePositions=nodep_open,\n                                                 InitEdges=edges_open,Lambda=Lambda,Mu=Mu,\n                                                 alpha=0.01,FinalEnergy='Penalized')\n    else:\n        egrl = elpigraph.computeElasticPrincipalCurve(X,n_nodes,InitNodePositions=nodep_open,\n                                                 InitEdges=edges_open,Lambda=Lambda,Mu=Mu,\n                                                 alpha=0.01,FinalEnergy='Penalized')\n    return egrl, starting_node\n\ndef subtract_cell_cycle_trajectory(X,partition):\n    points = range(X.shape[0])\n    r2scores = []\n    X1 = X[points,:]\n    X_ro = np.zeros((X.shape[0],X.shape[1]))\n    partition_points = partition[points]\n    inds = {}\n    for k in range(len(partition_points)):\n        j = partition_points[k][0]\n        if not j in inds:\n            inds[j] = [k]\n        else:\n            inds[j].append(k)\n    XT = X1.T\n    for j in range(X1.shape[0]):\n        k = partition_points[j][0]\n        ind = np.array(inds[k])\n        X_ro[j,:] = (XT[:,j] - np.mean(XT[:,ind],axis=1)).T\n    residue_matrix = X1 - X_ro\n    residues_var = np.var(residue_matrix,axis=0)\n    vrs = np.var(X1,axis=0)\n    r2scores = residues_var/vrs\n    return X_ro, residue_matrix, r2scores","metadata":{"execution":{"iopub.status.busy":"2021-06-30T08:45:33.222498Z","iopub.execute_input":"2021-06-30T08:45:33.222776Z","iopub.status.idle":"2021-06-30T08:45:33.27528Z","shell.execute_reply.started":"2021-06-30T08:45:33.222751Z","shell.execute_reply":"2021-06-30T08:45:33.274382Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"list_genes2include_mandotory = ['E2F1','FOXM1', 'ESRRB', 'NR5A2',  'HIST1H4C', 'FOXO3' , 'EZH1',  'ANLN' ]\nlist_genes2include_mandotory += ['PCNA','TOP2A' ]\n\nlist_genes2include_mandotory += ['WEE1','WEE2'] # Inhibitor of CDK1 https://en.wikipedia.org/wiki/Wee1\nlist_genes2include_mandotory += ['CDK1', 'UBE2C','TOP2A','HIST1H4E','HIST1H4C'] # Hsiao2020GenomeResearch 5 genes 'HIST1H4E' = H4C5,'HIST1H4C'=H4C3\nlist_genes2include_mandotory +=  ['MYBL2', 'TP53', 'KLF4', 'CDKN1A', 'CDKN2A', 'RB1', 'MDM2', 'MDM4', 'ZBTB20', 'GAPDH',  'EFNA5', 'LINC00511' ]\n# MYBL2 https://en.wikipedia.org/wiki/MYBL2 - transcription factor\n# The protein encoded by this gene, a member of the MYB family of transcription factor genes, is a nuclear protein involved in cell cycle progression. The encoded protein is phosphorylated by cyclin A/cyclin-dependent kinase 2 during the S-phase of the cell cycle and possesses both activator and repressor activities.\n# https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6788831/\n# Binding of FoxM1 to G2/M gene promoters is dependent upon B-Myb Christin F Down 1, Julie Millour, Eric W-F Lam, Roger J Watson\n#https://pubmed.ncbi.nlm.nih.gov/23555285/\n#Long noncoding RNA MALAT1 controls cell cycle progression by regulating the expression of oncogenic transcription factor B-MYB -\n\nlist_genes2include_mandotory += [ 'CCNE1', 'CCNE2', 'CDK2'] +  ['CCNB1','CCNB2','CCNB3'] + ['CCNA1',\n'CCNA2','CCNF', 'CDK1'] + ['CCND1', 'CCND2', 'CCND3'] + [ 'CDK4','CDK6']\n#                       'CCNA1','CCNA2', 'CCNE1', 'CCNE2',]\nlist_genes2include_mandotory += ['CDK11A', # All of the form: CDKnumber*  - 27\n                                 'CDK14', 'CDK17', 'CDK13', 'CDK5RAP1', 'CDK16', 'CDK6', 'CDK5RAP3', 'CDK2AP1', 'CDK18', 'CDK2', 'CDK8', 'CDK7', 'CDK4', 'CDK9', 'CDK5RAP2', 'CDK15', 'CDK19', 'CDK20', 'CDK5', 'CDK12', 'CDK2AP2', 'CDK1', 'CDK5R2', 'CDK5R1', 'CDK10', 'CDK11B']\nlist_genes2include_mandotory += ['CDKN3', # 10 CDKN* - A cyclin-dependent kinase inhibitor protein is a protein which inhibits the enzyme cyclin-dependent kinase (CDK). \n         'CDKN1B', 'CDKN2C', 'CDKN1A', 'CDKN2D', 'CDKN1C', 'CDKN2A', 'CDKN2AIP', 'CDKN2AIPNL', 'CDKN2B-AS1']\n# p16\tCDKN2A\tCyclin-dependent kinase 4, Cyclin-dependent kinase 6\n# p15\tCDKN2B\tCyclin-dependent kinase 4\n#p18\tCDKN2C\tCyclin-dependent kinase 4, Cyclin-dependent kinase 6\n#p19\tCDKN2D\tCyclin-dependent kinase 4, Cyclin-dependent kinase 6\n#p21 / WAF1\tCDKN1A[2]\tCyclin E1/Cyclin-dependent kinase 2\n#p27\tCDKN1B\tCyclin D3/Cyclin-dependent kinase 4, Cyclin E1/Cyclin-dependent kinase 2\n#p57\tCDKN1C\tCyclin E1/Cyclin-dependent kinase 2\n# https://en.wikipedia.org/wiki/Cyclin-dependent_kinase_inhibitor_protein\n    \nlist_genes2include_mandotory += ['CDKL3', # All starts with CDK - 43 \n                                 'CDKL5', 'CDK11A', 'CDK14', 'CDK17', 'CDK13', 'CDKL1', 'CDKN3', 'CDK5RAP1', 'CDK16', 'CDK6', 'CDK5RAP3', 'CDKN1B', 'CDK2AP1', 'CDK18', 'CDKN2C', 'CDK2', 'CDKN1A', 'CDKN2D', 'CDKN1C', 'CDK8', 'CDK7', 'CDK4', 'CDK9', 'CDK5RAP2', 'CDK15', 'CDKL2', 'CDKAL1', 'CDKN2A', 'CDK19', 'CDK20', 'CDK5', 'CDK12', 'CDK2AP2', 'CDKN2AIP', 'CDK1', 'CDK5R2', 'CDK5R1', 'CDK10', 'CDKL4', 'CDKN2AIPNL', 'CDKN2B-AS1', 'CDK11B']\n\nlist_genes2include_mandotory += ['CDC27', #  \"cdc\" in its name refers to \"cell division cycle\"\n 'CDC42', 'CDC14A', 'CDC14B', 'CDC45', 'CDC6', 'CDC23', 'CDC5L', 'CDC7', 'CDC34', 'CDC25B', 'CDC37', 'CDC37L1', 'CDCA3', 'CDC20', 'CDC42EP1', 'CDC16', 'CDC73', 'CDCA8', 'CDC42BPA', 'CDCA7', 'CDCA5', 'CDC42EP2', 'CDC123', 'CDC25C', 'CDC42SE2', 'CDC42EP3', 'CDCP1', 'CDC25A', 'CDC20B', 'CDCA7L', 'CDC42EP5', 'CDC40', 'CDCA4', 'CDC42BPG', 'CDC26', 'CDC42EP4', 'CDCA2', 'CDC42SE1', 'CDC42BPB', 'CDC42-IT1', 'CDC20P1', 'CDC42P6', 'CDC27P2', 'CDC42P5', 'CDC37L1-AS1', 'CDC27P3']\n# Cdc25 is a dual-specificity phosphatase first isolated from the yeast Schizosaccharomyces pombe as a cell cycle defective mutant.[2] As with other cell cycle proteins or genes such as Cdc2 and Cdc4, the \"cdc\" in its name refers to \"cell division cycle\". Dual-specificity phosphatases are considered a sub-class of protein tyrosine phosphatases. By removing inhibitory phosphate residues from target cyclin-dependent kinases (Cdks),[3] Cdc25 proteins control entry into and progression through various phases of the cell cycle, including mitosis and S (\"Synthesis\") phase.\n\nlist_genes2include_mandotory += ['E2F2', # 10 genes E2F transc. factors  \n     'E2F1', 'E2F3', 'E2F8', 'E2F5', 'E2F7', 'E2F6', 'E2F3P2', 'E2F4', 'E2F3-IT1']\n\nlist_genes2include_mandotory += ['GATA1', 'FOXA1', 'BRD4', 'KMT2A' ]\n# 'GATA1', 'FOXA1', 'BRD4', 'KMT2A' - transcription factors are retained in mitotic chromosomes (from paper \"Trans landscape of the human cell cycle\" )\n\nlist_histone_genes_from_wiki = ['H1F0' , 'H1FNT' , 'H1FOO' , 'H1FX' ,\n'HIST1H1A' , 'HIST1H1B' , 'HIST1H1C' , 'HIST1H1D' , 'HIST1H1E' , 'HIST1H1T' ,\n'H2AFB1' , 'H2AFB2' , 'H2AFB3' , 'H2AFJ' , 'H2AFV' , 'H2AFX' , 'H2AFY' , 'H2AFY2' , 'H2AFZ' ,\n'HIST1H2AA' , 'HIST1H2AB' , 'HIST1H2AC' , 'HIST1H2AD' , 'HIST1H2AE' , 'HIST1H2AG' , 'HIST1H2AI' , 'HIST1H2AJ' , 'HIST1H2AK' , 'HIST1H2AL' , 'HIST1H2AM' ,\n'HIST2H2AA3' , 'HIST2H2AC' ,\n'H2BFM' , 'H2BFS' , 'H2BFWT' ,\n'HIST1H2BA' , 'HIST1H2BB' , 'HIST1H2BC' , 'HIST1H2BD' , 'HIST1H2BE' , 'HIST1H2BF' , 'HIST1H2BG' , 'HIST1H2BH' , 'HIST1H2BI' ,\n'HIST1H2BJ' , 'HIST1H2BK' , 'HIST1H2BL' , 'HIST1H2BM' , 'HIST1H2BN' , 'HIST1H2BO' ,\n'HIST2H2BE' ,\n'HIST1H3A' , 'HIST1H3B' , 'HIST1H3C' , 'HIST1H3D' , 'HIST1H3E' , 'HIST1H3F' , 'HIST1H3G' , 'HIST1H3H' , 'HIST1H3I' , 'HIST1H3J' ,\n'HIST2H3C' ,\n'HIST3H3' ,\n'HIST1H4A' , 'HIST1H4B' , 'HIST1H4C' , 'HIST1H4D' , 'HIST1H4E' , 'HIST1H4F' , 'HIST1H4G' , 'HIST1H4H' , 'HIST1H4I' , 'HIST1H4J' , 'HIST1H4K' , 'HIST1H4L' ,\n'HIST4H4' ]\nlist_genes2include_mandotory += list_histone_genes_from_wiki","metadata":{"execution":{"iopub.status.busy":"2021-06-30T08:45:33.276269Z","iopub.execute_input":"2021-06-30T08:45:33.276608Z","iopub.status.idle":"2021-06-30T08:45:33.299841Z","shell.execute_reply.started":"2021-06-30T08:45:33.276579Z","shell.execute_reply":"2021-06-30T08:45:33.298801Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"print(adata_orig.X.sum()) # 547605500.0 - csv #  327662140.0\n","metadata":{"execution":{"iopub.status.busy":"2021-06-30T08:45:33.300914Z","iopub.execute_input":"2021-06-30T08:45:33.301272Z","iopub.status.idle":"2021-06-30T08:45:33.338032Z","shell.execute_reply.started":"2021-06-30T08:45:33.301243Z","shell.execute_reply":"2021-06-30T08:45:33.337134Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"adata_orig.var_names_make_unique()\nadata_orig.var_names_make_unique()","metadata":{"execution":{"iopub.status.busy":"2021-06-30T08:45:33.339101Z","iopub.execute_input":"2021-06-30T08:45:33.339357Z","iopub.status.idle":"2021-06-30T08:45:33.353979Z","shell.execute_reply.started":"2021-06-30T08:45:33.339332Z","shell.execute_reply":"2021-06-30T08:45:33.352878Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"## Preprocessing and conversion from sparse adata.X to dense adata.X using only 10 000 top expressed genes\n","metadata":{}},{"cell_type":"code","source":"from sklearn.neighbors import NearestNeighbors\n\n# First standard preprocessing\n# Then \"pooling\" - fighting with zeros \n\n# Params: \nn_top_genes_to_keep = 10000\n\nthreshold_pct_counts_mt = 40 \nmin_count = 1000 # Threshold on min total_counts\nmax_count = 10000# Threshold on max total_counts\n\nif str_data_inf == ' D085T8 txtfile':\n    threshold_pct_counts_mt = 4\n    min_count = 500\n    max_count = 20000# Threshold on max total_counts\nelif str_data_inf == ' D085T8 loomfile':\n    threshold_pct_counts_mt = 10\n    min_count = 500\n    max_count = 10000# Threshold on max total_counts\nelif str_data_inf == ' sciPlex2_A549 ':\n    threshold_pct_counts_mt = 15\n    min_count = 500\nelif str_data_inf == ' K562_HDACs+vehicle ':\n    threshold_pct_counts_mt = 20\n    min_count = 500\n    max_count = 3000# 12000\nelif 'Lundberg2020' in str_data_inf :\n    if flag_load_from_loom_file:\n        threshold_pct_counts_mt = 32\n        min_count = 150_000\n        max_count = 420_000# 12000\n    threshold_pct_counts_mt = 25\n    min_count = 300_000\n    max_count = 700_000# 12000\n    \n\n\nprint(adata_orig.X.sum())\n\n# ################################################################################################\n# Preprocessing first step: filter CELLs by  counts and level of MT-percent\n# thresholds are set visually looking on violin plots \n# Examples:\n# threshold_pct_counts_mt = 20 - 40  min_count = 500 - 1000; max_count = 10000 - 12000; \n# These thresholds depends on cell line dataset\n\n\nadata = adata_orig.copy()\nprint(adata)\n\nadata.var['mt'] = adata.var_names.str.startswith('MT-')  # annotate the group of mitochondrial genes as 'mt'\n#sv.pp.remove_duplicate_cells(adata)\nsc.pp.calculate_qc_metrics(adata, qc_vars=['mt'], percent_top=None, log1p=False, inplace=True)\nsc.pl.violin(adata, ['n_genes_by_counts', 'n.umi', 'total_counts', 'pct_counts_mt'],jitter=1.9, multi_panel=True)\nmedian_count = np.median(adata.obs['total_counts'])\nprint('Median total counts =',median_count)\nprint('min(n_genes_by_counts):',adata.obs['n_genes_by_counts'].min() )\nprint('min(total_counts):',adata.obs['total_counts'].min() )\n# min_count = np.max((median_count/2,5000))\n\n\nprint('Thresholds: min_count=',min_count,'max_count=',max_count)\n\nprint('Look at total_counts va MT-percent, expect some linear dependence - but does not happen: ')\nsc.pl.scatter(adata, x='total_counts', y='pct_counts_mt')\n\ninds1 = np.where((adata.obs['total_counts']>min_count) & (adata.obs['total_counts']<max_count))\ninds2 = np.where(adata.obs['pct_counts_mt']<threshold_pct_counts_mt)\nprint(len(inds1[0]),'samples pass the count filter')\nprint(len(inds2[0]),' samples pass the mt filter')\nind_samples = np.intersect1d(inds1[0],inds2[0])\nprint('Samples selected',len(ind_samples))\nadata.uns['ind_samples'] = ind_samples\n\n# Here we cut cells. Filtering out those with counts too low or too big\nadata = adata[ind_samples,:]\n\n# ################################################################################################\n# Preprocessing second step: \n# 1) normalization to some value, i.e. median of the total counts\n# 2) taking logs\n# 3) keeping only higly variable genes\n\nsc.pp.normalize_total(adata, target_sum=np.median(adata.obs[\"total_counts\"]))\nsc.pp.log1p(adata)\nsc.pp.highly_variable_genes(adata,n_top_genes=n_top_genes_to_keep,n_bins=20)\nind_genes = np.where(adata.var['highly_variable'])[0]\nind_genes2 = np.where(adata.var.index.isin(list_genes2include_mandotory ) )[0]\nind_genes = list( set( ind_genes )|set( ind_genes2 ) )\nadata = adata[:,ind_genes]\n\nprint('Violin plots after filtering cells and genes')\nsc.pl.violin(adata, ['n_genes_by_counts','n.umi', 'total_counts', 'pct_counts_mt'],jitter=1.9, multi_panel=True)\n# ################################################################################################\n# Preprocessing third step: \n# Mainly: Create non-sparse adata (otherwise next step - pooling works extremely slow)\n# 1) calculate PCA for adata previously calculated adata (which is cutted, normed, logged) - and keep it for use in pooling\n# 2) create new non-sparse adata with cells and genes, obtained by filters on previous steps\n\n#print('Check - we should see ints, otherwise data has been corrupted: ')\n#print(adata_orig.X[:4,:5].toarray())\n\nsc.tl.pca(adata,n_comps=30)\nX_pca = adata.obsm['X_pca']\nX_pca1 = X_pca.copy()\n\nadata_orig[ind_samples,ind_genes].var.shape\nif scipy.sparse.issparse(adata.X):\n    XX2 = adata_orig[ind_samples,ind_genes].X.toarray()\nelse:\n    XX2 = adata_orig[ind_samples,ind_genes].X.copy()\nadata =  anndata.AnnData(XX2,  obs=adata_orig[ind_samples,ind_genes].obs.copy(), \n                     var=adata_orig[ind_samples,ind_genes].var.copy() ) #\nprint('New non-sparse adata: ', adata)\n\n#sv.pp.remove_duplicate_cells(adata)\nadata.var['mt'] = adata.var_names.str.startswith('MT-')  # annotate the group of mitochondrial genes as 'mt'\nsc.pp.calculate_qc_metrics(adata, qc_vars=['mt'], percent_top=None, log1p=False, inplace=True)\n\nprint('Violin plots for new adata')\nsc.pl.violin(adata, ['n_genes_by_counts', 'n.umi', 'total_counts', 'pct_counts_mt'],jitter=1.9, multi_panel=True)\n\n\n# ################################################################################################\n# Preprocessing 4-th step: may take quite long - 10 minutes for 40K celss \n# Mainly: pooling - we are making pooling on original counts, but calculate KNN graph using PCA obtained from normed/log data\n# (that is why we forget adata obtained on the steps 1-2, and will do norm-log again here )\n# 1) pooling\n# 2) quality metrics calculations\n# 3) normalization \n# 4) log\n\nt0 = time.time()\n_smooth_adata_by_pooling(adata,X_pca,n_neighbours=20)\nprint(time.time()-t0,'seconds passed')\nsc.pp.calculate_qc_metrics(adata, percent_top=None, log1p=False, inplace=True)\nsc.pp.normalize_total(adata, target_sum=np.median(adata.obs[\"total_counts\"]))\nsc.pp.log1p(adata)\n#adata = adata[:,ind_genes]\nadata.uns['ind_genes'] = ind_genes\n\nsc.tl.pca(adata,n_comps=30)\nX_pca = adata.obsm['X_pca']\n\nadata","metadata":{"execution":{"iopub.status.busy":"2021-06-30T08:45:33.355641Z","iopub.execute_input":"2021-06-30T08:45:33.356031Z","iopub.status.idle":"2021-06-30T08:45:39.607871Z","shell.execute_reply.started":"2021-06-30T08:45:33.355991Z","shell.execute_reply":"2021-06-30T08:45:39.605854Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"type(adata.X)","metadata":{"execution":{"iopub.status.busy":"2021-06-30T08:45:39.609414Z","iopub.execute_input":"2021-06-30T08:45:39.609827Z","iopub.status.idle":"2021-06-30T08:45:39.616495Z","shell.execute_reply.started":"2021-06-30T08:45:39.609785Z","shell.execute_reply":"2021-06-30T08:45:39.615595Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# Correlation analysis","metadata":{}},{"cell_type":"code","source":"number_top_genes_to_consider = 1_000\nv = np.asarray(adata.X.sum(axis = 0 ) ).ravel()\nv.shape\nI = np.argsort(v)\nI = I[::-1]\nI = I[:number_top_genes_to_consider]\nnumber_top_genes_to_consider = 10_000\nv = np.asarray(adata.X.sum(axis = 0 ) ).ravel()\nv.shape\nI = np.argsort(v)\nI = I[::-1]\nI = I[:number_top_genes_to_consider]\n\nmask = np.ones( adata.X.shape[0]).astype(bool) \nt0 = time.time()\nif scipy.sparse.issparse(adata.X):\n    corr_matr = np.corrcoef(adata[mask].X[:,I].toarray().T) # Hint - use numpy , pandas is MUCH SLOWER   (df.corr() )\nelse:\n    corr_matr = np.corrcoef(adata[mask].X[:,I].T) # Hint - use numpy , pandas is MUCH SLOWER   (df.corr() )\nprint('Calculated correlation matrix. Shape: ',corr_matr.shape,'  seconds passed:' , np.round( time.time() - t0,1) )\n\nprint()\nprint(np.min(corr_matr ), 'minimal correlation' )\ncorr_matr_abs = np.abs( corr_matr )\nprint(np.mean(corr_matr_abs ), 'average absolute correlation' )\nprint(np.median(corr_matr_abs), 'median absolute correlation' )\nprint(np.min(corr_matr_abs ), 'min absolute correlation' )\nprint(np.std(corr_matr_abs ), 'std absolute correlation' )\nprint()\n\nv = corr_matr.flatten()\nplt.figure(figsize=(14,8))\nt0 = time.time()\nplt.hist(v, bins = 50)\nplt.title('correlation coefficients distribution')\nplt.show()\nprint(time.time() - t0, 'seconds passed')\n\nprint(np.min(corr_matr ), 'minimal correlation' )\nprint(np.mean(corr_matr_abs ), 'average absolute correlation' )\nprint(np.median(corr_matr_abs), 'median absolute correlation' )\nprint(np.min(corr_matr_abs ), 'min absolute correlation' )\nprint(np.std(corr_matr_abs ), 'std absolute correlation' )\nfor t in [0.5,0.6, 0.7,0.8,0.9,0.95,0.97,0.98,.99]:\n    print( ((np.abs(v) < 0.99999999) & (np.abs(v) > t)).sum()/2 , 'number of pairs correlated more than', t  )\nv.shape\n\nprint()\nprint()\nfor threshold_corr in [0.99,0.98,00.97,0.95,0.93,0.90,0.85,0.8,0.75,0.7,0.65,0.6,0.55,.5,0.45,0.4,0.35,0.3]: \n    a,b = np.where( np.triu(np.abs(corr_matr),1) > threshold_corr )\n    if len(a) > 50: break\nprint('for threshold:', threshold_corr,'we have ',len(a),' correlated pairs' )\n\n\nd = pd.DataFrame()\nfor i in range(len(a)):\n    Iai, Ibi =  I[a[i]], I[b[i]]\n    d.loc[i,'Gene1'] = adata[mask].var.index[Iai] \n    d.loc[i,'Gene2'] = adata[mask].var.index[Ibi ] \n    d.loc[i,'Correlation'] = corr_matr[a[i],b[i]] \n    d.loc[i,'Correlation Abs'] = np.abs(d.loc[i,'Correlation'] )\n    d.loc[i,'Gene1 Expression Sum'] = np.asarray(adata[mask].X[:,Iai].sum(axis = 0)).ravel()[0]\n    d.loc[i,'Gene2 Expression Sum'] = np.asarray(adata[mask].X[:,Ibi].sum(axis = 0)).ravel()[0]\n    \nprint(time.time() - t0, 'seconds passed')\n    \ndisplay( d.sort_values('Correlation Abs',ascending = False).head(30) )\n\n\n#clustermap\nimport seaborn as sns\nt0 = time.time()\nsns.clustermap(np.abs(corr_matr[:3000,:3000]),cmap='vlag');\nprint( np.round(time.time()- t0,1), ' seconds passed.')\n\n","metadata":{"execution":{"iopub.status.busy":"2021-06-30T08:45:39.618182Z","iopub.execute_input":"2021-06-30T08:45:39.618646Z","iopub.status.idle":"2021-06-30T08:46:48.873462Z","shell.execute_reply.started":"2021-06-30T08:45:39.618602Z","shell.execute_reply":"2021-06-30T08:46:48.872679Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"t0=time.time()\nw,v = np.linalg.eig(corr_matr[:2000,:2000])\nprint(time.time()-t0, 'seconds passed')\nw = np.real(w.ravel())\nw = np.sort(w)[::-1]\nplt.plot(w,'*-')\nplt.title('Eigenvalues of correlation matrix ')\nplt.show()","metadata":{"execution":{"iopub.status.busy":"2021-06-30T08:46:48.87453Z","iopub.execute_input":"2021-06-30T08:46:48.874942Z","iopub.status.idle":"2021-06-30T08:46:54.937862Z","shell.execute_reply.started":"2021-06-30T08:46:48.874909Z","shell.execute_reply":"2021-06-30T08:46:54.937048Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# Visualizatios","metadata":{}},{"cell_type":"markdown","source":"## Compare PCA before and after pooling ","metadata":{}},{"cell_type":"code","source":"mask = np.ones(len(adata)).astype(bool)\n\nfor color_by_mode in  ['cell_cycle_phase', 'n.umi', 'pct_counts_mt']:\n    if color_by_mode not in adata.obs:\n        color_by_mode = None\n    color_by_field_name = color_by_mode\n\n    fig = plt.figure(figsize = (25,12))\n    plt.suptitle(str_data_inf + ' PCA before pooling and after.  n_cells: ' + str(mask.sum()), fontsize = 20   )#' \n    \n    c = 0; \n    #for col in ['cell_type', 'n.umi', 'pct_counts_mt']:\n    if 1:\n        c+=1; fig.add_subplot(1, 2 , c)         \n        #color_by = adata[mask].obs.drug #color_by = adata[mask].obs.phase\n        #color_by = (adata[mask].obs[col])\n        #color_by = (adata.obs['product_name'])\n\n        if color_by_mode is None:\n            ax = sns.scatterplot(x=X_pca1[:,0],y=X_pca1[:,1])# ,   hue= color_by)#, alpha = 0.7, marker = '.', legend = None)\n        else:\n            color_by = (adata.obs[color_by_field_name])\n            ax = sns.scatterplot(x=X_pca1[:,0],y=X_pca1[:,1] ,   hue= color_by, palette = \"viridis\")# sns.color_palette(\"viridis\") )#, alpha = 0.7, marker = '.', legend = None)\n            plt.setp(ax.get_legend().get_texts(), fontsize='20') # for legend text\n            plt.setp(ax.get_legend().get_title(), fontsize='20') # for legend title\n\n        #plt.setp(ax.get_legend().get_texts(), fontsize='15') # for legend text\n        #plt.setp(ax.get_legend().get_title(), fontsize='15') # for legend title\n        plt.title('PCA without pooling. (ALL genes)', fontsize=20 )\n        plt.xlabel('PCA1', fontsize=20 )\n        plt.ylabel('PCA2', fontsize=20 )\n        c+=1; fig.add_subplot(1, 2 , c)         \n        #color_by = adata[mask].obs.drug #color_by = adata[mask].obs.phase\n        #color_by = (adata.obs['product_name'])\n        if color_by_mode is None:\n            ax = sns.scatterplot(x=X_pca[:,0],y=X_pca[:,1])# ,   hue= color_by)#, alpha = 0.7, marker = '.', legend = None)\n        else:\n            color_by = (adata.obs[color_by_field_name])\n            #color_by = (adata.obs[color_by_field_name]) > np.median( adata.obs[color_by_field_name].values ) \n            ax = sns.scatterplot(x=X_pca[:,0],y=X_pca[:,1] ,   hue= color_by, palette = \"viridis\")# sns.color_palette(\"viridis\") )#, alpha = 0.7, marker = '.', legend = None)\n            plt.setp(ax.get_legend().get_texts(), fontsize='20') # for legend text\n            plt.setp(ax.get_legend().get_title(), fontsize='20') # for legend title\n        #plt.setp(ax.get_legend().get_texts(), fontsize='15') # for legend text\n        #plt.setp(ax.get_legend().get_title(), fontsize='15') # for legend title\n        plt.title('PCA WITH pooling. (ALL genes)', fontsize=20 )\n        plt.xlabel('PCA1', fontsize=20 )\n        plt.ylabel('PCA2', fontsize=20 )","metadata":{"execution":{"iopub.status.busy":"2021-06-30T08:46:54.939206Z","iopub.execute_input":"2021-06-30T08:46:54.93959Z","iopub.status.idle":"2021-06-30T08:46:56.468605Z","shell.execute_reply.started":"2021-06-30T08:46:54.93955Z","shell.execute_reply":"2021-06-30T08:46:56.46783Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"## PCA plots in several axes","metadata":{}},{"cell_type":"code","source":"mask = np.ones(len(adata)).astype(bool)\nfrom sklearn.decomposition import PCA\npca1 = PCA(n_components=5)\n\nlist_genes = list(  adata.var.index )\nI = np.where(adata.var.index.isin(list_genes ) )[0]\nr = pca1.fit_transform(adata[mask].X[:,I])\n        \nfor color_by_mode in  ['n.umi']:# 'n.umi']: # , 'pct_counts_mt']:\n    if color_by_mode not in adata.obs:\n        color_by_mode = None\n    color_by_field_name = color_by_mode\n        \n\n    c = 0\n    for i,j in [(0,1), (0,2),(1,2),(0,3),(1,3),(2,3)]:\n        if c % 3 == 0:\n            fig = plt.figure(figsize = (25,12))\n            plt.suptitle(str_data_inf + ' Change  PCA axes.  PCA from all genes.  n_cells: '\\\n                 +str(mask.sum()) + ' n_genes: ' + str(len(I))  , fontsize = 20 )   #' \n            c = 0; \n        \n        c+=1; fig.add_subplot(1, 3 , c)\n        if color_by_mode is None:\n            ax = sns.scatterplot(x=r[:,i],y=r[:,j])# ,   hue= color_by)#, alpha = 0.7, marker = '.', legend = None)\n        else:\n            color_by = (adata.obs[color_by_field_name])\n            ax = sns.scatterplot(x=r[:,i],y=r[:,j] ,   hue= color_by, palette =\"viridis\")#  sns.color_palette(\"viridis\", as_cmap=True) )#, alpha = 0.7, marker = '.', legend = None)\n            plt.setp(ax.get_legend().get_texts(), fontsize='20') # for legend text\n            plt.setp(ax.get_legend().get_title(), fontsize='20') # for legend title\n        plt.xlabel('PCA'+str(i+1),  fontsize = 20)\n        plt.ylabel('PCA'+str(j+1),  fontsize = 20)\n        plt.title('PCA: '+ str(i+1) +' '+str(j+1),  fontsize = 20)\n        ","metadata":{"execution":{"iopub.status.busy":"2021-06-30T08:46:56.469583Z","iopub.execute_input":"2021-06-30T08:46:56.469956Z","iopub.status.idle":"2021-06-30T08:46:59.451725Z","shell.execute_reply.started":"2021-06-30T08:46:56.46993Z","shell.execute_reply":"2021-06-30T08:46:59.450732Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"## UMAP","metadata":{}},{"cell_type":"code","source":"import scipy\nimport umap \nt0 = time.time()\nreducer = umap.UMAP()\nr = reducer.fit_transform(adata.X)\n\nfig = plt.figure(figsize = (25,12))\nplt.title(str_data_inf + ' UMAP n_cells: ' + str(adata.X.shape[0]) +\\\n            ' n_genes: ' + str(adata.X.shape[1])  + ' '  , fontsize = 20   )#' \nif 'n.umi' not in adata.obs.columns: \n    ax = sns.scatterplot(x=v1, y = v2)#,  hue= color_by)#, palette = color_palette )#, palette = \"viridis\")# sns.color_palette(\"viridis\", as_cmap=True),\n                        #)# ,   alpha = 0.8, marker = '.')#, )#, legend=None)\nelse:\n    color_by = adata.obs['n.umi'].values\n    ax = sns.scatterplot(x=r[:,0], y = r[:,1],  hue= color_by)#, palette = color_palette )#, palette = \"viridis\")# sns.color_palette(\"viridis\", as_cmap=True),\n                        #)# ,   alpha = 0.8, marker = '.')#, )#, legend=None)\n    plt.setp(ax.get_legend().get_texts(), fontsize='20') # for legend text\n    plt.setp(ax.get_legend().get_title(), fontsize='20') # for legend title\n\nplt.xlabel('UMAP1' , fontsize = 20 )\nplt.ylabel('UMAP2' , fontsize = 20 )\nplt.show()\nprint( np.round( time.time() - t0,1) , ' seconds passed ' ) ","metadata":{"execution":{"iopub.status.busy":"2021-06-30T08:46:59.453079Z","iopub.execute_input":"2021-06-30T08:46:59.453378Z","iopub.status.idle":"2021-06-30T08:47:14.280102Z","shell.execute_reply.started":"2021-06-30T08:46:59.453349Z","shell.execute_reply":"2021-06-30T08:47:14.278975Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"import scipy\nimport umap \nt0 = time.time()\nn_neighbors = 50\nmin_dist = 0.7\nreducer = umap.UMAP(n_neighbors=n_neighbors, min_dist=min_dist,)\nr = reducer.fit_transform(adata.X)\n\nfig = plt.figure(figsize = (25,12))\nplt.title(str_data_inf + ' UMAP('+ str(n_neighbors)+','+str( min_dist )+ ') n_cells: ' + str(adata.X.shape[0]) +\\\n            ' n_genes: ' + str(adata.X.shape[1])  + ' '  , fontsize = 20   )#' \nif 'n.umi' not in adata.obs.columns: \n    ax = sns.scatterplot(x=v1, y = v2)#,  hue= color_by)#, palette = color_palette )#, palette = \"viridis\")# sns.color_palette(\"viridis\", as_cmap=True),\n                        #)# ,   alpha = 0.8, marker = '.')#, )#, legend=None)\nelse:\n    color_by = adata.obs['n.umi'].values\n    ax = sns.scatterplot(x=r[:,0], y = r[:,1],  hue= color_by)#, palette = color_palette )#, palette = \"viridis\")# sns.color_palette(\"viridis\", as_cmap=True),\n                        #)# ,   alpha = 0.8, marker = '.')#, )#, legend=None)\n    plt.setp(ax.get_legend().get_texts(), fontsize='20') # for legend text\n    plt.setp(ax.get_legend().get_title(), fontsize='20') # for legend title\n\nplt.xlabel('UMAP1' , fontsize = 20 )\nplt.ylabel('UMAP2' , fontsize = 20 )\nplt.show()\nprint( np.round( time.time() - t0,1) , ' seconds passed ' ) ","metadata":{"execution":{"iopub.status.busy":"2021-06-30T08:47:14.284394Z","iopub.execute_input":"2021-06-30T08:47:14.2848Z","iopub.status.idle":"2021-06-30T08:47:30.248132Z","shell.execute_reply.started":"2021-06-30T08:47:14.284759Z","shell.execute_reply":"2021-06-30T08:47:30.244623Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"## Plots in genes axes","metadata":{}},{"cell_type":"code","source":"mask = np.ones( adata.X.shape[0]).astype(bool) # \n\ngene1 = 'E2F1' \ngene2 = 'FOXM1'\nn_x_subplots = 4\n\nc = 0 \nfor gene1,gene2 in [ ('E2F1','FOXM1'), ('PCNA','TOP2A'), ('CCNB1', 'CCNB2'), ('CCNE1', 'CCNB1'),\n                   ('E2F1','PCNA'),('FOXM1','PCNA'), ('E2F1','TOP2A'),('FOXM1','TOP2A'), ]: \n    #,'PCNA','TOP2A', 'CCNE1', 'CCNE2', 'CDK2', 'CCNB1','CCNB2','CCNB3','CCNA2',  ]: # , 'pct_counts_mt']:\n    if gene1 not in adata.var.index : continue\n    if gene2 not in adata.var.index : continue\n        \n    color_by_mode = 'n.umi' # ,#'cell_cycle_phase', #  'PCA1ALL', 'n.umi', 'threshold_binarize_5_PCA1ALL',\n    if color_by_mode in adata.obs:\n        color_by_field_name = color_by_mode\n        color_by = (adata[mask].obs[color_by_field_name])\n    else:\n        color_by_mode = None\n    \n    if c % n_x_subplots == 0:\n        fig = plt.figure(figsize = (20,6) ); c = 0\n        plt.suptitle(str_data_inf + ' Genes axes   n_cells: ' + str(adata.X.shape[0]) +\\\n            ' n_genes: ' + str(adata.X.shape[1])  + ' color by '+ str(color_by_mode) , fontsize = 20   )#' \n        c = 0\n        \n    c += 1; fig.add_subplot(1,n_x_subplots ,c)\n    plt.title(gene1 + ' ' + gene2) # str(color_by_mode) + ' PCA '+str(i)+','+str(j))\n\n    plt.legend()\n    I_gene = np.where(adata.var.index == gene1)[0]\n    v = adata[mask].X[:,I_gene]\n    if scipy.sparse.issparse(adata.X):\n        v = np.asarray(v.toarray()).ravel()\n    else:\n        v = np.asarray(v).ravel()\n        \n    v1 = v\n    I_gene = np.where(adata.var.index == gene2)[0]\n    v = adata[mask].X[:,I_gene]\n    if scipy.sparse.issparse(adata.X):\n        v = np.asarray(v.toarray()).ravel()\n    else:\n        v = np.asarray(v).ravel()\n    v2 = v\n\n\n    if color_by_mode is None:\n        ax = sns.scatterplot(x=v1,y=v2)# ,   hue= color_by)#, alpha = 0.7, marker = '.', legend = None)\n    else:\n        ax = sns.scatterplot(x=v1,y=v2,   hue= color_by)#, alpha = 0.7, marker = '.', legend = None)\n        plt.setp(ax.get_legend().get_texts(), fontsize='15') # for legend text\n        plt.setp(ax.get_legend().get_title(), fontsize='15') # for legend title    plt.xlabel('PCA1' , fontsize = 20 )\n        ax.legend().set_visible(False)\n\n    plt.xlabel(gene1,  fontsize = 20)\n    plt.ylabel(gene2,  fontsize = 20)\nplt.show()  ","metadata":{"execution":{"iopub.status.busy":"2021-06-30T08:47:30.249812Z","iopub.execute_input":"2021-06-30T08:47:30.25024Z","iopub.status.idle":"2021-06-30T08:47:32.331387Z","shell.execute_reply.started":"2021-06-30T08:47:30.250194Z","shell.execute_reply":"2021-06-30T08:47:32.330469Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"print(np.round(time.time()-t0start,1),  np.round( (time.time()-t0start)/60,1),  np.round( (time.time()-t0start)/3600,1), \n      'seconds, munutes, hours  passed')","metadata":{"execution":{"iopub.status.busy":"2021-06-30T08:47:32.332955Z","iopub.execute_input":"2021-06-30T08:47:32.333334Z","iopub.status.idle":"2021-06-30T08:47:32.34025Z","shell.execute_reply.started":"2021-06-30T08:47:32.333294Z","shell.execute_reply":"2021-06-30T08:47:32.339118Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"","metadata":{},"execution_count":null,"outputs":[]}]}