{"metadata":{"kernelspec":{"language":"python","display_name":"Python 3","name":"python3"},"language_info":{"pygments_lexer":"ipython3","nbconvert_exporter":"python","version":"3.6.4","file_extension":".py","codemirror_mode":{"name":"ipython","version":3},"name":"python","mimetype":"text/x-python"}},"nbformat_minor":4,"nbformat":4,"cells":[{"cell_type":"markdown","source":"# What is about ?\n\nData - single RNA-seq genes expressions data. \nMainly subparts extracted from very huge https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE139944\nwhich does not fit into 16G memory.\n\nK562,MCF7, A549 - names of well-known cell-lines \n\nVersion 1: \nTake a file: fn = '/kaggle/input/scrnaseq-exposed-to-multiple-compounds/sciPlex3_MCF7_A549_K562_only_vehicle.h5ad'\nstr_data_inf = ' sciPlex3_Vehicle '\ncontaining only vehicle (no drugs) for the three cell lines and look at it. \n","metadata":{}},{"cell_type":"code","source":"# This Python 3 environment comes with many helpful analytics libraries installed\n# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python\n# For example, here's several helpful packages to load\n\nimport numpy as np # linear algebra\nimport pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)\n\n# Input data files are available in the read-only \"../input/\" directory\n# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory\n\nimport os\nfor dirname, _, filenames in os.walk('/kaggle/input'):\n    for filename in filenames:\n        print(os.path.join(dirname, filename))\n\n# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using \"Save & Run All\" \n# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session","metadata":{"_uuid":"8f2839f25d086af736a60e9eeb907d3b93b6e0e5","_cell_guid":"b1076dfc-b9ad-4769-8c92-a6c4dae69d19","execution":{"iopub.status.busy":"2021-05-25T16:22:29.682215Z","iopub.execute_input":"2021-05-25T16:22:29.682673Z","iopub.status.idle":"2021-05-25T16:22:29.701304Z","shell.execute_reply.started":"2021-05-25T16:22:29.682637Z","shell.execute_reply":"2021-05-25T16:22:29.700043Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"import time\nimport matplotlib.pyplot as plt\nimport seaborn as sns\n\n!pip install scanpy\n#!pip install scvelo\n#!pip install git+https://github.com/csgroen/scycle.git#egg=scycle\n#!pip install  --no-dependencies  git+https://github.com/j-bac/elpigraph-python.git\n# from elpigraph_ps_tools import *\n#import elpigraph\nimport scanpy as sc\nimport anndata\n\n\n\nimport importlib\n\nimport pandas as pd\nimport numpy as np\n#import scanpy as sc\n#import scycle as cc\n# import scvelo as sv\n#import anndata\nfrom sklearn.decomposition import PCA\n\nimport matplotlib as mpl\nimport matplotlib.pyplot as plt\nfrom sklearn.neighbors import NearestNeighbors\n\nmpl.rcParams['figure.dpi'] = 70\n\nimport os\n\nimport sys\n#DATA_PATH = '/mnt/c/Datas/SingleCellTranscriptomics/sciPlex/'\n#sys.path.append(DATA_PATH)\nplt.style.use('dark_background')\n\n\n\n","metadata":{"execution":{"iopub.status.busy":"2021-05-25T16:22:29.772325Z","iopub.execute_input":"2021-05-25T16:22:29.772908Z","iopub.status.idle":"2021-05-25T16:22:37.277926Z","shell.execute_reply.started":"2021-05-25T16:22:29.772858Z","shell.execute_reply":"2021-05-25T16:22:37.276557Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"fn = '/kaggle/input/scrnaseq-exposed-to-multiple-compounds/sciPlex3_MCF7_zero_dose.h5ad'\nstr_data_inf = ' MCF7dose0 '\nfn = '/kaggle/input/scrnaseq-exposed-to-multiple-compounds/sciPlex3_K562_zero_dose.h5ad'\nstr_data_inf = ' K562dose0 '\nfn = '/kaggle/input/scrnaseq-exposed-to-multiple-compounds/sciPlex3_A549_zero_dose.h5ad'\nstr_data_inf = ' A549dose0 '\nfn = '/kaggle/input/scrnaseq-exposed-to-multiple-compounds/sciPlex3_MCF7_HDACs.h5ad'\nstr_data_inf = ' MCF7_HDACs '\nfn = '/kaggle/input/scrnaseq-exposed-to-multiple-compounds/sciPlex3_A549_HDACs.h5ad'\nstr_data_inf = ' A549_HDACs '\n#fn = '/kaggle/input/scrnaseq-exposed-to-multiple-compounds/sciPlex3_K562_HDACs.h5ad'\n#str_data_inf = ' K562_HDACs '\nfn = '/kaggle/input/scrnaseq-exposed-to-multiple-compounds/sciPlex3_K562_HDACs_and_vehicle.h5ad'\nstr_data_inf = ' K562_HDACs+vehicle '\nfn = '/kaggle/input/scrnaseq-exposed-to-multiple-compounds/sciPlex3_A549_HDACs_and_vehicle.h5ad'\nstr_data_inf = ' A549_HDACs+vehicle '\nfn = '/kaggle/input/scrnaseq-exposed-to-multiple-compounds/sciPlex3_MCF7_HDACs_and_vehicle.h5ad'\nstr_data_inf = ' MCF7_HDACs+vehicle '\nfn = '/kaggle/input/scrnaseq-exposed-to-multiple-compounds/sciPlex3_K562_not_HDACs_and_vehicle.h5ad'\nstr_data_inf = ' K562_not_HDACs+vehicle '\nfn = '/kaggle/input/scrnaseq-exposed-to-multiple-compounds/sciPlex3_K562_HDACs_and_vehicle.h5ad'\nstr_data_inf = ' K562_HDACs+vehicle '\nfn = '/kaggle/input/scrnaseq-exposed-to-multiple-compounds/sciPlex3_A549_HDACs_and_vehicle.h5ad'\nstr_data_inf = ' A549_HDACs+vehicle '\nfn = '/kaggle/input/scrnaseq-exposed-to-multiple-compounds/sciPlex3_MCF7_HDACs_and_vehicle.h5ad'\nstr_data_inf = ' MCF7_HDACs+vehicle '\nfn = '/kaggle/input/scrnaseq-exposed-to-multiple-compounds/sciPlex2_A549_all4_drugs_and_doses.h5ad'\nstr_data_inf = ' sciPlex2_A549 '\nfn = '/kaggle/input/scrnaseq-exposed-to-multiple-compounds/sciPlex3_A549_JAK_Aurora Kinase_PARP_Histone Methyltransferase_and_vehicle.h5ad'\nstr_data_inf = ' sciPlex3_A549_JAK_AuroraKinase_PARP_Histone Methyltransferase '\nfn = '/kaggle/input/scrnaseq-exposed-to-multiple-compounds/sciPlex1_HEK293T.h5ad'\nstr_data_inf = ' sciPlex1_HEK293T '\nfn = '/kaggle/input/scrnaseq-exposed-to-multiple-compounds/sciPlex3_MCF7_JAK_Aurora Kinase_PARP_Histone Methyltransferase_and_vehicle.h5ad'\nstr_data_inf = ' sciPlex3_MCF7_JAK_AuroraKinase_PARP_Histone Methyltransferase '\nfn = '/kaggle/input/scrnaseq-exposed-to-multiple-compounds/sciPlex3_MCF7_A549_K562_only_vehicle.h5ad'\nstr_data_inf = ' sciPlex3_Vehicle '\n","metadata":{"execution":{"iopub.status.busy":"2021-05-25T16:22:37.28016Z","iopub.execute_input":"2021-05-25T16:22:37.280542Z","iopub.status.idle":"2021-05-25T16:22:37.290065Z","shell.execute_reply.started":"2021-05-25T16:22:37.280494Z","shell.execute_reply":"2021-05-25T16:22:37.288861Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# Load","metadata":{}},{"cell_type":"code","source":"fn = '/kaggle/input/scrnaseq-exposed-to-multiple-compounds/sciPlex3_MCF7_A549_K562_only_vehicle.h5ad'\nstr_data_inf = ' sciPlex3_Vehicle '\n\n\nadata_orig = sc.read_h5ad(fn)\nadata_orig","metadata":{"execution":{"iopub.status.busy":"2021-05-25T16:22:37.292421Z","iopub.execute_input":"2021-05-25T16:22:37.292814Z","iopub.status.idle":"2021-05-25T16:22:39.107783Z","shell.execute_reply.started":"2021-05-25T16:22:37.292773Z","shell.execute_reply":"2021-05-25T16:22:39.106521Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"adata_orig = adata_orig[:,:58347] # Cut some strange presence of mouse genes \nprint(adata_orig)\nadata_orig.var_names_make_unique()\nadata_orig.var_names_make_unique()\n","metadata":{"execution":{"iopub.status.busy":"2021-05-25T16:22:39.109587Z","iopub.execute_input":"2021-05-25T16:22:39.109927Z","iopub.status.idle":"2021-05-25T16:22:39.723034Z","shell.execute_reply.started":"2021-05-25T16:22:39.109892Z","shell.execute_reply":"2021-05-25T16:22:39.721851Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# Show some statistics ","metadata":{}},{"cell_type":"code","source":"vec_count_sum_orig = adata_orig.X.sum(axis = 1)\nvec_count_sum_orig = np.asarray(vec_count_sum_orig).ravel() \nprint(type(vec_count_sum_orig), vec_count_sum_orig.shape)\nplt.plot( np.sort(vec_count_sum_orig  ) )\nplt.xlabel('cell sorted')\nplt.ylabel('counts per cell')\nplt.title('Counts per cell')\nadata_orig.obs.count_orig = vec_count_sum_orig","metadata":{"execution":{"iopub.status.busy":"2021-05-25T16:22:39.724765Z","iopub.execute_input":"2021-05-25T16:22:39.725209Z","iopub.status.idle":"2021-05-25T16:22:39.92808Z","shell.execute_reply.started":"2021-05-25T16:22:39.725158Z","shell.execute_reply":"2021-05-25T16:22:39.926867Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# ","metadata":{}},{"cell_type":"markdown","source":"# Preprocessing and conversion from sparse adata.X to dense adata.X using only 10 000 top expressed genes ","metadata":{}},{"cell_type":"code","source":"# standard preprocessing and creating Dense adata with cutted cells and genes\n# preprocessing: 1) cutting cells 2) norm+log 3) cutting genes\n\n# Thresholds to cut cells: \nthreshold_pct_counts_mt = 50\nmin_count = 500\nmax_count = 12000\n# threshold for number of genes to keep\nn_top_genes_keep = 10001\n\nadata = adata_orig\nadata.var['mt'] = adata.var_names.str.startswith('MT-')  # annotate the group of mitochondrial genes as 'mt'\nsc.pp.calculate_qc_metrics(adata, qc_vars=['mt'], percent_top=None, log1p=False, inplace=True)\n#sc.pl.violin(adata, ['n_genes_by_counts', 'total_counts', 'pct_counts_mt'],jitter=0.4, multi_panel=True)\n#print('min_count=',min_count,'max_count=',max_count)\n#sc.pl.scatter(adata, x='total_counts', y='pct_counts_mt')\ninds1 = np.where((adata.obs['total_counts']>min_count) & (adata.obs['total_counts']<max_count))\ninds2 = np.where(adata.obs['pct_counts_mt']<threshold_pct_counts_mt)\n#print(len(inds1[0]),'samples pass the count filter')\n#print(len(inds2[0]),' samples pass the mt filter')\nind_samples = np.intersect1d(inds1[0],inds2[0])\n#print('Samples selected',len(ind_samples))\n#adata.uns['ind_samples'] = ind_samples\n# Here we cut cells. Filtering out those with counts too low or too big\nadata = adata[ind_samples,:]\n\n# Here we norm+log and cut genes\nsc.pp.normalize_total(adata, target_sum=np.median(adata.obs[\"total_counts\"]))\nsc.pp.log1p(adata)\nsc.pp.highly_variable_genes(adata,n_top_genes=n_top_genes_keep,n_bins=20)\nind_genes = np.where(adata.var['highly_variable'])[0]\nadata = adata[:,ind_genes]\n\n# Here we create dense matrix:\n# \n# adata_orig[ind_samples,ind_genes].var.shape\nXX2 = adata.X.toarray()\nadata =  anndata.AnnData(XX2,  obs=adata.obs.copy(), var=adata.var.copy() ) #\nadata\n\n","metadata":{"execution":{"iopub.status.busy":"2021-05-25T16:22:39.929884Z","iopub.execute_input":"2021-05-25T16:22:39.930347Z","iopub.status.idle":"2021-05-25T16:22:43.929981Z","shell.execute_reply.started":"2021-05-25T16:22:39.930299Z","shell.execute_reply":"2021-05-25T16:22:43.928881Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"print(adata_orig.X[:5,:5].toarray())\nprint(adata.X[:5,:5])\n","metadata":{"execution":{"iopub.status.busy":"2021-05-25T16:22:43.93159Z","iopub.execute_input":"2021-05-25T16:22:43.931961Z","iopub.status.idle":"2021-05-25T16:22:43.940184Z","shell.execute_reply.started":"2021-05-25T16:22:43.931922Z","shell.execute_reply":"2021-05-25T16:22:43.939022Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"import time\nt0 = time.time()\nsc.tl.pca(adata,n_comps=30)\nprint(time.time() - t0,'secs passed' )","metadata":{"execution":{"iopub.status.busy":"2021-05-25T16:22:43.942685Z","iopub.execute_input":"2021-05-25T16:22:43.943042Z","iopub.status.idle":"2021-05-25T16:22:59.325433Z","shell.execute_reply.started":"2021-05-25T16:22:43.943004Z","shell.execute_reply":"2021-05-25T16:22:59.324093Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"adata","metadata":{"execution":{"iopub.status.busy":"2021-05-25T16:33:49.600397Z","iopub.execute_input":"2021-05-25T16:33:49.601077Z","iopub.status.idle":"2021-05-25T16:33:49.610232Z","shell.execute_reply.started":"2021-05-25T16:33:49.601018Z","shell.execute_reply":"2021-05-25T16:33:49.608955Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"sc.pl.pca(adata,color= 'cell_type' )\nsc.pl.pca(adata,color='dose')\nsc.pl.pca(adata,color= 'time_point' )\nsc.pl.pca(adata,color='n.umi')\nsc.pl.pca(adata,color = 'pct_counts_mt' )\nsc.pl.pca(adata,color='MALAT1')\nsc.pl.pca(adata,color= 'product_name' )\n","metadata":{"execution":{"iopub.status.busy":"2021-05-25T16:37:13.001563Z","iopub.execute_input":"2021-05-25T16:37:13.002088Z","iopub.status.idle":"2021-05-25T16:37:14.849428Z","shell.execute_reply.started":"2021-05-25T16:37:13.002041Z","shell.execute_reply":"2021-05-25T16:37:14.848068Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"","metadata":{"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"import time\nt0=time.time()\ndrug = 'Vehicle'\nmask = adata.obs['product_name'] == drug\nadata_m = adata[mask]\nsc.pp.neighbors(adata_m)\nsc.tl.umap(adata_m)#,color='n.umi')\nsc.pl.umap(adata_m, color='n.umi')\nprint(time.time()-t0,'seconds passed')\n","metadata":{"execution":{"iopub.status.busy":"2021-05-25T16:23:00.699631Z","iopub.execute_input":"2021-05-25T16:23:00.699975Z","iopub.status.idle":"2021-05-25T16:23:29.05696Z","shell.execute_reply.started":"2021-05-25T16:23:00.69994Z","shell.execute_reply":"2021-05-25T16:23:29.055639Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"sc.pl.umap(adata_m, color='pct_counts_mt' )\n","metadata":{"execution":{"iopub.status.busy":"2021-05-25T16:23:29.058833Z","iopub.execute_input":"2021-05-25T16:23:29.059221Z","iopub.status.idle":"2021-05-25T16:23:29.326781Z","shell.execute_reply.started":"2021-05-25T16:23:29.059185Z","shell.execute_reply":"2021-05-25T16:23:29.325433Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"print(adata_m.obsm['X_umap'].shape)\nadata_m","metadata":{"execution":{"iopub.status.busy":"2021-05-25T16:23:29.328676Z","iopub.execute_input":"2021-05-25T16:23:29.329179Z","iopub.status.idle":"2021-05-25T16:23:29.338613Z","shell.execute_reply.started":"2021-05-25T16:23:29.329126Z","shell.execute_reply":"2021-05-25T16:23:29.337227Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"","metadata":{"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# Cell types plots\n","metadata":{}},{"cell_type":"code","source":"list_drugs = list(adata.obs['product_name'].value_counts().index)\ndrug = 'Vehicle'\n\ncol =  'n.umi' # 'dose'#  'n.umi' #\n\nn_subplots_x = 4\nfor cell_type in ['MCF7','K562','A549']:# list_drugs:# [:2]:\n    mask = adata.obs['cell_type'] == cell_type\n    #mask = mask & ( adata.obs['time_point'] == 24) # .obs[col]\n    X = adata.obsm['X_pca'][mask.values,:]# adata.obsm.X_pca[mask.values,:]\n    \n    fig = plt.figure(figsize = (20,5) ); c = 0      \n    plt.suptitle( cell_type + ' ' +str( mask.sum() ) )\n    \n    for (i,j) in [(0,1),(1,2),(0,4),(2,3)]:\n        c += 1; fig.add_subplot(1,n_subplots_x,c)\n        sns.scatterplot(x = X[:,i], y = X[:,j], hue = np.log10(adata.obs[col][mask]) )\n        plt.xlabel('PCA'+str(i))\n        plt.ylabel('PCA'+str(j))\n        plt.title('PCA '+(str(i)) + ' ' + str(j))\n    \n","metadata":{"execution":{"iopub.status.busy":"2021-05-25T16:23:29.34058Z","iopub.execute_input":"2021-05-25T16:23:29.341091Z","iopub.status.idle":"2021-05-25T16:23:36.401444Z","shell.execute_reply.started":"2021-05-25T16:23:29.34104Z","shell.execute_reply":"2021-05-25T16:23:36.40057Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"col =  'n.umi' # 'dose'#  'n.umi' #\n\n#n_subplots_x = 4\nfor cell_type in ['MCF7','K562','A549']:# list_drugs:# [:2]:\n    mask = adata.obs['cell_type'] == cell_type\n    #mask = mask & ( adata.obs['time_point'] == 24) # .obs[col]\n    X = adata.obsm['X_pca'][mask.values,:]# adata.obsm.X_pca[mask.values,:]\n    \n    adata_m = adata[mask]\n    sc.pp.neighbors(adata_m)\n    sc.tl.umap(adata_m)#,color='n.umi')\n    sc.pl.umap(adata_m, color='n.umi', title = cell_type+' '+'color by n.umi')\n    sc.pl.umap(adata_m, color='pct_counts_mt', title = cell_type+' '+'color by pct_counts_mt')\n    #sc.pl.umap(adata_m, color='dose', title = cell_type+' '+'color by dose' )\n\n    \n","metadata":{"execution":{"iopub.status.busy":"2021-05-25T16:23:36.402594Z","iopub.execute_input":"2021-05-25T16:23:36.403022Z","iopub.status.idle":"2021-05-25T16:24:39.520917Z","shell.execute_reply.started":"2021-05-25T16:23:36.402987Z","shell.execute_reply":"2021-05-25T16:24:39.519889Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"'MALAT1' in adata.var.index","metadata":{"execution":{"iopub.status.busy":"2021-05-25T16:32:14.033939Z","iopub.execute_input":"2021-05-25T16:32:14.034394Z","iopub.status.idle":"2021-05-25T16:32:14.042326Z","shell.execute_reply.started":"2021-05-25T16:32:14.034359Z","shell.execute_reply":"2021-05-25T16:32:14.041086Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"","metadata":{"execution":{"iopub.status.busy":"2021-05-25T16:32:56.03594Z","iopub.execute_input":"2021-05-25T16:32:56.036308Z","iopub.status.idle":"2021-05-25T16:32:56.294077Z","shell.execute_reply.started":"2021-05-25T16:32:56.036275Z","shell.execute_reply":"2021-05-25T16:32:56.29267Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"","metadata":{},"execution_count":null,"outputs":[]}]}