{"cells":[{"metadata":{"trusted":true},"cell_type":"code","source":"import numpy as np\nimport pandas as pd","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"import matplotlib.pyplot as plt\nimport seaborn as sns","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"df = pd.read_csv('../input/phenotype-genotype-integrator/PheGenI.csv')","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"df.sample(5)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"df.info()","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"df['P-Value']","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"# Converting P-Value column","execution_count":null},{"metadata":{},"cell_type":"markdown","source":"P-Values have both float and object types","execution_count":null},{"metadata":{"trusted":true},"cell_type":"code","source":"pv = df['P-Value'][df['P-Value'].apply(lambda x: isinstance(x, str))]","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"pv = pv.str.split('-')","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"pv = pd.to_numeric(pv.apply(lambda x: x[0][:-1])) * 10 ** (-pd.to_numeric(pv.apply(lambda x: x[1]), downcast='float'))","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"pv","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"df['P-Value'][pv.index] = pv.values","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"df['P-Value']","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"pd.to_numeric(df['P-Value'])","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"df['P-Value'][df['P-Value'].apply(lambda x: isinstance(x, str))]","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"df['P-Value'].sort_values()","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"df['P-Value'].sort_values().reset_index()['P-Value'][:].plot(figsize=(15,7))","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"df['P-Value'].sort_values().reset_index()['P-Value'][:134000].plot(figsize=(15,7))","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"# Analysis","execution_count":null},{"metadata":{"trusted":true},"cell_type":"code","source":"df['Trait'].unique().size","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"df.groupby('Context').count().sort_values('Gene', ascending=False)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"Below is the diagram describing various SNP types.","execution_count":null},{"metadata":{},"cell_type":"markdown","source":"<img src=\"SNP types.jpg\"/>","execution_count":null},{"metadata":{},"cell_type":"markdown","source":"1. The “Near Gene” region includes the mRNA region of the gene as well as arbitrary regions of 2K nucleotides upstream and 0.5K nucleotides down stream to allow for potential regulatory regions. (https://www.ncbi.nlm.nih.gov/books/NBK44455/)\n2. UTR-5 is the region that is directly upstream from the initiation codon which is transcribed to mRNA but not translated to Protein. UTR-3 is similar at downstream.\n3. A SNP will be classified as “splice-site” if the SNP’s position is one or two bases before the start of an exon or If the SNP is located one or two bases following the end of an exon.\n","execution_count":null},{"metadata":{"trusted":true},"cell_type":"code","source":"plt.figure(figsize=(15,5))\nsns.countplot(x='Context', data=df, order=df['Context'].value_counts().index)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"df[(df['Gene'] == df['Gene 2'])].groupby('Context').count().sort_values('Gene', ascending=False)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"If Gene and Gene 2 names are same then SNP is not between two genes.","execution_count":null},{"metadata":{"trusted":true},"cell_type":"code","source":"(df['Gene'] == df['Gene 2']).sum()","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"plt.figure(figsize=(15,5))\nsns.countplot(x='Context', data=df[(df['Gene'] == df['Gene 2'])], order=df['Context'].value_counts().index)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"df[(df['Gene'] != df['Gene 2'])].groupby('Context').count().sort_values('Gene', ascending=False)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"df[(df['P-Value']>0) & (df['P-Value']<10**-300)].sort_values(by='P-Value')","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"P value less than 5x10^-8 is commonly accepted as threshold. (https://en.wikipedia.org/wiki/Genome-wide_significance)","execution_count":null},{"metadata":{"trusted":true},"cell_type":"code","source":"df_p = df[df['P-Value'] < 5 * 10 ** -8]","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"df_p.groupby('Trait').count().sort_values('P-Value', ascending=False).head(10)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"df_p.groupby('Trait').count().sort_values('P-Value', ascending=False)['Gene'].plot(figsize=(18,7))","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"df_p['Trait'].unique().size","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"traits = df_p.groupby('Trait').count().sort_values('P-Value', ascending=False).index","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"traits[:50]","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"# Genes by Trait","execution_count":null},{"metadata":{},"cell_type":"markdown","source":"Function to search traits","execution_count":null},{"metadata":{"trusted":true},"cell_type":"code","source":"import difflib","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"matches = difflib.get_close_matches('atherosclerosis', traits, n=15, cutoff=.4)\nmatches","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"def genes_by_trait(trait):\n    temp = df_p[df_p['Trait']==trait]\n    return set(temp['Gene']).union(set(temp['Gene 2']))","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"len(genes_by_trait('Body Mass Index'))","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"list_1 = ['Blood Pressure', 'Stroke', 'Diabetes Mellitus','Diabetes Mellitus, Type 2','Diabetes Mellitus, Type 1', 'Myocardial Infarction', 'Atherosclerosis', 'Plaque, Atherosclerotic']","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"factors_paired = [(i,j) for i in list_1 for j in list_1]","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"common_genes = []\n\nfor i,j in factors_paired:\n    common_genes.append(len(genes_by_trait(i).intersection(genes_by_trait(j))))","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"common_genes = np.array(common_genes).reshape(len(list_1),len(list_1))","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"common_genes = pd.DataFrame(common_genes, index=list_1, columns=list_1)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"common_genes","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"plt.figure(figsize=(15,5))\ncommon_genes.style.background_gradient(cmap='YlOrRd', axis=0)","execution_count":null,"outputs":[]},{"metadata":{"jupyter":{"outputs_hidden":true},"trusted":true},"cell_type":"code","source":"common_genes = genes_by_trait('Stroke').intersection(genes_by_trait('Diabetes Mellitus')).intersection(genes_by_trait('Blood Pressure'))\nprint(len(common_genes))\ncommon_genes","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"matches = difflib.get_close_matches('inflammatory bowel', traits, n=15, cutoff=.4)\nmatches","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"list_2 = ['Multiple Sclerosis', 'Psoriasis', 'Lupus Erythematosus, Systemic', 'Crohn Disease', 'Inflammatory Bowel Diseases', 'Diabetes Mellitus, Type 1']","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"factors_paired = [(i,j) for i in list_2 for j in list_2]","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"common_genes = []\n\nfor i,j in factors_paired:\n    common_genes.append(len(genes_by_trait(i).intersection(genes_by_trait(j))))","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"common_genes = np.array(common_genes).reshape(len(list_2),len(list_2))","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"common_genes = pd.DataFrame(common_genes, index=list_2, columns=list_2)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"common_genes","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"plt.figure(figsize=(15,5))\ncommon_genes.style.background_gradient(cmap='YlOrRd', axis=0)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"Gene information from https://www.ensembl.org/index.html","execution_count":null},{"metadata":{"trusted":true},"cell_type":"code","source":"import requests, sys\nimport pprint","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"\nserver = \"https://rest.ensembl.org\"\next = \"/phenotype/gene/homo_sapiens/GCKR?include_associated=0\"\n     \nr = requests.get(server+ext, headers={ \"Content-Type\" : \"application/json\"})\n     \nif not r.ok:\n    r.raise_for_status()\n    sys.exit()\n     \ndecoded = r.json()\npprint.pprint(decoded)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"","execution_count":null,"outputs":[]}],"metadata":{"kernelspec":{"language":"python","display_name":"Python 3","name":"python3"},"language_info":{"pygments_lexer":"ipython3","nbconvert_exporter":"python","version":"3.6.4","file_extension":".py","codemirror_mode":{"name":"ipython","version":3},"name":"python","mimetype":"text/x-python"}},"nbformat":4,"nbformat_minor":4}