{"metadata":{"kernelspec":{"language":"python","display_name":"Python 3","name":"python3"},"language_info":{"pygments_lexer":"ipython3","nbconvert_exporter":"python","version":"3.6.4","file_extension":".py","codemirror_mode":{"name":"ipython","version":3},"name":"python","mimetype":"text/x-python"}},"nbformat_minor":4,"nbformat":4,"cells":[{"cell_type":"code","source":"# This Python 3 environment comes with many helpful analytics libraries installed\n# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python\n# For example, here's several helpful packages to load\n\nimport numpy as np # linear algebra\nimport pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)\n\n# Input data files are available in the read-only \"../input/\" directory\n# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory\n\nimport os\nfor dirname, _, filenames in os.walk('/kaggle/input'):\n    for filename in filenames:\n        print(os.path.join(dirname, filename))\n\n# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using \"Save & Run All\" \n# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session","metadata":{"_uuid":"8f2839f25d086af736a60e9eeb907d3b93b6e0e5","_cell_guid":"b1076dfc-b9ad-4769-8c92-a6c4dae69d19","execution":{"iopub.status.busy":"2021-08-25T10:14:04.334912Z","iopub.execute_input":"2021-08-25T10:14:04.335265Z","iopub.status.idle":"2021-08-25T10:14:04.345936Z","shell.execute_reply.started":"2021-08-25T10:14:04.335236Z","shell.execute_reply":"2021-08-25T10:14:04.344791Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"!pip install scanpy\nimport scanpy as sc\nimport anndata\n\nimport scipy \n\n\nimport time\nt0start = time.time()\n\nimport pandas as pd\nimport numpy as np\nimport os\nimport sys\n\nimport matplotlib.pyplot as plt\nimport matplotlib as mpl\nmpl.rcParams['figure.dpi'] = 70\nplt.style.use('dark_background')\n\nimport seaborn as sns\n\nfrom sklearn.decomposition import PCA","metadata":{"execution":{"iopub.status.busy":"2021-08-25T10:14:04.34759Z","iopub.execute_input":"2021-08-25T10:14:04.3479Z","iopub.status.idle":"2021-08-25T10:14:11.654238Z","shell.execute_reply.started":"2021-08-25T10:14:04.34787Z","shell.execute_reply":"2021-08-25T10:14:11.653245Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"    str_data_inf = ' GSE64016_human_H1hESC_Fucci_Y2014_460Cells_GEO_counts_Leng_Kendziorski '\n    path_and_fn = '/kaggle/input/scrnaseq-hesc-h1-cell-line-fucci-cell-cycle-label/GSE64016_human_H1hESC_Fucci_Y2014_460Cells_GEO_counts_Leng_Kendziorski.h5ad'\n    adata_orig = sc.read(path_and_fn)\n    adata = adata_orig.copy()\n    adata","metadata":{"execution":{"iopub.status.busy":"2021-08-25T10:14:11.656316Z","iopub.execute_input":"2021-08-25T10:14:11.656624Z","iopub.status.idle":"2021-08-25T10:14:12.236271Z","shell.execute_reply.started":"2021-08-25T10:14:11.656591Z","shell.execute_reply":"2021-08-25T10:14:12.235406Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"adata.obs.columns = ['Cell type', 'Cell cycle phase']# = adata.obs['Fucci cell cycle phase'] \nadata","metadata":{"execution":{"iopub.status.busy":"2021-08-25T10:14:12.237415Z","iopub.execute_input":"2021-08-25T10:14:12.237695Z","iopub.status.idle":"2021-08-25T10:14:12.245177Z","shell.execute_reply.started":"2021-08-25T10:14:12.237668Z","shell.execute_reply":"2021-08-25T10:14:12.244469Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"adata.obs[ 'Cell cycle phase'].value_counts()","metadata":{"execution":{"iopub.status.busy":"2021-08-25T10:14:12.246448Z","iopub.execute_input":"2021-08-25T10:14:12.246731Z","iopub.status.idle":"2021-08-25T10:14:12.262271Z","shell.execute_reply.started":"2021-08-25T10:14:12.246704Z","shell.execute_reply":"2021-08-25T10:14:12.261206Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"adata.uns['Cell cycle phase labels'] = True # \n\nadata.uns['Id'] = 'GSE64016'# 'GSE146773'  # Will be used for filename - can be GSE, GSM, 10xGenomics, etc \n\n#adata.uns['GSM'] =  ''\nadata.uns['GSE'] = 'GSE64016' # adata.uns['Id'] #   'GSE146773' # adata.uns['Id']\nadata.uns['Brief Info'] = 'H1 hESC Fucci and WT'# 'U2OS FUCCI' # Might be used for figures annotations\nadata.uns['Link'] = r'https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE64016'  #Link to data:  r'https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE146773'  \nadata.uns['Cellosaurus'] = r'https://web.expasy.org/cellosaurus/CVCL_9771'# r'https://web.expasy.org/cellosaurus/CVCL_0042'\n#adata.uns['Doubling time string'] = r''# r'Cellosaurus: ~25-30 hours (DSMZ).'\n#adata.uns['Doubling hours'] = np.nan# 27 #\n\n\nadata.uns['Date'] =  '2014 12 09'# '2020 03 11' # Earliest date \n\nadata.uns['Cancer Cells'] =  0# 1 or 0  cancer cell or not #\n\nadata.uns['Organism'] = 'Human' # 'Human' \nadata.uns['Cell type'] = 'H1hESC' # 'U2OS'\nadata.uns['Cell further info'] = 'FUCCU' # Additional info - drug, 'FUCCI' or other special condtion\nadata.uns['Year'] = 2014 # 2020 \nadata.uns['Cell Count'] = adata.X.shape[0]\nadata.uns['Source'] = 'GEO'# 'GEO' # where data were taken # \nadata.uns['Counts or preprocessing'] =  'Counts' # counts or preprocessed \nadata.uns['First author'] = 'Leng' # First author for paper if knownMahdessian\nadata.uns['Last author'] = 'Kendziorski' # Last author for paper if known Lundberg \n\n\nadata.uns['Paper Link'] = r'https://pubmed.ncbi.nlm.nih.gov/26301841/'# r'https://www.nature.com/articles/s41586-021-03232-9'\nadata.uns['Paper Title'] = 'Oscope identifies oscillatory genes in unsynchronized single-cell RNA-seq experiments'# 'Spatiotemporal dissection of the cell cycle with single-cell proteogenomics'\nadata.uns['Abstract'] = '''Oscillatory gene expression is fundamental to development, but technologies for monitoring expression oscillations are limited. We have developed a statistical approach called Oscope to identify and characterize the transcriptional dynamics of oscillating genes in single-cell RNA-seq data from an unsynchronized cell population. Applying Oscope to a number of data sets, we demonstrated its utility and also identified a potential artifact in the Fluidigm C1 platform.''' #  '''The cell cycle, over which cells grow and divide, is a fundamental process of life. Its dysregulation has devastating consequences, including cancer1,2,3. The cell cycle is driven by precise regulation of proteins in time and space, which creates variability between individual proliferating cells. To our knowledge, no systematic investigations of such cell-to-cell proteomic variability exist. Here we present a comprehensive, spatiotemporal map of human proteomic heterogeneity by integrating proteomics at subcellular resolution with single-cell transcriptomics and precise temporal measurements of individual cells in the cell cycle. We show that around one-fifth of the human proteome displays cell-to-cell variability, identify hundreds of proteins with previously unknown associations with mitosis and the cell cycle, and provide evidence that several of these proteins have oncogenic functions. Our results show that cell cycle progression explains less than half of all cell-to-cell variability, and that most cycling proteins are regulated post-translationally, rather than by transcriptomic cycling. These proteins are disproportionately phosphorylated by kinases that regulate cell fate, whereas non-cycling proteins that vary between cells are more likely to be modified by kinases that regulate metabolism. This spatially resolved proteomic map of the cell cycle is integrated into the Human Protein Atlas and will serve as a resource for accelerating molecular studies of the human cell cycle and cell proliferation.'''\nadata.uns['PMID'] = 26301841 # \n\nadata.uns['Comment'] = r'Reformating scripts to h5ad can be found at: https://www.kaggle.com/alexandervc/saving-adata-with-comments-and-long-fnames-u1-fuc' \nadata.uns['Link to our collection'] = r'https://www.kaggle.com/alexandervc/scrnaseq-datasets-collection' \n\ndef get_filename(adata):\n    fn = ''\n    for t in [ 'Id', 'Organism', 'Cell type', 'Cell further info', 'Year', 'Cell Count', 'Source', \n              'Counts or preprocessing', 'First author', 'Last author']:\n        if t == 'Year':\n            fn += 'Y'+str(adata.uns[t])\n        elif t == 'Cell Count':\n            fn += str(adata.uns[t]) + 'Cells'\n        else:\n            fn += adata.uns[t]\n        if t != 'Last author': fn += '_'\n\n    fn += '.h5ad'\n    return fn\n\nfn = get_filename(adata)\nfn","metadata":{"execution":{"iopub.status.busy":"2021-08-25T10:14:12.263561Z","iopub.execute_input":"2021-08-25T10:14:12.263845Z","iopub.status.idle":"2021-08-25T10:14:12.283571Z","shell.execute_reply.started":"2021-08-25T10:14:12.263817Z","shell.execute_reply":"2021-08-25T10:14:12.282442Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"adata.write_h5ad(fn,compression='gzip')\nadata","metadata":{"execution":{"iopub.status.busy":"2021-08-25T10:14:12.284707Z","iopub.execute_input":"2021-08-25T10:14:12.284985Z","iopub.status.idle":"2021-08-25T10:14:13.5107Z","shell.execute_reply.started":"2021-08-25T10:14:12.284958Z","shell.execute_reply":"2021-08-25T10:14:13.509494Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"","metadata":{},"execution_count":null,"outputs":[]}]}