{"metadata":{"kernelspec":{"language":"python","display_name":"Python 3","name":"python3"},"language_info":{"pygments_lexer":"ipython3","nbconvert_exporter":"python","version":"3.6.4","file_extension":".py","codemirror_mode":{"name":"ipython","version":3},"name":"python","mimetype":"text/x-python"}},"nbformat_minor":4,"nbformat":4,"cells":[{"cell_type":"markdown","source":"# Understanding phylogenetic trees\n\nIn this notebook, we will investigate how phylogenetic trees are shaped by the mutations that accumulate while bacteria multiply. \n\n#### Credits - code in this notebook was adapted from the sources below:\n* Sequence alignment: https://dmnfarrell.github.io/bioinformatics/bokeh-sequence-aligner\n* Sequence evolution simulation: https://hplgit.github.io/bioinf-py/doc/pub/html/main_bioinf.html. ","metadata":{}},{"cell_type":"code","source":"import random\nimport numpy as np\n\nfrom Bio.Alphabet import generic_dna\nfrom Bio.Seq import Seq\nfrom Bio.SeqRecord import SeqRecord\nfrom Bio.Align import MultipleSeqAlignment\nfrom Bio.Phylo.TreeConstruction import DistanceCalculator\nfrom Bio.Phylo.TreeConstruction import DistanceTreeConstructor\nfrom Bio import Phylo\nfrom io import StringIO\n\nimport re\n\nimport panel as pn\nimport panel.widgets as pnw\npn.extension()\n\nfrom bokeh.plotting import figure\nfrom bokeh.models import ColumnDataSource, Plot, Grid, Range1d\nfrom bokeh.models.glyphs import Text, Rect\nfrom bokeh.layouts import gridplot","metadata":{"_uuid":"d629ff2d2480ee46fbb7e2d37f6b5fab8052498a","_cell_guid":"79c7e3d0-c299-4dcb-8224-4455121ee9b0","_kg_hide-input":true,"_kg_hide-output":true,"execution":{"iopub.status.busy":"2021-08-10T20:13:01.87854Z","iopub.execute_input":"2021-08-10T20:13:01.878865Z","iopub.status.idle":"2021-08-10T20:13:01.942987Z","shell.execute_reply.started":"2021-08-10T20:13:01.878835Z","shell.execute_reply":"2021-08-10T20:13:01.941659Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# Setting up functions\n\nFor this explainer, I have created functions to generate random DNA sequences, simulate the evolution of these sequences over time, build a phylogenetic tree of the simulated sequences and view the differences in the sequences. If you want to explore them, click the 'Show hidden cell' buttons to examine the code. \n\nFor each mutation event, a random position in the sequence is chosen, and a random base is selected, with a 1 in 4 chance that the same base is selected so there is no observable mutation. ","metadata":{}},{"cell_type":"code","source":"# substitute a random base in a sequence n times\n\ndef mutate_n(dna, N):\n    dna = np.array(dna, dtype='c')  # array of characters\n    mutation_sites = random.sample(range(0, len(dna) - 1), N)\n    # Draw new bases - 1 in 4 chance of no mutation\n    new_bases = random.choices(list('ATCG'), k=N)\n    # Replace the mutated bases\n    dna[mutation_sites] = new_bases\n    return ''.join([bytes.decode(s) for s in np.frombuffer(dna, dtype='S1')])","metadata":{"execution":{"iopub.status.busy":"2021-08-10T20:13:01.944803Z","iopub.execute_input":"2021-08-10T20:13:01.945258Z","iopub.status.idle":"2021-08-10T20:13:01.952868Z","shell.execute_reply.started":"2021-08-10T20:13:01.945206Z","shell.execute_reply":"2021-08-10T20:13:01.951619Z"},"_kg_hide-input":true,"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"# visualize a sequence alignment\n# adapted from https://dmnfarrell.github.io/bioinformatics/bokeh-sequence-aligner\n\nimport panel as pn\nimport panel.widgets as pnw\npn.extension()\n\nfrom bokeh.plotting import figure\nfrom bokeh.models import ColumnDataSource, Plot, Grid, Range1d\nfrom bokeh.models.glyphs import Text, Rect\nfrom bokeh.layouts import gridplot\n\ndef view_alignment(aln, fontsize=\"9pt\", plot_width=800):\n    \"\"\"Bokeh sequence alignment view\"\"\"\n\n    #make sequence and id lists from the aln object\n    seqs = [rec.seq for rec in (aln)]\n    ids = [rec.id for rec in aln]    \n    text = [i for s in list(seqs) for i in s]\n    colors = get_colors(seqs)    \n    N = len(seqs[0])\n    S = len(seqs)    \n    width = .4\n\n    x = np.arange(0.5,N+0.5)\n    y = np.arange(0,S,1)\n    #creates a 2D grid of coords from the 1D arrays\n    xx, yy = np.meshgrid(x, y)\n    #flattens the arrays\n    gx = xx.ravel()\n    gy = yy.flatten()\n    #use recty for rect coords with an offset\n    recty = gy+.5\n    h= 1/S\n    #now we can create the ColumnDataSource with all the arrays\n    source = ColumnDataSource(dict(x=gx, y=gy, recty=recty, text=text, colors=colors))\n    plot_height = len(seqs)*15+50\n    x_range = Range1d(0,N+1, bounds='auto')\n    if N>100:\n        viewlen=100\n    else:\n        viewlen=N\n    #view_range is for the close up view\n    view_range = (0,viewlen)\n    tools=\"xpan, xwheel_zoom, reset, save\" \n\n    #sequence text view with ability to scroll along x axis\n    p = figure(title=None, plot_width=plot_width, plot_height=plot_height,\n                x_range=view_range, y_range=ids, tools=\"xpan,reset\",\n                min_border=0, toolbar_location='below')#, lod_factor=1)          \n    glyph = Text(x=\"x\", y=\"y\", text=\"text\", text_align='center',text_color=\"black\",\n                text_font=\"monospace\",text_font_size=fontsize)\n    rects = Rect(x=\"x\", y=\"recty\",  width=1, height=1, fill_color=\"colors\",\n                line_color=None, fill_alpha=0.4)\n    p.add_glyph(source, glyph)\n    p.add_glyph(source, rects)\n\n    p.grid.visible = False\n    p.xaxis.major_label_text_font_style = \"bold\"\n    p.yaxis.minor_tick_line_width = 0\n    p.yaxis.major_tick_line_width = 0\n\n    p = gridplot([[p]], toolbar_location='below')\n    return p\n\ndef get_colors(seqs):\n    \"\"\"make colors for bases in sequence\"\"\"\n    text = [i for s in list(seqs) for i in s]\n    clrs =  {'A':'red','T':'green','G':'orange','C':'blue','-':'white'}\n    colors = [clrs[i] for i in text]\n    return colors\n","metadata":{"_kg_hide-input":true,"execution":{"iopub.status.busy":"2021-08-10T20:13:02.060696Z","iopub.execute_input":"2021-08-10T20:13:02.061138Z","iopub.status.idle":"2021-08-10T20:13:02.140959Z","shell.execute_reply.started":"2021-08-10T20:13:02.061082Z","shell.execute_reply":"2021-08-10T20:13:02.139865Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"# build a phylogenetic tree from a list of sequences\n\ndef build_tree(sequences, method='upgma'):\n    alignment = MultipleSeqAlignment(sequences)\n    \n    calculator = DistanceCalculator('identity')\n    dm = calculator.get_distance(alignment)\n    \n    constructor = DistanceTreeConstructor()\n    if method is 'upgma':\n        tree = constructor.upgma(dm)\n    if method is 'nj':\n        tree = constructor.nj(dm)\n    \n    tree.root_with_outgroup('Ancestor')\n    tree.ladderize() # place longer branches on the bottom\n    edit_tree = re.sub('Inner\\\\d+', '', tree.format('newick')) # remove the inner node labels\n    handle = StringIO(edit_tree)\n    tree = Phylo.read(handle, 'newick')\n    Phylo.draw(tree)\n    \n    return alignment\n","metadata":{"_kg_hide-input":true,"execution":{"iopub.status.busy":"2021-08-10T20:13:02.14288Z","iopub.execute_input":"2021-08-10T20:13:02.143149Z","iopub.status.idle":"2021-08-10T20:13:02.151209Z","shell.execute_reply.started":"2021-08-10T20:13:02.143119Z","shell.execute_reply":"2021-08-10T20:13:02.150503Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"# simulate sequence evolution\n\ndef random_DNA(length):\n    DNA=\"\"\n    for count in range(length):\n        DNA+=random.choice(\"CGTA\")\n    return DNA\n\ndef make_random_seqs(num_mutations):\n    seq_recs_random = [SeqRecord(Seq(seq1, generic_dna), id=\"Ancestor\")]\n    for i in range(n_seqs):\n        num = i + 1\n        num_mutations = random.randint(0, 10)\n        print('Generating sequence ' + str(num) + ': ' + str(num_mutations) + ' mutations from the Ancestor')\n        sequence = mutate_n(seq1, num_mutations)\n        seq_recs_random.append(SeqRecord(Seq(sequence, generic_dna), id=\"Sequence_\" + str(num)))\n    return seq_recs_random\n\ndef make_serial_seqs(num_mutations): \n    seq_recs_serial = [SeqRecord(Seq(seq1, generic_dna), id=\"Ancestor\")]\n    for i in range(n_seqs):\n        num = i + 1\n        print('Generating sequence ' + str(num) + ': ' + str(num_mutations) + ' mutations from sequence ' + str(i))\n        sequence = mutate_n(str(seq_recs_serial[i].seq), num_mutations)\n        seq_recs_serial.append(SeqRecord(Seq(sequence, generic_dna), id=\"Sequence_\" + str(num)))\n    return seq_recs_serial\n\ndef make_bifurcating_seqs(num_mutations):\n    seq_recs_bifurcating = {'Ancestor': SeqRecord(Seq(seq1, generic_dna), id=\"Ancestor\")}\n    gen_names = {'0' : ['Ancestor']}\n    \n    for gen in range(4): # model four generations\n        ids = ['A', 'B']\n        ancestors = gen_names[str(gen)]\n        names = []\n        for ancestor in ancestors:\n            offspring = [ancestor + \".\" + id for id in ids]\n            names.extend(offspring)\n            for i in offspring:\n                sequence = mutate_n(str(seq_recs_bifurcating[ancestor].seq), num_mutations)\n                seq_recs_bifurcating[i] = SeqRecord(Seq(sequence, generic_dna), id=i)\n        gen_names[str(gen+1)] = names\n    seq_recs_bifurcating = list(seq_recs_bifurcating.values())\n    return seq_recs_bifurcating\n\ndef make_binary_fission(num_mutations, num_gens=4):\n    seq_recs_bifurcating = {'Ancestor': SeqRecord(Seq(seq1, generic_dna), id=\"Ancestor\")}\n    gen_names = {'0' : ['Ancestor']}\n    \n    for gen in range(num_gens): # model num_gens generations\n        ancestors = gen_names[str(gen)]\n        names = []\n        for ancestor in ancestors:\n            print('Generating sequences for generation ' + str(gen) + ': ' + str(num_mutations) + ' mutations from the previous generation')\n            offspring = [ancestor + 'same', ancestor + 'mut']\n            names.extend(offspring)\n            seq_recs_bifurcating[offspring[0]] = SeqRecord(Seq(str(seq_recs_bifurcating[ancestor].seq), generic_dna), id=offspring[0])\n            sequence = mutate_n(str(seq_recs_bifurcating[ancestor].seq), num_mutations)\n            seq_recs_bifurcating[offspring[1]] = SeqRecord(Seq(sequence, generic_dna), id=offspring[1])\n        gen_names[str(gen+1)] = names\n    seq_recs_bifurcating = list(seq_recs_bifurcating.values())\n    return seq_recs_bifurcating\n","metadata":{"execution":{"iopub.status.busy":"2021-08-10T20:22:43.744219Z","iopub.execute_input":"2021-08-10T20:22:43.744697Z","iopub.status.idle":"2021-08-10T20:22:43.772852Z","shell.execute_reply.started":"2021-08-10T20:22:43.744656Z","shell.execute_reply":"2021-08-10T20:22:43.771159Z"},"_kg_hide-input":true,"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"Here, we define the DNA sequence of our bacterial ancestor ('seq1'), the number of sequences to generate ('n_seqs'), and the number of mutations to simulate per step ('num_mutations'). These parameters can all be changed to explore the impact of each factor on the appearance of the tree.","metadata":{}},{"cell_type":"code","source":"seq1 = random_DNA(length=40)\nprint(\"DNA sequence of our ancestor: \" + seq1)\nn_seqs = 20\nnum_mutations = 2 # number of mutations to simulate per step (if applicable)","metadata":{"execution":{"iopub.status.busy":"2021-08-10T20:13:02.180249Z","iopub.execute_input":"2021-08-10T20:13:02.18076Z","iopub.status.idle":"2021-08-10T20:13:02.202088Z","shell.execute_reply.started":"2021-08-10T20:13:02.180718Z","shell.execute_reply":"2021-08-10T20:13:02.200961Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# Creating bacteria with a fixed number of mutations compared to their ancestor\n\nHere, each sequence is a mutated version of the previous one. This results in a steady movement of sequences from the left to the right as they become more different to the ancestor. ","metadata":{}},{"cell_type":"code","source":"sequences = make_serial_seqs(num_mutations)\n\nalignment = build_tree(sequences, method='nj')\np = view_alignment(alignment, plot_width=1200)\npn.pane.Bokeh(p)","metadata":{"execution":{"iopub.status.busy":"2021-08-10T20:14:17.375742Z","iopub.execute_input":"2021-08-10T20:14:17.376099Z","iopub.status.idle":"2021-08-10T20:14:17.902219Z","shell.execute_reply.started":"2021-08-10T20:14:17.376066Z","shell.execute_reply":"2021-08-10T20:14:17.900737Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# Simulating a growing population of bacteria\n\nIn this scenario, we look at a scenario that reflects the way bacteria replicate through binary fission. During this process, the ancestor's double-stranded DNA is peeled apart, and each strand is used as a template to create a complementary copy, resulting in two copies of the genome. During this process, mistakes can occur which makes the offspring's DNA different from the ancestor. \n\n<img src=\"https://cdn.kastatic.org/ka-perseus-images/3af458a8d1139517c80af9fff471aca53bab0dbf.png\" width=\"400px\">\n\nIn this simulation, we capture the genome of the ancestor, plus both of their offspring, which each have a chance of accumulating mutations during the copying of their ancestor's DNA, plus two of their offspring, and so on...","metadata":{}},{"cell_type":"code","source":"sequences = make_bifurcating_seqs(num_mutations)\n\nalignment = build_tree(sequences, method='nj')\np = view_alignment(alignment, plot_width=1200)\npn.pane.Bokeh(p)","metadata":{"execution":{"iopub.status.busy":"2021-08-10T20:14:30.010824Z","iopub.execute_input":"2021-08-10T20:14:30.011217Z","iopub.status.idle":"2021-08-10T20:14:30.577696Z","shell.execute_reply.started":"2021-08-10T20:14:30.011183Z","shell.execute_reply":"2021-08-10T20:14:30.576486Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"Some of this variation is reduced by simulating a longer genome and fewer mutations between samples.","metadata":{}},{"cell_type":"code","source":"seq1 = random_DNA(length=100)\nprint(\"DNA sequence of our ancestor: \" + seq1)\nn_seqs = 20\nnum_mutations = 2 # number of mutations to simulate per step\n\nsequences = make_bifurcating_seqs(num_mutations)\n\nalignment = build_tree(sequences, method='nj')\np = view_alignment(alignment, plot_width=1200)\npn.pane.Bokeh(p)","metadata":{"execution":{"iopub.status.busy":"2021-08-10T20:18:56.219485Z","iopub.execute_input":"2021-08-10T20:18:56.219817Z","iopub.status.idle":"2021-08-10T20:18:56.843122Z","shell.execute_reply.started":"2021-08-10T20:18:56.219785Z","shell.execute_reply":"2021-08-10T20:18:56.842078Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# Creating bacteria with a random number of mutations compared to our ancestor\n\nHere, we generate a collection of sequences with a random number of mutation events compared to our ancestor sequence. This is a better representation of the data we normally collect when we randomly collect a sample of bacteria, e.g. from a patient population or from the environment. \n\nIn this scenario, we expect to see the ancestor on the left hand side of the tree, representing 0 mutations. We expect to see each generated sequence a certain distance from the Ancestor, with sequences on the same part of the x-axis if they are identical to the ancestor, and sequences with the most chances to mutate sitting further to the right. \n\nThe number of mutation opportunities and the distance from the ancestor won't line up perfectly for two reasons:\n1. Each opportunity to mutate has a 1 in 4 chance of not resulting in a change in the DNA\n2. In some cases, an early mutation might change the ancestral sequence and a later mutation might change it back","metadata":{}},{"cell_type":"code","source":"sequences = make_random_seqs(num_mutations)\n\nalignment = build_tree(sequences, method='nj')\np = view_alignment(alignment, plot_width=1200)\npn.pane.Bokeh(p)","metadata":{"execution":{"iopub.status.busy":"2021-08-10T20:13:53.050585Z","iopub.execute_input":"2021-08-10T20:13:53.050945Z","iopub.status.idle":"2021-08-10T20:13:53.514353Z","shell.execute_reply.started":"2021-08-10T20:13:53.050913Z","shell.execute_reply":"2021-08-10T20:13:53.51361Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"","metadata":{},"execution_count":null,"outputs":[]}]}