{"cells":[{"metadata":{},"cell_type":"markdown","source":"# What is about ?\n\nVisualization via dimensional reductions  of Stoeckius et al 2017 single cell RNA sequences data - 1,774 cells in total by 11757 genes\n\nThe dataset contains counts (i.e. integers - no log-transforms or whaever )\n\n\nPS\n\nnotebook based on:\n\nhttps://www.kaggle.com/alexandervc/dim-reduction-and-plots-scrnaseq-nestorova16\n\nhttps://www.kaggle.com/alexandervc/moa-data-visualization-via-dimensional-reduct\n","execution_count":null},{"metadata":{"_uuid":"8f2839f25d086af736a60e9eeb907d3b93b6e0e5","_cell_guid":"b1076dfc-b9ad-4769-8c92-a6c4dae69d19","trusted":true},"cell_type":"code","source":"# This Python 3 environment comes with many helpful analytics libraries installed\n# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python\n# For example, here's several helpful packages to load\n\nimport numpy as np # linear algebra\nimport pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)\n\n# Input data files are available in the read-only \"../input/\" directory\n# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory\n\nimport os\nfor dirname, _, filenames in os.walk('/kaggle/input'):\n    for filename in filenames:\n        print(os.path.join(dirname, filename))\n\n# You can write up to 5GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using \"Save & Run All\" \n# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session","execution_count":null,"outputs":[]},{"metadata":{"_uuid":"d629ff2d2480ee46fbb7e2d37f6b5fab8052498a","_cell_guid":"79c7e3d0-c299-4dcb-8224-4455121ee9b0","trusted":true},"cell_type":"code","source":"","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"import time\nimport matplotlib.pyplot as plt\nimport seaborn as sns\nimport umap\n\n","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"# Prepare data","execution_count":null},{"metadata":{"trusted":true},"cell_type":"code","source":"df = pd.read_csv('/kaggle/input/single-cell-rna-seq-from-stoeckius-et-al-2017/GSE100866_CD8_merged-RNA_umi.csv', index_col = 0)\n# The dataset is single cell rna seq from the Cite-seq paper. \n# Cells were classified by flow cytometry as low, medium, or high cd8 expression cells (columns), with rows corresponding to RNA counts. 1,774 cells in total by 11757 genes. \n# Stoeckius et al 2017.\n\ndf = df.T\n\nX = df.values.copy(); X_original_save = X.copy()\n\nprint(X.shape)\n\ndf","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"y = pd.Series(dtype = 'object')\ny.loc[1] = '1'","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"# Extract CD8 type from inde \nv = df.index\n\ny = pd.Series(dtype = 'object')\nfor i, t in enumerate(v):\n    if 'high' in t:\n        #print(t)\n        y[t] = 'CD8_high'\n    if 'off' in t:\n        #print(t)\n        y[t] = 'CD8_off'\n    if 'intermediate' in t:\n        #print(t)\n        y[t] = 'CD8_intermediate'\n    if 'low' in t:\n        y[t] = 'CD8_low'\nprint( y.shape, df.shape )\ny","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"v = np.sum(X,axis = 0 )\nplt.plot(v)\n((X - X.astype(int)) != 0 ).sum()","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"# PCA","execution_count":null},{"metadata":{"trusted":true},"cell_type":"code","source":"\n\nimport matplotlib.pyplot as plt\nimport seaborn as sns\nfrom sklearn.decomposition import PCA\n\nX = df.values.copy()\n\npca = PCA()\nt0 = time.time()\nr = pca.fit_transform(X.copy())\nprint(time.time()-t0 )\n\nplt.figure(figsize = (15,7) )\nsns.scatterplot(x=r[:,0], y=r[:,1], hue = y  )\nplt.show()\n\nfig = plt.figure(figsize = (15,7) )\nfig.add_subplot(1, 2, 1) \nplt.plot(pca.singular_values_,'o-')\nplt.title('Singular values')\nfig.add_subplot(1, 2, 2) \nplt.plot(pca.explained_variance_ratio_,'o-')\nplt.title('explained variance')\n","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"# UMAP","execution_count":null},{"metadata":{"trusted":true},"cell_type":"code","source":"import matplotlib.pyplot as plt\nimport seaborn as sns\nimport umap\n\nX = df.values\n\nt0 = time.time()\nr = umap.UMAP().fit_transform(X.copy())\nprint(time.time()-t0 )\n#plt.scatter(r[:,0],r[:,1],c = adata.obs['cell_types'].values )\nplt.figure(figsize = (15,7) )\nsns.scatterplot(x=r[:,0], y=r[:,1], hue = y  )\nplt.show()\n\n","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"import umap\nimport time\n\n\nfig = plt.figure(figsize=(15,15))\n\n#perplexities = [5, 30, 50, 100]\nc=0\nfor i,n_neighbors in enumerate( [5,200] ) :\n  for min_dist in [0,0.5,1] :\n    c += 1\n    t0 = time.time()\n    #tsne = TSNE(n_components=2, init='random', random_state=0, perplexity=perplexity)\n    #r = tsne.fit_transform(X)\n    r = umap.UMAP(n_neighbors= n_neighbors, min_dist = min_dist ).fit_transform(X)\n    td = time.time()-t0\n    print(td, 'secs passed')\n    ax = fig.add_subplot(2, 3,c )\n    sns.scatterplot(x=r[:,0], y=r[:,1], hue = y  )\n    ax.set_title(\"n_neighbors=%d min_dist=%.1f.  %.2f secs\" % (n_neighbors, min_dist, td ) )","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"# Visualize data by various dimensional reduction methods","execution_count":null},{"metadata":{},"cell_type":"markdown","source":"# Fast methods","execution_count":null},{"metadata":{"trusted":true},"cell_type":"code","source":"# Based on: \n# https://scikit-learn.org/stable/auto_examples/manifold/plot_compare_methods.html#sphx-glr-auto-examples-manifold-plot-compare-methods-py\n# See also:\n# https://scikit-learn.org/stable/auto_examples/manifold/plot_lle_digits.html#\n\n\n\nimport umap \nfrom sklearn import manifold\nfrom sklearn.decomposition import PCA\nfrom sklearn.decomposition import FactorAnalysis\nfrom sklearn.decomposition import NMF\nfrom sklearn.decomposition import FastICA\nfrom sklearn.decomposition import FactorAnalysis\nfrom sklearn.decomposition import LatentDirichletAllocation\nfrom sklearn.ensemble import RandomTreesEmbedding\nfrom sklearn.random_projection import SparseRandomProjection\nfrom sklearn.discriminant_analysis import LinearDiscriminantAnalysis\n\nfrom sklearn.pipeline import make_pipeline\nfrom sklearn.decomposition import TruncatedSVD\n\n\nfrom collections import OrderedDict\nfrom functools import partial\nfrom matplotlib.ticker import NullFormatter\n\n\nn_neighbors = 10\nn_components = 2\n# Set-up manifold methods\nLLE = partial(manifold.LocallyLinearEmbedding,\n              n_neighbors, n_components, eigen_solver='auto')\n\nmethods = OrderedDict()\nmethods['PCA'] = PCA()\nmethods['umap'] = umap.UMAP(n_components = n_components)\nmethods['t-SNE'] = manifold.TSNE(n_components=n_components, init='pca', random_state=0)\nmethods['ICA'] = FastICA(n_components=n_components,         random_state=0)\nmethods['FA'] = FactorAnalysis(n_components=n_components, random_state=0)\nmethods['LLE'] = LLE(method='standard')\nmethods['Modified LLE'] = LLE(method='modified')\nmethods['Isomap'] = manifold.Isomap(n_neighbors, n_components)\nmethods['MDS'] = manifold.MDS(n_components, max_iter=100, n_init=1)\nmethods['SE'] = manifold.SpectralEmbedding(n_components=n_components,\n                                           n_neighbors=n_neighbors)\nmethods['NMF'] = NMF(n_components=n_components,  init='random', random_state=0) \nmethods['RandProj'] = SparseRandomProjection(n_components=n_components, random_state=42)\n\nrand_trees_embed = make_pipeline(RandomTreesEmbedding(n_estimators=200, random_state=0, max_depth=5), TruncatedSVD(n_components=n_components) )\nmethods['RandTrees'] = rand_trees_embed\nmethods['LatDirAll'] = LatentDirichletAllocation(n_components=n_components,  random_state=0)\nmethods['LTSA'] = LLE(method='ltsa') \nmethods['Hessian LLE'] = LLE(method='hessian') \n\nlist_fast_methods = ['PCA','umap','FA', 'ICA','NMF','RandProj','RandTrees']\nlist_slow_methods = ['t-SNE','LLE','Modified LLE','Isomap','MDS','SE','LatDirAll','LTSA','Hessian LLE']\n\n# transformer = NeighborhoodComponentsAnalysis(init='random',  n_components=2, random_state=0) # Cannot be applied since supervised - requires y \n# methods['LinDisA'] = LinearDiscriminantAnalysis(n_components=n_components)# Cannot be applied since supervised - requires y \n\n\n# Create figure\nfig = plt.figure(figsize=(25, 16))\n\n# Plot results\nc = 0\nfor i, (label, method) in enumerate(methods.items()):\n    if label not in  list_fast_methods :\n        continue\n        \n    t0 = time.time()\n    try:\n        r = method.fit_transform(X.copy())\n    except:\n        print('Got Exception', label )\n        continue \n    t1 = time.time()\n    print(\"%s: %.2g sec\" % (label, t1 - t0))\n    c+=1\n    fig.add_subplot(3, 3 , c) \n    sns.scatterplot(x=r[:,0], y=r[:,1],hue =  y )\n    plt.title(label )\n    plt.legend('')\n\nplt.show()\n","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"# All methods including slow\n\nto speed up we cut X to smaller size","execution_count":null},{"metadata":{"trusted":true},"cell_type":"code","source":"# To speed-up reduce dimensions by PCA first and cut the size\nX_save = X.copy( )\nr = pca.fit_transform(X)\ni_cut = 5000\nX = r[:i_cut,:50]\n\n\nlist_slow_methods = ['t-SNE','LLE','Modified LLE','Isomap','MDS','SE','LatDirAll','LTSA','Hessian LLE']\n\n# transformer = NeighborhoodComponentsAnalysis(init='random',  n_components=2, random_state=0) # Cannot be applied since supervised - requires y \n# methods['LinDisA'] = LinearDiscriminantAnalysis(n_components=n_components)# Cannot be applied since supervised - requires y \n\n\n# Create figure\nfig = plt.figure(figsize=(25, 4))\n\n# Plot results\nc = 0\nfor i, (label, method) in enumerate(methods.items()):\n    #if label not in list_slow_methods: # list_fast_methods :\n    #    continue\n        \n    t0 = time.time()\n    try:\n        r = method.fit_transform(X.copy())\n    except:\n        print('Got Exception', label )\n        continue \n    t1 = time.time()\n    print(\"%s: %.2g sec\" % (label, t1 - t0))\n    c+=1\n    fig.add_subplot(1, 4 , c) \n    sns.scatterplot(x=r[:,0], y=r[:,1],hue =  y.iloc[:i_cut])\n    plt.title(label )\n    plt.legend('')\n    if c%4 == 0:\n        c = 0\n        plt.show()\n        fig = plt.figure(figsize=(25, 4))\n        \n    \n\nplt.show()\nX = X_save.copy()","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"","execution_count":null,"outputs":[]}],"metadata":{"kernelspec":{"language":"python","display_name":"Python 3","name":"python3"},"language_info":{"pygments_lexer":"ipython3","nbconvert_exporter":"python","version":"3.6.4","file_extension":".py","codemirror_mode":{"name":"ipython","version":3},"name":"python","mimetype":"text/x-python"}},"nbformat":4,"nbformat_minor":4}