{"metadata":{"kernelspec":{"language":"python","display_name":"Python 3","name":"python3"},"language_info":{"pygments_lexer":"ipython3","nbconvert_exporter":"python","version":"3.6.4","file_extension":".py","codemirror_mode":{"name":"ipython","version":3},"name":"python","mimetype":"text/x-python"}},"nbformat_minor":4,"nbformat":4,"cells":[{"cell_type":"markdown","source":"# \n## Searching new genetic causes of ALS by aberrant expression analysis \n### When the outlier is the signal","metadata":{"_uuid":"8f2839f25d086af736a60e9eeb907d3b93b6e0e5","_cell_guid":"b1076dfc-b9ad-4769-8c92-a6c4dae69d19"}},{"cell_type":"markdown","source":"Felix Brechtmann 1 , M. Hasan Çelik 2 , Julien Gagneur 1 , Florian Hölzlwimmer 1 , Michaela Müller 1 ,\nNils Wagner 1 , Xiaohui Xie 2 , Vicente Yépez 1 , Michael Zech 1\n\n1 Technical University of Munich\n\n2 University of California, Irvine","metadata":{}},{"cell_type":"markdown","source":"We asked whether combining DNA and RNA data could be employed to discover novel genes and biological pathways implicated in ALS. Given the heterogeneity of genetic causes of ALS reported in the literature, we assumed that no single gene or pathway would collectively explain disease cases. Therefore, instead of comparing cases against controls, we searched for potential aberrations that could be specific for a single or a few individuals.\nTo this end, we first identified expression outliers, which are defined as genes exhibiting aberrant expression levels. We therefore used the machine learning technique of a denoising autoencoder, that we previously adapted to work with RNA-sequencing data [1]. To ensure that the outliers are not technical artefacts of some sort but driven by genetics, we filtered them down to genes that carried rare genetic variants that could be likely to disrupt gene expression. Specifically, we identified and narrowed down the set of variants based on whether they were predicted to have a high or moderate impact or to affect splicing. This analysis yielded a total of 109 genetically-supported outlier genes.\nAmong these, four genes corresponded to catalogued ALS mutations or known ALS genes, supporting the validity of the approach and possibly yielding a more detailed genetic diagnosis of those patients. \n\nIn order to put potential ALS gene candidates into perspective, we searched for new ALS genes using a two-fold strategy. First, we considered genes that are functionally linked to established ALS genes by applying network diffusion on the STRING gene network. This led to the identification of 16 genetically-supported outliers which are functionally linked to ALS genes or are well established disease genes. Second, we asked whether new potential pathways would emerge as clusters of functionally related genetically-supported outliers. This second analysis revealed gene groups that are members of nucleopores, kinetochores, DNA repair and machinery, as well as ribosomal RNA biogenesis.\n\nAltogether, these results demonstrate the applicability of expression outlier and network diffusion analysis to genetic predispositions of ALS. Our results indicate the involvement of new genes and pathways and supports the view of a complex genetic architecture of ALS. \n\n[1] Brechtmann, F., Mertes, C., Matusevičiūtė, A., Yépez, V. A., Avsec, Ž., Herzog, M., ... & Gagneur, J. (2018). OUTRIDER: A statistical method for detecting aberrantly expressed genes in RNA sequencing data. The American Journal of Human Genetics, 103(6), 907-917.","metadata":{}},{"cell_type":"markdown","source":"**Disclaimer**\n\nWe have used a combination of Python, R and other tools (e.g., VEP). The running time of the tools are longer than the allowed running time at kaggle. Also, we used publicly avaliable genomics datasets (gnomAD, CADD, etc.). \n\nOur pre-computed results were generated using code freely accessible at the following repository: https://github.com/gagneurlab/ALS-kaggle\n\nWe summarized our results in this presentation: https://github.com/gagneurlab/ALS-kaggle/blob/master/reports/ALS_slides.pdf\n","metadata":{}}]}