{"cells":[{"metadata":{"_uuid":"f1a88de32ef105d6df15c22750fbbec2b0d7cb70","_cell_guid":"d6f80b20-a3ca-4018-a59f-0ddf15ffc81a"},"cell_type":"markdown","source":"# Analysis of Molecular Subtypes in Breast Cancer Tumors using Data Visualization\nThis data set contains information on breast cancer patients with 4 major molecular subtypes of tumors; Luminal A, Luminal B, HER2 Enriched, and Basal(triple negative).  \n\n#### This is a project for Data Visualization course for Master of Science in Data Science course for Dr.Bora Pajo. This can also be used as a guide for Data visualization\n\n## Table of Contents\n\nPart 1: Introduction\n\nPart 2: Information on Dataset\n\nPart 3: Visualization using Matplot on Clinical Dataset\n\nPart 4: Visualization using seaborn on Breast Cancer Proteomes\n\n# Part 1: Introduction \n\nBreast Cancer is a multifactorial disease that forms in the cells of the breast. Breast cancer can occur in both men and women, but it's far more common in women. Breast cancer is the most common form of cancer in American women, the average risk of an American Women developing breast cancer in her life time is about 1 in 7 or about 12%. Of that 12 percent of women who do develop breast cancer, the cancer that a breast cancer patient will die from her disease is 2.7%.    \n\n**Density map of worldwide incidents of Breast Cancer**\n![breastcancerstatsworldwide](http://www.worldwidebreastcancer.com/wp-content/uploads/2011/08/breastcancerstatsworldwide.jpg)\n\nBreast cancer is often referred to as one disease, however there are many different types of breast cancers based on varying types of tumors. Tumors can vary in location, size, shape, and grade(severity). These characteristics, along with hormone receptor status and HER2 status affect prognosis. \n\n**Breast Cancer by the Numbers** \n![bythenumbers](http://www.julep.com/blog/wp-content/uploads/2014/10/FINAL-2014-Infographic-Julep_EdithSanford_5x5-1.jpg)\n\nThere are many factors that increase one’s chances in developing breast cancer, mutations in DNA are often the root causes of breast cancer. Other causes that may influence one’s susceptibility to developing breast cancer are mutations to genes that assist in cell differentiation (Proto-Oncogenes), and mutations to genes that modulate cell division, cellular repair, apoptosis (Tumor Suppressor Genes).  \n\nBreast cancer awareness and research funding has helped created advances in the diagnosis and treatment of breast cancer. Medical advances and Innovation in treating breast cancer has increased survival rates, lower remission rates, and lower the number of deaths associated with the disease. More recently, the introduction of precision medicine and gene therapy has the potential to transform how physician treat breast cancer. Precision medicine refers to the tailoring of medical treatment based on the cellular profile of a disease and the patient’s genome. \n\nResearch has identified 4 major molecular subtypes within breast cancer tumors, these subtypes are based on the genes that cancer cell express: Luminal A, Luminal B, HER2 Enriched, and Basal (triple negative). Identifying and studying these subtypes has potential in planning more effective treatment and developing new therapies. Currently, Prognosis and treatment decisions are guided mainly by tumor stage, tumor grade, hormone receptor status and HER2 status. Molecular subtypes are mostly used in research settings; they are not part of a patients report and are not used to guide treatment. However, the use of molecular subtypes has greatly expanded, based on determining what genes are expressed in tumor samples, identifying subtypes of tumors can improve prognosis.\n\n**The scope of this project is to visualize what the differences in gene expression among molecular subtypes of breast cancer, overall survive time, average age at initial pathological diagnosis.** \n \n\n# Part 2: Information on Dataset\n\n\n1. 77_cancer_proteomes_CPTAC_itraq.csv\n\n\tIncludes 12553 unique genes from a total of 83 breast cancer patients\n\n\n2. clinical_data_breast_cancer.csv\n\n\tContains Clinical information from 105 breast cancer patients \n    \n    Variables: **Complete TCGA ID',\n 'Gender',\n 'Age at Initial Pathologic Diagnosis',\n 'ER Status',\n 'PR Status',\n 'HER2 Final Status',\n 'Tumor',\n 'Tumor--T1 Coded',\n 'Node',\n 'Node-Coded',\n 'Metastasis',\n 'Metastasis-Coded',\n 'AJCC Stage',\n 'Converted Stage',\n 'Survival Data Form',\n 'Vital Status',\n 'Days to Date of Last Contact',\n 'Days to date of Death',\n 'OS event',\n 'OS Time',\n 'PAM50 mRNA',\n 'SigClust Unsupervised mRNA',\n 'SigClust Intrinsic mRNA',\n 'miRNA Clusters',\n 'methylation Clusters',\n 'RPPA Clusters',\n 'CN Clusters',\n 'Integrated Clusters (with PAM50)',\n 'Integrated Clusters (no exp)',\n 'Integrated Clusters (unsup exp)'**\n\n\n3. BCGENES.csv \n\n    This dataset was generated using subset reduction technique to trim down noisy gene variables and dylpr to inner_\\join variables from the 77_cancer_proteomes_CPTAC_itraq.csv and clinical_data_breast_cancer.csv\n    \n    Variables: **'PAM50 mRNA',\n 'myoferlin isoform a',\n 'heat shock protein HSP 90-beta isoform a',\n 'keratin, type II cytoskeletal 72 isoform 1',\n 'dedicator of cytokinesis protein 1',\n 'keratin, type I cytoskeletal 23',\n '1-phosphatidylinositol 4,5-bisphosphate phosphodiesterase beta-3 isoform 1',\n 'TBC1 domain family member 1 isoform 2',\n 'kinesin-like protein KIF21A isoform 1',\n 'ubiquitin carboxyl-terminal hydrolase 4 isoform a',\n 'PREDICTED: myomegalin-like',\n 'KN motif and ankyrin repeat domain-containing protein 1 isoform a',\n 'spermatogenesis-associated protein 5',\n 'TBC1 domain family member 9',\n 'receptor tyrosine-protein kinase erbB-2 isoform a precursor',\n 'epidermal growth factor receptor isoform a precursor',\n 'UPF0505 protein C16orf62',\n 'MICAL-like protein 1',\n 'UTP--glucose-1-phosphate uridylyltransferase isoform a',\n 'probable ATP-dependent RNA helicase DDX6',\n 'L-lactate dehydrogenase B chain',\n 'myelin expression factor 2',\n 'histone-lysine N-methyltransferase NSD3 isoform long',\n 'histone-lysine N-methyltransferase NSD3 isoform short',\n 'signal transducer and activator of transcription 6 isoform 1',\n 'WD repeat-containing protein 91',\n 'UPF0553 protein C9orf64',\n 'ran-binding protein 9',\n 'arfaptin-1 isoform 2',\n 'tonsoku-like protein',\n 'protein LSM14 homolog B',\n 'oxysterol-binding protein-related protein 2 isoform 2',\n 'PDZ domain-containing protein GIPC2',\n 'calpain small subunit 2',\n 'condensin-2 complex subunit G2',\n 'telomere length regulation protein TEL2 homolog',\n 'guanine nucleotide-binding protein G(olf) subunit alpha isoform 1',\n 'squalene synthase',\n 'cathepsin B preproprotein',\n \"5'-nucleotidase domain-containing protein 2 isoform 1\",\n 'COBW domain-containing protein 1 isoform 2',\n 'transcription elongation factor A protein-like 5',\n 'DNA repair protein XRCC4 isoform 1',\n 'transmembrane protein 132A isoform a precursor',\n 'alpha-methylacyl-CoA racemase isoform 3',\n 'trans-acting T-cell-specific transcription factor GATA-3 isoform 1',\n 'uncharacterized protein C7orf43',\n 'molybdopterin synthase catalytic subunit large subunit MOCS2B',\n 'Golli-MBP isoform 1',\n 'hepatocyte nuclear factor 3-gamma',\n '39S ribosomal protein L40, mitochondrial',\n 'migration and invasion enhancer 1',\n 'ubiquitin-conjugating enzyme E2 E3',\n 'protein S100-A13',\n 'transcription initiation factor TFIID subunit 8',\n 'phosphoribosyltransferase domain-containing protein 1',\n 'polyadenylate-binding protein-interacting protein 2'].**\n\n### Data manipulation of Breast Cancer Proteomes in Rstudio\n\nThe data used in this portion of the project was gathered from kaggle. The original dataset includes 12553 unique genes from a total of 83 breast cancer patients. For visualization purposes, it will be difficult to apply data visualization techniques to all 12553 unique genes. \n\nIt is common that genetic datasets exhibit cases where n, the number of observations is larger than the p, the number or predictors. When p >n, there is no longer a unique least squares coefficient estimate. Often when p>n, high\nvariance and overfitting are a major concern in this setting. Thus,\nsimple, highly regularized approaches often become the methods of choice. To address this issue, there are many techniques (Subset selection), (Lasso and Ridge) and (Dimension reduction) to exclude irrelevant variables from a regression or a dataset. \n\nUsing a Shrinkage Method known as Lasso in Rstudio, we fit a model containing all 12553 predictors\nthat constrains or regularizes the coefficient estimates, or equivalently, that\nshrinks the coefficient estimates towards zero. Variables with a coefficient estimate of zero are left out of the final model. **The results of the lasso method on the 77_cancer_proteomes_CPTAC_itraq.csv data indicate 57 relevant variables that will be used in predicting tumor type and for data visualization.**\n\nThe significant 57 genes from a sample of breast cancer patients along with the associated tumor types are then queried out in RStudio into dataset BCGENES.csv. **BCGENES.csv is the main dataset which will be used in seaborn for data visualization purposes.** A variable importance plot from Rstudio of the cancer data indicates the top 5 most influential genes we will use for data visualization purposes.\n\n### Datasets used for Visualization \n\nMatplot used the on Clinical Dataset, **clinical_data_breast_cancer.csv**\n\nSeaborn used on the dataset containing the significant 57 genes from a sample of breast cancer patients. **BCGENES.csv **"},{"metadata":{"_kg_hide-input":true,"_uuid":"d6466982815b01479bb48e73822b6b50b913268d","_cell_guid":"0b7549e1-574e-4787-9ebb-75cdb60682f5","trusted":true},"cell_type":"code","source":"# Library required and Dataset is imported\n# We need matplot to plot the data in for Iris dataset\nimport matplotlib.pyplot as plt\n\n# We need numpy to import and work with the data\nimport numpy as np\n\n# We need seaborn as it's a Python graphing library\nimport seaborn as sns\n\n# Let's import python first as it an important package for data processing\nimport pandas as pd\n\n# This method produces a lot of warning, so let's import the package and ignore those warnings \nimport warnings \nwarnings.filterwarnings(\"ignore\")\n\n# This handles  the error of jupyter not plotting\n%matplotlib inline\n\nClinical = pd.read_csv('../input/breastcancerproteomes/clinical_data_breast_cancer.csv')\nClinical.head(5)","execution_count":null,"outputs":[]},{"metadata":{"_uuid":"c26ea1e7973775f3f9e20a0b012f3efb0bc05b27","_cell_guid":"a25c1a60-3fd5-4f0b-9406-50989adbe545","trusted":true},"cell_type":"code","source":"# Check the data for any null values\n# We elect to not change any of the null values to numeric 0 or remove the data, in this dataset a null value indicates no information on a patient record\n# The numeric 0 can not be used to represent no information.\n#0 represents a measurement and a null value represents no information. \nClinical.isnull().values.any()","execution_count":null,"outputs":[]},{"metadata":{"_uuid":"b7d6c43adeb44b8b70139fe27418333106655d8c","_cell_guid":"ad48522d-3428-4f29-be56-e5bcb99d8bbf","trusted":true},"cell_type":"code","source":"# To get a list of the variables that might be of interest for data visualization \nlist(Clinical)","execution_count":null,"outputs":[]},{"metadata":{"_uuid":"646d1f202dafa13d075ff05e22b769a8ad0241c5","_cell_guid":"cb01e3cf-e80b-4c16-8ca7-f51eed895d55"},"cell_type":"markdown","source":"# Proportion Of Different Tumor Types Using a Pie Chart\nA Pie Chart is a standardize method for llustrating the numerical proportion of the data. The pie chart below displays the numerical proportion of tumor types within our data."},{"metadata":{"_uuid":"6516cd6d68f65ebe4805d1f1abf6e82cd36fe6e7","_cell_guid":"3656d863-4655-463a-8fc8-15adbc3c426e","trusted":true},"cell_type":"code","source":"# The pie chart below displays the numerical proportion of tumor types that make up out data.\n#Luminal_B is the most common type of tumor, follow by Luminal_A,Basal_like and Her2_enriched\n\nfig = plt.figure()\nfig,ax = plt.subplots(figsize=(8,8))\nplt.rcParams['font.size'] = 15.0\nClinical['PAM50 mRNA'].value_counts(sort=False).plot(kind='pie',autopct='%1.0f%%')\nplt.title('Numerical Proportion of Tumor Types',fontsize=20)\nfig.savefig('PIE.png')","execution_count":null,"outputs":[]},{"metadata":{"_uuid":"0914b03bf6c114655168cf3280de805eca3969a5","_cell_guid":"8ec7836f-c56b-4832-b01d-81ac2019791a"},"cell_type":"markdown","source":"## Barplot of Overall Survival Time in different Cancer Tumor Types\n\nA bar chart is a graphical representation of categorical data with rectangular bars proportional to the values that they represent. The bars can be plotted vertically or horizontally. \n\nSource:https://en.wikipedia.org/wiki/Bar_chart\n\nIn cancer treament Overall Survival is a measurement of the length of time from either the date of diagnosis or the start of treatment for a disease, such as cancer, that patients diagnosed with the disease are still alive. In a clinical trial, measuring the overall survival is one way to see how well a new treatment works.\n\n**Let’s use a bar chart to visually represent the average overall survival based on tumor type.**\n\nSource:https://www.cancer.gov/publications/dictionaries/cancer-terms?cdrid=655245"},{"metadata":{"_uuid":"9119d539da216e0dd99bb82611f0636ad53cecff","_cell_guid":"40037944-3e1d-4642-94dc-766f36781450","trusted":true},"cell_type":"code","source":"#Before we can produce a barplot of Overall Survival Time, we need to group the breast cancer tumor types.\n#Grouping the tumor types allows for one to get the mean Overall Survival Time of our patients. \nTumorOS=Clinical.groupby(by='PAM50 mRNA').mean()[['OS Time']]\nTumorOS","execution_count":null,"outputs":[]},{"metadata":{"collapsed":true,"_uuid":"6fc1f87ba938a4908e53b34191c9dce3a02c4d73","_cell_guid":"78d33f53-208f-4519-8d31-48ac0922117f"},"cell_type":"markdown","source":"Consistent with current research, as HER2_Enriched tumors are known to be the more aggressive of the tumor types. Overall survival time among HER2_Enriched tumors have always been considered significantly shorter than other tumor types."},{"metadata":{"_uuid":"a4b6a5f273bf65cfd3e57fe446d3fc4bc6a92a57","_cell_guid":"5fe99198-d9cc-4a32-bb19-cf3b54b35811","trusted":true},"cell_type":"code","source":"# The plot below indicates that HER2_Enriched tumors on average have an lower overall survival time.\nfig = plt.figure()\nTumorOS.plot(kind='bar',figsize=(12,8),grid=True,title='Comparing the average overall survival (OS) times among different tumor types')\nplt.ylabel('Days',fontsize=15)\nfig = plt.figure()\n\nfig.savefig('BP1.png')","execution_count":null,"outputs":[]},{"metadata":{"_uuid":"0c858919d265a6112bfae3d0e37d500d317b4bfa","_cell_guid":"647baff6-18f6-4942-b545-3c89ee314262"},"cell_type":"markdown","source":"## Line Chart Of Average Age Of Initial Pathologic Diagnosis Of Breast Cancer\n\nA line chart or line graph is a type of graphical representation that displays information as a series of data points, these data points are connected by straight line segments. \n\nClinical research has established a positive correlation among age and breast cancer. Age is a risk factor for breast cancer, the older a woman is, there is a high chance that she will develop breast cancer. Rates of breast cancer are low in women under 40, less than 5 percent of breast cancer patients are younger than 40. The median age of diagnosis of breast cancer is 62. \n\n**Below is a line graph where we can display the average age of initial pathologic diagnosis of cancer based on tumor type.**"},{"metadata":{"_uuid":"e77f70136122202295f97be275d2ce8fb4192103","_cell_guid":"1bdb05bc-7469-49ed-931e-fa3dfcb70cda","trusted":true},"cell_type":"code","source":"# Using the line graph we can display the average age of initial pathologic diagnosis of cancer based on tumor type\n# We are grouping by tumor type and comparing average age of initial pathologic diagnosis of cancer. \n# Our visualization indicates that Basel like tumor types are diagnosis at a younger age and is rarer. \nfig = plt.figure()\ndiatime=Clinical.groupby(by='PAM50 mRNA').mean()[['Age at Initial Pathologic Diagnosis']]\ndiatime.plot(kind='line',figsize=(13,9),grid=True,title='Age at Initial Pathologic Diagnosis')\n\nfig.savefig('LINE.png')","execution_count":null,"outputs":[]},{"metadata":{"_uuid":"489fbaab193c1bd8907ae28374734803ea35401b","_cell_guid":"06803edd-6c8b-4ef0-b8bb-6f6b3c2a2aee"},"cell_type":"markdown","source":"**Part 4: Visualization using seaborn on Breast Cancer Proteomes**"},{"metadata":{"_uuid":"c55338f3e6446d4a4ca3954db149d22af841852d","_cell_guid":"6e8e1ab7-aaac-4363-aab1-20858b90c337","trusted":true},"cell_type":"code","source":"#Loading the BCGENES.csv for data Visualization\n\nBC = pd.read_csv('../input/bcgenes/BCGENES.csv')\nBC.head(5)","execution_count":null,"outputs":[]},{"metadata":{"_uuid":"23031979fb8add2e59dc7cb3b370b3c0eebffbd5","_cell_guid":"79abb747-0d11-40a6-b45b-b1557a0f5b66","trusted":true},"cell_type":"code","source":"# Check the data for any null values\nBC.isnull().values.any()","execution_count":null,"outputs":[]},{"metadata":{"_uuid":"e8fa77dd6db0d48347e3fee72207517903ec6cbf","_cell_guid":"32d2b614-3877-465e-ba2f-49de9da392b6","trusted":true},"cell_type":"code","source":"# To get the names of the genes\nlist(BC)","execution_count":null,"outputs":[]},{"metadata":{"_uuid":"dea1bf33cb19502663906e31fd8e6ec119196a1b","_cell_guid":"0c68df09-0647-4f74-b28b-318b53e68237"},"cell_type":"markdown","source":"From Rstudio we identified and listed in descending order of the most influential gene in determining breast cancer tumor type. The following code queries and compares the 5 most influential genes from the Breast Cancer Proteomes csv among tumor types."},{"metadata":{"_uuid":"76d2f9f2939e767dfcf968d6e75c0036f54c21a3","_cell_guid":"e3ee8b34-8090-4c60-a1e6-d387dea3fdfe","trusted":true},"cell_type":"code","source":"# We randomly select 4 of the top ten most influential protein from the Breast Cancer Proteomes csv \n#and assign them to a new dataframe. \n\ndv=BC[['PAM50 mRNA','MICAL-like protein 1',\n    'hepatocyte nuclear factor 3-gamma',\n   'keratin, type I cytoskeletal 23','TBC1 domain family member 9']] \n\n\n#Re-assigning proteins to gene name\ndv.columns = ['PAM50 mRNA','MICAL-1','FOXA3','KRT23',\n                     'TBC1D9']","execution_count":null,"outputs":[]},{"metadata":{"_uuid":"93e55a87df9bde3a270a9562036c0ab621241d3c","_cell_guid":"6ef7790e-b41b-4792-af8e-83935eb379d4"},"cell_type":"markdown","source":"## Pairplot Of The Four Most Influential Genes in Breast Cancer\n\nWe use a PairPlot function from Seaborn to identify where different tumors tend to cluster on a scatterplot in respect to certain gene measurements. Scatter plots are like line graphs; they show how much one variable is affected by another and how the data is dispersed when used to compare groups. A scatter plot can suggest various kinds of correlations between variables.\n\nBelow is a PairPlot containing scatterplots for the following genes : 'MICAL-like protein 1' (**MICAL-1**), 'hepatocyte nuclear factor 3-gamma' (**FOXA3**'), 'keratin, type I cytoskeletal 23' (**KRT23**), and 'TBC1 domain family member 9' (**TBC1D9**).\nImportant conclusions from Pairplot are that: Luminal_A and Luminal_B tumors have similar means in measurements of 'TBC1 domain family member 9', the mean of Luminal_A,Luminal_B,and HER2_Enriched tumors are similar and have overlapping measurements. Whereas measurements from Basel like tumors tend to stand out on their own with very little over against the other tumor types.\n"},{"metadata":{"_uuid":"bd1000f4d9c970699a4f063a38521c1fbe4ed430","_cell_guid":"2043b39b-0239-44c5-9108-730afce113f5","trusted":true,"_kg_hide-output":false,"_kg_hide-input":false},"cell_type":"code","source":"#PairPlot\npairplot = sns.pairplot(dv.iloc[:,0:6],\n                        hue=\"PAM50 mRNA\",\n    \n                        diag_kind=\"kde\",\n                        diag_kws=dict(shade=True),\n                        plot_kws = {'alpha': 0.6, 's': 80, 'edgecolor': 'k'}\n                       \n                       );\n\n\npairplot.savefig(\"pairplot.png\")","execution_count":null,"outputs":[]},{"metadata":{"_uuid":"0d76141d3611fd75c75f5a229466df5c826764d2"},"cell_type":"markdown","source":""},{"metadata":{"_uuid":"f9782f842c3e1a67a0ae5c0903bb8c35fcb42e42"},"cell_type":"markdown","source":""},{"metadata":{"_uuid":"431b7f4420f735368c5a79dc12387437365116db","_cell_guid":"982eca79-9653-45ee-9f18-85e52ac996d4"},"cell_type":"markdown","source":"## Heatmap Of The Four Most Influential Genes in Breast Cancer \n\nThe heat map tumor of feature measurements indicates which measurements are highly correlated/related to one another. \n\nImportant conclusions of the heatmap are that, 'hepatocyte nuclear factor 3-gamma' and 'TBC1 domain family member 9' are highly positively correlated with another. \nFurther research has confirmed that 'hepatocyte nuclear factor 3-gamma' regulate the transcription of a diverse groups of genes into proteins and that TBC1 is a protein that has a role in regulating cell growth. **Perhaps mutations of the 'hepatocyte nuclear factor 3-gamma' genes leads to an over expression of the TBC1 protein causing a pro tumor environment and unregulated cell growth which is known as cancer.**\n\nSource: https://en.wikipedia.org/wiki/Hepatocyte_nuclear_factors \n\nOther positively correlated predictors are 'keratin, type I cytoskeletal 23' and 'MICAL-like protein 1'."},{"metadata":{"_uuid":"5c0d5623afc72a2cb3af98a733e8375df5dabb67","_cell_guid":"0f434a3d-836f-4608-b411-5390abf17e4a","trusted":true},"cell_type":"code","source":"#Heatmap\n\nimport seaborn as sns\ncorr = dv.iloc[:,0:5].corr()\nheat1=sns.heatmap(corr, \n            xticklabels=corr.columns.values,\n            yticklabels=corr.columns.values)\n\nfigure = heat1.get_figure()\nfigure.savefig('heat1.png', dpi=400)","execution_count":null,"outputs":[]},{"metadata":{"_uuid":"ae802d789a7d06f800c18741f1220ea4193eb39c","_cell_guid":"e99bb7e1-7194-4465-bbd3-91494193e8af"},"cell_type":"markdown","source":"## Seaborn FacetGrid of 'MICAL-like protein 1' gene  and 'hepatocyte nuclear factor 3-gamma' Gene Among Tumor Type\n\nThe facetgrid from seaborn is similar to a scatterplot. \n\nScatter plots are like line graphs; they show how much one variable is affected by another and how the data is dispersed when used to compare groups. A scatter plot can suggest various kinds of correlations between variables. In Rstudio we identified that 'MICAL-like protein 1' and 'hepatocyte nuclear factor 3-gamma' are the most influential genes in determining tumor type.  We plot 'MICAL-like protein 1' and 'hepatocyte nuclear factor 3-gamma' among tumor types on a scatter plot. \n\nData points from Luminal_A,Luminal_B,and HER2_Enriched tumors are similar and have overlapping measurements."},{"metadata":{"_uuid":"6fb97a80599751439c3ae4d7251b7ea6584a2826","_cell_guid":"4f4b50f4-38f2-4985-9e9f-0d295f5bbb6a","trusted":true},"cell_type":"code","source":"SP = sns.FacetGrid(dv, hue=\"PAM50 mRNA\", size=8) \\\n   .map(plt.scatter, 'MICAL-1','FOXA3',)\\\n   .add_legend()\n\nSP.savefig(\"SP.png\")","execution_count":null,"outputs":[]},{"metadata":{"_uuid":"f1747db47172926934fbb26c962cdacd29787bd5","_cell_guid":"85be81a6-2ca8-4297-9b23-2f829b803421"},"cell_type":"markdown","source":"## Boxplot of the most Influential predictor : MICAL-like protein 1\n\nBoxplots is a standardized method for graphically plotting groups of numeric data based on the distribution of the data. Boxplots are extremely helpful in visualizing the minimum, first quartile, median, third quartile and maximum of different groups. The advantage of using boxplots is its ability to quickly display one or more sets of data graphically.\n\nSource: https://en.wikipedia.org/wiki/Box_plot \n \nBelow is a boxplot of the MICAL-like protein 1 gene from the four tumor types: Luminal A, Luminal B, HER2 Enriched, and Basal-like (triple negative).\n\nMICAL-proteins participate in various cellular activates, they regulate axonal growth cone repulsion, membrane trafficking, and apoptosis in cells. Potentially mutations genes regulating the expression of MICAL-proteins has a pro-cancer effect on cells. \n\nBased on the boxplot of the MICAL-like protein 1 gene, Basal_like tumors have the highest mean of the tumor types; they are strikingly different in terms of expression of MICAL-like protein 1 gene. The mean from HER2_enriched tumors and Luminal_B closely resemble each other. \n\nSource: https://www.ncbi.nlm.nih.gov/pubmed/23834433"},{"metadata":{"_uuid":"32bde6e758fd3726f11cf165e18ecef9d306c094","_cell_guid":"4d581db6-098d-452e-a6ed-fb7693e1a04d","trusted":true},"cell_type":"code","source":"ax = sns.boxplot(x=\"PAM50 mRNA\", y='MICAL-like protein 1', data=dv)\nax = sns.stripplot(x=\"PAM50 mRNA\", y='MICAL-like protein 1', data=dv, jitter=True, edgecolor=\"gray\")\n\nfigure = ax.get_figure()\nfigure.savefig('boxplot.png', dpi=400)","execution_count":null,"outputs":[]},{"metadata":{"_uuid":"4135a62a517a660b9a4e5c8aa652815d5072134e","_cell_guid":"dba6a2ca-0339-4065-a26d-b0ac2837fee4"},"cell_type":"markdown","source":"## Violin Plot of the most Influential predictor : MICAL-like protein 1\n\nFrom the Violin plots of the MICAL-like protein 1 gene, the probability density of Basal_like tumors seem platykurtic, a statistical distribution with dispersed observations along the x axis that results in wider and flatter peak, whereas HER2_enriched, Luminal_B and Luminal_A seem leptokurtic or normal, data points that sit clustered resulting in a higher and narrower peak.\n\n![leptokurtic](https://img.tfd.com/mk/K/X2604-K-11.png) \n\nSource: https://en.wikipedia.org/wiki/Violin_plot\n\nThe previous box plot indicated that the mean from HER2_enriched tumors and Luminal_B closely resemble each other; the Violin plot indicates that the position and relative amplitude of peaks are similar in Luminal_B and Luminal_A tumors. Other observations indicates that HER2_enriched  and Basal_like tumors have a larger ranges than Luminal_B and Luminal_A tumors. "},{"metadata":{"collapsed":true,"_uuid":"a8256e77c8b0b421c6e6700a34f63e7b09c9b33b","_cell_guid":"aabe1ff2-ea6e-4b40-b7bc-27cc20e8f1c3","trusted":false},"cell_type":"code","source":"snsviolinplot = sns.violinplot(x=\"PAM50 mRNA\", y='MICAL-like protein 1', data=dv, size=10)\nfigure = snsviolinplot.get_figure()","execution_count":null,"outputs":[]}],"metadata":{"language_info":{"name":"python","version":"3.6.6","mimetype":"text/x-python","codemirror_mode":{"name":"ipython","version":3},"pygments_lexer":"ipython3","nbconvert_exporter":"python","file_extension":".py"},"kernelspec":{"display_name":"Python 3","language":"python","name":"python3"}},"nbformat":4,"nbformat_minor":1}