{"cells":[{"outputs":[],"source":"<h2>My first attempt at bioinformatic analysis (besides a few Rosalind challenges). I have an undergraduate+ understanding of biology and will try to explain things as I go along. I will tackle the questions:</h2>\n    \n\n - We have a number of different gene prediction systems in the dataset,\n   how do they compare to each other? How do they compare to the mRNA\n   data?\n\n<h3>I hope to investigate these issues by</h3> \n\n 1. Gaining understanding of our data sets\n 2. Comparing gene prediction sets to mRNA data set\n","cell_type":"markdown","execution_count":null,"metadata":{"_execution_state":"idle","_uuid":"e421ee5e517d049e2e478b3ed09f2ecceafcc11a","_cell_guid":"1758ca6d-ac05-4251-ab4c-bf78852c2fe0","collapsed":false}},{"outputs":[],"source":"import numpy as np\nimport pandas as pd \nimport matplotlib.pyplot as plt\n%matplotlib inline\nimport seaborn as sb\nimport Bio\n\nprint(Bio.__version__)","cell_type":"code","execution_count":null,"metadata":{"_execution_state":"idle","_uuid":"995b707fbc0fb1f61842a536b5279f1cd0073231","_cell_guid":"ab252dd7-da0e-493d-97bc-4af5495cd655","trusted":false}},{"outputs":[],"source":"# Loading  data\n\nagsts = pd.read_csv('../input/genes-augustus.csv')\ngnscn = pd.read_csv('../input/genes-genscan.csv')\nrfsq = pd.read_csv('../input/genes-refseq.csv')\nxrfsq = pd.read_csv('../input/genes-xeno-refseq.csv')\n\n### NOTE: cannot find gene names when searched on ensembl.org!!!\nensmbl = pd.read_csv('../input/genes-ensembl.csv')\npredictions = {'AUGUSTUS':agsts, 'GENSCAN':gnscn, 'REFSEQ':rfsq, 'XENOREFSEQ':xrfsq, 'ENSEMBL':ensmbl}\nagsts.head()","cell_type":"code","execution_count":null,"metadata":{"_execution_state":"idle","_uuid":"2ee31dd26cd6623c66ed841bf70e4ef01c844e64","_cell_guid":"74ee2674-490a-46fc-99a5-e48e9a27d636","trusted":false,"collapsed":false}},{"outputs":[],"source":"<h2>A bit on what all that means.</h2>\n\nWe can find out some things from [https://www.kaggle.com/mylesoneill/drosophila-melanogaster-genome][1],\nbut some of the descriptions are a bit opaque. \n\n\n  - **Bin,** from what I can tell, is of no consequence.\n  - **txStart, txEnd, cdsStart, cdsEnd** stand for transcription start/end and coding sequence start/end. \n    the *transcription sequence* is what is read by the [ribosome][2], and the coding sequence is everything that is read by the ribosome **AND** translated into RNA.\n  - **Score** I have no clue, sorry.\n  - **cdsEndStat** I think is bookkeeping; is it known where (or if) the coding sequence ends?\n  - **exonFrame** from the <a href=\"https://www.ncbi.nlm.nih.gov/pmc/articles/PMC441517/bin/nar_32_suppl-2_W309__index.html\">literature</a> this is *\"the number of\nbases of the exon (0, 1 or 2) upstream of the most upstream codon boundary of the exon\"*, so basically does the beginning of this codon match with the beginning of its first full <a href=\"https://en.wikipedia.org/wiki/Genetic_code\">codon</\n\n  [1]: https://www.kaggle.com/mylesoneill/drosophila-melanogaster-genome\n  [2]: https://en.wikipedia.org/wiki/Ribosome \"ribosome\"","cell_type":"markdown","execution_count":null,"metadata":{"_execution_state":"idle","_uuid":"88add75d6ed79682aad41f5a927efbbbad26858d","_cell_guid":"733d8821-578a-4390-80a6-38554161ca37","collapsed":false}},{"outputs":[],"source":"gnscn.head()","cell_type":"code","execution_count":null,"metadata":{"_execution_state":"idle","_uuid":"c16b106a4ccace75a6ad57f1127ddf0fe2ab9d1f","_cell_guid":"aaf7bb82-39bc-41cf-8eb6-2b4fdf6889b8","trusted":false,"collapsed":false}},{"outputs":[],"source":"<h3>Hey this is pretty similar to what's in the AUGUSTUS file (thanks Myles!).</h3>","cell_type":"markdown","execution_count":null,"metadata":{"_execution_state":"idle","_uuid":"67f597072a566dfd1742a12023a4c72a5ba4d78c","_cell_guid":"aaf10fa0-aec8-42fc-a742-96e2fdf78068","collapsed":false}},{"outputs":[],"source":"# It would be interesting to sort by transcription start to see the first genes of some chromosomes\nagsts.sort_values(by='txStart')","cell_type":"code","execution_count":null,"metadata":{"_execution_state":"idle","_uuid":"bf642296027e5d42ed51d0af5744bf0cb32506b3","_cell_guid":"314d1ff8-f03d-4e1c-90b4-fb29473c6cd8","trusted":false,"collapsed":false}},{"outputs":[],"source":"<h2>Now what is this?!</h2>\n\nThose are a bunch of unmapped sections of chromosomes. I don't know if the transcription *really* starts at the first base-which seems unlikely given that the ends of chromosomes are usually capped with <a href=\"https://en.wikipedia.org/wiki/Telomere\">telomeres</a> (at least in humans...). Maybe the transcription 'starts' at 0 for these few sequences because they are unmapped and we don't know much about them.\n\n<h3>Time to compare our predicted gene data sets. First, how many genes did each set predict?</h3>\n","cell_type":"markdown","execution_count":null,"metadata":{"_execution_state":"idle","_uuid":"9b363e8dbff0c22b8d5fc4cf97285aec0d001ba5","_cell_guid":"71070ab2-5be2-47ec-b335-4c7c1cd68ece","collapsed":false}},{"outputs":[],"source":"for method, dataset in predictions.items():\n    print(method + \" # of genes: \" + str(dataset.shape[0]) + '\\n')","cell_type":"code","execution_count":null,"metadata":{"_execution_state":"idle","_uuid":"6fa68215c97950c9e7736186d29e1a039fa279e4","_cell_guid":"85bc8447-ae53-4759-9c52-a6adf2b721d8","trusted":false,"collapsed":false}},{"outputs":[],"source":"That pretty much makes sense. AUGUSTUS and GENSCAN are older while the other methods are based on an entire database of sequences and are updated often. Thus I only focus only on RefSeq and Ensembl.\n\n","cell_type":"markdown","execution_count":null,"metadata":{"_execution_state":"idle","_uuid":"340fa7ad3e5f489f7df2fe2d00111769f1ec35ff","_cell_guid":"eca95c13-a2b2-4e3d-9e55-611e2d2cda6f","collapsed":false}},{"outputs":[],"source":"refseq = pd.read_csv('../input/genes-refseq.csv')\nrefseq = refseq.groupby('chrom')['chrom'].count()\nrefseq = refseq.loc[refseq.values > 10]\n\nensmbl = pd.read_csv('../input/genes-ensembl.csv')\nensmbl = ensmbl.groupby('chrom')['chrom'].count()\nensmbl = ensmbl.loc[ensmbl.values > 10]\n\ndata = {'rfsq': refseq, 'ensmbl': ensmbl}\ncounts = pd.DataFrame.from_dict(data=data)\ncounts\n","cell_type":"code","execution_count":null,"metadata":{"_execution_state":"idle","_uuid":"126b498dc68c74403ae740b6356661e5b1e07de7","_cell_guid":"1673785e-688e-46d6-b7df-43c586719aa9","trusted":false,"collapsed":false}},{"outputs":[],"source":"<h2>to be continued...</h2>","cell_type":"markdown","execution_count":null,"metadata":{"_execution_state":"idle","_uuid":"07380e666f8a64d7c828fad3c3a69f2eb31fc9f7","_cell_guid":"5745089f-f095-4ed6-ab79-2664023e6e86","collapsed":false}}],"nbformat":4,"nbformat_minor":0,"metadata":{"language_info":{"file_extension":".py","version":"3.6.1","codemirror_mode":{"version":3,"name":"ipython"},"nbconvert_exporter":"python","pygments_lexer":"ipython3","mimetype":"text/x-python","name":"python"},"kernelspec":{"language":"python","display_name":"Python 3","name":"python3"}}}