{"cells":[{"metadata":{"id":"27F2Vpla1I3V"},"cell_type":"markdown","source":"# Task 3 - CORD Hackathon Submission"},{"metadata":{"trusted":true},"cell_type":"code","source":"from IPython.display import Image\n","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"Image(\"/kaggle/input/task3-data/Images/Images/Cover2.PNG\")","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"\n## Introduction\n\nCovid-19 has plunged the world into an unprecedented crisis and the global medical community is working tirelessly to not only contain the pandemic but also prepare to manage the disease effectively in the long term. Key to this endeavor is a robust fundamental understanding of the disease itself – how it originated, what do the genetics tell us and can we learn and predict its evolution. There is a myriad of existing research that contains crucial knowledge, but a lack of standard nomenclature or a scheme of organization makes it a difficult resource for the scientific community to leverage. \nTo alleviate this constraint, our team has worked towards delivering a scalable solution that makes it easier for the medical researchers and global scientific community globally to find what they need from this ever-expanding trove of disease-related research. \n"},{"metadata":{},"cell_type":"markdown","source":"## Task Details\n\nWhat do we know about virus genetics, origin, and evolution? \nWhat do we know about the virus origin and management measures at the human-animal interface?\n\nSpecifically, we want to know what the literature reports about:\n\nReal-time tracking of whole genomes and a mechanism for coordinating the rapid dissemination of that information to inform the development of diagnostics and therapeutics and to track variations of the virus over time.\nAccess to geographic and temporal diverse sample sets to understand geographic distribution and genomic differences, and determine whether there is more than one strain in circulation. Multi-lateral agreements such as the Nagoya Protocol could be leveraged.\nEvidence that livestock could be infected (e.g., field surveillance, genetic sequencing, receptor binding) and serve as a reservoir after the epidemic appears to be over.\nEvidence of whether farmers are infected, and whether farmers could have played a role in the origin.\nSurveillance of mixed wildlife- livestock farms for SARS-CoV-2 and other coronaviruses in Southeast Asia.\nExperimental infections to test host range for this pathogen.\nAnimal host(s) and any evidence of continued spill-over to humans\nSocioeconomic and behavioral risk factors for this spill-over\nSustainable risk reduction strategies"},{"metadata":{},"cell_type":"markdown","source":"## What are we delivering?\nThe following Notebook will provide a data extraction mechanism for researchers to readily find relevant publications regarding the overall Task 3 objective"},{"metadata":{},"cell_type":"markdown","source":"## Approach\nThe tool that the Team had to create must be powerful enough to extract data from nearly 60,000 available publications to be searched regarding coronavirus or the more specific COVID-19. General “Find” or “Google” searches were not powerful enough to support this type of data-mining platform, so the team had to be creative in order to be successful. \nThe data set used is the COVID-19 Complete Dataset via Kaggle.\n\nBelow is an outline of this team’s effort to provide a viable research tool from scientists to utilize to cross-reference other publications across the world.\n\n#### Section 1 : Real-time tracking of whole genomes to track variations of the virus over time\n\n#### Section 2: Access to geographic and temporal diverse sample sets to understand geographic distribution and genomic differences\n\n#### Section 3 : Evidence that livestock could be infected\n    3.1 Evidence that livestock could be infected (e.g., field surveillance, genetic sequencing, receptor binding) and serve as a reservoir after the epidemic appears to be over: Evidence of whether farmers are infected, and whether farmers could have played a role in the origin.\n\n    3.2 Evidence that livestock could be infected (e.g., field surveillance, genetic sequencing, receptor binding) and serve as a reservoir after the epidemic appears to be over.: Surveillance of mixed wildlife- livestock farms for SARS-CoV-2 and other coronaviruses in Southeast Asia.\n\n    3.3 Evidence that livestock could be infected (e.g., field surveillance, genetic sequencing, receptor binding) and serve as a reservoir after the epidemic appears to be over.: Experimental infections to test host range for this pathogen.\n\n#### Section 4:  What do we know about virus genetics, origin, and evolution? What do we know about the virus origin and management measures at the human-animal interface?\n\n    Task 4.1- Animal host(s) and any evidence of continued spill-over to humans\n    Task 4.2- Sustainable risk reduction strategies\n    Task 4.3- Socioeconomic and behavioral risk factors for this spill-over\n    \n\n#### Section 5:  Retrieve relevant articles using BERT\n\n"},{"metadata":{},"cell_type":"markdown","source":"### Note :  A few cells are commented to save time. Processed files are saved In order to view processing steps, plese un comment the relevant code"},{"metadata":{"id":"r5txqgAW1I3v"},"cell_type":"markdown","source":"## Import Packages\n\nPython was chosen as the platform to build the tool for its versatility and ubiquity. The final deliverable involved complex text processing, array manipulation and search, and creative visualizations. For this purpose, several specialized Python libraries were utilized including pandas, numpy, scipy and sklearn. "},{"metadata":{"trusted":true},"cell_type":"code","source":"\n!pip install bert-extractive-summarizer\n#!pip install nxviz","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"!pip install -U sentence-transformers","execution_count":null,"outputs":[]},{"metadata":{"id":"dDyvaeLrKzu0","outputId":"07f4ca19-e37b-44f4-99e6-98a3761d7d5e","trusted":true},"cell_type":"code","source":"import pandas as pd\nimport os\nimport numpy as np\nfrom tqdm.notebook import tqdm\nimport scipy\n\n\nimport textwrap\nimport json\nimport logging\nimport pickle\nimport warnings\nwarnings.simplefilter('ignore')\n\nimport re\nimport nltk\nnltk.download('stopwords')\nnltk.download('punkt')\nfrom nltk.corpus import stopwords\nfrom nltk.stem.porter import PorterStemmer\nnltk.download('wordnet') \nfrom nltk.stem.wordnet import WordNetLemmatizer\nfrom nltk.tokenize.treebank import TreebankWordDetokenizer\nfrom nltk import tokenize\nfrom nltk.stem.wordnet import WordNetLemmatizer\nfrom nltk.tokenize import RegexpTokenizer, word_tokenize\nfrom nltk.translate.bleu_score import sentence_bleu\n\nimport json\nimport glob\nimport string\n\n\nfrom scipy.spatial.distance import cdist\nfrom sklearn.metrics.pairwise import cosine_similarity\nfrom sklearn.feature_extraction.text import TfidfVectorizer, CountVectorizer\nfrom IPython.core.display import display, HTML\nimport torch\nfrom sentence_transformers import SentenceTransformer\nfrom transformers import AutoTokenizer, AutoModel\n#from sentence_transformers import models, SentenceTransformer\nimport shutil\n\nimport torch\nfrom transformers import BertTokenizer, BertModel\n\nimport pandas as pd\n\nfrom nltk.tokenize import word_tokenize\nimport numpy as np\n\n# Load the library with the CountVectorizer method\nfrom sklearn.feature_extraction.text import CountVectorizer\nimport numpy as np\n\nimport seaborn as sns\nsns.set_style('whitegrid')\n%matplotlib inline\n\nfrom summarizer import Summarizer\nimport math\n\nimport os, stat\nimport ntpath\nfrom string import ascii_uppercase\n#import nxviz\n","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"%matplotlib inline\nimport numpy as np\nimport math\nimport matplotlib\nimport matplotlib.pyplot as plt\nimport os, stat\nimport ntpath\nfrom string import ascii_uppercase","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"import spacy\nfrom spacy.matcher import PhraseMatcher #import PhraseMatcher class\n","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"import matplotlib\nimport matplotlib.pyplot as plt\nfrom wordcloud import WordCloud","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"## Download Data (In case you want a local copy of the data)"},{"metadata":{"trusted":true},"cell_type":"code","source":"# Specify the Kaggle Username and Key to use the Kaggle Api\n\n# os.environ['KAGGLE_USERNAME'] = '*************'\n# os.environ['KAGGLE_KEY'] = '****************'","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"# from kaggle.api.kaggle_api_extended import KaggleApi\n\n# api = KaggleApi()\n# api.authenticate()\n\n# api.dataset_download_files(dataset=\"allen-institute-for-ai/CORD-19-research-challenge\", path=DATA_PATH, unzip=True)\n\n# HTML('''<script>\n# code_show=true; \n# function code_toggle() {\n#  if (code_show){\n#  $('div.input').hide();\n#  } else {\n#  $('div.input').show();\n#  }\n#  code_show = !code_show\n# } \n# $( document ).ready(code_toggle);\n# </script>\n# The raw code for this IPython notebook is by default hidden for easier reading.\n# To toggle on/off the raw code, click <a href=\"javascript:code_toggle()\">here</a>.''')","execution_count":null,"outputs":[]},{"metadata":{"id":"ooc1W4oz1I31"},"cell_type":"markdown","source":"## Data Pre-processing\n\nFor the ease of searching and indexing, all the research articles were converted into flat text files and stored in a table format. In this step, all the articles were parsed and filtered for relevance. As a framework, we used categories of keywords. For example, to address animal source, the keywords were categorized as animal words, spill-over words and virus words. Through this exercise, we were able to narrow down the number of relevant articles to around 8000. This framework was also utilized for the other two sub-tasks. "},{"metadata":{},"cell_type":"markdown","source":"### Specify Data Folders"},{"metadata":{"trusted":true},"cell_type":"code","source":"DATA_PATH = os.getcwd()+'/kaggle/input/CORD-19-research-challenge/'","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"### Combine articles to a dataframe"},{"metadata":{},"cell_type":"markdown","source":"To save time, combined data is stored as a csv 'fin_df.csv'. That is loaded and used. Uncooment the below lines to combine articles to a DataFrame"},{"metadata":{"trusted":true},"cell_type":"code","source":"#Get data from path '/data1/cov19/kaggle_data_0331/'\n'''\nbio_path = '/kaggle/input/CORD-19-research-challenge//biorxiv_medrxiv/'\ncomm_path = '/kaggle/input/CORD-19-research-challenge//comm_use_subset/'\nnon_comm_path = '/kaggle/input/CORD-19-research-challenge/noncomm_use_subset/'\ncustom_path = '/kaggle/input/CORD-19-research-challenge/custom_license/'\njournals = {\"BIORXIV_MEDRXIV\": bio_path,\n             \"COMMON_USE_SUB\" : comm_path,\n             \"NON_COMMON_USE_SUB\" : non_comm_path,\n             \"CUSTOM_LICENSE\" : custom_path}\n'''","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"'''\ndef parse_each_json_file(file_path,journal):\n    inp = None\n    with open(file_path) as f:\n        inp = json.load(f)\n    rec = {}\n    rec['sha'] = inp['paper_id'] or None\n    rec['title'] = inp['metadata']['title'] or None\n    abstract = \"\\n \".join([inp['abstract'][_]['text'] for _ in range(len(inp['abstract']) - 1)]) \n    rec['abstract'] = abstract or None\n    full_text = []\n    for _ in range(len(inp['body_text'])):\n        try:\n            full_text.append(inp['body_text'][_]['text'])\n        except:\n            pass\n\n    rec['full_text'] = \"\\n \".join(full_text) or None\n    rec['source'] =  journal     or None    \n    return rec\n\ndef parse_json_and_create_csv(journals):\n    journal_dfs = []\n    for journal, path in journals.items():\n        parsed_rcds = []  \n        json_files = glob.glob('{}/**/*.json'.format(path), recursive=True)\n        for file_name in json_files:\n            rec = parse_each_json_file(file_name,journal)\n            parsed_rcds.append(rec)\n        df = pd.DataFrame(parsed_rcds)\n        journal_dfs.append(df)\n    return pd.concat(journal_dfs)\n\n\nfin_df = parse_json_and_create_csv(journals=journals)\nfin_df.head()\n\n'''","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"fin_df = pd.read_csv('/kaggle/input/task3-data/fin_df.csv')","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"fin_df.head()","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"## Section 1 : Real-time tracking of whole genomes to track variations of the virus over time\n \n\n### Method to measure variations in different Genome sequences"},{"metadata":{},"cell_type":"markdown","source":"### 1.1 Approach\n\n\nTo accomplish this task we follow the steps below:oRetreive the genome suquences from publicly available realtime databases   such   as China   National   GeneBank& GISAID EpiCoV™ DatabaseoUse  same  algorithm  from  Task  3.2  to classify  wether  the  new samples or sequences belong to exsiting or new groups oDetermine wether  or  not  if  new  sequences  are  changing  over time•This task tracks temporal variations •Bootstrap confidence interval of the measured metric•Hierarchical classification of the sample sequenceThis subtask relates to the evolution of the genome sequence in time.  To study these  variations,we  use  the  same  approach  as  described  in  section  3.2.    The only two differences are as follows.  1.First,   the   input sequences   must   be   specifiedfor   an   individual geographical  location,  to  remove  the  spatial  dependency.    Also,  these sequences  must  be  specified  for  a  given  period  of  time  and  sorted  by date in ascending order.  2.The algorithm of section 3.2 is then appliedbut is modified so that it only analyzes the consecutive pairs of sequences, e.g. 1st & 2nd sequences; 2nd & 3rd; 3rd & 4th, ..., (Nth-1)& Nth, where N is the number of input sequences.\n\n\nOur approach is to convert the genome sequences to waveforms and determine a metric that measures how much different or similar two  sequences are.    In order to do this, we find the confidence interval of this metric  and compare it to different threshold levels"},{"metadata":{"trusted":true},"cell_type":"markdown","source":"#### Method executed with two genome sequences\n\nGenome Sequence ConversionIn this step, the input genome sequence is converted to a random signal.  This is  useful,  since  the  new waveform  is  much  easier  to  visualize and  manipulate than  letter  patterns.    Furthermore,  this  allows  us  to  use  signal  correlation  in order to find how different or similar the two signals are.    The Figure below shows the signal conversion for the two input sequences that will  be  analyzed.  Each  of  the  characters  of  the  sequence  is  translated  to  a complex random number where the same random number is assigned to every character occurrence."},{"metadata":{"trusted":true},"cell_type":"code","source":"Image(\"/kaggle/input/task3-data/Images/Images/SequenceConversion.PNG\")","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"#### Reading and Assembling Genome sequence file"},{"metadata":{"trusted":true},"cell_type":"code","source":"###-------------------------------------------------------------\n### READING AND ASSEMBLING GENOME SEQUENCE FILE\n###-------------------------------------------------------------\n#File 1 to compare\nfname = open(\"/kaggle/input/task3-data/gnome_data_countries/gnome_data_countries/CHINA_SHENZHEN_MN938384.1.fasta\",\"r\")\nA = fname.read().replace('\\n','')\nfname.close()\nsequence1 = (list(A))\n\n#File2 to compare\nfname2 = open(\"/kaggle/input/task3-data/gnome_data_countries/gnome_data_countries/italy_MT066156.1.fasta\",\"r\")\nB = fname2.read().replace('\\n','')\nfname.close()\nsequence2 = (list(B))","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"#### Conversion of Sequence into a time domain Complex signal\n\nSince  the  input  sequences  are  about  30,000  samples  long,  the  signal  is divided in equal length segments.  The length of the segment selected in our study is 1024 but it can be changed to any given value.  Figure below showshow the signal is segmented.  In this example,we get 30 segments of size 1024 for the ~30,000 samples of the input sequence."},{"metadata":{"trusted":true},"cell_type":"code","source":"Image(\"/kaggle/input/task3-data/Images/Images/SignalSegmentation.PNG\")","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"Image(\"/kaggle/input/task3-data/Images/Images/SC1.PNG\")","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"Image(\"/kaggle/input/task3-data/Images/Images/SC2.PNG\")","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"Image(\"/kaggle/input/task3-data/Images/Images/SC3.PNG\")","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"### Add here"},{"metadata":{"trusted":true},"cell_type":"code","source":"###-------------------------------------------------------------\n### corrERSION OF SEQUENCE INTO A TIME DOMAIN RANDOM COMPLEX SIGNAL\n###-------------------------------------------------------------\nfrom string import ascii_uppercase\n\nseq_len0 = len(sequence1)\nseq_len1 = len(sequence2)\n    \nsignal0 = np.zeros(seq_len0, dtype=np.complex_)\nsignal1 = np.zeros(seq_len1, dtype=np.complex_)\n    \n    \nmax_seq_len=max(seq_len0,seq_len1);\n### 26 is total number of alphabets\nL = 26\n\n\nRANDOM_VEC = (1/np.sqrt(2))*((np.random.randn(1,L))+1j*(np.random.randn(1,L)))\n\nprob0 = np.zeros(L)\nprob1 = np.zeros(L)\n\nfor i in ascii_uppercase:\n    prob0[ord(i)-65] =  sequence1.count(i)\n    prob1[ord(i)-65] =  sequence2.count(i)\n  \n       \nfor i in range (seq_len0):\n    signal0[i] = RANDOM_VEC[0][ord(sequence1[i])-65]\n    #print('signal0[', i, ']=', signal0[i], '\\n')\n        \nfor i in range (seq_len1):\n    signal1[i] = RANDOM_VEC[0][ord(sequence2[i])-65]","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"#### TIME DOMAIN SIGNAL SEGMENTATION - FFT SIZE"},{"metadata":{"trusted":true},"cell_type":"code","source":"###------------------------------------------------\n###TIME DOMAIN SIGNAL SEGMENTATION - FFT SIZE\n###------------------------------------------------\nfft_size = 1024;\nimport math\nrows = math.ceil(max_seq_len/fft_size)\ncols = fft_size\n\n###print(\"seq_len=\", seq_len, \" rows=\", rows, \" columns=\", cols)\n\nseq_arr0=np.zeros((rows,cols), dtype=np.complex_)\nseq_arr1=np.zeros((rows,cols), dtype=np.complex_)\n\n\nfor row_pointer in range (rows):\n    start_index = fft_size*(row_pointer)\n    end_index = fft_size*(row_pointer+1)\n    if(end_index <= len(sequence1)):\n        seq_arr0[row_pointer,:] = signal0[start_index:end_index]\n    else:\n        if (start_index >= len(sequence1)):\n            ###fill all 0\n            seq_arr0[row_pointer,:] = np.zeros(1,fft_size);\n                \n        else:\n            ###copy till seq_len and then all 0s\n            ###print('row pointer:', row_pointer, ' start:',start_index, ' end:',end_index, ' seq_len:', seq_len0)\n            seq_arr0[row_pointer,0:(seq_len0-start_index)] = signal0[start_index:seq_len0]\n            seq_arr0[row_pointer,(seq_len0-start_index):] = np.zeros((1,end_index-seq_len0), dtype=np.complex_)\n                \nfor row_pointer in range (rows):\n    start_index = fft_size*(row_pointer)\n    end_index = fft_size*(row_pointer+1)\n    if(end_index <= len(sequence2)):\n        seq_arr1[row_pointer,:] = signal1[start_index:end_index]\n    else:\n        if (start_index >= len(sequence2)):\n            ###fill all 0\n            seq_arr1[row_pointer,:] = np.zeros(1,fft_size);\n                \n        else:\n            ###copy till seq_len and then all 0s\n            seq_arr1[row_pointer,0:(seq_len1-start_index)] = signal1[start_index:seq_len1]\n            seq_arr1[row_pointer,(seq_len1-start_index):] = np.zeros((1,end_index-seq_len1), dtype=np.complex_)\n","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"#### SIGNAL IN FREQ DOMAIN "},{"metadata":{"trusted":true},"cell_type":"code","source":"###------------------------------------------------\n### SIGNAL IN FREQ DOMAIN  -- CHECK\n###------------------------------------------------\n\nSEQ_ARR0=np.zeros((rows,cols), dtype=np.complex_)\nSEQ_ARR1=np.zeros((rows,cols), dtype=np.complex_)\nfor j in range (rows):\n    SEQ_ARR0[j,:]=np.fft.fft(seq_arr0[j,:])\n    SEQ_ARR1[j,:]=np.fft.fft(seq_arr1[j,:])\n","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"#### SIGNAL CORRELATION IN FREQ DOMAIN"},{"metadata":{"trusted":true},"cell_type":"code","source":"###------------------------------------------------\n###SIGNAL CORRELATION IN FREQ DOMAIN\n###------------------------------------------------\nnrof_segments=rows\n    \nCORR_CELL=np.zeros((nrof_segments,2,2,cols), dtype=np.complex_)\n    \nfor m in range (nrof_segments):\n    CORR_CELL[m, 0, 0, :]= np.multiply( (SEQ_ARR0[m,:]),(np.conj(SEQ_ARR0[m,:])) )\n    CORR_CELL[m, 0, 1, :]= np.multiply( (SEQ_ARR0[m,:]),(np.conj(SEQ_ARR1[m,:])) )\n    CORR_CELL[m, 1, 0, :]= np.multiply( (SEQ_ARR1[m,:]),(np.conj(SEQ_ARR0[m,:])) )\n    CORR_CELL[m, 1, 1, :]= np.multiply( (SEQ_ARR1[m,:]),(np.conj(SEQ_ARR1[m,:])) )","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"#### BACK TO TIME DOMAIN - correlation of piece wise time domain signal"},{"metadata":{"trusted":true},"cell_type":"code","source":"###------------------------------------------------\n### BACK TO TIME DOMAIN - correlation of piece wise time domain signal\n###------------------------------------------------\n\ncorr_cell=np.zeros((nrof_segments,2,2,cols), dtype=np.complex_)\ncoef0 = np.zeros(nrof_segments);\ncoef1 = np.zeros(nrof_segments);\nfor m in range (nrof_segments):\n    for i in range (2):\n        for j in range(2):\n            corr_cell[m, i, j, :]= np.fft.ifft(CORR_CELL[m,i,j])\n    coef0[m]=np.absolute(corr_cell[m, 0, 0, 0])\n    coef1[m]=np.absolute(corr_cell[m, 1, 0, 0])","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"#### Plot Data"},{"metadata":{"trusted":true},"cell_type":"code","source":"###------------------------------------------------\n###PLOT DATA\n###------------------------------------------------\nimport matplotlib.pyplot as plt\nmetric1_arr = np.divide(coef0,coef1)\nmetric1 = np.sum(np.absolute(metric1_arr) - 1)\nplt.rcParams['figure.figsize'] = [15,10]\nfig, axs = plt.subplots(2,1, constrained_layout=True)\n\nfig = plt.figure()\naxs[0].plot(range(nrof_segments), metric1_arr, 'cs')\naxs[0].set_title('metric1 : coef1/coef2 - 1 : '+str(metric1));\naxs[0].set_xlabel('number of segments');\naxs[0].set_ylabel('C1/C2-1');\naxs[0].grid()\nfig.suptitle('COVID19 GENOME Variation Plot', fontsize=16)\n    \nmetric2_arr = np.subtract(coef0,coef1)\nmetric2 = np.sum(np.absolute(metric2_arr))\naxs[1].plot(range(nrof_segments), metric2_arr, 'gs')\naxs[1].set_title('metric2 : coef1-coef2 : '+str(metric2));\naxs[1].set_xlabel('number of segments');\naxs[1].set_ylabel('C1-C2');\naxs[1].grid()\n","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"#### Reading Sequence Files"},{"metadata":{"trusted":true},"cell_type":"code","source":"files = []\n\n#currdir = os.getcwd()\n#path = currdir + '/' + 'gnome_data'\n\npath = '/kaggle/input/task3-data/gnome_data_countries/gnome_data_countries'\n# r=root, d=directories, f = files\nfor r, d, f in os.walk(path):\n    for file in f:\n        if '.txt' in file or '.fasta' in file:\n            files.append(os.path.join(r, file))\n\ng_fname = files    ","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"for f in g_fname:\n    print(f)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":""},{"metadata":{"trusted":true},"cell_type":"code","source":"out_file_path = os.getcwd()","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"\n## Section 2: Access to geographic and temporal diverse sample sets to understand geographic distribution and genomic differences"},{"metadata":{},"cell_type":"markdown","source":"#### Processing Genome files"},{"metadata":{"trusted":true},"cell_type":"code","source":"def run_corr_2files(num1, num2):\n    \n    \n    ###-------------------------------------------------------------\n    ### corrERSION OF SEQUENCE INTO A TIME DOMAIN RANDOM COMPLEX SIGNAL\n    ###-------------------------------------------------------------\n    \n    seq_len0 = len(sequence[num1])\n    seq_len1 = len(sequence[num2])\n    \n    \n    ###max_seq_len=max(seq_len0,seq_len1);\n    min_seq_len = min(seq_len0,seq_len1)\n    ### 26 is total number of alphabets\n    L = 26\n    \n    B=100\n    metric_samples=np.zeros(B)\n    for b in range(B):\n        \n        signal0 = np.zeros(seq_len0, dtype=np.complex_)\n        signal1 = np.zeros(seq_len1, dtype=np.complex_)\n    \n        RANDOM_VEC = (1/np.sqrt(2))*((np.random.randn(1,L))+1j*(np.random.randn(1,L)))\n        \n        prob0 = np.zeros(L)\n        prob1 = np.zeros(L)\n        \n        for i in ascii_uppercase:\n            prob0[ord(i)-65] =  sequence[num1].count(i)\n            prob1[ord(i)-65] =  sequence[num2].count(i)\n         \n          \n        for i in range (seq_len0):\n            signal0[i] = RANDOM_VEC[0][ord(sequence[num1][i])-65]\n            #print('signal0[', i, ']=', signal0[i], '\\n')\n            \n            \n        for i in range (seq_len1):\n            signal1[i] = RANDOM_VEC[0][ord(sequence[num2][i])-65]\n        \n        ###------------------------------------------------\n        ###TIME DOMAIN SIGNAL SEGMENTATION - FFT SIZE\n        ###------------------------------------------------\n        fft_size = 1024\n    \n        rows = math.ceil(min_seq_len/fft_size)\n        cols = fft_size\n        ###print(\"seq_len=\", seq_len, \" rows=\", rows, \" columns=\", cols)\n        \n        seq_arr0=np.zeros((rows,cols), dtype=np.complex_)\n        seq_arr1=np.zeros((rows,cols), dtype=np.complex_)\n        \n        for row_pointer in range (rows):\n            start_index = fft_size*(row_pointer)\n            end_index = fft_size*(row_pointer+1)\n            if(end_index <= len(sequence[num1])):\n                seq_arr0[row_pointer,:] = signal0[start_index:end_index]\n            else:\n                if (start_index >= len(sequence[num1])):\n                    ###fill all 0\n                    seq_arr0[row_pointer,:] = np.zeros((1,fft_size), dtype=np.complex_);\n                    \n                else:\n                    ###copy till seq_len and then all 0s\n                    ###print('row pointer:', row_pointer, ' start:',start_index, ' end:',end_index, ' seq_len:', seq_len0)\n                    seq_arr0[row_pointer,0:(seq_len0-start_index)] = signal0[start_index:seq_len0]\n                    seq_arr0[row_pointer,(seq_len0-start_index):] = np.zeros((1,end_index-seq_len0), dtype=np.complex_)\n                    \n        for row_pointer in range (rows):\n            start_index = fft_size*(row_pointer)\n            end_index = fft_size*(row_pointer+1)\n            if(end_index <= len(sequence[num2])):\n                seq_arr1[row_pointer,:] = signal1[start_index:end_index]\n            else:\n                if (start_index >= len(sequence[num2])):\n                    ###fill all 0\n                    seq_arr1[row_pointer,:] = np.zeros((1,fft_size), dtype=np.complex_);\n                    \n                else:\n                    ###copy till seq_len and then all 0s\n                    seq_arr1[row_pointer,0:(seq_len1-start_index)] = signal1[start_index:seq_len1]\n                    seq_arr1[row_pointer,(seq_len1-start_index):] = np.zeros((1,end_index-seq_len1), dtype=np.complex_)\n                              \n    \n        ###------------------------------------------------\n        ###SIGNAL IN FREQ DOMAIN  -- CHECK\n        ###------------------------------------------------\n        \n        SEQ_ARR0=np.zeros((rows,cols), dtype=np.complex_)\n        SEQ_ARR1=np.zeros((rows,cols), dtype=np.complex_)\n        for j in range (rows):\n            SEQ_ARR0[j,:]=np.fft.fft(seq_arr0[j,:])\n            SEQ_ARR1[j,:]=np.fft.fft(seq_arr1[j,:])\n        \n        ###------------------------------------------------\n        ###SIGNAL CORRELATION IN FREQ DOMAIN\n        ###------------------------------------------------\n        nrof_segments=rows\n        \n        CORR_CELL=np.zeros((nrof_segments,1,2,cols), dtype=np.complex_)\n        \n        for m in range (nrof_segments):\n            CORR_CELL[m, 0, 0, :]= np.multiply( (SEQ_ARR0[m,:]),(np.conj(SEQ_ARR0[m,:])) )\n            CORR_CELL[m, 0, 1, :]= np.multiply( (SEQ_ARR0[m,:]),(np.conj(SEQ_ARR1[m,:])) )\n                   \n        \n        ###------------------------------------------------\n        ###BACK TO TIME DOMAIN - correlation of piece wise time domain signal\n        ###------------------------------------------------\n        \n        corr_cell=np.zeros((nrof_segments,1,2,cols), dtype=np.complex_)\n        coef0 = np.zeros(nrof_segments);\n        coef1 = np.zeros(nrof_segments);\n        for m in range (nrof_segments):\n            corr_cell[m, 0, 0, :]= np.fft.ifft(CORR_CELL[m,0,0])\n            corr_cell[m, 0, 1, :]= np.fft.ifft(CORR_CELL[m,0,1])\n            coef0[m]=max(np.absolute(corr_cell[m, 0, 0]))\n            coef1[m]=max(np.absolute(corr_cell[m, 0, 1]))\n            \n        coef = np.divide(coef0,coef1) - 1\n        \n        metric_samples[b] = np.sum(np.absolute(coef))\n        \n              \n    np.sort(metric_samples)\n    #print('sorted: ', metric_samples,'\\n')\n    return(metric_samples[89])","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"#print(g_fname)\n    \nsequence = [] \ncheck_var1 = 0\n\n\ni = 0\n\nfor j in range (len(g_fname)):\n    A=''\n    with open(g_fname[j]) as fhandler:\n        first_line = fhandler.readline()\n        if (first_line.find('genome') != -1):\n            ###print(first_line)\n            print(\"Discarding first line of:\", g_fname[j],'\\n')\n            ##messagebox.showwarning('Warning', 'Discarding first line: '+first_line)\n        else :\n            A = first_line\n\n        for line in fhandler:\n            A += line   \n\n        A = A.replace('\\n','') \n\n        sequence.append([])\n        #####-----------------------------\n        #####DNA to Amino\n        #####-----------------------------          \n        if(check_var1 == 1):\n            #####-----------------------------\n            #####DNA to RNA\n            #####-----------------------------\n            A = A.replace('T','U') \n\n            ####------------------------------\n            #### RNA to amino Acid\n            ####------------------------------                \n            key = ''\n            for k in range(0,3*math.floor(len(A)/3),3):\n                key = A[k]+A[k+1]+A[k+2]\n                ##print(\"here:\",k, A[k],A[k+1],A[k+2],'\\n')\n                if(rna_to_amino.get(key) == None):\n                    print('Data Corrupted for file...abanding:\\n', g_fname[j])\n                    messagebox.showerror('Error', 'Data corrupted for file: '+g_fname[j]+' => expected only A,C,G and T in the DNA sequence')\n\n\n                if(rna_to_amino[key] != '*'):\n                    sequence[i].append(rna_to_amino[key])\n        else :\n            sequence[i] = list(A)\n\n    i = i+1","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"#### Generating Genome Outputs"},{"metadata":{"trusted":true},"cell_type":"code","source":"global out_file_path\nnrof_samples = len(g_fname)\nnrof_calc_metrics = int((nrof_samples*(nrof_samples-1))/2)\n\nmetric_arr = np.zeros(nrof_calc_metrics)\nsample_loc_arr = np.chararray((nrof_calc_metrics,1,2), itemsize=100, unicode=True)\nsample_number_arr = np.zeros((nrof_calc_metrics,3))\nrecord_arr = np.chararray((nrof_calc_metrics), itemsize=100, unicode=True)\nloc_arr = np.chararray((nrof_calc_metrics), itemsize=100, unicode=True)\n\nk=0\nfor i in range(nrof_samples):\n    for j in range((i+1),nrof_samples):\n        if(len(sequence[i]) > len(sequence[j])):\n            metric_arr[k] = run_corr_2files(i, j)\n        else :\n            metric_arr[k] = run_corr_2files(j, i)\n\n        sample_loc_arr[k,0,0] = ntpath.basename(g_fname[i]).replace('.fasta','')\n        sample_loc_arr[k,0,1] = ntpath.basename(g_fname[j]).replace('.fasta','')\n        record_arr[k] = sample_loc_arr[k,0,0]+', '+sample_loc_arr[k,0,1]+' = '+str(metric_arr[k]) \n        loc_arr[k] = sample_loc_arr[k,0,0]+', '+sample_loc_arr[k,0,1]\n        print(record_arr[k]+'\\n')\n        sample_number_arr[k,0] = i;\n        sample_number_arr[k,1] = j;\n        sample_number_arr[k,2] = metric_arr[k];\n        k = k+1\n\n#-----------\n#OUtput file\n#-----------\nsort_metric_arr = np.sort(metric_arr, axis=0)\nindex = np.argsort(metric_arr, axis=0)\nsort_record_arr = []\nsort_loc_arr = []\nfor i in index:\n    sort_record_arr.append(record_arr[i])\n    sort_loc_arr.append(loc_arr[i])\n\n","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"#remove previous created output folder\n#!rmdir 'Outpur'\n!mkdir 'Output'","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"filename_output= 'Output/output_metrics1.csv'\nfilename_output_raw= 'Output/output_raw_data1.csv'\nfilename_output_loc= 'Output/output_loc1.txt'\nfilename_output_sample_number= 'Output//sample_number1.txt'\n\nfileID = open(filename_output,'w');\nfileID2 = open(filename_output_raw,'w');\nfileID3 = open(filename_output_loc,'w');\nfileID4 = open(filename_output_sample_number,'w');\nfor k in range(nrof_calc_metrics):    \n    fileID.write(str(sort_metric_arr[k])+'\\n')  \n    fileID2.write(sort_record_arr[k]+'\\n')   \n    fileID3.write(sort_loc_arr[k]+'\\n')\n    fileID4.write(str(sample_number_arr[k,:])+'\\n')\n\nfileID.close()\nfileID2.close()\nfileID3.close()\nfileID4.close()","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"filename_output","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"filename_output_raw","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"filename_output_loc","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"filename_output_sample_number","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"#### Output Description\n\n-------------------------------------------------------------------------------------\n\n\n**output_metrics.txt:**\n\n(M,1) vector,   where M = N* (N-1)/2   # of combinations of the input sequences and N is the number of input sequences N.\n\n[metric ]\n[metric ]\n\n\n\n**output_raw_data.csv:**   \n\n(M,3) vector\n=\n[seq1, seq2,metric]\n[seq1, seq2,metric]\n…\n\n\n**output_loc.txt:**\n\n(M,3) vector \nseq1,seq2]\n[seq1,seq2]\n..\n\n\n\n**sample_number.txt:**\n\n\n (M,3) vector, with elements:  (combination index, ordinal number for seq1, ordinal number for seq2)\n=\n[1 1 2]\n[2 1 2]\n[3 1 3]\n…\n\n[m 2 3]\n[m 2 4]\n….\n\n…\n\n[n  N-1, N]\n\n\n--------------------------------------------------------------------------\n\n"},{"metadata":{"trusted":true},"cell_type":"code","source":"files = []\n\npath = 'Output/'\n\nfor r, d, f in os.walk(out_file_path):\n    for file in f:\n        print(file)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"### Method Applied to samples form same location in diferent times."},{"metadata":{},"cell_type":"markdown","source":"#### Read files"},{"metadata":{"trusted":true},"cell_type":"code","source":"files = []\n\n#currdir = os.getcwd()\n#path = currdir + '/' + 'gnome_data'\n\npath = './' + 'gnome_data_time'\n# r=root, d=directories, f = files\nfor r, d, f in os.walk(path):\n    for file in f:\n        if '.txt' in file or '.fasta' in file:\n            files.append(os.path.join(r, file))\n\ng_fname = files    ","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"for f in g_fname:\n    print(f)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"# Output File Location\nout_file_path = 'Output/'","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"#### Process files"},{"metadata":{"trusted":true},"cell_type":"code","source":"def run_corr_2files(num1, num2):\n    \n    \n    ###-------------------------------------------------------------\n    ### corrERSION OF SEQUENCE INTO A TIME DOMAIN RANDOM COMPLEX SIGNAL\n    ###-------------------------------------------------------------\n    \n    seq_len0 = len(sequence[num1])\n    seq_len1 = len(sequence[num2])\n    \n    \n    ###max_seq_len=max(seq_len0,seq_len1);\n    min_seq_len = min(seq_len0,seq_len1)\n    ### 26 is total number of alphabets\n    L = 26\n    \n    B=100\n    metric_samples=np.zeros(B)\n    for b in range(B):\n        \n        signal0 = np.zeros(seq_len0, dtype=np.complex_)\n        signal1 = np.zeros(seq_len1, dtype=np.complex_)\n    \n        RANDOM_VEC = (1/np.sqrt(2))*((np.random.randn(1,L))+1j*(np.random.randn(1,L)))\n        \n        prob0 = np.zeros(L)\n        prob1 = np.zeros(L)\n        \n        for i in ascii_uppercase:\n            prob0[ord(i)-65] =  sequence[num1].count(i)\n            prob1[ord(i)-65] =  sequence[num2].count(i)\n         \n          \n        for i in range (seq_len0):\n            signal0[i] = RANDOM_VEC[0][ord(sequence[num1][i])-65]\n            #print('signal0[', i, ']=', signal0[i], '\\n')\n            \n            \n        for i in range (seq_len1):\n            signal1[i] = RANDOM_VEC[0][ord(sequence[num2][i])-65]\n        \n        ###------------------------------------------------\n        ###TIME DOMAIN SIGNAL SEGMENTATION - FFT SIZE\n        ###------------------------------------------------\n        fft_size = 1024\n    \n        rows = math.ceil(min_seq_len/fft_size)\n        cols = fft_size\n        ###print(\"seq_len=\", seq_len, \" rows=\", rows, \" columns=\", cols)\n        \n        seq_arr0=np.zeros((rows,cols), dtype=np.complex_)\n        seq_arr1=np.zeros((rows,cols), dtype=np.complex_)\n        \n        for row_pointer in range (rows):\n            start_index = fft_size*(row_pointer)\n            end_index = fft_size*(row_pointer+1)\n            if(end_index <= len(sequence[num1])):\n                seq_arr0[row_pointer,:] = signal0[start_index:end_index]\n            else:\n                if (start_index >= len(sequence[num1])):\n                    ###fill all 0\n                    seq_arr0[row_pointer,:] = np.zeros((1,fft_size), dtype=np.complex_);\n                    \n                else:\n                    ###copy till seq_len and then all 0s\n                    ###print('row pointer:', row_pointer, ' start:',start_index, ' end:',end_index, ' seq_len:', seq_len0)\n                    seq_arr0[row_pointer,0:(seq_len0-start_index)] = signal0[start_index:seq_len0]\n                    seq_arr0[row_pointer,(seq_len0-start_index):] = np.zeros((1,end_index-seq_len0), dtype=np.complex_)\n                    \n        for row_pointer in range (rows):\n            start_index = fft_size*(row_pointer)\n            end_index = fft_size*(row_pointer+1)\n            if(end_index <= len(sequence[num2])):\n                seq_arr1[row_pointer,:] = signal1[start_index:end_index]\n            else:\n                if (start_index >= len(sequence[num2])):\n                    ###fill all 0\n                    seq_arr1[row_pointer,:] = np.zeros((1,fft_size), dtype=np.complex_);\n                    \n                else:\n                    ###copy till seq_len and then all 0s\n                    seq_arr1[row_pointer,0:(seq_len1-start_index)] = signal1[start_index:seq_len1]\n                    seq_arr1[row_pointer,(seq_len1-start_index):] = np.zeros((1,end_index-seq_len1), dtype=np.complex_)\n                              \n    \n        ###------------------------------------------------\n        ###SIGNAL IN FREQ DOMAIN  -- CHECK\n        ###------------------------------------------------\n        \n        SEQ_ARR0=np.zeros((rows,cols), dtype=np.complex_)\n        SEQ_ARR1=np.zeros((rows,cols), dtype=np.complex_)\n        for j in range (rows):\n            SEQ_ARR0[j,:]=np.fft.fft(seq_arr0[j,:])\n            SEQ_ARR1[j,:]=np.fft.fft(seq_arr1[j,:])\n        \n        ###------------------------------------------------\n        ###SIGNAL CORRELATION IN FREQ DOMAIN\n        ###------------------------------------------------\n        nrof_segments=rows\n        \n        CORR_CELL=np.zeros((nrof_segments,1,2,cols), dtype=np.complex_)\n        \n        for m in range (nrof_segments):\n            CORR_CELL[m, 0, 0, :]= np.multiply( (SEQ_ARR0[m,:]),(np.conj(SEQ_ARR0[m,:])) )\n            CORR_CELL[m, 0, 1, :]= np.multiply( (SEQ_ARR0[m,:]),(np.conj(SEQ_ARR1[m,:])) )\n                   \n        \n        ###------------------------------------------------\n        ###BACK TO TIME DOMAIN - correlation of piece wise time domain signal\n        ###------------------------------------------------\n        \n        corr_cell=np.zeros((nrof_segments,1,2,cols), dtype=np.complex_)\n        coef0 = np.zeros(nrof_segments);\n        coef1 = np.zeros(nrof_segments);\n        for m in range (nrof_segments):\n            corr_cell[m, 0, 0, :]= np.fft.ifft(CORR_CELL[m,0,0])\n            corr_cell[m, 0, 1, :]= np.fft.ifft(CORR_CELL[m,0,1])\n            coef0[m]=max(np.absolute(corr_cell[m, 0, 0]))\n            coef1[m]=max(np.absolute(corr_cell[m, 0, 1]))\n            \n        coef = np.divide(coef0,coef1) - 1\n        \n        metric_samples[b] = np.sum(np.absolute(coef))\n        \n              \n    np.sort(metric_samples)\n    #print('sorted: ', metric_samples,'\\n')\n    return(metric_samples[89])","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"sequence = [] \ncheck_var1 = 0\ni = 0\n\nfor j in range (len(g_fname)):\n    A=''\n    with open(g_fname[j]) as fhandler:\n        first_line = fhandler.readline()\n        if (first_line.find('genome') != -1):\n            ###print(first_line)\n            print(\"Discarding first line of:\", g_fname[j],'\\n')\n            messagebox.showwarning('Warning', 'Discarding first line: '+first_line)\n        else :\n            A = first_line\n\n        for line in fhandler:\n            A += line   \n\n        A = A.replace('\\n','') \n\n        if ((A.count('A')+A.count('T')+A.count('C')+A.count('G')) != len(A)):\n            print ('Error :Data corrupted for file: '+g_fname[j]+' => expected only A,C,G and T in the DNA sequence')\n            ##root.destroy()\n            ##return()\n\n        sequence.append([])\n        #####-----------------------------\n        #####DNA to Amino\n        #####-----------------------------          \n        if(check_var1 == 1):\n            #####-----------------------------\n            #####DNA to RNA\n            #####-----------------------------\n            A = A.replace('T','U') \n\n            ####------------------------------\n            #### RNA to amino Acid\n            ####------------------------------                \n            key = ''\n            for k in range(0,3*math.floor(len(A)/3),3):\n                key = A[k]+A[k+1]+A[k+2]\n                ##print(\"here:\",k, A[k],A[k+1],A[k+2],'\\n')\n                if(rna_to_amino.get(key) == None):\n                    print('Data Corrupted for file...abanding:\\n', g_fname[j])\n                    #messagebox.showerror('Error', 'Data corrupted for file: '+g_fname[j]+' => expected only A,C,G and T in the DNA sequence')\n                    ##root.destroy()\n                    ##return()\n\n                if(rna_to_amino[key] != '*'):\n                    sequence[i].append(rna_to_amino[key])\n        else :\n            sequence[i] = list(A)\n\n    i = i+1","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"#### Generating Output files"},{"metadata":{"trusted":true},"cell_type":"code","source":"global out_file_path\nnrof_samples = len(g_fname)\nnrof_calc_metrics = int((nrof_samples*(nrof_samples-1))/2)\n\nmetric_arr = np.zeros(nrof_calc_metrics)\nsample_loc_arr = np.chararray((nrof_calc_metrics,1,2), itemsize=100, unicode=True)\nsample_number_arr = np.zeros((nrof_calc_metrics,3))\nrecord_arr = np.chararray((nrof_calc_metrics), itemsize=100, unicode=True)\nloc_arr = np.chararray((nrof_calc_metrics), itemsize=100, unicode=True)\n\nk=0\nfor i in range(nrof_samples):\n    for j in range((i+1),nrof_samples):\n        if(len(sequence[i]) > len(sequence[j])):\n            metric_arr[k] = run_corr_2files(i, j)\n        else :\n            metric_arr[k] = run_corr_2files(j, i)\n\n        sample_loc_arr[k,0,0] = ntpath.basename(g_fname[i]).replace('.fasta','')\n        sample_loc_arr[k,0,1] = ntpath.basename(g_fname[j]).replace('.fasta','')\n        record_arr[k] = sample_loc_arr[k,0,0]+', '+sample_loc_arr[k,0,1]+' = '+str(metric_arr[k]) \n        loc_arr[k] = sample_loc_arr[k,0,0]+', '+sample_loc_arr[k,0,1]\n        print(record_arr[k]+'\\n')\n        sample_number_arr[k,0] = i;\n        sample_number_arr[k,1] = j;\n        sample_number_arr[k,2] = metric_arr[k];\n        k = k+1\n\n#-----------\n#OUtput file\n#-----------\nsort_metric_arr = np.sort(metric_arr, axis=0)\nindex = np.argsort(metric_arr, axis=0)\nsort_record_arr = []\nsort_loc_arr = []\nfor i in index:\n    sort_record_arr.append(record_arr[i])\n    sort_loc_arr.append(loc_arr[i])\n\nfilename_output= out_file_path+'/output_metrics1.csv'\nfilename_output_raw= out_file_path+'/output_raw_data1.csv'\nfilename_output_loc= out_file_path+'/output_loc1.txt'\nfilename_output_sample_number= out_file_path+'/sample_number1.txt'\n","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"filename_output","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"fileID = open(filename_output,'w');\nfileID2 = open(filename_output_raw,'w');\nfileID3 = open(filename_output_loc,'w');\nfileID4 = open(filename_output_sample_number,'w');\nfor k in range(nrof_calc_metrics):    \n    fileID.write(str(sort_metric_arr[k])+'\\n')  \n    fileID2.write(sort_record_arr[k]+'\\n')   \n    fileID3.write(sort_loc_arr[k]+'\\n')\n    fileID4.write(str(sample_number_arr[k,:])+'\\n')\n\nfileID.close()\nfileID2.close()\nfileID3.close()\nfileID4.close()","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"### Ouput files Description\n\n-------------------------------------------------------------------------------------\n\n\n**output_metrics.txt:**\n\n(M,1) vector,   where M = N* (N-1)/2   # of combinations of the input sequences and N is the number of input sequences N.\n\n[metric ]\n[metric ]\n\n\n\n**output_raw_data.csv:**   \n\n(M,3) vector\n=\n[seq1, seq2,metric]\n[seq1, seq2,metric]\n…\n\n\n**output_loc.txt:**\n\n(M,3) vector \nseq1,seq2]\n[seq1,seq2]\n..\n\n\n\n**sample_number.txt:**\n\n\n (M,3) vector, with elements:  (combination index, ordinal number for seq1, ordinal number for seq2)\n=\n[1 1 2]\n[2 1 2]\n[3 1 3]\n…\n\n[m 2 3]\n[m 2 4]\n….\n\n…\n\n[n  N-1, N]"},{"metadata":{"trusted":true},"cell_type":"code","source":"Image(\"/kaggle/input/task3-data/Images/Images/Res1.PNG\")","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"Image(\"/kaggle/input/task3-data/Images/Images/Res2.PNG\")","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"Image(\"/kaggle/input/task3-data/Images/Images/Res3.PNG\")","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"Image(\"/kaggle/input/task3-data/Images/Images/Res4.PNG\")","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"Image(\"/kaggle/input/task3-data/Images/Images/Res5.PNG\")","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"# Section 1 : Summary\n\n### Section 1: Real-time tracking of whole genomes to track variations of the virus over time\n\n\nSummary:\n•\tTo accomplish this task we follow the steps below: \no\tRetrieve the genome sequences from publicly available realtime databases such as China National GeneBank & GISAID EpiCoV™ Database\no\tUse same algorithm from Task 3.2 to classify whether the new samples or sequences belong to exsiting or new groups \no\tDetermine whether or not if new sequences are changing over time\n•\tThis task tracks temporal variations \n\n•\tBootstrap confidence interval of the measured metric\n\n•\tHierarchical classification of the sample sequence\nThis subtask relates to the evolution of the genome sequence in time.  To study these variations, we use the same approach as described in section 3.2.  The only two differences are as follows.  \n\n1.\tFirst, the input sequences must be specified for an individual geographical location, to remove the spatial dependency.  Also, these sequences must be specified for a given period of time and sorted by date in ascending order.  \n2.\tThe algorithm of section 3.2 is then applied but is modified so that it only analyzes the consecutive pairs of sequences, e.g. 1st & 2nd sequences; 2nd & 3rd; 3rd & 4th, …, (Nth-1) & Nth, where N is the number of input sequences.\n\nThe results below show the genome evolution for China, from end of December 2019 to beginning of February 2020. Figure 1 shows the ordinate the measured metric (that estimates the difference between two consecutive sequences) and in the abscissa the time stamp of the sequence.   \nFigure 1 shows some similar variation for the first three sequences, which were each taken about 10 days apart.  However, there is a larger variation for sequence 4th, with respect to 3rd and 5th, this could indicate that sequence 4th corresponds to a different strain of the virus within the same location of China. Finally, there is almost no variation in the last two sequences.\n\n\n### Section 2: Access to geographic and temporal diverse sample sets to understand geographic distribution and genomic differences\n\nSummary:\n•\tGenome sequence conversion (obtained from publicly available realtime databases such as China National GeneBank & GISAID EpiCoV™ Database) to a random signal\n•\tSignal segmentation \n•\tSignal correlation \n•\tResults\n•\tThis task tracks geographical variations\n\nTask 3.2: Genome Sequence Conversion\nIn this step, the input genome sequence is converted to a random signal.  This is useful, since the new waveform is much easier to visualize and manipulate than letter patterns.  Furthermore, this allows us to use signal correlation in order to find how different or similar the two signals are.     \n\nThe Figure below shows the signal conversion for the two input sequences that will be analyzed. Each of the characters of the sequence is translated to a complex random number where the same random number is assigned to every character occurrence.\n\nTask 3.2: Signal Segmentation \nSince the input sequences are about 30,000 samples long, the signal is divided in equal length segments.  The length of the segment selected in our study is 1024 but it can be changed to any given value.  Figure below shows how the signal is segmented.  In this example, we get 30 segments of size 1024 for the ~30,000 samples of the input sequence.\n\nTask 3.2: Signal Correlation – Part 1\n\nIn order to perform this step, we do the following. \nLet s_1,1,s_1,2,s_1,3,…..,s_(1,n) be the n  segments of signal s_1, the signal corresponding to the first sequence and s_2,1,s_2,2,s_2,3,…..,s_(2,n) the n  segments of signal s_2 be the signal corresponding to the second sequence, then for each segment:\n\n\tCalculate the FFT of the signal to obtain the frequency signal (see Figure)\n\tS_1,1,S_1,2,S_1,3,…..,S_(1,n) for signal s_1\n\tS_2,1,S_2,2,S_2,3,…..,S_(2,n) for signal s_2\n\tLet S_1,1,S_1,2,S_1,3,…..,S_(1,n) be the n segments of signal S_1\n\tLet S_2,1,S_2,2,S_2,3,…..,S_(2,n) be the n segments of signal S_2\n\tCalculate the following vectors as shown in Figure 5\n\tR_11,1=S_1,1 〖 S〗_1,1^*  ,R_11,2=S_1,2 〖 S〗_1,2^*  ,…..R_(11,n)=S_(1,n) 〖 S〗_(1,n)^*     \n\tR_12,1=S_1,1 〖 S〗_2,1^*  ,R_12,2=S_1,2 〖 S〗_2,2^*  ,…..R_(12,n)=S_(1,n) 〖 S〗_(2,n)^*     \n\tPerform IFFT to obtain the autocorrelation and cross correlation vectors rxx and rxy (see Figure) and set rxx,ryy as the ith segment entry of the diagonal of matrix r (below figure)\n\tThen we calculate the correlation vectors as follows (see Figure 6):\n\t r_11,1=〖IFFT(R〗_11,1),r_11,2=〖IFFT(R〗_11,2),…,r_(11,n)=〖IFFT(R〗_(11,n))   \n\tr_21,1=〖IFFT(R〗_21,1),r_21,2=〖IFFT(R〗_21,2),…,r_(21,n)=〖IFFT(R〗_(21,n))   \n\nTask 3.2: Correlation Coefficients and Metrics\n\nWe obtain c_1,〖  c〗_2, which are given by the sum of the peak values of the correlation vectors over all the segments i=(1,…,n) for each sequence.  This is shown below\nc_1=∑_(i=0)^n▒〖〖max⁡(r〗_(11,i))〗,    c_2=∑_(i=0)^n▒〖〖max⁡(r〗_(12,i))〗\nRefer Figure 7.\n\nFinally, we propose the following metric to measure the differences between a pair of sequences\nm_12= (c_1/c_2 -1)\n\nFurthermore, we calculate the α=90% confidence interval of  m_12 using bootstrap and if this value is greater than a given threshold, then the sequences have significant variations\n  〖(m_12)〗_α>  T\n\nRefer Figure 8.\n\nFigure 9 illustrates the input sequences on the left and the associated sequence pair to perform the comparison.  The number of calculated metrics values is p=q(q-1)/2   Which corresponds to all the possible combinations of the input sequences.\n\nTask 3.2: Algorithm 2: Genome Strains Classification\nFigure 10 shows different Thresholds values T_1,T_2, T_3.\nWhich can be adjusted to see different degrees of variations.  Also, the number of thresholds needed could be optimized so that only significant variations are detected. In our case, we have selected three levels.  Finally Figure 10, shows the tree structure where each sequence is ordered according to their confidence interval metric estimates and Threshold levels.  The data is ordered in time from oldest to the top of the tree to the most recent to the bottom of the tree.  \nThe largest threshold level captures the most significant differences between a pair of sequences, the medium threshold level captures medium differences and the lower threshold level captures small differences between a pair of sequences.\n\n\n### Pros and cons of the approach \n\nTask 3.1\nThis section describes the pros and cons for Task 3.1.\n\nPros\n•\tThe results are very easy to interpret since it shows the difference between consecutive samples.  The measured data represents the time evolution for the specified geographical location. \n\nCons\n•\tThe input sequences need to be collected manually from the publicly available databases. It would be useful to have a tool that can do this automatically, by specifying the geographical location and dates one is interested to analyze.\n\n4.2\tTask 3.2\nThis section describes the pros and cons for Task 3.2.\n\nPros \n•\tThe method is flexible since it allows one to specify the branch tree depths, it does not need alignment in the input sequences, and it can be easily adapted to measure different features of the sequence.  We have analyzed samples from seven countries, and we have been able to identify different variations and their evolution paths.\n\nCons \n•\tThe variations that this method has identified may not coincide with the definition of a strain but can be adapted to satisfy the strain criteria. \n\n•\tThe current solution does not implement the sorting algorithm in the tree structure of the input data sequences.  This is now done manually.  For large number of input sequences this task is complex.  \n\n•\tA post-processing tool that classifies this data will make visualization and interpretation of results more efficient.\n\n•\tThe results of our study are very limited since the number of samples used was rather small.  A much larger set of samples is needed to get more meaningful results."},{"metadata":{},"cell_type":"markdown","source":"## Section 3 : Evidence that livestock could be infected\n#### 3.1\nEvidence that livestock could be infected (e.g., field surveillance, genetic sequencing, receptor binding) and serve as a reservoir after the epidemic appears to be over: Evidence of whether farmers are infected, and whether farmers could have played a role in the origin.\n\n#### 3.2 \nEvidence that livestock could be infected (e.g., field surveillance, genetic sequencing, receptor binding) and serve as a reservoir after the epidemic appears to be over.: Surveillance of mixed wildlife- livestock farms for SARS-CoV-2 and other coronaviruses in Southeast Asia.\n\n#### 3.3 \nEvidence that livestock could be infected (e.g., field surveillance, genetic sequencing, receptor binding) and serve as a reservoir after the epidemic appears to be over.: Experimental infections to test host range for this pathogen."},{"metadata":{},"cell_type":"markdown","source":"### Visualization Functions"},{"metadata":{"trusted":true},"cell_type":"code","source":"def wordcloud_draw(text, color = 'white'):\n    \"\"\"\n    Plots wordcloud of string text after removing stopwords\n    \"\"\"\n    cleaned_word = \" \".join([word for word in text.split()])\n    wordcloud = WordCloud(stopwords=STOPWORDS,\n                      background_color=color,\n                      width=1000,\n                      height=1000\n                     ).generate(cleaned_word)\n    plt.figure(1,figsize=(15, 15))\n    plt.imshow(wordcloud)\n    plt.axis('off')\n    display(plt.show())","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"##### Load animal List"},{"metadata":{"trusted":true},"cell_type":"code","source":"animals = []\nanimalList = []\nwith open('/kaggle/input/task3-data/animals.txt', \"r\") as f:\n    animals = f.readlines()\nanimalList = [s.replace('\\n', '') for s in animals]\nanimalList.append('pangolin')\nanimalList.append('mice')\nanimalList.append('animal')\nanimalList = [string for string in animalList if string != \"\"]\nanimalList = list(map(lambda x:x.lower(), animalList))\nanimalList.remove('human')\nanimalList.remove('discus')\npluralList = ['{0}s'.format(elem) for elem in animalList]\nanimalList = animalList + pluralList\n#animalList = [' {0} '.format(elem) for elem in animalList]\nanimalList[0:5]","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"covid19_list = []\nwith open('/kaggle/input/task3-data/covid19.txt', \"r\") as f:\n    words = f.readlines()\ncovid19_list = [s.replace('\\n', '') for s in words]","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"covid19_list[0:5]","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"### Explore Articles"},{"metadata":{"trusted":true},"cell_type":"code","source":"#Load Meta data\nmetadata = pd.read_csv(\"/kaggle/input/CORD-19-research-challenge/metadata.csv\")\nmetadata.head()","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"#### Merging Articles with Meta data. Loading data from already merged csv\n\nLeft_join = pd.merge(fin_df,  \n                     metadata,  \n                     on ='sha',  \n                     how ='left') \nLeft_join = Left_join.drop(columns=['cord_uid', 'doi', 'title_y', 'pmcid', 'abstract_y', 'Microsoft Academic Paper ID', 'WHO #Covidence',\n                       'full_text_file', 'url', 'pubmed_id'])\nLeft_join.head()","execution_count":null,"outputs":[]},{"metadata":{"id":"Tt4Mz-dZ1I41","trusted":true},"cell_type":"code","source":"Left_join['title_x'].fillna(\"NoTitle\", inplace = True)\nLeft_join['abstract_x'].fillna(\"NoAbstract\", inplace = True)\nLeft_join['full_text'].fillna(\"NoText\", inplace = True)\nLeft_join['combined'] = Left_join['title_x'] + ' ' + Left_join['abstract_x'] + ' ' + Left_join['full_text']\n","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"Left_join.head()","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"#### Explore Livestock articles"},{"metadata":{"trusted":true},"cell_type":"code","source":"cond2 = Left_join['abstract_x'].str.contains('livestock')\nprint(sum(cond2))","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"abstract_livestock = Left_join[cond2]\nabstract_livestock.shape\nabstract_livestock.head(5)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"#### Explore Livestock Articles related to farmer\n"},{"metadata":{"trusted":true},"cell_type":"code","source":"cond3 = abstract_livestock['abstract_x'].str.contains('farmer')\nabstract_livestock_farmer = abstract_livestock[cond3]\nabstract_livestock_farmer.head(5)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"## Section 4:  What do we know about virus genetics, origin, and evolution? What do we know about the virus origin and management measures at the human-animal interface?\n\n###\tTask 4.1- Animal host(s) and any evidence of continued spill-over to humans\n###\tTask 4.2- Sustainable risk reduction strategies\n###\tTask 4.3- Socioeconomic and behavioral risk factors for this spill-over\n\n"},{"metadata":{"id":"gp5pf1jF1I5A","trusted":true},"cell_type":"code","source":"def clean_text(article):\n    clean1 = re.sub(r'['+string.punctuation + '’—”'+']', \"\", article.lower())\n    return re.sub(r'\\W+', ' ', clean1)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"### Tokenize combined data"},{"metadata":{},"cell_type":"markdown","source":"The following cells takes time to execute. After execution, the file is saved as 'animal_articles.csv'. This file is loaded in again. Uncomment the cell if need to process again.\n\n"},{"metadata":{"id":"hspJOhs41I5G","trusted":true},"cell_type":"code","source":"import re\n\n#Left_join['tokenized'] = Left_join['combined'].map(lambda x: clean_text(x))","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"\nimport nltk\n'''\nw_tokenizer = nltk.tokenize.WhitespaceTokenizer()\nlemmatizer = nltk.stem.WordNetLemmatizer()\n\ndef lemmatize_text(text):\n    return [lemmatizer.lemmatize(w) for w in w_tokenizer.tokenize(text)]\nLeft_join['tokenized'] = Left_join['tokenized'].apply(lemmatize_text)\n'''","execution_count":null,"outputs":[]},{"metadata":{"id":"UlmPV0LxMm1y","outputId":"f35fb880-badf-4201-ff97-25797c5558f9","scrolled":true,"trusted":true},"cell_type":"markdown","source":"## Let's free up some space"},{"metadata":{"trusted":true},"cell_type":"code","source":"## If we need space\n#del Left_join\n","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"def find_spillover_wds(content, wds = ['transfer','spillover','pass on','transmit','contract',\n                                       'distribute','progress','incubate','spread','disseminate','zoonosis','zoonotic']):\n    found = ''\n    for w in wds:\n        if w in content:\n            found += w + ' '\n    return found","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"def find_human_wds(content, wds = ['human','people','man','child','kid']):\n    found = ''\n    for w in wds:\n        if w in content:\n            found += w + ' '\n    return found","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"def find_covid_wds(content, wds = ['covid-19','covid','cov','coronavirus','corona']):\n    found = ''\n    for w in wds:\n        if w in content:\n            found += w + ' '\n    return found","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"def find_evidence_wds(content, wds = ['domestic animal', 'backyard*livestock', 'wet markets', 'meat markets','seafood markets', 'bites', 'bitten', \n                                      'laboratory*accident', 'fairs', 'petting*zoo',  'trading*wild animal', \n                                      'destruction*habitat', 'wild animal*food', 'lifestock pathogens','genetic mutations',\n                                      'animal testing', ' hunting ', 'industrial*farming', ' pet ', 'butcher', ' eat ', ' meat ']):\n    found = ''\n    for w in wds:\n        if w in content:\n            found += w + ','\n    return found","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"'''\nLeft_join['evidence_wds'] = Left_join['combined'].apply(find_evidence_wds)\nLeft_join.evidence_wds.unique()\n'''","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"### 4.1 Animal Source (Task 4.1)\nIn this step, we took the narrowed down list of 8000 articles and applied an additional level of precision to the search. To prevent skewing of the analysis due to irrelevant terms, we began with pulling together a manually curated list of 22 animals that were most likely related to the origin and spread of COVID-19.\nThe next logical step was to determine which topics each of the 22 animals were related to. For this purpose, we identified the most frequently used words and phrases for each animal term and demonstrated the relative frequency through wordcloud visualizations. \nWe also observed that many articles often referenced a pair of animals in describing the traits of the disease. Based on this, we built a circos-plot visualization that would represent which animal pairs appeared together most frequently in the articles.\n"},{"metadata":{"trusted":true},"cell_type":"code","source":"'''\nLeft_join['evidence_wds'].replace('', np.nan, inplace=True)\n\n\nLeft_join['animal_wds'] = Left_join['tokenized'].apply(find_animal_wds)\nLeft_join['spillover_wds'] = Left_join['tokenized'].apply(find_spillover_wds)\nLeft_join['human_wds'] = Left_join['tokenized'].apply(find_human_wds)\nLeft_join['virus_wds'] = Left_join['tokenized'].apply(find_covid_wds)\nLeft_join['evidence_wds'] = Left_join['combined'].apply(find_evidence_wds)\n\nLeft_join['animal_wds'] = Left_join['animal_wds'].str.replace('discus', '')\nLeft_join['animal_wds'].replace('', np.nan, inplace=True)\nLeft_join['spillover_wds'].replace('', np.nan, inplace=True)\nLeft_join['human_wds'].replace('', np.nan, inplace=True)\nLeft_join['virus_wds'].replace('', np.nan, inplace=True)\nLeft_join['evidence_wds'].replace('', np.nan, inplace=True)\n\narticlesAnimal = Left_join[Left_join[['animal_wds', 'spillover_wds', 'human_wds', 'virus_wds','evidence_wds']].notnull().all(1)]\narticlesAnimal['animal_wds'] = articlesAnimal['animal_wds'].map(lambda x: clean_text(x))\narticlesAnimal['animal_wds'].replace(' ', np.nan, inplace=True)\narticlesAnimal = articlesAnimal.dropna(subset=['animal_wds'])\narticlesAnimal\n'''","execution_count":null,"outputs":[]},{"metadata":{"id":"gqAv9eds1I6H","trusted":true},"cell_type":"code","source":"'''\nprint('Total articles containing animal keywords: ' + str(len(articlesAnimal)))\nprint('Total articles available: ' + str(len(fin_df)))\nprint(str(float(len(articlesAnimal)/len(fin_df)*100)) + '% of the articles available contains animal keywords')\n'''","execution_count":null,"outputs":[]},{"metadata":{"id":"_OHSpHaQ1I6L","trusted":true},"cell_type":"code","source":"'''\ndef find_country_wds(content, wds = ['transfer','spillover','pass on','transmit','contract',\n                                       'distribute','progress','incubate','spread','disseminate','zoonosis']):\n    found = ''\n    for w in wds:\n        if w in content:\n            found += w + ' '\n    return found\n'''","execution_count":null,"outputs":[]},{"metadata":{"id":"z7TuQJqH1I6P","trusted":true},"cell_type":"code","source":"#articlesAnimal.to_csv('animal_articles.csv', sep=',', encoding='utf-8')","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"#### Since we have saved 'animal_articles.csv', in the previous step, we can delete 'Left_join' to free up space. We only need 'animal_articles' going forward"},{"metadata":{"trusted":true},"cell_type":"code","source":"del Left_join","execution_count":null,"outputs":[]},{"metadata":{"id":"VNbBFiuT1I6Y"},"cell_type":"markdown","source":"### Import Processed Data"},{"metadata":{"id":"J2dCGqyk1I6Y","outputId":"f50780fd-6755-4cc0-a963-20010bfed8a9","trusted":true},"cell_type":"code","source":"articlesAnalysis = pd.read_csv(\"/kaggle/input/task3-data/animal_articles.csv\")\n#articlesAnalysis['combined'] = articlesAnalysis['combined'].str.replace(r'\\b(\\w{1,2})\\b', '')\narticlesAnalysis.head()","execution_count":null,"outputs":[]},{"metadata":{"id":"PHcprxlWUz8n","outputId":"ae97edd2-5255-4f87-f648-a736c50c481b","trusted":true},"cell_type":"code","source":"# Join the different processed titles together.\nanimal_string = ''.join(list(articlesAnalysis['animal_wds'].values))\n# Create a WordCloud object\nwordcloud = WordCloud(width = 600, height = 400, background_color=\"white\", max_words=5000, contour_width=3, contour_color='steelblue', collocations=False)\n# Generate a word cloud\nwordcloud.generate(animal_string)\n# Visualize the word cloud\nwordcloud.to_image()","execution_count":null,"outputs":[]},{"metadata":{"id":"30EAlGBX1I6b"},"cell_type":"markdown","source":"## Spillover Analysis"},{"metadata":{"id":"o4QWPSZP1I6c","trusted":true},"cell_type":"code","source":"def find_animal_wds(content, wds = [\n ' dog ',\n ' cat ',\n 'pig',\n 'mouse',\n 'bird',\n ' bat ',\n 'horse',\n ' rat ',\n 'sheep',\n 'chicken',\n 'rabbit',\n 'insect',\n 'goat',\n 'monkey',\n 'fox',\n 'fish',\n 'cow',\n 'ferret',\n 'deer',\n 'fly',\n 'raccoon',\n 'camel',\n 'hamster',\n 'bear',\n'pangolin',\n'tiger']):\n    found = ''\n    for w in wds:\n        if w in content:\n            found += w + ','\n            if found.count(',') == 5:\n                break\n    return found","execution_count":null,"outputs":[]},{"metadata":{"id":"cdtQj3Iz1I6g","outputId":"1fb3e4bf-5dc7-4fd7-b6bd-c61f64546610","trusted":true},"cell_type":"code","source":"articlesAnalysis['animal_wds'] = articlesAnalysis['combined'].apply(find_animal_wds)\narticlesAnalysis['animal_wds'] = articlesAnalysis['animal_wds'].str.replace(',', ' ')\narticlesAnalysis['animal_wds'].replace('', np.nan, inplace=True)\narticlesAnalysis = articlesAnalysis.dropna(subset=['animal_wds'])\narticlesAnalysis","execution_count":null,"outputs":[]},{"metadata":{"id":"xyVoxt6l1I6n","outputId":"d28fa562-b20d-478f-ffaa-0a853de1cecf","trusted":true},"cell_type":"code","source":"# Join the different processed titles together.\nanimal_string = ''.join(list(articlesAnalysis['animal_wds'].values))\n# Create a WordCloud object\nwordcloud = WordCloud(width = 600, height = 400, background_color=\"white\", max_words=5000, contour_width=3, contour_color='steelblue', collocations=False)\n# Generate a word cloud\nwordcloud.generate(animal_string)\n# Visualize the word cloud\nwordcloud.to_image()","execution_count":null,"outputs":[]},{"metadata":{"id":"An_dfqqL1I6t","trusted":true},"cell_type":"code","source":"def plot_50_most_common_words(count_data, count_vectorizer):\n    import matplotlib.pyplot as plt\n    words = count_vectorizer.get_feature_names()\n    total_counts = np.zeros(len(words))\n    for t in count_data:\n        total_counts+=t.toarray()[0]\n    \n    count_dict = (zip(words, total_counts))\n    count_dict = sorted(count_dict, key=lambda x:x[1], reverse=True)[0:20]\n    words = [w[0] for w in count_dict]\n    counts = [w[1] for w in count_dict]\n    x_pos = np.arange(len(words)) \n    \n    plt.figure(2, figsize=(10, 8/1.6180))\n    plt.subplot(title='20 most common words')\n    sns.set_context(\"notebook\", font_scale=1.25, rc={\"lines.linewidth\": 2.5})\n    sns.barplot(x_pos, counts, palette='husl')\n    plt.xticks(x_pos, words, rotation=90) \n    plt.xlabel('words')\n    plt.ylabel('counts')\n    plt.show()","execution_count":null,"outputs":[]},{"metadata":{"id":"jZYkQ3UM1I6w","outputId":"fc459e43-5c9e-4948-ea09-e860eb7781f4","trusted":true},"cell_type":"code","source":"# Initialise the count vectorizer with the English stop words\ncount_vectorizer = CountVectorizer(stop_words='english')\n# Fit and transform the processed titles\ncount_data2 = count_vectorizer.fit_transform(articlesAnalysis['animal_wds'])\n# Visualise the 50 most common words\nplot_50_most_common_words(count_data2, count_vectorizer)","execution_count":null,"outputs":[]},{"metadata":{"id":"g_u_K_XR1I61","outputId":"4cfc1ac0-7f46-42ff-bf85-d820c2c1883b","trusted":true},"cell_type":"code","source":"import re\ndef listToString(s):  \n    # initialize an empty string \n    str1 = \" \" \n    # return string   \n    return (str1.join(s)) \n\narticlesAnalysisTiger= articlesAnalysis[articlesAnalysis['animal_wds'].str.contains(\"tiger\")]\n\narticlesAnalysisTiger['animal_sentence'] = ''\nfor i in range(0,len(articlesAnalysisTiger)):\n    articlesAnalysisTiger['animal_sentence'].iloc[i] = listToString(re.findall(r\"([^.]*? tiger[^.]*\\.)\",'.' + articlesAnalysisTiger['combined'].iloc[i])).lower()\n\narticlesAnalysisTiger","execution_count":null,"outputs":[]},{"metadata":{"id":"Y9Kc8GJp1I68","outputId":"34f40b81-6ff5-459d-dc3f-6aefd1cf6581","trusted":true},"cell_type":"code","source":"import re\ndef listToString(s):  \n    # initialize an empty string \n    str1 = \" \" \n    # return string   \n    return (str1.join(s)) \n\narticlesAnalysisBat= articlesAnalysis[articlesAnalysis['animal_wds'].str.contains(\"bat\")]\n\narticlesAnalysisBat['animal_sentence'] = ''\nfor i in range(0,len(articlesAnalysisBat)):\n    articlesAnalysisBat['animal_sentence'].iloc[i] = listToString(re.findall(r\"([^.]*? bats [^.]*\\.)\",'.' + articlesAnalysisBat['combined'].iloc[i])).lower()\n\narticlesAnalysisBat","execution_count":null,"outputs":[]},{"metadata":{"id":"5ZAHx_oAXwm0","outputId":"159cc1ed-dc0f-41da-dbaf-9e8098569e8d","trusted":true},"cell_type":"code","source":"import re\ndef listToString(s):  \n    # initialize an empty string \n    str1 = \" \" \n    # return string   \n    return (str1.join(s)) \n\narticlesAnalysisPangolin= articlesAnalysis[articlesAnalysis['combined'].str.contains(\"pangolin\")]\n\narticlesAnalysisPangolin['animal_sentence'] = ''\nfor i in range(0,len(articlesAnalysisPangolin)):\n    articlesAnalysisPangolin['animal_sentence'].iloc[i] = listToString(re.findall(r\"([^.]*? pangolin[^.]*\\.)\",'.' + articlesAnalysisPangolin['combined'].iloc[i])).lower()\n\narticlesAnalysisPangolin","execution_count":null,"outputs":[]},{"metadata":{"id":"DAhgN9H0X0gl","outputId":"f6379cb2-242a-4e29-c216-09808f411be7","trusted":true},"cell_type":"code","source":"import re\ndef listToString(s):  \n    # initialize an empty string \n    str1 = \" \" \n    # return string   \n    return (str1.join(s)) \n\narticlesAnalysisCat= articlesAnalysis[articlesAnalysis['animal_wds'].str.contains(\"cat\")]\n\narticlesAnalysisCat['animal_sentence'] = ''\nfor i in range(0,len(articlesAnalysisCat)):\n    articlesAnalysisCat['animal_sentence'].iloc[i] = listToString(re.findall(r\"([^.]*? cat [^.]*\\.)\",'.' + articlesAnalysisCat['combined'].iloc[i])).lower()\n\narticlesAnalysisCat","execution_count":null,"outputs":[]},{"metadata":{"id":"Qo3VeD6C1I7A","outputId":"c723f396-e1ed-41fd-ab62-b48d68725d8d","trusted":true},"cell_type":"code","source":"articlesAnalysisTigerCov = articlesAnalysisTiger[articlesAnalysisTiger['animal_sentence'].str.contains(\"cov\" or \"corona\")]\narticlesAnalysisTigerCov","execution_count":null,"outputs":[]},{"metadata":{"id":"CwKmv4gr1I7M","outputId":"9ad4dc7a-fde6-4a13-9086-99c8c983fa92","trusted":true},"cell_type":"code","source":"# Join the different processed titles together.\nanimal_string = ''.join(list(articlesAnalysisBat['animal_sentence'].values))\n# Create a WordCloud object\nwordcloud = WordCloud(width = 600, height = 400, background_color=\"white\", max_words=5000, contour_width=3, contour_color='steelblue', collocations=False)\n# Generate a word cloud\nwordcloud.generate(animal_string)\n# Visualize the word cloud\nwordcloud.to_image()","execution_count":null,"outputs":[]},{"metadata":{"id":"D12bFIrY1I7O","outputId":"dab40ec1-1f2e-4ad6-beda-fb1d3b74412a","trusted":true},"cell_type":"code","source":"# Initialise the count vectorizer with the English stop words\ncount_vectorizer = CountVectorizer(stop_words='english')\n# Fit and transform the processed titles\ncount_data = count_vectorizer.fit_transform(articlesAnalysisBat['animal_sentence'])\n# Visualise the 50 most common words\nplot_50_most_common_words(count_data, count_vectorizer)","execution_count":null,"outputs":[]},{"metadata":{"id":"v89CEtZm1I7W","outputId":"2be26786-ca7e-4274-ed05-f098c15148ad","trusted":true},"cell_type":"code","source":"from sklearn.feature_extraction.text import TfidfTransformer\n \ntfidf_transformer=TfidfTransformer(smooth_idf=True,use_idf=True)\ntfidf_transformer.fit(count_data)\n\ndocs_test=articlesAnalysisBat['animal_sentence'].tolist()\n\ndef sort_coo(coo_matrix):\n    tuples = zip(coo_matrix.col, coo_matrix.data)\n    return sorted(tuples, key=lambda x: (x[1], x[0]), reverse=True)\n \ndef extract_topn_from_vector(feature_names, sorted_items, topn=10):\n    \"\"\"get the feature names and tf-idf score of top n items\"\"\"\n    \n    #use only topn items from vector\n    sorted_items = sorted_items[:topn]\n \n    score_vals = []\n    feature_vals = []\n    \n    # word index and corresponding tf-idf score\n    for idx, score in sorted_items:\n        \n        #keep track of feature name and its corresponding score\n        score_vals.append(round(score, 3))\n        feature_vals.append(feature_names[idx])\n \n    #create a tuples of feature,score\n    #results = zip(feature_vals,score_vals)\n    results= {}\n    for idx in range(len(feature_vals)):\n        results[feature_vals[idx]]=score_vals[idx]\n    \n    return results\nfeature_names=count_vectorizer.get_feature_names()\n \n# get the document that we want to extract keywords from\ndoc=listToString(docs_test)\n \n#generate tf-idf for the given document\ntf_idf_vector=tfidf_transformer.transform(count_vectorizer.transform([doc]))\n \n#sort the tf-idf vectors by descending order of scores\nsorted_items=sort_coo(tf_idf_vector.tocoo())\n \n#extract only the top n; n here is 10\nkeywords=extract_topn_from_vector(feature_names,sorted_items,10)\n \n# now print the results\nprint(\"\\n=====Doc=====\")\nprint(doc)\nprint(\"\\n===Keywords===\")\nfor k in keywords:\n    print(k,keywords[k])\n","execution_count":null,"outputs":[]},{"metadata":{"id":"7spiTZ_51I7d","trusted":true},"cell_type":"code","source":"from collections import Counter\n\nevidence = ['backyard*livestock', 'wet markets', 'meat markets','seafood markets', 'bit*', \n                                      'laboratory*accident', 'fairs', 'petting*zoo',  'trading*wild animal', \n                                      'destruction*habitat', 'wild animal*food', 'lifestock pathogens','genetic mutations',\n                                      'animal testing', ' hunting ', 'industrial*farming', ' pet ', 'butcher', ' eat ', ' meat ']\nfrom operator import itemgetter\narticlesAnalysis['evidence_wds'] = articlesAnalysis['evidence_wds'].str.replace('domestic animal','')\nflat_list = [item for sublist in articlesAnalysis['evidence_wds'].str.split(',') for item in sublist]\nflat_list.remove('')\ntest_dict = Counter(flat_list)\ndel test_dict['']","execution_count":null,"outputs":[]},{"metadata":{"id":"vm5DG5iF1I7f","outputId":"22eb820c-849d-40d4-943c-c94e8e60eea7","trusted":true},"cell_type":"code","source":"test_dict","execution_count":null,"outputs":[]},{"metadata":{"id":"pDYeMc9g1I7m","outputId":"ba1b38ed-f9f6-4073-d4b3-529f1e4a17c1","trusted":true},"cell_type":"code","source":"import matplotlib.pyplot as plt\nfrom matplotlib.pyplot import figure\nfigure(num=None, figsize=(8, 6), dpi=80, facecolor='w', edgecolor='k')\n\n\n#explsion\nexplode = (0.05,0.05,0.05,0.05,0.05,0.05,0.05,0.05,0.05,0.05,0.05,0.05)\n \nplt.pie(test_dict.values(), labels=test_dict, autopct='%1.1f%%', startangle=90, pctdistance=0.85, explode = explode)\n#draw circle\ncentre_circle = plt.Circle((0,0),0.70,fc='white')\nfig = plt.gcf()\nfig.gca().add_artist(centre_circle)\n# Equal aspect ratio ensures that pie is drawn as a circle\nplt.title('Spillover Source')\nplt.tight_layout()\nplt.show()\n","execution_count":null,"outputs":[]},{"metadata":{"id":"yk0LtIh61I7s","outputId":"f4ab90a5-96ea-4bfa-f6fc-fd126a7f7169","trusted":true},"cell_type":"code","source":"articlesAnalysis['animal_wds'] = articlesAnalysis['animal_wds'].str.replace('animal','')\n#articlesAnalysisAnimal = articlesAnalysis[articlesAnalysis['animal_wds'] == \"bat \"]\n#articlesAnalysisAnimal = articlesAnalysis[articlesAnalysis['animal_wds'].str.contains(\"dog\")]\narticlesAnalysisAnimal = articlesAnalysis[articlesAnalysis['evidence_wds'].str.contains(\"meat\")]\narticlesAnalysisAnimal = articlesAnalysisAnimal[articlesAnalysisAnimal['evidence_wds'].str.contains(\" eat \")]\narticlesAnalysisAnimal = articlesAnalysisAnimal[articlesAnalysisAnimal['spillover_wds'].str.contains(\"zoono\")]\narticlesAnalysisAnimal","execution_count":null,"outputs":[]},{"metadata":{"id":"gvQaZvq51I70","outputId":"d840089c-d2e7-4f45-c424-f1ea3ca98f1a","trusted":true},"cell_type":"code","source":"# Initialise the count vectorizer with the English stop words\ncount_vectorizer = CountVectorizer(stop_words='english')\n# Fit and transform the processed titles\ncount_data2 = count_vectorizer.fit_transform(articlesAnalysisAnimal['animal_wds'])\n# Visualise the 50 most common words\nplot_50_most_common_words(count_data2, count_vectorizer)","execution_count":null,"outputs":[]},{"metadata":{"id":"_ZWINL9s1I8A","outputId":"25e8e7f4-82e0-4828-acfa-b01273b6fcdf","trusted":true},"cell_type":"code","source":"import matplotlib as mpl\nfrom matplotlib.pyplot import figure\nmpl.rcParams['font.size'] = 9.0\nfigure(num=None, figsize=(8, 6), dpi=80, facecolor='w', edgecolor='k')\n\nflat_list = [item for sublist in articlesAnalysisAnimal['evidence_wds'].str.split(',') for item in sublist]\nflat_list.remove('')\ntest_dict = Counter(flat_list)\ndel test_dict['']\n\n#explsion\nexplode = (0.05,0.05,0.05,0.05,0.05,0.05,0.05,0.05,0.05,0.05,0.05,0.05,0.05)\n \nplt.pie(test_dict.values(), labels=test_dict, autopct='%1.1f%%', startangle=90, pctdistance=0.85, explode = None, textprops={'fontsize': 8})\n#draw circle\ncentre_circle = plt.Circle((0,0),0.70,fc='white')\nfig = plt.gcf()\nfig.gca().add_artist(centre_circle)\n# Equal aspect ratio ensures that pie is drawn as a circle\nplt.title('Spillover Source')\nplt.tight_layout()\nplt.show()","execution_count":null,"outputs":[]},{"metadata":{"id":"zWJQVaQM1I8b","outputId":"fe1a39f1-3c34-4a30-c8a8-89a3fa102a95","trusted":true},"cell_type":"code","source":"import datetime as dt\narticlesAnalysis['publish_time'] = pd.to_datetime(articlesAnalysis['publish_time'])\narticlesAnalysis.head()","execution_count":null,"outputs":[]},{"metadata":{"id":"EAj6-uyI1I8f","outputId":"4fd192cb-e189-4663-db38-1c3fd0537cee","trusted":true},"cell_type":"code","source":"import datetime as dt\n\narticlesAnalysis3 = articlesAnalysis[articlesAnalysis['evidence_wds'].str.contains(\" eat \")]\narticlesAnalysis3 = articlesAnalysis3[articlesAnalysis3['publish_time'].dt.year > 2018]\narticlesAnalysis3 = articlesAnalysis3[~articlesAnalysis3['evidence_wds'].str.contains(\"bit\")]\narticlesAnalysis3 = articlesAnalysis3[articlesAnalysis3['spillover_wds'].str.contains(\"zoono\")]\n\narticlesAnalysis3\n","execution_count":null,"outputs":[]},{"metadata":{"id":"_6iDzoNd1I8m","outputId":"8b4e0ede-407a-46e7-9fdd-94d25fb90d81","trusted":true},"cell_type":"code","source":"# Initialise the count vectorizer with the English stop words\ncount_vectorizer = CountVectorizer(stop_words='english')\n# Fit and transform the processed titles\ncount_data4 = count_vectorizer.fit_transform(articlesAnalysis3['animal_wds'])\n# Visualise the 50 most common words\nplot_50_most_common_words(count_data4, count_vectorizer)","execution_count":null,"outputs":[]},{"metadata":{"id":"I95x2Ul11I87","outputId":"41dd2e51-9c12-4d43-a0c2-4f0621ce2e20","trusted":true},"cell_type":"code","source":"articlesAnalysis2 = articlesAnalysis[articlesAnalysis['combined'].str.contains(\"zoono\")]\narticlesAnalysis2","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"## LDA "},{"metadata":{"id":"5ZOK41_81I9E","outputId":"3fb7301a-b41b-4e7b-ed79-910b1c1cae02","trusted":true},"cell_type":"code","source":"\nfrom sklearn.decomposition import LatentDirichletAllocation as LDA\n\n# Initialise the count vectorizer with the English stop words\ncount_vectorizer = CountVectorizer(stop_words='english')\n# Fit and transform the processed titles\ncount_data = count_vectorizer.fit_transform(articlesAnalysis2['combined']) \n# Helper function\ndef print_topics(model, count_vectorizer, n_top_words):\n    words = count_vectorizer.get_feature_names()\n    for topic_idx, topic in enumerate(model.components_):\n        print(\"\\nTopic #%d:\" % topic_idx)\n        print(\" \".join([words[i]\n                        for i in topic.argsort()[:-n_top_words - 1:-1]]))\n        \n# Tweak the two parameters below\nnumber_topics = 10\nnumber_words = 10\n# Create and fit the LDA model\nlda = LDA(n_components=number_topics, n_jobs=-1)\nlda.fit(count_data)\n# Print the topics found by the LDA model\nprint(\"Topics found via LDA:\")\nprint_topics(lda, count_vectorizer, number_words)\n","execution_count":null,"outputs":[]},{"metadata":{"id":"UhtMPRVP1I9K","outputId":"122d9ed8-439e-4e76-8e39-6d628fad8e8d","trusted":true},"cell_type":"code","source":"# Initialise the count vectorizer with the English stop words\ncount_vectorizer = CountVectorizer(stop_words='english')\n# Fit and transform the processed titles\ncount_data = count_vectorizer.fit_transform(articlesAnalysis2['combined'])\n# Visualise the 50 most common words\nplot_50_most_common_words(count_data, count_vectorizer)","execution_count":null,"outputs":[]},{"metadata":{"id":"yBiXiuNf1I9O","outputId":"52744f82-2248-4353-a365-8530fa5af6df","trusted":true},"cell_type":"code","source":"from sklearn.feature_extraction.text import TfidfTransformer\n \ntfidf_transformer=TfidfTransformer(smooth_idf=True,use_idf=True)\ntfidf_transformer.fit(count_data)","execution_count":null,"outputs":[]},{"metadata":{"id":"5vEdjnjq1I9R","trusted":true},"cell_type":"code","source":"docs_test=articlesAnalysis2['combined'].tolist()","execution_count":null,"outputs":[]},{"metadata":{"id":"dODU52Rf1I9T","trusted":true},"cell_type":"code","source":"def sort_coo(coo_matrix):\n    tuples = zip(coo_matrix.col, coo_matrix.data)\n    return sorted(tuples, key=lambda x: (x[1], x[0]), reverse=True)\n \ndef extract_topn_from_vector(feature_names, sorted_items, topn=10):\n    \"\"\"get the feature names and tf-idf score of top n items\"\"\"\n    \n    #use only topn items from vector\n    sorted_items = sorted_items[:topn]\n \n    score_vals = []\n    feature_vals = []\n    \n    # word index and corresponding tf-idf score\n    for idx, score in sorted_items:\n        \n        #keep track of feature name and its corresponding score\n        score_vals.append(round(score, 3))\n        feature_vals.append(feature_names[idx])\n \n    #create a tuples of feature,score\n    #results = zip(feature_vals,score_vals)\n    results= {}\n    for idx in range(len(feature_vals)):\n        results[feature_vals[idx]]=score_vals[idx]\n    \n    return results","execution_count":null,"outputs":[]},{"metadata":{"id":"3SO2lJKY1I9U","outputId":"46e2d9a5-d636-4d62-ff26-4c967f8ddd8d","trusted":true},"cell_type":"code","source":"feature_names=count_vectorizer.get_feature_names()\n \n# get the document that we want to extract keywords from\ndoc=docs_test[0]\n \n#generate tf-idf for the given document\ntf_idf_vector=tfidf_transformer.transform(count_vectorizer.transform([doc]))\n \n#sort the tf-idf vectors by descending order of scores\nsorted_items=sort_coo(tf_idf_vector.tocoo())\n \n#extract only the top n; n here is 10\nkeywords=extract_topn_from_vector(feature_names,sorted_items,10)\n \n# now print the results\nprint(\"\\n=====Doc=====\")\nprint(doc)\nprint(\"\\n===Keywords===\")\nfor k in keywords:\n    print(k,keywords[k])\n ","execution_count":null,"outputs":[]},{"metadata":{"id":"pz6TtdGv1I9X","trusted":true},"cell_type":"code","source":"def find_africa_wds(content, wds = ['Africa','Burundi', 'Comoros', 'Djibouti', 'Eritrea', 'Ethiopia', 'Kenya', 'Madagascar', 'Malawi', 'Mauritius', 'Mayotte’,  ‘Mozambique', 'Reunion', 'Rwanda', 'Seychelles', 'Somalia', 'Tanzania', 'United Republic of Uganda', 'Zambia', 'Zimbabwe', 'Angola', 'Cameroon', 'Chad', 'Congo', 'Algeria', 'Egypt', 'Libyan Arab Jamahiriya', 'Morroco', 'South Sudan', 'Sudan', 'Tunisia', 'Western Sahara', 'Botswana', 'Eswatini’, ’Swaziland', 'Lesotho', 'Namibia', 'South Africa', 'Benin', 'Burkina Faso', 'Cape Verde', 'Ivory Coast', 'Gambia', 'Ghana', 'Guinea', 'Guinea-Bissau', 'Liberia', 'Mali', 'Mauritania', 'Niger', 'Nigeria', 'Saint Helena', 'Senegal', 'Sierra Leone', 'Togo']):\n    found = ''\n    for w in wds:\n        if w in content:\n            found += w + ' '\n    return found","execution_count":null,"outputs":[]},{"metadata":{"id":"oDizE7uP1I9c","trusted":true},"cell_type":"code","source":"def find_asia_wds(content, wds = ['Asia','Afganistan', 'Armenia', 'Azerbaijan', 'Bangladesh', 'Bhutan', 'Brunei Darussalam', 'Cambodia', 'China', 'Georgia', 'Hong Kong', 'India', 'Indonesia', 'Japan', 'Kazakhstan', 'North Korea’, “South Korea', 'Kyrgyzstan', 'Laos', 'Macao', 'Malaysia', 'Maldives', 'Mongolia', 'Myanmar', 'Nepal', 'Pakistan', 'Phillipines', 'Singapore', 'Sri Lanka', 'Taiwan', 'Tajikistan', 'Thailand', 'Turkmenistan', 'Uzbekistan', 'Vietnam']):\n    found = ''\n    for w in wds:\n        if w in content:\n            found += w + ' '\n    return found","execution_count":null,"outputs":[]},{"metadata":{"id":"voUJCUqz1I9g","trusted":true},"cell_type":"code","source":"def find_america_wds(content, wds = ['Bermuda', 'Canada', 'Greenland', 'United States', 'U.S.A.', 'USA', 'US', 'Argentina', 'Bolivia', 'Brazil', 'Chile', 'Colombia', 'Ecuador', 'Guyana', 'Paraguay', 'Peru', 'Uruguay', 'Venezuela', 'Anguilla', 'Antigua', 'Barbuda', 'Aruba', 'Bahamas', 'Barbados', 'Bonaire', 'British Virgin Islands', 'Cayman Islands', 'Cuba', 'Curaçao', 'Dominican Republic', 'Grenada', 'Guadeloupe', 'Haiti', 'Jamaica', 'Martinique', 'Monserrat', 'Puerto Rico', 'Saint Lucia', 'Saint Martin', 'Saint Vincent and the Grenadines', 'Sint Maarten', 'Trinidad and Tobago', 'Turks and Caicos Islands', 'Virgin Islands (US)', 'Belize', 'Costa Rica', 'El Salvador', 'Guatemala', 'Honduras', 'Mexico', 'Nicaragua', 'Panama']):\n    found = ''\n    for w in wds:\n        if w in content:\n            found += w + ' '\n    return found\n\ndef find_europe_wds(content, wds = ['Albania', 'Andorra', 'Belarus', 'Bosnia', 'Croatia', 'European Union', 'Faroe Islands', 'Gibraltar’,  ‘Iceland', 'Jersey', 'Kosovo', 'Liechtenstein', 'Moldova', 'Monaco', 'Montenegro', 'North Macedonia', 'Norway', 'Russia', 'San Marino', 'Serbia', 'Switzerland', 'Turkey', 'Ukraine']):\n    found = ''\n    for w in wds:\n        if w in content:\n            found += w + ' '\n    return found\n\ndef find_middleeast_wds(content, wds = ['Bahrain', 'Iraq', 'Iran', 'Israel', 'Jordan', 'Kuwait', 'Lebanon', 'Oman', 'Palestine', 'Qatar', 'Saudi Arabia', 'Syria', 'United Arab Emirates', 'Yemen']):\n    found = ''\n    for w in wds:\n        if w in content:\n            found += w + ' '\n    return found\n\ndef find_oceania_wds(content, wds = ['Australia', 'Fiji', 'French Polynesia', 'Guam', 'Kiribati', 'Marshall Islands', 'Micronesia', 'New Caledonia', 'New Zealand', 'Papua New Guinea', 'Samoa', 'Samoa, American', 'Solomon, Islands', 'Tonga', 'Vanuatu']):\n    found = ''\n    for w in wds:\n        if w in content:\n            found += w + ' '\n    return found","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"articlesAnalysis['africa'] = articlesAnalysis['combined'].apply(find_africa_wds)\narticlesAnalysis['africa'].replace('', np.nan, inplace=True)\narticlesAnalysis['asia'] = articlesAnalysis['combined'].apply(find_asia_wds)\narticlesAnalysis['asia'].replace('', np.nan, inplace=True)\narticlesAnalysis['america'] = articlesAnalysis['combined'].apply(find_america_wds)\narticlesAnalysis['america'].replace('', np.nan, inplace=True)\narticlesAnalysis['europe'] = articlesAnalysis['combined'].apply(find_europe_wds)\narticlesAnalysis['europe'].replace('', np.nan, inplace=True)\narticlesAnalysis['middleeast'] = articlesAnalysis['combined'].apply(find_middleeast_wds)\narticlesAnalysis['middleeast'].replace('', np.nan, inplace=True)\narticlesAnalysis['oceania'] = articlesAnalysis['combined'].apply(find_oceania_wds)\narticlesAnalysis['oceania'].replace('', np.nan, inplace=True)\narticlesAnalysis\n\narticlesAnalysisAmerica = articlesAnalysis.dropna(subset=['america'])\narticlesAnalysisAmerica\n\narticlesAnalysisUSA = articlesAnalysisAmerica[articlesAnalysisAmerica['america'].str.contains('USA' or 'United States')]\narticlesAnalysisUSA = articlesAnalysisUSA[articlesAnalysisUSA[['africa', 'asia', 'europe', 'middleeast','oceania']].isnull().all(1)]\narticlesAnalysisUSA","execution_count":null,"outputs":[]},{"metadata":{"id":"mn1va-Ph1I9p","outputId":"b3d3ce16-ff12-4f95-e4fb-a74fcc405899","trusted":true},"cell_type":"code","source":"flat_list = [item for sublist in articlesAnalysisUSA['evidence_wds'].str.split(',') for item in sublist]\nflat_list.remove('')\ntest_dict = Counter(flat_list)\ndel test_dict['']\ntest_dict","execution_count":null,"outputs":[]},{"metadata":{"id":"1RQoxtCB1I9u","outputId":"8fbbe69e-15f3-4ce9-f571-a571e9a6acb1","trusted":true},"cell_type":"code","source":"figure(num=None, figsize=(8, 6), dpi=80, facecolor='w', edgecolor='k')\n#explsion\nexplode = (0.05,0.05,0.05,0.05,0.05,0.05,0.05,0.05,0.05,0.05)\n \nplt.pie(test_dict.values(), labels=test_dict, autopct='%1.1f%%', startangle=90, pctdistance=0.85, explode = None)\n#draw circle\ncentre_circle = plt.Circle((0,0),0.70,fc='white')\nfig = plt.gcf()\nfig.gca().add_artist(centre_circle)\n# Equal aspect ratio ensures that pie is drawn as a circle\nplt.title('Spillover Source')\nplt.tight_layout()\nplt.show()","execution_count":null,"outputs":[]},{"metadata":{"id":"u9EX4l4S1I9z","outputId":"1db51a98-3d31-4228-a606-1985766c1df7","trusted":true},"cell_type":"code","source":"# Initialise the count vectorizer with the English stop words\ncount_vectorizer = CountVectorizer(stop_words='english')\n# Fit and transform the processed titles\ncount_data4 = count_vectorizer.fit_transform(articlesAnalysisUSA['animal_wds'])\n# Visualise the 50 most common words\nplot_50_most_common_words(count_data4, count_vectorizer)","execution_count":null,"outputs":[]},{"metadata":{"id":"IB2YsTa31I-G"},"cell_type":"markdown","source":"### 4.2 Risk Reduction\n\nRisk Reduction Strategies (Task 4.2)\nSustainable risk mitigation strategies are critical to prevent future instances of similar viral outbreaks involving animals. Research and implementation of such strategies would require very specific data mining using keywords related to risk.\n•\tRisk reduction strategy key search terms: education, economy, socioeconomic, unemployment, school closure, population density, religion, discrimination, behavioral risk, racism, financial status\n\nAdditional weight is added to sentences with “risk reduction” phrase in them. \n"},{"metadata":{"id":"mphTIBEE1I-G","outputId":"54711569-8673-4e94-9561-ad241bcf9e49","trusted":true},"cell_type":"code","source":"### Filtering Dataset\n\narticlesAnalysisRiskReduction = articlesAnalysis[articlesAnalysis['combined'].str.contains('risk reduction')]\narticlesAnalysisRiskReduction['risk_reduction_sentence'] = ''\nfor i in range(0,len(articlesAnalysisRiskReduction)):\n    articlesAnalysisRiskReduction['risk_reduction_sentence'].iloc[i] = listToString(re.findall(r\"([^.]*?[^.]*?risk reduction[^.]*\\.[^.]*\\.)\",'.' + articlesAnalysisRiskReduction['combined'].iloc[i])).lower()\narticlesAnalysisRiskReduction['risk_reduction_sentence'].replace('', np.nan, inplace=True)\narticlesAnalysisRiskReduction = articlesAnalysisRiskReduction.dropna(subset=['risk_reduction_sentence'])\narticlesAnalysisRiskReduction","execution_count":null,"outputs":[]},{"metadata":{"id":"PIVT4rMN1I-I","outputId":"683308c8-c651-4ce9-e282-9214b9665651","trusted":true},"cell_type":"code","source":"### Keywords Visualization\n\n# Initialise the count vectorizer with the English stop words\ncount_vectorizer = CountVectorizer(stop_words='english')\n# Fit and transform the processed titles\ncount_data10 = count_vectorizer.fit_transform(articlesAnalysisRiskReduction['risk_reduction_sentence'])\n# Visualise the 50 most common words\nplot_50_most_common_words(count_data10, count_vectorizer)\n\n# Join the different processed titles together.\nanimal_string = ''.join(list(articlesAnalysisRiskReduction['risk_reduction_sentence'].values))\n# Create a WordCloud object\nwordcloud = WordCloud(width = 600, height = 400, background_color=\"white\", max_words=5000, contour_width=3, contour_color='steelblue', collocations=False)\n# Generate a word cloud\nwordcloud.generate(animal_string)\n# Visualize the word cloud\nwordcloud.to_image()","execution_count":null,"outputs":[]},{"metadata":{"id":"xo_z0eqP1I-a"},"cell_type":"markdown","source":"### 4.3 Socioeconomic Risk\n\n4.3. Socioeconomic and Behavioral Risk Factors (Task 4.3)\nIn order to analyze socioeconomic impacts of the potential future transfer of the virus from animals to humans, further keywords were utilized to further refine search results.\n•\tSocioeconomic terms used:  socioeconomic, behavioral, economy, population density, school closure, daily wage, crowd, wet market, discrimination, racism, religion, financial status, education\n\nAdditional weight is added to sentences with “socioeconomic” and “behavioral” word in them. \n"},{"metadata":{"id":"jsWOz5061I-b","outputId":"562d83b4-03de-41b5-8294-430497d21cac","trusted":true},"cell_type":"code","source":"### Filtering Dataset\n\ndef find_socioeconomic_wds(content, wds = ['socioeconomic','economy', 'behavioral risk', 'unemployment', 'population density', 'school closure', 'daily wage', 'discrimination', 'racism', 'religion', 'financial status', 'education']):\n    found = ''\n    for w in wds:\n        if w in content:\n            found += w + ', '\n    return found\n\narticlesAnalysis['social_wds'] = articlesAnalysis['combined'].apply(find_socioeconomic_wds)\narticlesAnalysis['social_wds'].replace('', np.nan, inplace=True)\narticlesAnalysisSocial = articlesAnalysis.dropna(subset=['social_wds'])\narticlesAnalysisSocio = articlesAnalysisSocial[articlesAnalysisSocial['social_wds'].str.contains('socioeconomic')]\narticlesAnalysisSocio","execution_count":null,"outputs":[]},{"metadata":{"id":"FwEggWuc1I-g","outputId":"4c29f55e-ec34-47ab-ecce-4fb715ab27f2","trusted":true},"cell_type":"code","source":"### Keywords Visualization\n\nimport matplotlib.pyplot as plt\nfrom matplotlib.pyplot import figure\nfigure(num=None, figsize=(8, 6), dpi=80, facecolor='w', edgecolor='k')\n\nfrom operator import itemgetter\nflat_list = [item for sublist in articlesAnalysisSocial['social_wds'].str.split(', ') for item in sublist]\ntest_dict = Counter(flat_list)\ndel test_dict['']\n\n#explsion\nexplode = (0.05,0.05,0.05,0.05,0.05,0.05,0.05,0.05,0.05,0.05,0.05)\n \nplt.pie(test_dict.values(), labels=test_dict, autopct='%1.1f%%', startangle=90, pctdistance=0.85, explode = None, textprops={'fontsize': 8})\n#draw circle\ncentre_circle = plt.Circle((0,0),0.70,fc='white')\nfig = plt.gcf()\nfig.gca().add_artist(centre_circle)\n# Equal aspect ratio ensures that pie is drawn as a circle\nplt.title('Socioeconomic Keywords')\nplt.tight_layout()\nplt.show()","execution_count":null,"outputs":[]},{"metadata":{"id":"Y1FerpQV1I-j","outputId":"bfe19822-21f2-4586-cf48-6e0c035afb78","trusted":true},"cell_type":"code","source":"test_dict","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"import re\ndef listToString(s):  \n    # initialize an empty string \n    str1 = \" \" \n    # return string   \n    return (str1.join(s)) \n\narticlesAnalysisSocio['socioeconomic_sentence'] = ''\nfor i in range(0,len(articlesAnalysisSocio)):\n    articlesAnalysisSocio['socioeconomic_sentence'].iloc[i] = listToString(re.findall(r\"([^.]*?socioeconomic[^.]*\\.[^.]*\\.[^.]*\\.[^.]*\\.)\",'.' + articlesAnalysisSocio['combined'].iloc[i])).lower()\narticlesAnalysisSocio['socioeconomic_sentence'].replace('', np.nan, inplace=True)\narticlesAnalysisSocialEco = articlesAnalysisSocio.dropna(subset=['socioeconomic_sentence'])\narticlesAnalysisSocialEco\n\narticlesAnalysisSocio['socioeconomic_sentence'].iloc[0]\n\n# Join the different processed titles together.\nanimal_string = ''.join(list(articlesAnalysisSocio['socioeconomic_sentence'].values))\n# Create a WordCloud object\nwordcloud = WordCloud(width = 600, height = 400, background_color=\"white\", max_words=5000, contour_width=3, contour_color='steelblue', collocations=False)\n# Generate a word cloud\nwordcloud.generate(animal_string)\n# Visualize the word cloud\nwordcloud.to_image()","execution_count":null,"outputs":[]},{"metadata":{"id":"V0kWeR6V1I_P"},"cell_type":"markdown","source":"## Section 5 : Retrieve Relevant articles using BERT\n\nIn the spirit of building a scalable and reusable solution, we utilized BERT that allows a user to repurpose the code to find relevant articles for any phrase. As an output, the code will return the relevant articles along with the summary and a similarity score as a measure of relevance."},{"metadata":{"id":"gcIiZVT5LGmz","outputId":"f6f32cf7-5f40-4929-cf51-ec51a7550db5","trusted":false},"cell_type":"markdown","source":"fin_df = pd.read_csv('/kaggle/input/task3-data/fin_df.csv')\nfin_df.head()"},{"metadata":{"trusted":true},"cell_type":"code","source":"\nfin_df = pd.read_csv('/kaggle/input/task3-data/fin_df.csv')\nfin_df.head()","execution_count":null,"outputs":[]},{"metadata":{"id":"ZhFgdG-CLSKs","outputId":"c4f6cffa-09e8-49ec-8e02-e8cf42479a75","trusted":false},"cell_type":"markdown","source":"fin_df.isnull().sum()"},{"metadata":{"id":"lkoHC_GZTMUh","trusted":true},"cell_type":"code","source":"fin_df['title'].fillna(\"NoTitle\", inplace = True)\nfin_df['abstract'].fillna(\"NoAbstract\", inplace = True)\nfin_df['full_text'].fillna(\"NoText\", inplace = True)\nfin_df['combined'] = 'Title : '+fin_df['title'] + '; Abstract ' + fin_df['abstract'] + '; Full Text ' + fin_df['full_text']","execution_count":null,"outputs":[]},{"metadata":{"id":"5nS5fIrATv06","outputId":"1b4d76f9-f77e-4ba9-fb4c-696f4858f4ca","trusted":true},"cell_type":"code","source":"\n\n# Load the BERT model. Various models trained on Natural Language Inference (NLI) https://github.com/UKPLab/sentence-transformers/blob/master/docs/pretrained-models/nli-models.md and \n# Semantic Textual Similarity are available https://github.com/UKPLab/sentence-transformers/blob/master/docs/pretrained-models/sts-models.md\n\nmodel = SentenceTransformer('bert-base-nli-mean-tokens')","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"### Word embeddings on full article text"},{"metadata":{"id":"ikS8NeIQc_ny","outputId":"6be33ca3-21d3-49b9-a108-722a9a1724df","trusted":true},"cell_type":"code","source":"#Uncomment line below if you want to embed\n#content = [re.sub(' \\n\\n\\n ','',x) for x in fin_df['full_text'].to_list()]\n","execution_count":null,"outputs":[]},{"metadata":{"id":"87j37IRg1I_j"},"cell_type":"markdown","source":"*** The Cell below Encodes the articles. It takes quite a while to run. The encoded embedding is then saved as a pickle file. For Testing on New queries, please load the pickle file.\nGo To section 'Test New Queries'\n***"},{"metadata":{},"cell_type":"markdown","source":"### NOTE : THE CELL BELOW WILL TAKE A LONG TIME TO EXECUTE. SKIP AND LOAD SAVED MODEL\n"},{"metadata":{"id":"aK5m_CIcXxEF","outputId":"27529f07-5c53-493d-97d5-f01aeb91c1e9","trusted":true},"cell_type":"code","source":"### Embed the article contents\n'''\nembedding = model.encode(content, show_progress_bar=True)\n'''","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"#Save embeddings\n'''\n\nwith open('full_text_embeddings.pkl', 'wb') as embed:\n    pickle.dump(embedding, embed)\n'''","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"### Load Picked Embedding file"},{"metadata":{"trusted":true},"cell_type":"code","source":"with open('/kaggle/input/task3-data/full_text_embeddings.pkl','rb') as f:\n    embedding = pickle.load(f)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"### List Questions"},{"metadata":{"trusted":true},"cell_type":"code","source":"queries = ['How is livestock affected due to Corona virus?',\n          'How are farmers afftected due to Coronoa vurus?',\n          'How is the spread of corona virus?',\n          'Has corona virus infected animals?',\n          'How does corona virus transfer?']","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"### Embed Queries"},{"metadata":{"id":"QdueU3mKYjVu","trusted":true},"cell_type":"code","source":"query_embeddings = model.encode(queries)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"fin_df.head()","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"df = pd.DataFrame(columns=['Query','Cosine Similarity','Summary','Article Full Text'])","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"\nimport scipy as sc\ntop_n_selects = 1\nfor query, query_embedding in zip(queries, query_embeddings):\n    distances = sc.spatial.distance.cdist([query_embedding], embedding, \"cosine\")[0]\n\n    results = zip(range(len(distances)), distances)\n    results = sorted(results, key=lambda x: x[1])\n    \n    \n    \n    print('Query : ',query)\n    print('###########################################')\n    \n    for idx, distance in results[0:top_n_selects]:\n        print(\"\\nCosine Similarity (Score: %.4f)\" % (1-distance),\"\\n\")\n        \n        body = fin_df['combined'][idx].strip() \n        summary_model = Summarizer()\n        result = summary_model(body, min_length=60)\n        summary = ''.join(result)\n        print('Summary:',summary)\n        print('\\n')\n        print('Article:',body)\n        \n        similarity = (1-distance)\n        to_append = [query, similarity, summary, body]\n        a_series = pd.Series(to_append, index = df.columns)\n        df = df.append(a_series, ignore_index=True)\n\n        \n        print('_________________________________________')","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"df.head()","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"## Search Based on New Query"},{"metadata":{"id":"opqUFXMS1I_4","outputId":"d0d49566-0d25-4524-c802-0d47bc51ff1c","trusted":false},"cell_type":"code","source":"'''\nqueries = []\nquery = input('Enter you query:')\nqueries.append(query)\nqueries\n'''","execution_count":0,"outputs":[]},{"metadata":{"id":"yK2Iatwr1I_7","trusted":false},"cell_type":"code","source":"# Load the BERT model. Various models trained on Natural Language Inference (NLI) https://github.com/UKPLab/sentence-transformers/blob/master/docs/pretrained-models/nli-models.md and \n# Semantic Textual Similarity are available https://github.com/UKPLab/sentence-transformers/blob/master/docs/pretrained-models/sts-models.md\n\n#model = SentenceTransformer('bert-base-nli-mean-tokens')","execution_count":0,"outputs":[]},{"metadata":{"id":"4S_hM27d1I_9"},"cell_type":"markdown","source":"### Embed the query"},{"metadata":{"id":"Y6DfIieT1I_9","trusted":false},"cell_type":"code","source":"#query_embeddings = model.encode(queries)","execution_count":0,"outputs":[]},{"metadata":{"id":"Sh-I7flT1I_-"},"cell_type":"markdown","source":"## Load the Pickled embedding file"},{"metadata":{"id":"r_oeTNwd1I__","trusted":false},"cell_type":"code","source":"'''\nwith open('/kaggle/input/task3-data/full_text_embeddings.pkl','rb') as f:\n    embedding = pickle.load(f)\n'''","execution_count":0,"outputs":[]},{"metadata":{"id":"7f4je8QN1JAB"},"cell_type":"markdown","source":"## Retrieve Questions based on new question "},{"metadata":{"id":"9lFZkchy1JAC","trusted":false},"cell_type":"code","source":"#summary_model = Summarizer()","execution_count":0,"outputs":[]},{"metadata":{"id":"_RRKyEJI1JAF","trusted":false},"cell_type":"code","source":"#df = pd.DataFrame(columns=['Query','Cosine Similarity','Summary','Article'])","execution_count":0,"outputs":[]},{"metadata":{"id":"9kmsj0TT1JAH","outputId":"585134c1-6ac8-47bc-a729-13802247b470","trusted":false},"cell_type":"code","source":"'''\ntop_n_selects = 2\nimport scipy as sc\n\nfor query, query_embedding in zip(queries, query_embeddings):\n    distances = sc.spatial.distance.cdist([query_embedding], embedding, \"cosine\")[0]\n\n    results = zip(range(len(distances)), distances)\n    results = sorted(results, key=lambda x: x[1])\n\n    print('Query : ',query)\n    print('###########################################')\n    \n    for idx, distance in results[0:top_n_selects]:\n        print(\"\\nCosine Similarity (Score: %.4f)\" % (1-distance),\"\\n\")\n        \n        body = fin_df['combined'][idx].strip() \n        summary_model = Summarizer()\n        result = summary_model(body, min_length=60)\n        summary = ''.join(result)\n        similarity = (1-distance)\n        to_append = [query, similarity, summary, body]\n        a_series = pd.Series(to_append, index = df.columns)\n        df = df.append(a_series, ignore_index=True\n'''","execution_count":0,"outputs":[]},{"metadata":{"id":"4NH-Xg_r1JAJ","trusted":false},"cell_type":"code","source":"#df.head()","execution_count":0,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"## Summary\n\nThe medical and science community is at the fore-front of the fight against COVID-19, and as technologists the best we can do is to enable and empower them in this critical endeavor. With this motivation, our team has strived to design and deliver a solution that makes existing knowledge more readily accessible through the use of technology. The project required many crucial design and approach decisions and we relied on a few key tenets for guidance - practicality, user-friendliness and scalability. As we deliver this project, we hope it adds value in this battle against Corona virus.\n"},{"metadata":{},"cell_type":"markdown","source":"## Reference "},{"metadata":{},"cell_type":"markdown","source":"#### External References \n1. Center for Disease Prevention (CDC)\n\n2. NextstrainNextstrain is an open-source project to harness the scientific and public health  potential  of  pathogen  genome  data.  We  provide  a continually updatedview of publicly available data alongside powerful analytic and visualization tools for use by the community. \n\n3. China National GeneBankFrom the web page of the China National GeneBank, we can get the full coronavirus genome sequence. They have data from 2020-01-03 to 2020-03-15\n\n4. GISAID EpiCoV™ DatabaseGISAID  EpiCoV™  Database  Until  2020-03-27  01:16:12  CST,  the  total number of viruses is 1836.  China National GeneBank DataBase (CNGBdb) is an official partner of the GISAID Initiative. It provides access to EpiCoV \n2020-04-05PA1UenRev PA1©Ericsson AB 202025(30)Commercial in Confidenceand features the most complete collection of hCoV-19 genome sequences along with related clinical and epidemiological data\n\n5. China National GeneBankoFrom the web page of the China National GeneBank, we can get the  full  coronavirus  genome  sequence.  They  have  data  from 2020-01-03 to 2020-03-15\n\n6. GISAID EpiCoV™ DatabaseoGISAID  EpiCoV™  Database  Until  2020-03-27  01:16:12  CST, the total number of viruses is 1836.  China National GeneBank DataBase   (CNGBdb)   is   an   official   partner   of   the   GISAID Initiative.  It  provides  access  to  EpiCoV  and  features  the  most complete  collection  of  hCoV-19  genome  sequences  along  with related clinical and epidemiological data.\n\n7.\tBERT: https://arxiv.org/abs/1810.04805\n\n8.\tBERT Code: https://github.com/google-research/bert\n\n9.\tScibert: https://arxiv.org/pdf/1903.10676.pdf\n\n10.\tCosine Similarity: https://en.wikipedia.org/wiki/Cosine_similarity\n\n11.\tBERT Summarizer: https://pypi.org/project/bert-extractive-summarizer/\n\n12.\tBERT Summarizer : https://arxiv.org/abs/1906.04165\n\n13.\thttps://github.com/huggingface/neuralcoref\n"}],"metadata":{"accelerator":"TPU","colab":{"collapsed_sections":["V0kWeR6V1I_P","4S_hM27d1I_9","Sh-I7flT1I_-","7f4je8QN1JAB"],"machine_shape":"hm","name":"Task3Team3.ipynb","provenance":[]},"kernelspec":{"display_name":"Python 3","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.7.6"}},"nbformat":4,"nbformat_minor":4}