{"cells":[{"metadata":{"_uuid":"8f2839f25d086af736a60e9eeb907d3b93b6e0e5","_cell_guid":"b1076dfc-b9ad-4769-8c92-a6c4dae69d19","trusted":true},"cell_type":"code","source":"# This Python 3 environment comes with many helpful analytics libraries installed\n# It is defined by the kaggle/python docker image: https://github.com/kaggle/docker-python\n# For example, here's several helpful packages to load in \n\nimport numpy as np # linear algebra\nimport pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)\nimport seaborn as sns\n# Input data files are available in the \"../input/\" directory.\n# For example, running this (by clicking run or pressing Shift+Enter) will list the files in the input directory\n\nimport os\nprint(os.listdir(\"../input\"))\n\n# Any results you write to the current directory are saved as output.","execution_count":null,"outputs":[]},{"metadata":{"_cell_guid":"79c7e3d0-c299-4dcb-8224-4455121ee9b0","_uuid":"d629ff2d2480ee46fbb7e2d37f6b5fab8052498a","trusted":true},"cell_type":"code","source":"# AUGUSTUS is a piece of software that predicts genes ab initio using Hidden Markov Models.\nfile_path_augustus = os.path.join('..', 'input', 'genes-augustus.csv')\naugustus_df = pd.read_csv(file_path_augustus)\n\naugustus_df.head()","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"augustus_df.info()","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"txStart: transcription start site usually includes UTR(untranslated regions - doesn't contibute to protein sequence)\n\ncdsStart: downstream to the txStart, start of coding sequence - this is what ACTUALLY get translated to protein\n\nexonCount: mRNA fragments that are joined together during splicing to form full length CDS"},{"metadata":{},"cell_type":"markdown","source":"**Lets find out what is the average length of 5' UTRs**. \n\n***5'UTR = txStart - cdsStart***"},{"metadata":{"trusted":true},"cell_type":"code","source":"UTR_5_df = abs(augustus_df['txStart'] - augustus_df['cdsStart'])\nUTR_5_df.head()","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"sns.set_style('whitegrid')\nsns.set_context('talk')\nsns.set_palette('husl')\ng = sns.boxenplot(data=UTR_5_df)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"g = sns.boxenplot(data=UTR_5_df)\ng.set_yscale('log')","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"**Lets check the exon count**"},{"metadata":{"trusted":true},"cell_type":"code","source":"sns.set_palette('bright')\ng = sns.boxenplot(data=augustus_df['exonCount'])","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"Hmm  there are a few outliers with more than 20 exons"},{"metadata":{},"cell_type":"markdown","source":"Lets check if there is a correlation between mRNA length and exon number\n\nmRNA length = abs(txEnd - txStart)"},{"metadata":{"trusted":true},"cell_type":"code","source":"mRNA_df = abs(augustus_df['txEnd'] - augustus_df['txStart'])\nnormalization_factor = mRNA_df.max()/augustus_df['exonCount'].max()\nmRNA_df = mRNA_df / mRNA_df.max()\nexon_num_df = augustus_df['exonCount']*normalization_factor\nexon_num_df = exon_num_df/exon_num_df.max()","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"\ncorr = mRNA_df.corr(exon_num_df)\nprint(corr)\n","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"exon_mRNA_df = pd.concat([exon_num_df, mRNA_df], axis=1)\nexon_mRNA_df.columns = ['exonCount', 'mRNA_len']\nexon_mRNA_df.head()","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"import matplotlib.pyplot as plt\nsns.set_palette('bright')\nax = sns.scatterplot(x='mRNA_len', y='exonCount', data=exon_mRNA_df)\n","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"augustus_df['exonCount'].describe()","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"There exists a strong correlation upto certain mRNA length. After which the effect of mRNA length on exon number is minimal.\n\nIt is likely that there was some slection pressure against indefinite increase in exon number"},{"metadata":{"trusted":true},"cell_type":"code","source":"","execution_count":null,"outputs":[]}],"metadata":{"kernelspec":{"display_name":"Python 3","language":"python","name":"python3"},"language_info":{"name":"python","version":"3.6.4","mimetype":"text/x-python","codemirror_mode":{"name":"ipython","version":3},"pygments_lexer":"ipython3","nbconvert_exporter":"python","file_extension":".py"}},"nbformat":4,"nbformat_minor":1}