{"cells":[{"cell_type":"markdown","metadata":{"_cell_guid":"1bbf5de7-de3f-2147-194b-c1c06210e9b3"},"source":"This notebook is an example of how Pairwise Alignment can be used in Biopython to match one sequence within another sequence. Biopython has a module \"pairwise2\" which does just this.\n\nInfo on Sequence Alignment: https://en.wikipedia.org/wiki/Sequence_alignment\n\nInfo on biopython pairwise2: http://biopython.org/DIST/docs/api/Bio.pairwise2-module.html"},{"cell_type":"code","execution_count":null,"metadata":{"_cell_guid":"578c29d2-2ec4-1d63-0695-b137c7db4a45"},"outputs":[],"source":"import numpy as np # linear algebra\nimport pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)\nimport Bio\nfrom Bio import pairwise2"},{"cell_type":"markdown","metadata":{"_cell_guid":"9b681beb-0d99-e913-4fb9-a615046089ff"},"source":"## Lets start with the examples from the documentation: ##"},{"cell_type":"code","execution_count":null,"metadata":{"_cell_guid":"53e45ee7-b92e-a33d-ed5c-413dea5a1634"},"outputs":[],"source":"# Find the best global alignment between the two sequences.\n# Identical characters are given 1 point.  No points are deducted\n# for mismatches or gaps.\n\nalignments = pairwise2.align.globalxx(\"ACCGT\", \"ACG\")\nfor a in pairwise2.align.globalxx(\"ACCGT\", \"ACG\"):\n    print(pairwise2.format_alignment(*a))"},{"cell_type":"code","execution_count":null,"metadata":{"_cell_guid":"accd23a6-5123-967c-81c6-b4c3d66e8f0e"},"outputs":[],"source":"# Same thing as before, but with a local alignment.\n\nfor a in pairwise2.align.localxx(\"ACCGT\", \"ACG\"):\n     print(pairwise2.format_alignment(*a))"},{"cell_type":"code","execution_count":null,"metadata":{"_cell_guid":"def720c4-c942-5d7d-ac04-2a7cc5eefb0f"},"outputs":[],"source":"# Do a global alignment.  Identical characters are given 2 points,\n# 1 point is deducted for each non-identical character.\nfor a in pairwise2.align.globalmx(\"ACCGT\", \"ACG\", 2, -1):\n    print(pairwise2.format_alignment(*a))"},{"cell_type":"code","execution_count":null,"metadata":{"_cell_guid":"5ff499db-e3ae-f154-b528-7783cfcb6d61"},"outputs":[],"source":"# Same as above, except now 0.5 points are deducted when opening a\n# gap, and 0.1 points are deducted when extending it.\nfor a in pairwise2.align.globalms(\"ACCGT\", \"ACG\", 2, -1, -.5, -.1):\n     print(pairwise2.format_alignment(*a))"},{"cell_type":"code","execution_count":null,"metadata":{"_cell_guid":"e74f27a2-de8e-14a7-75de-c0d964ba41f9"},"outputs":[],"source":"#The alignment function can also use known matrices already included in Biopython ( Bio.SubsMat -> MatrixInfo ):\n\nfrom Bio.SubsMat import MatrixInfo as matlist\nmatrix = matlist.blosum62\nfor a in pairwise2.align.globaldx(\"KEVLA\", \"EVL\", matrix):\n     print(pairwise2.format_alignment(*a))"},{"cell_type":"code","execution_count":null,"metadata":{"_cell_guid":"9d565371-bffa-819a-019b-0d7fe8e56536"},"outputs":[],"source":"## Next lets try using the actual data from our dataset: ##"},{"cell_type":"code","execution_count":null,"metadata":{"_cell_guid":"dfccbfe0-3172-6281-d3a8-72d8a7bdde1e"},"outputs":[],"source":"# First lets import our chromosomes from genome.fa\n\nfrom Bio import SeqIO\ncount = 0\nsequences = []\n\nfor seq_record in SeqIO.parse(\"../input/genome.fa\", \"fasta\"):\n    if (count < 6):\n        sequences.append(seq_record)\n        print(\"Id: \" + seq_record.id + \" \\t \" + \"Length: \" + str(\"{:,d}\".format(len(seq_record))) )\n        print(repr(seq_record.seq) + \"\\n\")\n        count = count + 1\n        \nchr2L = sequences[0].seq\nchr2R = sequences[1].seq\nchr3L = sequences[2].seq\nchr3R = sequences[3].seq\nchr4 = sequences[4].seq\nchrM = sequences[5].seq"},{"cell_type":"code","execution_count":null,"metadata":{"_cell_guid":"67f9fa79-c0eb-e7ca-2813-902a7a0020a4"},"outputs":[],"source":"# Next lets grab some mRNA sequences to test from mrna-genbank.fa\n\ncount = 0\nmrna_sequences = []\n\nfor seq_record in SeqIO.parse(\"../input/mrna-genbank.fa\", \"fasta\"):\n    if (count < 6):\n        mrna_sequences.append(seq_record)\n        print(\"Id: \" + seq_record.id + \" \\t \" + \"Length: \" + str(\"{:,d}\".format(len(seq_record))) )\n        print(repr(seq_record.seq) + \"\\n\")\n        count = count + 1\n        \nmRNA1 = mrna_sequences[0].seq\nmRNA2 = mrna_sequences[1].seq\nmRNA3 = mrna_sequences[2].seq\nmRNA4 = mrna_sequences[3].seq\nmRNA5 = mrna_sequences[4].seq\nmRNA1 = sequences[5].seq"}],"metadata":{"_change_revision":0,"_is_fork":false,"kernelspec":{"display_name":"Python 3","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.5.2"}},"nbformat":4,"nbformat_minor":0}