{"cells":[{"metadata":{},"cell_type":"markdown","source":"# Nucleotide\n\n![NCBI](https://www.ncbi.nlm.nih.gov/Class/MLACourse/Original8Hour/Genetics/nucleotide2.gif)\n\n\n\nDNA and Nucleotides(NCBI,1999)\n\n\nCoVs are enveloped, positive-stranded RNA viruses with nucleocapsid. For addressing pathogenetic mechanisms of SARS-CoV-2, its viral structure, and genome must be considerations. In CoVs, the genomic structure is organized in a +ssRNA of approximately 30 kb in length — the largest known RNA viruses — and with a 5′-cap structure and 3′-poly-A tail. (https://www.ncbi.nlm.nih.gov/books/NBK554776/)\n<br/>\n<br/>\n<br/>\n![pathways](https://mms.businesswire.com/media/20200311005738/en/779142/5/COVID-19s_Virus_Genomic_Evolution_Pathway_March_10_2020.jpg?download=1)\n<br/>\nCOVID-19's Virus Genomic Evolution Pathway.\nhttps://www.businesswire.com/news/home/20200311005738/en/Next-Gen-Artificial-Intelligence-AI-Company-Graphen-Monitors"},{"metadata":{"_uuid":"8f2839f25d086af736a60e9eeb907d3b93b6e0e5","_cell_guid":"b1076dfc-b9ad-4769-8c92-a6c4dae69d19","trusted":true},"cell_type":"code","source":"# This Python 3 environment comes with many helpful analytics libraries installed\n# It is defined by the kaggle/python docker image: https://github.com/kaggle/docker-python\n# For example, here's several helpful packages to load in \n\nimport numpy as np # linear algebra\nimport pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)\nimport matplotlib.pyplot as plt\nimport seaborn as sns\n\n# Input data files are available in the \"../input/\" directory.\n# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory\n\nimport os\nfor dirname, _, filenames in os.walk('/kaggle/input'):\n    for filename in filenames:\n        print(os.path.join(dirname, filename))\n\n# Any results you write to the current directory are saved as output.","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"In this notebook, we will see the similarity or disimarity on the nucleotide."},{"metadata":{"_uuid":"d629ff2d2480ee46fbb7e2d37f6b5fab8052498a","_cell_guid":"79c7e3d0-c299-4dcb-8224-4455121ee9b0","trusted":true},"cell_type":"code","source":"data = pd.read_csv('/kaggle/input/sars-coronavirus-accession/MN997409.1-4NY0T82X016-Alignment-HitTable.csv')","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"data.head()","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"data = data.rename(columns={\"MN997409.1\": \"query acc.ver\", \"MN997409.1.1\": \"subject acc.ver\",\n                            \"100.000\":\"% identity\",\"29882\":\"alignment length\",\"0\":\"mismatches\",\n                            \"0.1\":\"gap opens\",\"1\":\"q. start\",\"29882.1\":\"q. end\",\"1.1\":\"s. start\",\n                           \"29882.2\t\":\"s. end\",\"0.0\":\"evalue\",\"55182\":\"bit score\"})","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"data.head()","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"Seems that the first row carried one of the information. We will recover it and replace the column name with proper name."},{"metadata":{"trusted":true},"cell_type":"code","source":"data = data.append(pd.Series(['MN997409.1',\t'MN997409.1.1',\t\n                    100.000,\t29882,\t\n                    0,\t0.1,\t1,\t29882.1,\t\n                    1.1,\t29882.2,\t0.0,\t55182], index=data.columns), ignore_index=True)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"data.tail()","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"data.shape","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"data.describe()","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"data.corr()","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"plt.figure(figsize=(10,10))\nsns.heatmap(data.corr(), vmax=1, square=True,annot=True,cmap='viridis')\n\nplt.title('Correlation between different fearures')","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"# K-Means Clustering\n\nWe will see the genome information clusters"},{"metadata":{"trusted":true},"cell_type":"code","source":"from sklearn.cluster import KMeans","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"make dummy variable"},{"metadata":{"trusted":true},"cell_type":"code","source":"data['subject acc.ver'] = data['subject acc.ver'].astype(\"category\").cat.codes","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"data.head()","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"X = data.iloc[:,1:].values","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"## Using Silhouette Score for Evaluating Clusters\n\nSilhouette analysis can be used to study the separation distance between the resulting clusters. The silhouette plot displays a measure of how close each point in one cluster is to points in the neighboring clusters and thus provides a way to assess parameters like number of clusters visually. This measure has a range of [-1, 1].\n\nThis part was based on [Scikit-learn documentation.](https://scikit-learn.org/stable/auto_examples/cluster/plot_kmeans_silhouette_analysis.html)"},{"metadata":{"trusted":true},"cell_type":"code","source":"from sklearn.metrics import silhouette_samples, silhouette_score\nimport matplotlib.cm as cm","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"from sklearn.preprocessing import StandardScaler\nstandardized_data = StandardScaler().fit_transform(X)\nprint(standardized_data.shape)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"range_n_clusters = [2, 3, 4, 5, 6, 7, 8, 9, 10]\n\nfor n_clusters in range_n_clusters:\n    # Create a subplot with 1 row and 2 columns\n    fig, (ax1, ax2) = plt.subplots(1, 2)\n    fig.set_size_inches(18, 7)\n\n    # The 1st subplot is the silhouette plot\n    # The silhouette coefficient can range from -1, 1 but in this example all\n    # lie within [-0.1, 1]\n    ax1.set_xlim([-0.1, 1])\n    # The (n_clusters+1)*10 is for inserting blank space between silhouette\n    # plots of individual clusters, to demarcate them clearly.\n    ax1.set_ylim([0, len(X) + (n_clusters + 1) * 10])\n\n    # Initialize the clusterer with n_clusters value and a random generator\n    # seed of 10 for reproducibility.\n    clusterer = KMeans(n_clusters=n_clusters, random_state=10)\n    cluster_labels = clusterer.fit_predict(X)\n\n    # The silhouette_score gives the average value for all the samples.\n    # This gives a perspective into the density and separation of the formed\n    # clusters\n    silhouette_avg = silhouette_score(X, cluster_labels)\n    print(\"For n_clusters =\", n_clusters,\n          \"The average silhouette_score is :\", silhouette_avg)\n\n    # Compute the silhouette scores for each sample\n    sample_silhouette_values = silhouette_samples(X, cluster_labels)\n\n    y_lower = 10\n    for i in range(n_clusters):\n        # Aggregate the silhouette scores for samples belonging to\n        # cluster i, and sort them\n        ith_cluster_silhouette_values = \\\n            sample_silhouette_values[cluster_labels == i]\n\n        ith_cluster_silhouette_values.sort()\n\n        size_cluster_i = ith_cluster_silhouette_values.shape[0]\n        y_upper = y_lower + size_cluster_i\n\n        color = cm.nipy_spectral(float(i) / n_clusters)\n        ax1.fill_betweenx(np.arange(y_lower, y_upper),\n                          0, ith_cluster_silhouette_values,\n                          facecolor=color, edgecolor=color, alpha=0.7)\n\n        # Label the silhouette plots with their cluster numbers at the middle\n        ax1.text(-0.05, y_lower + 0.5 * size_cluster_i, str(i))\n\n        # Compute the new y_lower for next plot\n        y_lower = y_upper + 10  # 10 for the 0 samples\n\n    ax1.set_title(\"The silhouette plot for the various clusters.\")\n    ax1.set_xlabel(\"The silhouette coefficient values\")\n    ax1.set_ylabel(\"Cluster label\")\n\n    # The vertical line for average silhouette score of all the values\n    ax1.axvline(x=silhouette_avg, color=\"red\", linestyle=\"--\")\n\n    ax1.set_yticks([])  # Clear the yaxis labels / ticks\n    ax1.set_xticks([-0.1, 0, 0.2, 0.4, 0.6, 0.8, 1])\n\n    # 2nd Plot showing the actual clusters formed\n    colors = cm.nipy_spectral(cluster_labels.astype(float) / n_clusters)\n    ax2.scatter(X[:, 0], X[:, 1], marker='.', s=30, lw=0, alpha=0.7,\n                c=colors, edgecolor='k')\n\n    # Labeling the clusters\n    centers = clusterer.cluster_centers_\n    # Draw white circles at cluster centers\n    ax2.scatter(centers[:, 0], centers[:, 1], marker='o',\n                c=\"white\", alpha=1, s=200, edgecolor='k')\n\n    for i, c in enumerate(centers):\n        ax2.scatter(c[0], c[1], marker='$%d$' % i, alpha=1,\n                    s=50, edgecolor='k')\n\n    ax2.set_title(\"The visualization of the clustered data.\")\n    ax2.set_xlabel(\"Feature space for the 1st feature\")\n    ax2.set_ylabel(\"Feature space for the 2nd feature\")\n\n    plt.suptitle((\"Silhouette analysis for KMeans clustering on sample data \"\n                  \"with n_clusters = %d\" % n_clusters),\n                 fontsize=14, fontweight='bold')\n\nplt.show()","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"# PCA Visualization\n\nPCA used for dimension reduction for the visualization of high dimensional data.\n<br/>\n<br/>\n\n![](https://miro.medium.com/max/602/0*PnqMbZEdnuL9yHuo.png)\n\n![](https://miro.medium.com/max/562/0*cgmVlr-0eTeJrudC.png)\n\n[Image source](https://medium.com/analytics-vidhya/principal-component-analysis-pca-with-code-on-mnist-dataset-da7de0d07c22)\n\nI found it helpful [here](https://medium.com/analytics-vidhya/principal-component-analysis-pca-with-code-on-mnist-dataset-da7de0d07c22)"},{"metadata":{},"cell_type":"markdown","source":"### Make covariance matrix"},{"metadata":{"trusted":true},"cell_type":"code","source":"covar_matrix = np.matmul(standardized_data.T , standardized_data)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"covar_matrix.shape","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"### Eigenvalue & eigenvector"},{"metadata":{"trusted":true},"cell_type":"code","source":"from scipy.linalg import eigh ","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"#top2 eigenvalue\nvalues, vectors = eigh(covar_matrix, eigvals=(9,10))","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"vectors.shape","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"#transpose\nvectors = vectors.T\nvectors.shape","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"new_coordinates = np.matmul(vectors, standardized_data.T)\nprint (\"Resultant at new data shape: \", vectors.shape, \"*\", standardized_data.T.shape,\" = \", new_coordinates.shape)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"new_coordinates = np.vstack((new_coordinates)).T\n\ndf = pd.DataFrame(data=new_coordinates, columns=(\"1st_principal\", \"2nd_principal\"))\nprint(df.head())","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"\nsns.set()\nsns.FacetGrid(df, size=6).map(plt.scatter, '1st_principal', '2nd_principal').add_legend()\nplt.title('PCA visualization of sequences')\nplt.show()","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"### We found 5 different class of genome features from the PCA. To be added, the explanation about them."},{"metadata":{"trusted":true},"cell_type":"code","source":"","execution_count":null,"outputs":[]}],"metadata":{"kernelspec":{"language":"python","display_name":"Python 3","name":"python3"},"language_info":{"pygments_lexer":"ipython3","nbconvert_exporter":"python","version":"3.6.4","file_extension":".py","codemirror_mode":{"name":"ipython","version":3},"name":"python","mimetype":"text/x-python"}},"nbformat":4,"nbformat_minor":4}